// @er/primeng/esm2015/lib/components/notify.alert/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { BaseComponent } from '@er/core';
import { NotifyLevel, NotifyType, StateNames, UiPosition } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { MessageService } from 'primeng/api';
import { filter } from 'rxjs/operators';
/** @type {?} */
export const DEFAULT_ALERT_PROPS = {
    animation: true,
    duration: 3000,
    immutable: true,
    closable: true,
    autoZIndex: true,
    position: UiPosition.rightTop
};
export class PngAlertComponent extends BaseComponent {
    /**
     * @param {?} messageService
     * @param {?} cd
     */
    constructor(messageService, cd) {
        super();
        this.messageService = messageService;
        this.cd = cd;
        this.alert = Object.assign({}, DEFAULT_ALERT_PROPS, ConfigUtils.getConfig().alert || {});
        this.defaultNotifyType = ConfigUtils.getConfig().notify || NotifyType.ALERT;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.$subscriptions = StatesUtils.observe(StateNames.notify)
            .pipe(filter(notify => notify && notify.message), filter(notify => notify.type === NotifyType.ALERT || (!notify.type && this.defaultNotifyType === NotifyType.ALERT))).subscribe(notify => {
            /** @type {?} */
            const model = this.getNotifyModel(notify);
            this.messageService.add(model);
            this.cd.detectChanges();
        });
    }
    /**
     * @param {?} notify
     * @return {?}
     */
    getNotifyModel(notify) {
        return {
            key: notify.key,
            severity: notify.level || NotifyLevel.INFO,
            summary: notify.title,
            detail: notify.message,
            sticky: notify.sticky || false,
            life: notify.duration || notify.level === NotifyLevel.ERROR ? DEFAULT_ALERT_PROPS.duration * 2 : DEFAULT_ALERT_PROPS.duration,
            closable: notify.closable || true,
            data: notify.data
        };
    }
}
PngAlertComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-alert',
                template: `
    <p-toast [styleClass]="alert.styleClass"
             [style]="alert.style"
             [position]="alert.position"
             [autoZIndex]="alert.autoZIndex"
    ></p-toast>
  `,
                providers: [MessageService],
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngAlertComponent.ctorParameters = () => [
    { type: MessageService },
    { type: ChangeDetectorRef }
];
if (false) {
    /** @type {?} */
    PngAlertComponent.prototype.alert;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.defaultNotifyType;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.cd;
}
