// @er/primeng/esm2015/lib/components/list/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_LIST_BOX_PROPS } from './type';
export class PngListBoxComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.templateNames = {
            header: 'header',
            footer: 'footer',
            item: 'item'
        };
        this.$defaultProps = DEFAULT_PNG_LIST_BOX_PROPS;
    }
}
PngListBoxComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-list-box',
                template: `
    <p-listbox
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      [erApiData]="apiDataProps"
      [options]="data$|async"
      (onChange)="updateTouch()"
    >
      <p-header *ngIf="template(templateNames.header)">
        <ng-container *ngTemplateOutlet="template(templateNames.header)"></ng-container>
      </p-header>

      <p-header *ngIf="header">
        <div [innerHTML]="header"></div>
      </p-header>

      <p-footer *ngIf="template(templateNames.footer)">
        <ng-container *ngTemplateOutlet="template(templateNames.footer)"></ng-container>
      </p-footer>

      <p-footer *ngIf="$props.$ext['showValue'] || footer">
        <ng-container *ngIf="$props.$ext['showValue']">
          选中值：{{value | erValue: $props.$ext['field']}}
        </ng-container>
        <ng-container *ngIf="footer">
          <div [innerHTML]="footer"></div>
        </ng-container>
      </p-footer>

      <ng-template let-item let-i="index" pTemplate="item">
        <er-data-item [customerTpl]="template(templateNames.item)"
                      [item]="item.value"
                      [props]="$props.$ext.dataItemProps">
        </er-data-item>
      </ng-template>
    </p-listbox>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngListBoxComponent),
                        multi: true
                    }]
            }] }
];
PngListBoxComponent.propDecorators = {
    header: [{ type: Input }],
    footer: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngListBoxComponent.prototype.header;
    /** @type {?} */
    PngListBoxComponent.prototype.footer;
    /** @type {?} */
    PngListBoxComponent.prototype.templateNames;
    /** @type {?} */
    PngListBoxComponent.prototype.$defaultProps;
}
