// @er/primeng/esm2015/lib/components/list/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngListBoxProps() { }
if (false) {
    /** @type {?|undefined} */
    PngListBoxProps.prototype.multiple;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.listStyle;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.readonly;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.disabled;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.checkbox;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.filter;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.filterMode;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.metaKeySelection;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.dataKey;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.showToggleAll;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.optionLabel;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.onDblClick;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.onChange;
    /** @type {?|undefined} */
    PngListBoxProps.prototype.$ext;
}
/** @type {?} */
export const DEFAULT_PNG_LIST_BOX_PROPS = {
    optionLabel: 'label',
    checkbox: true,
    multiple: true,
    showToggleAll: true,
    $ext: {
        dataItemProps: {
            itemStyleClass: 'd-inline-flex ml-2'
        }
    }
};
