// @er/primeng/esm5/lib/components/radio/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { DEFAULT_RADIO_PROPS } from './type';
var PngRadioComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngRadioComponent, _super);
    function PngRadioComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.valueParser = _this.parseValue;
        _this.modelFormatter = _this.formatModel;
        _this.$defaultProps = DEFAULT_RADIO_PROPS;
        return _this;
    }
    Object.defineProperty(PngRadioComponent.prototype, "rbValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rbValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            var _this = this;
            this._rbValue = value;
            /** @type {?} */
            var values = this.data$.value;
            if (CommonsUtils.isArray(values)) {
                /** @type {?} */
                var matched = ((/** @type {?} */ (values))).filter(function (v) { return v[_this.dataKey] === value; });
                if (matched.length === 1) {
                    this.value = DataItemUtils.getItemValue(matched[0], this.valueKey);
                }
                else {
                    this.value = null;
                }
            }
            else {
                this.value = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngRadioComponent.prototype, "dataKey", {
        get: /**
         * @return {?}
         */
        function () {
            return CommonsUtils.get(this.$props, '$ext.dataItemProps.idKey', 'value');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngRadioComponent.prototype, "valueKey", {
        get: /**
         * @return {?}
         */
        function () {
            return CommonsUtils.get(this.$props, '$ext.dataItemProps.valueKey', 'value');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} model
     * @return {?}
     */
    PngRadioComponent.prototype.formatModel = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        if (CommonsUtils.isJson(model)) {
            this._rbValue = DataItemUtils.getItemValue(model, this.dataKey);
        }
        return model;
    };
    PngRadioComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-radio',
                    template: "\n    <div class=\"d-flex pb-2\" [erApiData]=\"apiDataProps\" [ngClass]=\"$props.$ext['inline']===false?'flex-column':'flex-wrap'\">\n      <ng-container *ngFor=\"let option of data$|async|erDataItems:$props.$ext.dataItemProps;let i=index\">\n       <span class=\"d-flex align-items-center\">\n        <p-radioButton\n          class=\"m-2\"\n          [(ngModel)]=\"rbValue\"\n          erPropsBind\n          [props]=\"$props\"\n          [context]=\"ctx\"\n          [inputId]=\"i\"\n          [value]=\"option.data|erValue:dataKey\"\n          (onClick)=\"updateTouch()\"\n        >\n        </p-radioButton>\n        <span class=\"mr-4\" [innerHTML]=\"option.label|erSafeHtml\"></span>\n       </span>\n      </ng-container>\n    </div>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngRadioComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngRadioComponent;
}(DataAwareControlWrapperComponent));
export { PngRadioComponent };
if (false) {
    /** @type {?} */
    PngRadioComponent.prototype.valueParser;
    /** @type {?} */
    PngRadioComponent.prototype.modelFormatter;
    /** @type {?} */
    PngRadioComponent.prototype.$defaultProps;
    /** @type {?} */
    PngRadioComponent.prototype._rbValue;
}
