// @er/primeng/esm5/lib/components/table/component/caption/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { OverlayPanel } from 'primeng/primeng';
import { PngTableComponent } from '../table/component';
var PngTableCaptionComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableCaptionComponent, _super);
    function PngTableCaptionComponent(pngTable) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        _this.onSimpleSearch = new EventEmitter();
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    PngTableCaptionComponent.prototype.search = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onSimpleSearch.emit(event);
    };
    /**
     * @return {?}
     */
    PngTableCaptionComponent.prototype.openPanel = /**
     * @return {?}
     */
    function () {
        StatesUtils.update(StateNames.popup, (/** @type {?} */ ((/** @type {?} */ ({
            header: '查询条件定义',
            content: this.searchBuilderPanel,
            visible: true,
            minWidth: 700,
            contentMaxHeight: 400
        })))));
    };
    /**
     * @return {?}
     */
    PngTableCaptionComponent.prototype.refresh = /**
     * @return {?}
     */
    function () {
        this.pngTable.load();
    };
    PngTableCaptionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-table-caption',
                    template: "\n    <ng-container *ngIf=\"customerTpl;else default\">\n      <ng-container *ngTemplateOutlet=\"customerTpl;context:{$implicit:pngTable}\"></ng-container>\n    </ng-container>\n    <ng-template #default>\n      <div>\n        <span class=\"pull-left table-caption\" [innerHTML]=\"caption|erSafeHtml\"></span>\n        <span class=\"pull-right d-inline\"><i class=\"fa fa-refresh text-primary\" title=\"\u5237\u65B0\u8868\u683C\" (click)=\"refresh()\"></i></span>\n      </div>\n      <div class=\"clearfix\"></div>\n      <div *ngIf=\"hasSearchBox || hasAddAction || hasSearchInput\">\n        <div class=\"d-flex align-items-start\">\n          <div class=\"pl-3 py-3\" *ngIf=\"hasAddAction\">\n            <png-button erPropsBind [props]=\"addButton\" [outputContext]=\"{pngTable: pngTable}\"></png-button>\n          </div>\n          <div class=\"ml-auto py-3\" *ngIf=\"hasSearchInput && !hasSearchBox\">\n            <png-search-box title=\"\u7EFC\u5408\u641C\u7D22\uFF0C\u8F93\u5165\u56DE\u8F66\u952E\u6267\u884C\" (onSearch)=\"search($event)\"></png-search-box>\n          </div>\n          <div class=\"ml-auto\" *ngIf=\"hasSearchBox\">\n            <button pButton icon=\"fa fa-search\" label=\"\u9AD8\u7EA7\u67E5\u8BE2\" (click)=\"openPanel()\"></button>\n            <ng-template #searchBuilderPanel>\n              <png-table-query-builder></png-table-query-builder>\n            </ng-template>\n          </div>\n        </div>\n      </div>\n    </ng-template>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngTableCaptionComponent.ctorParameters = function () { return [
        { type: PngTableComponent }
    ]; };
    PngTableCaptionComponent.propDecorators = {
        caption: [{ type: Input }],
        hasAddAction: [{ type: Input }],
        hasSearchInput: [{ type: Input }],
        hasSearchBox: [{ type: Input }],
        addButton: [{ type: Input }],
        onSimpleSearch: [{ type: Output }],
        searchBuilderPanel: [{ type: ViewChild, args: ['searchBuilderPanel',] }]
    };
    return PngTableCaptionComponent;
}(BaseComponent));
export { PngTableCaptionComponent };
if (false) {
    /** @type {?} */
    PngTableCaptionComponent.prototype.caption;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasAddAction;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasSearchInput;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasSearchBox;
    /** @type {?} */
    PngTableCaptionComponent.prototype.addButton;
    /** @type {?} */
    PngTableCaptionComponent.prototype.onSimpleSearch;
    /** @type {?} */
    PngTableCaptionComponent.prototype.searchBuilderPanel;
    /** @type {?} */
    PngTableCaptionComponent.prototype.pngTable;
}
