// @er/primeng/esm2015/lib/components/overlay-panel/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { OverlayPanel } from 'primeng/primeng';
export class PngOverlayPanelComponent extends BaseComponent {
    /**
     * @param {?} event
     * @return {?}
     */
    show(event) {
        this.panel.show(event);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    toggle(event) {
        this.panel.toggle(event);
    }
    /**
     * @return {?}
     */
    hide() {
        this.panel.hide();
    }
}
PngOverlayPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-overlay-panel',
                template: "<p-overlayPanel #overlayPanel [showCloseIcon]=\"true\" [dismissable]=\"false\">\n   <er-dyna-content [content]=\"content\" [context]=\"context\"></er-dyna-content>\n</p-overlayPanel>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PngOverlayPanelComponent.propDecorators = {
    content: [{ type: Input }],
    context: [{ type: Input }],
    panel: [{ type: ViewChild, args: [OverlayPanel,] }]
};
if (false) {
    /** @type {?} */
    PngOverlayPanelComponent.prototype.content;
    /** @type {?} */
    PngOverlayPanelComponent.prototype.context;
    /** @type {?} */
    PngOverlayPanelComponent.prototype.panel;
}
