// @er/formly-primeng/esm5/lib/field-types/dyna/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { DynaContentComponent } from '@er/core';
import { distinctUntilChanged } from 'rxjs/operators';
import { PngFormlyFieldType } from '../../base/field-type';
/**
 * @record
 */
export function DynaProps() { }
if (false) {
    /** @type {?} */
    DynaProps.prototype.content;
    /** @type {?} */
    DynaProps.prototype.refreshOnFieldChanges;
    /** @type {?} */
    DynaProps.prototype.refreshOnFormChanges;
}
var PngFormlyFieldDynaComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngFormlyFieldDynaComponent, _super);
    function PngFormlyFieldDynaComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    PngFormlyFieldDynaComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngOnInit.call(this);
        if (this.props['refreshOnForm']) {
            /** @type {?} */
            var form = this.field.formlyForm.form;
            this.subscription = form.valueChanges
                .pipe(distinctUntilChanged())
                .subscribe(function (_) {
                _this.comp.refresh();
            });
        }
        else if (this.props['refreshOnField']) {
            /** @type {?} */
            var formControl = this.field.formlyForm.formControl(this.props['refreshOnField']);
            this.subscription = formControl.valueChanges
                .pipe(distinctUntilChanged())
                .subscribe(function (_) {
                _this.comp.refresh();
            });
        }
        else {
            this.comp.refresh();
        }
    };
    /**
     * @return {?}
     */
    PngFormlyFieldDynaComponent.prototype.unsubscribe = /**
     * @return {?}
     */
    function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    /**
     * @return {?}
     */
    PngFormlyFieldDynaComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe();
    };
    PngFormlyFieldDynaComponent.decorators = [
        { type: Component, args: [{
                    template: "\n    <er-dyna-content [content]=\"props['content']\" [context]=\"{field: field}\"></er-dyna-content>\n  "
                }] }
    ];
    PngFormlyFieldDynaComponent.propDecorators = {
        comp: [{ type: ViewChild, args: [DynaContentComponent,] }]
    };
    return PngFormlyFieldDynaComponent;
}(PngFormlyFieldType));
export { PngFormlyFieldDynaComponent };
if (false) {
    /** @type {?} */
    PngFormlyFieldDynaComponent.prototype.comp;
    /** @type {?} */
    PngFormlyFieldDynaComponent.prototype.subscription;
}
