// @er/primeng/esm2015/lib/components/notify.announce/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { BaseComponent } from '@er/core';
import { NotifyLevel, NotifyType, StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { MessageService } from 'primeng/api';
import { filter } from 'rxjs/operators';
/** @type {?} */
export const DEFAULT_ANNOUNCE_PROPS = {
    level: NotifyLevel.INFO,
    animation: true,
    closable: true,
    baseZIndex: 2000
};
export class PngAnnounceComponent extends BaseComponent {
    /**
     * @param {?} messageService
     * @param {?} cd
     */
    constructor(messageService, cd) {
        super();
        this.messageService = messageService;
        this.cd = cd;
        this.announce = Object.assign({}, DEFAULT_ANNOUNCE_PROPS, ConfigUtils.getConfig().announce || {});
        this.defaultNotifyType = ConfigUtils.getConfig().notify;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.$subscriptions = StatesUtils.observe(StateNames.notify)
            .pipe(filter(notify => notify && notify.message), filter(notify => notify.type === NotifyType.ANNOUNCE || (!notify.type && this.defaultNotifyType === NotifyType.ANNOUNCE))).subscribe(notify => {
            /** @type {?} */
            const model = this.getNotifyModel(notify);
            this.messageService.add(model);
            this.cd.detectChanges();
        });
    }
    /**
     * @param {?} notify
     * @return {?}
     */
    getNotifyModel(notify) {
        return {
            key: notify.key,
            severity: notify.level || NotifyLevel.INFO,
            summary: notify.title,
            detail: notify.message,
            closable: notify.closable,
            data: notify.data
        };
    }
}
PngAnnounceComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-announce',
                template: `
    <p-messages [styleClass]="announce.styleClass"
                [style]="announce.style"
                [closable]="announce.closable"
    ></p-messages>
  `,
                providers: [MessageService],
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngAnnounceComponent.ctorParameters = () => [
    { type: MessageService },
    { type: ChangeDetectorRef }
];
if (false) {
    /** @type {?} */
    PngAnnounceComponent.prototype.announce;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.defaultNotifyType;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.cd;
}
