// @er/core/esm2015/lib/directives/text-tag/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { ColorUtils, CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
export class TagDirective extends BaseDirective {
    /**
     * @param {?} el
     * @param {?} render
     */
    constructor(el, render) {
        super();
        this.el = el;
        this.render = render;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!CommonsUtils.isEmpty(this.color)) {
            this.render.setStyle(this.el.nativeElement, 'color', this.color);
        }
        if (!CommonsUtils.isEmpty(this.bg)) {
            this.render.setStyle(this.el.nativeElement, 'background-color', this.bg);
        }
        if (CommonsUtils.isEmpty(this.color) && CommonsUtils.isEmpty(this.bg)) {
            if (!this.text) {
                this.text = this.el.nativeElement.innerHTML;
            }
            this.render.setStyle(this.el.nativeElement, 'background', ColorUtils.getColor(this.text));
            this.render.setStyle(this.el.nativeElement, 'color', 'white');
        }
        this.render.addClass(this.el.nativeElement, 'badge-pill');
        this.render.addClass(this.el.nativeElement, 'badge');
    }
}
TagDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erTag]'
            },] }
];
/** @nocollapse */
TagDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
TagDirective.propDecorators = {
    text: [{ type: Input }],
    bg: [{ type: Input }],
    color: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    TagDirective.prototype.text;
    /** @type {?} */
    TagDirective.prototype.bg;
    /** @type {?} */
    TagDirective.prototype.color;
    /**
     * @type {?}
     * @private
     */
    TagDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    TagDirective.prototype.render;
}
