// @er/primeng/esm2015/lib/components/chart/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { ApiUtils, BaseComponent } from '@er/core';
import { ChartType } from '@er/types';
import { ChartUtils, CommonsUtils, IdUtils } from '@er/utils';
import { BehaviorSubject } from 'rxjs';
export class PngChartComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.type = ChartType.BAR;
        this.plugins = [];
        this.responsive = true;
        this.switchable = true;
        this.templateNames = {
            header: 'header'
        };
        this.onDataSelect = new EventEmitter();
        this.data$ = new BehaviorSubject(undefined);
        this.chartTypes = [
            { label: '柱状图', value: ChartType.BAR, icon: 'fa fa-bar-chart' },
            { label: '圆饼图', value: ChartType.PIE, icon: 'fa fa-pie-chart' },
            { label: '曲线图', value: ChartType.LINE, icon: 'fa fa-line-chart' },
            { label: '雷达图', value: ChartType.RADAR, icon: 'fa fa-bar' },
            { label: '区域图', value: ChartType.AREA, icon: 'fa fa-area-chart' },
            { label: '环形图', value: ChartType.DOUGHNUT, icon: 'fa fa-circle' }
        ];
    }
    /**
     * @return {?}
     */
    get data() {
        return this.data$.value;
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set data(data) {
        this.data$.next(data || {});
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.rawData = undefined;
        this.buildChart();
    }
    /**
     * @return {?}
     */
    changeType() {
        this.buildChart();
    }
    /**
     * @return {?}
     */
    reBuildChart() {
        this.rawData = undefined;
        this.buildChart();
    }
    /**
     * @return {?}
     */
    buildChart() {
        if (!this.$id) {
            this.$id = IdUtils.getRandom();
        }
        if (this.rawData) {
            this.resolvedRawData(this.rawData);
        }
        else if (this.dataSource) {
            /** @type {?} */
            let ds = this.dataSource;
            if (!ds.hasOwnProperty('apiDataProps')) {
                /** @type {?} */
                const apiDataProps = ds;
                ds = {};
                ds['apiDataProps'] = apiDataProps;
            }
            if (this.resolveDataSource) {
                ds = this.resolveDataSource.method((/** @type {?} */ (ds)));
            }
            ApiUtils.batchFetch(ds, this.$id)
                .subscribe(data => {
                this.rawData = data;
                this.resolvedRawData(this.rawData);
            });
        }
    }
    /**
     * @param {?} rawData
     * @return {?}
     */
    resolvedRawData(rawData) {
        if (!CommonsUtils.isEmpty(rawData)) {
            /** @type {?} */
            let data;
            /** @type {?} */
            const datasets = [];
            /** @type {?} */
            let labels = [];
            if (CommonsUtils.isFunction(this.dimensions)) {
                data = ((/** @type {?} */ (this.dimensions)))(rawData);
            }
            else {
                if (this.dimensions) {
                    /** @type {?} */
                    const dimensions = CommonsUtils.getArrayValue(this.dimensions);
                    dimensions.forEach((dimension, index) => {
                        if (this.resolveDimension) {
                            dimension = this.resolveDimension.method(dimension);
                        }
                        /** @type {?} */
                        const chartData = ChartUtils.buildChartData(rawData, {
                            type: this.type,
                            dimension: dimension,
                            dataItemProps: this.dataItemProps
                        }, index);
                        datasets.push(...chartData.datasets);
                        if (labels.length === 0) {
                            labels = chartData.labels;
                        }
                    });
                    data = { labels, datasets };
                }
                else {
                    data = ChartUtils.buildChartData(rawData, {
                        type: this.type,
                        dataItemProps: this.dataItemProps
                    });
                }
            }
            this.data = data;
        }
        else {
            this.data = {};
        }
    }
}
PngChartComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-chart',
                template: `
    <p-panel [toggleable]="true" [showHeader]="!!caption">
      <p-header>
        <i [ngClass]="captionIcon || 'fa fa-bar-chart'"></i>
        <span [innerHTML]="caption|erSafeHtml"></span>
        <ng-container *ngTemplateOutlet="templates[templateNames.header]"></ng-container>
      </p-header>
      <er-http-status class="mr-2" [tag]="$id"></er-http-status>
      <div class="p-3" *ngIf="switchable">
        <p-selectButton erPosition="center" [options]="chartTypes" [(ngModel)]="type" (onChange)="changeType()"></p-selectButton>
      </div>
      <p-chart
        [type]="type"
        [options]="options"
        [data]="data$|async"
        [plugins]="plugins"
        [width]="width "
        [height]="height"
        [responsive]="responsive"
        (onDataSelect)="onDataSelect.emit($event)"
      >
      </p-chart>
    </p-panel>

  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PngChartComponent.propDecorators = {
    $id: [{ type: Input }],
    type: [{ type: Input }],
    rawData: [{ type: Input }],
    options: [{ type: Input }],
    caption: [{ type: Input }],
    captionIcon: [{ type: Input }],
    query: [{ type: Input }],
    plugins: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    responsive: [{ type: Input }],
    switchable: [{ type: Input }],
    dataSource: [{ type: Input }],
    dataItemProps: [{ type: Input }],
    dimensions: [{ type: Input }],
    resolveDataSource: [{ type: Input }],
    resolveDimension: [{ type: Input }],
    onDataSelect: [{ type: Output }],
    data: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngChartComponent.prototype.$id;
    /** @type {?} */
    PngChartComponent.prototype.type;
    /** @type {?} */
    PngChartComponent.prototype.rawData;
    /** @type {?} */
    PngChartComponent.prototype.options;
    /** @type {?} */
    PngChartComponent.prototype.caption;
    /** @type {?} */
    PngChartComponent.prototype.captionIcon;
    /** @type {?} */
    PngChartComponent.prototype.query;
    /** @type {?} */
    PngChartComponent.prototype.plugins;
    /** @type {?} */
    PngChartComponent.prototype.width;
    /** @type {?} */
    PngChartComponent.prototype.height;
    /** @type {?} */
    PngChartComponent.prototype.responsive;
    /** @type {?} */
    PngChartComponent.prototype.switchable;
    /** @type {?} */
    PngChartComponent.prototype.dataSource;
    /** @type {?} */
    PngChartComponent.prototype.dataItemProps;
    /** @type {?} */
    PngChartComponent.prototype.dimensions;
    /** @type {?} */
    PngChartComponent.prototype.resolveDataSource;
    /** @type {?} */
    PngChartComponent.prototype.resolveDimension;
    /** @type {?} */
    PngChartComponent.prototype.templateNames;
    /** @type {?} */
    PngChartComponent.prototype.onDataSelect;
    /** @type {?} */
    PngChartComponent.prototype.data$;
    /** @type {?} */
    PngChartComponent.prototype.chartTypes;
}
