// @er/formly/esm2015/lib/base/field-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { QueryList, ViewChild, ViewChildren, ViewContainerRef } from '@angular/core';
import { ComponentUtils } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { ValidationMessageDirective } from '@er/validate';
import { FieldType } from '@ngx-formly/core';
import { FormlyPropsBindDirective } from '../directives/props/directive';
/**
 * @abstract
 */
export class ErFormlyFieldType extends FieldType {
    /**
     * @return {?}
     */
    get attrs() {
        return this.field['attrs'] || {};
    }
    /**
     * @return {?}
     */
    get props() {
        return this.field['props'] || {};
    }
    /**
     * @return {?}
     */
    get context() {
        return this.field['context'] || {};
    }
    /**
     * @return {?}
     */
    get value() {
        return this.formControl.value;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set value(val) {
        this.formControl.setValue(val);
    }
    /**
     * @param {?=} key
     * @return {?}
     */
    component(key) {
        return this.form['components'] &&
            this.form['components'][key || this.field.key];
    }
    /**
     * @param {?=} key
     * @return {?}
     */
    data$(key) {
        return this.form['components'] &&
            this.form['components'][key || this.field.key] &&
            this.form['components'][key || this.field.key].data$;
    }
    /**
     * @param {?=} key
     * @return {?}
     */
    control(key) {
        return this.form.get(key);
    }
    /**
     * @param {?} key
     * @return {?}
     */
    fieldProps(key) {
        /** @type {?} */
        const props = this.field['formProps'].fields.filter(prop => prop.key === key);
        return (props && props.length === 1) ? props[0] : undefined;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        /** @type {?} */
        const _widgets = this.widgets.toArray();
        /** @type {?} */
        let component;
        if (_widgets.length === 1) {
            component = ComponentUtils.getHostComponentInst(_widgets[0]);
        }
        else if (_widgets.length > 1) {
            component = _widgets.map(widget => ComponentUtils.getHostComponentInst(widget));
        }
        if (this.validateMessage) {
            this.validateMessage.showError = this.showError || true;
        }
        if (!this.form['components']) {
            this.form['components'] = {};
        }
        this.form['components'][this.field.key] = component;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.field['props'] = this.field['props'] || {};
        if (this.defaultProps) {
            this.field['props'] = CommonsUtils.defaults({}, this.field['props'], this.defaultProps);
        }
        this.field['formlyField'] = this;
    }
}
ErFormlyFieldType.propDecorators = {
    widgets: [{ type: ViewChildren, args: [FormlyPropsBindDirective, { read: ViewContainerRef },] }],
    validateMessage: [{ type: ViewChild, args: [ValidationMessageDirective,] }]
};
if (false) {
    /** @type {?} */
    ErFormlyFieldType.prototype.defaultProps;
    /** @type {?} */
    ErFormlyFieldType.prototype.widgets;
    /** @type {?} */
    ErFormlyFieldType.prototype.validateMessage;
}
