// @er/formly-primeng/esm5/lib/field-types/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PngFormlyAutoCompleteComponent } from './auto-complete/component';
import { PngFormlyButtonComponent } from './button/component';
import { PngFormlyCalendarComponent } from './calendar/component';
import { PngFormlyCheckboxComponent } from './checkbox/component';
import { PngFormlyChipsComponent } from './chips/component';
import { PngFormlyFieldDynaComponent } from './dyna/component';
import { PngFormlyEditorComponent } from './editor/component';
import { PngFormlyFileUploadComponent } from './file-upload/component';
import { PngFormlyInputComponent } from './input/component';
import { PngFormlyKeyFilterComponent } from './key-filter/component';
import { PngFormlyListBoxComponent } from './list-box/component';
import { PngFormlyMaskComponent } from './mask/component';
import { PngFormlyMultiSelectComponent } from './multi-select/component';
import { PngFormlyRadioComponent } from './radio/component';
import { PngFormlyRatingComponent } from './rating/component';
import { PngFormlySelectButtonComponent } from './select-button/component';
import { PngFormlySelectComponent } from './select/component';
import { PngFormlySliderComponent } from './slider/component';
import { PngFormlySpinnerComponent } from './spinner/component';
import { PngFormlySwitchComponent } from './switch/component';
import { PngFormlyTableComponent } from './table/component';
import { PngFormlyTextAreaComponent } from './text-area/component';
import { PngFormlyTreeComponent } from './tree/component';
import { PngFormlyFieldValueComponent } from './value/component';
export { PngFormlyInputComponent } from './input/component';
export { PngFormlyCheckboxComponent } from './checkbox/component';
export { PngFormlyRadioComponent } from './radio/component';
export { PngFormlySelectComponent } from './select/component';
export { PngFormlyTextAreaComponent } from './text-area/component';
export { PngFormlyAutoCompleteComponent } from './auto-complete/component';
export { PngFormlyButtonComponent } from './button/component';
export { PngFormlyCalendarComponent } from './calendar/component';
export { PngFormlyChipsComponent } from './chips/component';
export { PngFormlyEditorComponent } from './editor/component';
export { PngFormlyFileUploadComponent } from './file-upload/component';
export { PngFormlyKeyFilterComponent } from './key-filter/component';
export { PngFormlyListBoxComponent } from './list-box/component';
export { PngFormlyMaskComponent } from './mask/component';
export { PngFormlyMultiSelectComponent } from './multi-select/component';
export { PngFormlyRatingComponent } from './rating/component';
export { PngFormlySelectButtonComponent } from './select-button/component';
export { PngFormlySliderComponent } from './slider/component';
export { PngFormlySpinnerComponent } from './spinner/component';
export { PngFormlySwitchComponent } from './switch/component';
export { PngFormlyTableComponent } from './table/component';
export { PngFormlyTreeComponent } from './tree/component';
export { PngFormlyFieldValueComponent } from './value/component';
export { PngFormlyFieldDynaComponent } from './dyna/component';
/** @type {?} */
export var PNG_FORMLY_TYPE_COMPONENTS = [
    PngFormlyTextAreaComponent,
    PngFormlySelectComponent,
    PngFormlyRadioComponent,
    PngFormlyCheckboxComponent,
    PngFormlyInputComponent,
    PngFormlyAutoCompleteComponent,
    PngFormlyMultiSelectComponent,
    PngFormlyCalendarComponent,
    PngFormlyChipsComponent,
    PngFormlyEditorComponent,
    PngFormlyKeyFilterComponent,
    PngFormlyListBoxComponent,
    PngFormlyMaskComponent,
    PngFormlyRatingComponent,
    PngFormlySliderComponent,
    PngFormlySelectButtonComponent,
    PngFormlyFileUploadComponent,
    PngFormlyButtonComponent,
    PngFormlySwitchComponent,
    PngFormlySpinnerComponent,
    PngFormlyTableComponent,
    PngFormlyTreeComponent,
    PngFormlyFieldValueComponent,
    PngFormlyFieldDynaComponent
];
