// @er/primeng/esm5/lib/components/table/component/body-row/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { TableCheckbox } from 'primeng/table';
import { PngTableComponent } from '../table/component';
var PngTableBodyRowComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableBodyRowComponent, _super);
    function PngTableBodyRowComponent(pngTable, cd) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        _this.cd = cd;
        return _this;
    }
    Object.defineProperty(PngTableBodyRowComponent.prototype, "rowContext", {
        get: /**
         * @return {?}
         */
        function () {
            return {
                columns: this.columns,
                row: this.row,
                rowIndex: this.rowIndex,
                expanded: this.expanded,
                context: this.pngTable
            };
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PngTableBodyRowComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.chk) {
            this.$subscriptions = this.chk.tableService.selectionSource$.subscribe(function () {
                _this.cd.detectChanges();
            });
        }
    };
    /**
     * @param {?} col
     * @return {?}
     */
    PngTableBodyRowComponent.prototype.colContext = /**
     * @param {?} col
     * @return {?}
     */
    function (col) {
        return tslib_1.__assign({}, this.rowContext, col);
    };
    /**
     * @param {?} col
     * @return {?}
     */
    PngTableBodyRowComponent.prototype.getColStyleClass = /**
     * @param {?} col
     * @return {?}
     */
    function (col) {
        return CommonsUtils.getValue(col.styleClass, this.colContext(col));
    };
    /**
     * @param {?} col
     * @return {?}
     */
    PngTableBodyRowComponent.prototype.getColStyle = /**
     * @param {?} col
     * @return {?}
     */
    function (col) {
        return CommonsUtils.getValue(col.style, this.colContext(col));
    };
    PngTableBodyRowComponent.decorators = [
        { type: Component, args: [{
                    selector: '[png-table-body-row]',
                    template: "\n    <ng-container *ngIf=\"customerTpl;else default\">\n      <ng-container *ngTemplateOutlet=\"customerTpl;context:rowContext\"></ng-container>\n    </ng-container>\n    <ng-template #default>\n      <tr *ngIf=\"rowHeaderTpl\">\n        <ng-container *ngTemplateOutlet=\"rowHeaderTpl;context:rowContext\"></ng-container>\n      </tr>\n      <td *ngIf=\"pngTable.$props.$ext.rowExpandable\" class=\"col-expandable\">\n        <a href=\"#\" [pRowToggler]=\"row\">\n          <i [ngClass]=\"expanded ? 'pi pi-chevron-down' : 'pi pi-chevron-right'\"></i>\n        </a>\n      </td>\n      <td *ngIf=\"pngTable.$props.$ext.hasRowCheckbox\" class=\"col-checkbox\">\n        <p-tableCheckbox #chk [value]=\"row\"></p-tableCheckbox>\n      </td>\n      <td *ngIf=\"pngTable.$props.$ext.hasRowRadioButton\" class=\"col-radio\">\n        <p-tableRadioButton [value]=\"row\"></p-tableRadioButton>\n      </td>\n      <ng-container *ngFor=\"let col of columns\">\n        <td png-table-body-cell class=\"col-data col-data-{{col.key}}\"\n            [ngClass]=\"getColStyleClass(col)\"\n            [ngStyle]=\"getColStyle(col)\"\n            [customerTpl]=\"cellTpl\"\n            [column]=\"col\"\n            [row]=\"row\"\n            [rowIndex]=\"rowIndex\"\n            [dataCellTpl]=\"dataCellTpl\"\n        >\n        </td>\n      </ng-container>\n      <tr *ngIf=\"rowFooterTpl\">\n        <ng-container *ngTemplateOutlet=\"rowFooterTpl;context:rowContext\"></ng-container>\n      </tr>\n    </ng-template>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngTableBodyRowComponent.ctorParameters = function () { return [
        { type: PngTableComponent },
        { type: ChangeDetectorRef }
    ]; };
    PngTableBodyRowComponent.propDecorators = {
        columns: [{ type: Input }],
        row: [{ type: Input }],
        rowIndex: [{ type: Input }],
        expanded: [{ type: Input }],
        rowHeaderTpl: [{ type: Input }],
        rowFooterTpl: [{ type: Input }],
        cellTpl: [{ type: Input }],
        dataCellTpl: [{ type: Input }],
        chk: [{ type: ViewChild, args: ['chk',] }]
    };
    return PngTableBodyRowComponent;
}(BaseComponent));
export { PngTableBodyRowComponent };
if (false) {
    /** @type {?} */
    PngTableBodyRowComponent.prototype.columns;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.row;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowIndex;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.expanded;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowHeaderTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowFooterTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.cellTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.dataCellTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.chk;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.pngTable;
    /**
     * @type {?}
     * @private
     */
    PngTableBodyRowComponent.prototype.cd;
}
