// @er/primeng/esm5/lib/components/calendar/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ControlWrapperComponent } from '@er/core';
import { DateUtils } from '@er/utils';
import { DEFAULT_CALENDAR_PROPS } from './type';
var PngCalendarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngCalendarComponent, _super);
    function PngCalendarComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.modelFormatter = _this.formatter;
        _this.today = DateUtils.getDay();
        _this.templateNames = {
            header: 'header',
            footer: 'footer',
            date: 'date'
        };
        _this.$defaultProps = DEFAULT_CALENDAR_PROPS;
        return _this;
    }
    /**
     * @param {?} model
     * @return {?}
     */
    PngCalendarComponent.prototype.formatter = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        if (model) {
            return DateUtils.toDate(model);
        }
    };
    PngCalendarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-calendar',
                    template: "\n    <p-calendar\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      [monthNavigator]=\"$props.monthNavigator\"\n      [yearNavigator]=\"$props.yearNavigator\"\n      [yearRange]=\"$props.yearRange\"\n      (onBlur)=\"updateTouch()\"\n    >\n      <p-header *ngIf=\"template(templateNames.header)\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.header)\"></ng-container>\n      </p-header>\n\n      <p-footer *ngIf=\"template(templateNames.footer)\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.footer)\"></ng-container>\n      </p-footer>\n      \n      <ng-template pTemplate=\"date\" let-date>\n        <ng-container *ngTemplateOutlet=\"template(templateNames.date, dateTpl);context:{$implicit: date}\"></ng-container>\n      </ng-template>\n    </p-calendar>\n\n    <ng-template #dateTpl let-date>\n      <span [ngClass]=\"{'bg-red white': (date.day === today)}\"\n            style=\"border-radius:50%;\">{{date.day}}</span>\n    </ng-template>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngCalendarComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngCalendarComponent;
}(ControlWrapperComponent));
export { PngCalendarComponent };
if (false) {
    /** @type {?} */
    PngCalendarComponent.prototype.modelFormatter;
    /** @type {?} */
    PngCalendarComponent.prototype.today;
    /** @type {?} */
    PngCalendarComponent.prototype.templateNames;
    /** @type {?} */
    PngCalendarComponent.prototype.$defaultProps;
}
