// @er/primeng/esm5/lib/components/calendar/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var CALENDAR_LOCAL_CN = {
    firstDayOfWeek: 0,
    dayNames: ['周日', '周一', '周二', '周三', '周四', '周五', '周六'],
    dayNamesShort: ['周日', '周一', '周二', '周三', '周四', '周五', '周六'],
    dayNamesMin: ['周日', '周一', '周二', '周三', '周四', '周五', '周六'],
    monthNames: ['一月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '十一月', '十二月'],
    monthNamesShort: ['一月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '十一月', '十二月'],
    today: '今天',
    clear: '清除'
};
/**
 * @record
 */
export function PngCalendarProps() { }
if (false) {
    /** @type {?|undefined} */
    PngCalendarProps.prototype.defaultDate;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.inputStyle;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.inputId;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.inputStyleClass;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.placeholder;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.dateFormat;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.inline;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showOtherMonths;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.selectOtherMonths;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showIcon;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.icon;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.locale;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.appendTo;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.readonlyInput;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.shortYearCutoff;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.monthNavigator;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.yearNavigator;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.yearRange;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.hourFormat;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.timeOnly;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showTime;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.stepHour;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.stepMinute;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.stepSecond;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showSeconds;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showOnFocus;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.dataType;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.selectionMode;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.maxDateCount;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.showButtonBar;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.todayButtonStyleClass;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.clearButtonStyleClass;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.autoZIndex;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.baseZIndex;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.panelStyleClass;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.panelStyle;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.keepInvalid;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.hideOnDateTimeSelect;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.numberOfMonths;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.view;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.touchUI;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onSelect;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onBlur;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onFocus;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onClose;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onInput;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onTodayClick;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onClearClick;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onMonthChange;
    /** @type {?|undefined} */
    PngCalendarProps.prototype.onYearChange;
}
/** @type {?} */
export var DEFAULT_CALENDAR_PROPS = {
    dateFormat: 'yy-mm-dd',
    hourFormat: '24',
    locale: CALENDAR_LOCAL_CN,
    monthNavigator: true,
    yearNavigator: true,
    showIcon: true,
    readonlyInput: false,
    yearRange: '2000:2030',
    showButtonBar: true
};
