// @er/validate/esm5/lib/directives/control-on-blur/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Optional } from '@angular/core';
import { NgControl } from '@angular/forms';
var ValidateOnBlurMessageDirective = /** @class */ (function () {
    function ValidateOnBlurMessageDirective(model) {
        this.model = model;
    }
    Object.defineProperty(ValidateOnBlurMessageDirective.prototype, "control", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.model && this.model.control) {
                return this.model.control;
            }
            else {
                return undefined;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ValidateOnBlurMessageDirective.prototype.onFocus = /**
     * @return {?}
     */
    function () {
        if (this.control) {
            this.validators = this.control.validator;
            this.asyncValidators = this.control.asyncValidator;
            this.control.clearAsyncValidators();
            this.control.clearValidators();
        }
    };
    /**
     * @return {?}
     */
    ValidateOnBlurMessageDirective.prototype.onBlur = /**
     * @return {?}
     */
    function () {
        if (this.control) {
            this.control.setAsyncValidators(this.asyncValidators);
            this.control.setValidators(this.validators);
            this.control.updateValueAndValidity();
        }
    };
    ValidateOnBlurMessageDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erValidateOnBlur]'
                },] }
    ];
    /** @nocollapse */
    ValidateOnBlurMessageDirective.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }] }
    ]; };
    ValidateOnBlurMessageDirective.propDecorators = {
        onFocus: [{ type: HostListener, args: ['focus',] }],
        onBlur: [{ type: HostListener, args: ['blur',] }]
    };
    return ValidateOnBlurMessageDirective;
}());
export { ValidateOnBlurMessageDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.validators;
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.asyncValidators;
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.model;
}
