// @er/core/esm5/lib/interceptors/error/interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpErrorResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { NotifyLevel, StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { throwError, TimeoutError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { RouteUtils } from '../../utils/nav';
var HttpErrorHandlerInterceptor = /** @class */ (function () {
    function HttpErrorHandlerInterceptor() {
    }
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    HttpErrorHandlerInterceptor.prototype.intercept = /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    function (req, next) {
        var _this = this;
        return next.handle(req)
            .pipe(catchError(function (err) {
            if (err instanceof HttpErrorResponse) {
                if (err.status === 401) {
                    RouteUtils.nav(ConfigUtils.getConfig().auth.loginRouter);
                }
                else if (err.status === 403) {
                    StatesUtils.create(StateNames.notify, {
                        message: '没有操作权限！',
                        level: NotifyLevel.ERROR
                    });
                }
                else {
                    _this.errorHandler(err);
                }
            }
            else if (err instanceof TimeoutError) {
                StatesUtils.create(StateNames.notify, {
                    message: '网络访问超时！',
                    level: NotifyLevel.ERROR
                });
            }
            return throwError(err);
        }));
    };
    /**
     * @private
     * @param {?} err
     * @return {?}
     */
    HttpErrorHandlerInterceptor.prototype.errorHandler = /**
     * @private
     * @param {?} err
     * @return {?}
     */
    function (err) {
        /** @type {?} */
        var message = '';
        if (err.error instanceof Error) {
            message = err.error.message;
        }
        else {
            if (err.status === 0) {
                message = '网络链接已断开';
            }
            else if (err.status === 404) {
                message = '请求地址不存在';
            }
            else if (err.status === 500) {
                message = '服务器程序错误';
            }
            else {
                message = '数据请求错误';
            }
        }
        StatesUtils.create(StateNames.notify, {
            level: NotifyLevel.ERROR,
            message: message + ",\u9519\u8BEF\u4EE3\u7801:\u3010" + err.status + "\u3011",
            title: '网络访问失败'
        });
    };
    HttpErrorHandlerInterceptor.decorators = [
        { type: Injectable }
    ];
    return HttpErrorHandlerInterceptor;
}());
export { HttpErrorHandlerInterceptor };
