// @er/core/esm5/lib/services/log/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { LogLevel } from '@er/types';
import { CommonsUtils, ConfigUtils, DateUtils, DialogUtils } from '@er/utils';
import { HttpUtils } from '../../utils/http';
import * as i0 from "@angular/core";
var LogService = /** @class */ (function () {
    function LogService() {
    }
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    LogService.prototype.debug = /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    function (from, message) {
        var context = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            context[_i - 2] = arguments[_i];
        }
        if (!ConfigUtils.getConfig().isProduction
            && ConfigUtils.getConfig().log.enableApp
            && ConfigUtils.getConfig().log.level <= LogLevel.DEBUG) {
            if ((typeof from) !== 'string' && from.constructor) {
                from = from.constructor.name;
            }
            if (ConfigUtils.getConfig().log.blackLList.indexOf(from) < 0) {
                this.getLogger().log("[" + DateUtils.getFormattedTime() + "] %c DEBUG ", 'color: white; background: green; font-size: 18px;', this.getLogData(tslib_1.__spread([from, message], context)));
            }
        }
    };
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    LogService.prototype.info = /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    function (from, message) {
        var context = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            context[_i - 2] = arguments[_i];
        }
        if (ConfigUtils.getConfig().log.level <= LogLevel.INFO) {
            if (typeof from !== 'string' && from.constructor) {
                from = from.constructor.name;
            }
            if (ConfigUtils.getConfig().log.blackLList.indexOf(from) < 0) {
                this.getLogger().log("[" + DateUtils.getFormattedTime() + "] %c INFO ", 'color: white; background: blue; font-size: 18px;', this.getLogData(tslib_1.__spread([from, message], context)));
            }
        }
    };
    /**
     * @param {?} messages
     * @return {?}
     */
    LogService.prototype.warn = /**
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        if (ConfigUtils.getConfig().log.level <= LogLevel.WARN) {
            this.getLogger().log("[" + DateUtils.getFormattedTime() + "] %c WARN ", 'color: white; background: blue; font-size: 18px;', this.getLogData(messages));
        }
    };
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    LogService.prototype.error = /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    function (from, message) {
        var context = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            context[_i - 2] = arguments[_i];
        }
        DialogUtils.error('错误', message);
        if (ConfigUtils.getConfig().log.level <= LogLevel.ERROR) {
            this.getLogger().log("[" + DateUtils.getFormattedTime() + "] %c ERROR ", 'color: white; background: red; font-size: 18px;', this.getLogData(tslib_1.__spread([from, message], context)));
            if (ConfigUtils.getConfig().log.httpPost) {
                if (CommonsUtils.isFunction(from)) {
                    from = from.constructor.name;
                }
                this.getHttpLogger().log(DateUtils.getFormattedDateTime(), this.getLogData([from, message]));
            }
        }
    };
    /**
     * @protected
     * @return {?}
     */
    LogService.prototype.getLogger = /**
     * @protected
     * @return {?}
     */
    function () {
        return console;
    };
    /**
     * @protected
     * @return {?}
     */
    LogService.prototype.getHttpLogger = /**
     * @protected
     * @return {?}
     */
    function () {
        return (/** @type {?} */ ({
            log: function (message) {
                var optionalParams = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    optionalParams[_i - 1] = arguments[_i];
                }
                HttpUtils.request(tslib_1.__assign({}, ConfigUtils.getConfig().logHttpOptions, { body: tslib_1.__assign({ message: message }, optionalParams) }));
            }
        }));
    };
    /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    LogService.prototype.getLogData = /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        return tslib_1.__assign({}, messages);
    };
    LogService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */ LogService.ngInjectableDef = i0.defineInjectable({ factory: function LogService_Factory() { return new LogService(); }, token: LogService, providedIn: "root" });
    return LogService;
}());
export { LogService };
