// @er/core/esm5/lib/components/dyna-component/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewContainerRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { ComponentUtils } from '../../utils/component';
import { DynaComponentUtils } from '../../utils/dyna';
var DynaCompComponent = /** @class */ (function () {
    function DynaCompComponent(viewContainerRef, el, renderer) {
        this.viewContainerRef = viewContainerRef;
        this.el = el;
        this.renderer = renderer;
        this.onCreated = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    DynaCompComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.type) {
            this.createComponent();
        }
    };
    /**
     * @return {?}
     */
    DynaCompComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.component) {
            this.component.destroy();
            this.component = null;
        }
    };
    /**
     * @private
     * @return {?}
     */
    DynaCompComponent.prototype.createComponent = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.type) {
            if (CommonsUtils.isString(this.type)) {
                this.component = DynaComponentUtils.createComponentByTemplate(this.viewContainerRef, (/** @type {?} */ (this.type)));
            }
            else {
                this.component = DynaComponentUtils.createComponentByType(this.viewContainerRef, (/** @type {?} */ (this.type)));
            }
            ComponentUtils.bind(this.component.instance, this.props);
            this.onCreated.emit(this.component);
            /** @type {?} */
            var parent_1 = this.el.nativeElement.parentNode;
            this.renderer.removeChild(parent_1, this.el.nativeElement);
        }
    };
    DynaCompComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-dyna-component',
                    template: '',
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynaCompComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    DynaCompComponent.propDecorators = {
        type: [{ type: Input }],
        props: [{ type: Input }],
        onCreated: [{ type: Output }]
    };
    return DynaCompComponent;
}());
export { DynaCompComponent };
if (false) {
    /** @type {?} */
    DynaCompComponent.prototype.type;
    /** @type {?} */
    DynaCompComponent.prototype.props;
    /** @type {?} */
    DynaCompComponent.prototype.onCreated;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.component;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.renderer;
}
