// @er/core/esm2015/lib/services/strategy/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// https://itnext.io/cache-components-with-angular-routereusestrategy-3e4c8b174d5f
export class CachedRouteReuseStrategy {
    constructor() {
        this.storedRouteHandles = new Map();
        this.allowRetrieveCache = {};
    }
    /**
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    shouldReuseRoute(future, curr) {
        if (this.isInOutlet(future['_routerState'].url) || this.isInOutlet(curr['_routerState'].url)) {
            return future.routeConfig === curr.routeConfig;
        }
        return future.routeConfig === curr.routeConfig && (false !== (future.data && future.data.$reuse));
    }
    /**
     * @param {?} route
     * @return {?}
     */
    retrieve(route) {
        return (/** @type {?} */ (this.storedRouteHandles.get(this.getPath(route))));
    }
    /**
     * @param {?} route
     * @return {?}
     */
    shouldAttach(route) {
        /** @type {?} */
        const path = this.getPath(route);
        if (this.allowRetrieveCache[path]) {
            return this.storedRouteHandles.has(this.getPath(route));
        }
        return false;
    }
    /**
     * @param {?} route
     * @return {?}
     */
    shouldDetach(route) {
        /** @type {?} */
        const path = this.getPath(route);
        if (this.allowRetrieveCache.hasOwnProperty(path)) {
            return true;
        }
        return false;
    }
    /**
     * @param {?} route
     * @param {?} detachedTree
     * @return {?}
     */
    store(route, detachedTree) {
        this.storedRouteHandles.set(this.getPath(route), detachedTree);
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    isInOutlet(url) {
        return /.*\(.*:.*\).*/.test(url);
    }
    /**
     * @private
     * @param {?} route
     * @return {?}
     */
    getPath(route) {
        if (route.routeConfig !== null && route.routeConfig.path !== null) {
            return route.routeConfig.path;
        }
        return '';
    }
}
if (false) {
    /** @type {?} */
    CachedRouteReuseStrategy.prototype.storedRouteHandles;
    /** @type {?} */
    CachedRouteReuseStrategy.prototype.allowRetrieveCache;
}
