// @er/primeng/esm5/lib/components/table/component/query/fields-builder/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { BaseComponent } from '@er/core';
import { EsQueryMethods } from '@er/types';
import { PngTableComponent } from '../../table/component';
var PngTableQueryBuilderComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableQueryBuilderComponent, _super);
    function PngTableQueryBuilderComponent(pngTable) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        _this.qbConfig = {};
        _this.switches = [
            { label: '并且', value: 'and', icon: 'fa fa-fw fa-cc-paypal' },
            { label: '或者', value: 'or', icon: 'fa fa-fw fa-cc-visa' },
            { label: '非', value: 'not', icon: 'fa fa-fw fa-cc-mastercard' }
        ];
        return _this;
    }
    /**
     * @return {?}
     */
    PngTableQueryBuilderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.qbConfig.fields = this.pngTable.$props.columns
            .filter(function (column) { return column.dataField; })
            .map(function (column) { return column.dataField; });
        this.qbConfig.operators = function (field) { return Object.keys(EsQueryMethods).map(function (key) {
            return {
                label: EsQueryMethods[key].label,
                value: EsQueryMethods[key].key
            };
        }); };
    };
    PngTableQueryBuilderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-table-query-builder',
                    template: "\n    <er-query-builder [config]=\"qbConfig\">\n      <!--<ng-template erTypedTemplate=\"switchGroup\" let-qb=\"qb\">-->\n      <!--<p-selectButton [options]=\"switches\" [ngModel]=\"qb.value.condition\"></p-selectButton>-->\n      <!--</ng-template>-->\n      <!--<ng-template erTypedTemplate=\"buttonGroup\" let-qb=\"qb\">-->\n      <!--<p-button label=\"\u6761\u4EF6\" icon=\"fa fa-plus\" (onClick)=\"qb.addRule()\"></p-button>-->\n      <!--<p-button label=\"\u7EC4\" icon=\"fa fa-group\" (onClick)=\"qb.addRuleSet()\"></p-button>-->\n      <!--<ng-container *ngIf=\"!!parentRuleSet && qb.config.allowRuleSet\">-->\n      <!--<p-button icon=\"fa fa-minus\" (onClick)=\"qb.removeRuleSet()\"></p-button>-->\n      <!--</ng-container>-->\n      <!--</ng-template>-->\n      <ng-template erTypedTemplate=\"input\" let-rule let-qb=\"qb\">\n        <png-query-input [options]=\"pngTable.$props.$ext\"></png-query-input>\n      </ng-template>\n    </er-query-builder>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngTableQueryBuilderComponent; }),
                            multi: true
                        }],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngTableQueryBuilderComponent.ctorParameters = function () { return [
        { type: PngTableComponent }
    ]; };
    return PngTableQueryBuilderComponent;
}(BaseComponent));
export { PngTableQueryBuilderComponent };
if (false) {
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.qbConfig;
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.switches;
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.pngTable;
}
