// @er/core/esm5/lib/components/query-builder/defaults.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var DEFAULT_QUERY_BUILDER_STYLE_CLASSES = {
    removeIcon: 'q-icon q-remove-icon',
    addIcon: 'q-icon q-add-icon',
    button: 'q-button',
    buttonGroup: 'q-button-group',
    removeButton: 'q-remove-button',
    switchGroup: 'q-switch-group',
    switchLabel: 'q-switch-label',
    switchRadio: 'q-switch-radio',
    rightAlign: 'q-right-align',
    transition: 'q-transition',
    tree: 'q-tree',
    row: 'q-row',
    connector: 'q-connector',
    rule: 'q-rule',
    ruleSet: 'q-ruleset',
    invalidRuleSet: 'q-invalid-ruleset',
    emptyWarning: 'q-empty-warning',
    fieldControl: 'q-field-control',
    fieldControlSize: 'q-control-size',
    entityControl: 'q-entity-control',
    entityControlSize: 'q-control-size',
    operatorControl: 'q-operator-control',
    operatorControlSize: 'q-control-size',
    inputControl: 'q-input-control',
    inputControlSize: 'q-control-size'
};
/** @type {?} */
export var DEFAULT_QUERY_BUILDER_BOOTSTRAP_STYLES_CLASS = {
    removeIcon: 'fa fa-minus',
    addIcon: 'fa fa-plus',
    button: 'btn',
    buttonGroup: 'btn-group',
    rightAlign: 'order-12 ml-auto',
    switchRow: 'd-flex px-2',
    switchGroup: 'd-flex align-items-center',
    switchRadio: 'custom-control-input',
    switchLabel: 'custom-control-label',
    switchControl: 'custom-control custom-radio custom-control-inline',
    row: 'row p-2 m-1',
    connector: 'q-connector',
    rule: 'border',
    ruleSet: 'border',
    invalidRuleSet: 'alert alert-danger',
    emptyWarning: 'text-danger mx-auto',
    operatorControl: 'form-control',
    operatorControlSize: 'col-auto pr-0',
    fieldControl: 'form-control',
    fieldControlSize: 'col-auto pr-0',
    entityControl: 'form-control',
    entityControlSize: 'col-auto pr-0',
    inputControl: 'form-control',
    inputControlSize: 'col-auto'
};
// export const DEFAULT_QUERY_BUILDER_TYPE_OPERATORS = {
//   string: ['=', '!=', 'contains', 'like'],
//   number: ['=', '!=', '>', '>=', '<', '<='],
//   time: ['=', '!=', '>', '>=', '<', '<='],
//   date: ['=', '!=', '>', '>=', '<', '<='],
//   category: ['=', '!=', 'in', 'not in'],
//   boolean: ['=']
// };
/** @type {?} */
export var EMPTY_QUERY_RULE_SET = { condition: 'AND', rules: [] };
/** @type {?} */
export var DEFAULT_QUERY_BUILDER_PROPS = {
    fields: [],
    styleClasses: DEFAULT_QUERY_BUILDER_BOOTSTRAP_STYLES_CLASS,
    allowRuleSet: true
};
