// @er/types/esm2015/lib/menu/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
const MenuOrientation = {
    STATIC: 0,
    OVERLAY: 1,
    SLIM: 2,
    HORIZONTAL: 3,
};
export { MenuOrientation };
MenuOrientation[MenuOrientation.STATIC] = 'STATIC';
MenuOrientation[MenuOrientation.OVERLAY] = 'OVERLAY';
MenuOrientation[MenuOrientation.SLIM] = 'SLIM';
MenuOrientation[MenuOrientation.HORIZONTAL] = 'HORIZONTAL';
/**
 * @record
 */
export function Menu() { }
if (false) {
    /** @type {?|undefined} */
    Menu.prototype.orientation;
    /** @type {?|undefined} */
    Menu.prototype.dark;
    /** @type {?|undefined} */
    Menu.prototype.menuItems;
}
/**
 * @record
 */
export function MenuItem() { }
if (false) {
    /** @type {?|undefined} */
    MenuItem.prototype.id;
    /** @type {?|undefined} */
    MenuItem.prototype.name;
    /** @type {?|undefined} */
    MenuItem.prototype.label;
    /** @type {?|undefined} */
    MenuItem.prototype.title;
    /** @type {?|undefined} */
    MenuItem.prototype.tag;
    /** @type {?|undefined} */
    MenuItem.prototype.icon;
    /** @type {?|undefined} */
    MenuItem.prototype.command;
    /** @type {?|undefined} */
    MenuItem.prototype.url;
    /** @type {?|undefined} */
    MenuItem.prototype.routerLink;
    /** @type {?|undefined} */
    MenuItem.prototype.routerLinkActiveOptions;
    /** @type {?|undefined} */
    MenuItem.prototype.queryParams;
    /** @type {?|undefined} */
    MenuItem.prototype.fragment;
    /** @type {?|undefined} */
    MenuItem.prototype.items;
    /** @type {?|undefined} */
    MenuItem.prototype.expanded;
    /** @type {?|undefined} */
    MenuItem.prototype.disabled;
    /** @type {?|undefined} */
    MenuItem.prototype.visible;
    /** @type {?|undefined} */
    MenuItem.prototype.target;
    /** @type {?|undefined} */
    MenuItem.prototype.separator;
    /** @type {?|undefined} */
    MenuItem.prototype.style;
    /** @type {?|undefined} */
    MenuItem.prototype.styleClass;
    /** @type {?|undefined} */
    MenuItem.prototype.badge;
    /** @type {?|undefined} */
    MenuItem.prototype.badgeStyleClass;
    /** @type {?|undefined} */
    MenuItem.prototype.badgeTheme;
    /** @type {?|undefined} */
    MenuItem.prototype.automationId;
    /** @type {?|undefined} */
    MenuItem.prototype.roles;
    /** @type {?|undefined} */
    MenuItem.prototype.permissions;
}
