// @er/bootstrap/esm2015/lib/components/layout/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { BaseComponent } from '@er/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
export class BsNavPageComponent extends BaseComponent {
    /**
     * @return {?}
     */
    ngOnInit() {
        this.title = StatesUtils.getValue(StateNames.appTitle);
    }
}
BsNavPageComponent.decorators = [
    { type: Component, args: [{
                selector: 'bs-nav-page',
                template: "<div class=\"container h-100\">\n  <div class=\"row\">\n    <div class=\"col-12 col-md-3 col-xl-3 py-md-4 bd-sidebar\">\n      <bs-side-nav></bs-side-nav>\n    </div>\n    <main class=\"col-12 col-md-9 col-xl-8 py-md-4 pl-md-4 bd-content\">\n      <div class=\"d-flex\">\n        <div style=\"flex:1;\">\n          <h2 class=\"bd-title\">{{title}}</h2>\n        </div>\n        <div style=\"align-self:center;\">\n        </div>\n      </div>\n\n      <div class=\"card\">\n        <div class=\"card-body\">\n          <router-outlet></router-outlet>\n        </div>\n      </div>\n    </main>\n  </div>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
if (false) {
    /** @type {?} */
    BsNavPageComponent.prototype.title;
}
