// @er/primeng/esm2015/lib/components/table/default.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApiUtils } from '@er/core';
import { DataAttr } from '@er/types';
import { StatesUtils } from '@er/utils';
/** @type {?} */
export const defaultPngColumnProps = {
    cellStyleClass: 'text-right',
    sortable: true,
    filterable: true,
    aggable: true,
    width: '150px'
};
/** @type {?} */
export const defaultPngTableProps = {
    value: [],
    tableStyleClass: 'rounded shadow',
    dataKey: 'id',
    alwaysShowPaginator: true,
    autoLayout: true,
    columnResizeMode: 'expand',
    compareSelectionBy: 'equals',
    defaultSortOrder: 1,
    resizableColumns: true,
    reorderableColumns: true,
    responsive: true,
    filterDelay: 300,
    filters: {},
    rowHover: true,
    loadingIcon: 'fa fa-spin fa-2x fa-circle-o-notch',
    lazy: true,
    lazyLoadOnInit: false,
    rows: 10,
    pageLinks: 5,
    rowsPerPageOptions: [10, 30, 50],
    paginator: true,
    paginatorPosition: 'both',
    rowExpandMode: 'multiple',
    sortMode: 'single',
    sortOrder: 1,
    $ext: {
        emptyMessage: '没有显示内容',
        hasColumnSelector: true,
        columnSelectorStyleClass: 'd-inline-block align-middle pull-right',
        rowHeight: '30px',
        rowActionsPlaceAhead: false,
        frozenActionColumn: false,
        hasSearchInput: true,
        hasSearchBox: false,
        hasRowIndex: false,
        hasRowCheckbox: false,
        hasHeaderSummary: false,
        hasPaginatorLeft: true,
        hasPaginatorRight: false,
        rowExpandable: false,
        hasAddAction: false,
        hasDeleteAction: false,
        hasAuditAction: false,
        hasDownloadAction: false,
        routerOutlet: 'aux',
        addRouter: 'add',
        viewRouter: 'view',
        editRouter: 'edit',
        auditRouter: 'audit',
        downloadRouter: 'download'
    }
};
/** @type {?} */
export const defaultRowIndexColumn = (/** @type {?} */ ({
    label: '行号',
    width: '50px',
    cellClass: 'text-right',
    sortable: false,
    filterable: false,
    aggable: false,
    order: -100,
    isDataColumn: false,
    cellContent: (data) => data.rowIndex + 1
}));
/** @type {?} */
export const defaultOrderedRowIndexColumn = {
    label: '行号',
    width: '50px',
    cellClass: 'text-right',
    sortable: false,
    filterable: false,
    aggable: false,
    order: -100,
    isDataColumn: false,
    cellContent: (data) => data.rowIndex + 1,
    cellStyleClass: (ctx) => {
        if (ctx.rowIndex === 0) {
            return 'badge badge-pill white bg-red No-1';
        }
        else if (ctx.rowIndex === 1) {
            return 'badge badge-pill black bg-yellow No-2';
        }
        else if (ctx.rowIndex === 2) {
            return 'badge badge-pill white bg-blue No-3';
        }
        else {
            return 'badge badge-pill badge-dark';
        }
    }
};
/** @type {?} */
export const defaultActionColumn = (/** @type {?} */ ({
    label: '操作',
    width: '150px',
    cellStyleClass: 'text-left',
    sortable: false,
    filterable: false,
    aggable: false,
    dataAttr: DataAttr.BUTTONS,
    isActionColumn: true,
    cellContent: (/** @type {?} */ ([]))
}));
/** @type {?} */
export const defaultAddButton = (/** @type {?} */ ({
    $id: 'add',
    title: '新建一条信息',
    label: '新建',
    icon: 'fa fa-plus',
    styleClass: 'ui-button-rounded ui-button-success',
    onClick: (event) => {
        event.pngTable.navigate(event.pngTable.$props.$ext.addRouter, [], { tableId: event.pngTable.$id });
    }
}));
/** @type {?} */
export const defaultViewButton = (/** @type {?} */ ({
    $id: 'view',
    title: '查看',
    icon: 'fa fa-edit',
    styleClass: 'ui-button-rounded ui-button-info',
    onClick: (event) => {
        event.pngTable.navigate(event.pngTable.$props.$ext.viewRouter, [event.row['id']]);
    }
}));
/** @type {?} */
export const defaultEditButton = (/** @type {?} */ ({
    $id: 'edit',
    title: '修改',
    icon: 'fa fa-edit',
    styleClass: 'ui-button-rounded ui-button-warning',
    onClick: (event) => {
        event.pngTable.navigate(event.pngTable.$props.$ext.editRouter, [event.row['id']], { tableId: event.pngTable.$id });
    }
}));
/** @type {?} */
export const defaultDeleteButton = (/** @type {?} */ ({
    $id: 'delete',
    title: '删除',
    icon: 'fa fa-trash',
    styleClass: 'ui-button-rounded ui-button-danger',
    onClick: (event) => {
        ApiUtils.deleteById(event.pngTable.$props.$ext.apiDataProps.apiEntry, event.row['id'])
            .subscribe(() => {
            event.pngTable.load();
            StatesUtils.update(event.pngTable.$props.$id, { deleted: event.row['id'] });
        });
    }
}));
/** @type {?} */
export const defaultAuditButton = (/** @type {?} */ ({
    $id: 'audit',
    title: '审核',
    icon: 'fa fa-check',
    styleClass: 'ui-button-rounded ui-button-success',
    onClick: (event) => {
        event.pngTable.navigate(event.pngTable.$props.$ext.auditRouter, [event.row['id']], { action: 'audit', tableId: event.pngTable.$id });
    }
}));
/** @type {?} */
export const defaultDownloadButton = (/** @type {?} */ ({
    $id: 'download',
    title: '下载',
    icon: 'fa fa-download',
    overlay: true,
    styleClass: 'ui-button-rounded ui-button-success',
    onClick: (event) => {
        event.pngTable.navigate(event.pngTable.$props.$ext.downloadRouter, [event.row['id']]);
    }
}));
/** @type {?} */
export const defaultOverlayColumnProps = {
    sortable: false,
    filterable: false,
    aggable: false
};
/** @type {?} */
export const defaultOverlayTableProps = {
    alwaysShowPaginator: false,
    scrollable: true,
    scrollWidth: '800px',
    paginatorPosition: 'bottom'
};
