// @er/types/esm5/lib/state/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var StateNames = {
    rootHttp: 'rootHttp',
    rootMenu: 'rootMenu',
    appTitle: 'appTitle',
    httpTimer: 'httpTimer',
    currentRoute: 'currentRoute',
    resources: 'resources',
    module: 'module',
    auth: 'auth',
    check: 'check',
    user: 'user',
    http: 'http',
    socket: 'socket',
    nav: 'nav',
    route: 'route',
    menu: 'menu',
    tree: 'tree',
    table: 'table',
    notify: 'notify',
    modal: 'modal',
    alert: 'alert',
    popup: 'popup',
    confirm: 'confirm',
    form: 'form',
    ocr: 'ocr',
    dict: 'dict',
    serverTime: 'serverTime',
};
export { StateNames };
/**
 * @record
 */
export function StoreItem() { }
if (false) {
    /** @type {?|undefined} */
    StoreItem.prototype.state;
    /** @type {?|undefined} */
    StoreItem.prototype.createdAt;
    /** @type {?|undefined} */
    StoreItem.prototype.preserved;
}
/**
 * @record
 */
export function Store() { }
