// @er/core/esm5/lib/services/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AuthGuard, DeactivateGuard } from './auth/guards';
import { GlobalErrorHandler } from './error/service';
import { LogService } from './log/service';
import { ConfigurablePreloadingStrategy } from './pre-loading/service';
import { PathPropsDataResolver } from './resolver/page-model/service';
import { ResourceResolver } from './resource/service';
import { RouterStateService } from './router/service';
import { TaskService } from './task/service';
export { GlobalErrorHandler } from './error/service';
export { AuthGuard, DeactivateGuard } from './auth/guards';
export { ConfigurablePreloadingStrategy } from './pre-loading/service';
export { TaskService, BACKGROUND_TASK_TOKEN } from './task/service';
export { RouterStateService } from './router/service';
export { ResourceResolver } from './resource/service';
export { LogService } from './log/service';
export { PathPropsDataResolver } from './resolver/page-model/service';
/** @type {?} */
export var CoreServices = [
    LogService,
    ResourceResolver,
    TaskService,
    RouterStateService,
    AuthGuard,
    DeactivateGuard,
    GlobalErrorHandler,
    ConfigurablePreloadingStrategy,
    PathPropsDataResolver
];
