// @er/core/esm2015/lib/bases/abstract/app-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BrowserOsUtils, CommonsUtils, ConfigUtils } from '@er/utils';
import { HTTP_STATE$, ROUTE_STATE$ } from '../../states/all';
import { AuthUtils } from '../../utils/auth';
import { PermissionUtils } from '../../utils/permission';
/**
 * @abstract
 */
export class AppConfigComponent {
    /**
     * @param {?} rootViewContainerRef
     */
    constructor(rootViewContainerRef) {
        this.rootViewContainerRef = rootViewContainerRef;
    }
    /**
     * @return {?}
     */
    get loadingClass() {
        return ConfigUtils.getConfig().onLoading;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const browserSupport = BrowserOsUtils.checkBrowser();
        this.resolveBrowserSupport(browserSupport);
        ConfigUtils.configApp(this.appConfig());
        setTimeout(() => {
            this.httpState$ = HTTP_STATE$;
            this.routeState$ = ROUTE_STATE$;
        }, 0);
        /** @type {?} */
        const user = AuthUtils.getCurrentUser();
        if (!CommonsUtils.isEmpty(user)) {
            PermissionUtils.addPermissions(user.permissions);
            PermissionUtils.addRoles(user.roles);
            if (user.roles && ConfigUtils.getConfig().auth.defaultRules) {
                PermissionUtils.addRoles(ConfigUtils.getConfig().auth.defaultRules);
            }
        }
    }
    /**
     * @param {?} browserSupport
     * @return {?}
     */
    resolveBrowserSupport(browserSupport) {
        if (!browserSupport.version) {
            alert('当前浏览器版本太低，请升级浏览器!');
        }
        else if (!browserSupport.features) {
            alert('当前浏览器功能不全，请更换浏览器!');
        }
    }
}
if (false) {
    /** @type {?} */
    AppConfigComponent.prototype.httpState$;
    /** @type {?} */
    AppConfigComponent.prototype.routeState$;
    /** @type {?} */
    AppConfigComponent.prototype.inValidBrowser;
    /** @type {?} */
    AppConfigComponent.prototype.rootViewContainerRef;
    /**
     * @abstract
     * @return {?}
     */
    AppConfigComponent.prototype.appConfig = function () { };
}
