// @er/core/esm2015/lib/bases/abstract/auth.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateNames } from '@er/types';
import { ConfigUtils } from '@er/utils';
import { AuthUtils } from '../../utils/auth';
import { RouteUtils } from '../../utils/nav';
import { BaseComponent } from '../component';
export class AuthBaseComponent extends BaseComponent {
    constructor() {
        super();
        this.userToken = ConfigUtils.getConfig().auth.userTokenKey;
        this.password = ConfigUtils.getConfig().auth.passwordKey;
        this.rememberMe = ConfigUtils.getConfig().auth.rememberMeKey;
        this.appName = ConfigUtils.getConfig().appName;
        this.stateKey = StateNames.auth;
    }
    /**
     * @return {?}
     */
    login() {
        this.error = undefined;
        AuthUtils.login(this.getSubmitData(this.form.value))
            .subscribe(response => {
            if (response.success === false) {
                this.error = response.message;
            }
        });
    }
    /**
     * @return {?}
     */
    register() {
        this.error = undefined;
        /** @type {?} */
        const data = this.form.value;
        delete data['rePassword'];
        AuthUtils.register(this.getSubmitData(this.form.value))
            .subscribe(response => {
            if (response.success === false) {
                this.error = response.message;
            }
            else {
                RouteUtils.nav(ConfigUtils.getConfig().auth.loginSuccessRouter);
            }
        });
    }
    /**
     * @param {?} data
     * @return {?}
     */
    getSubmitData(data) {
        return data;
    }
}
if (false) {
    /** @type {?} */
    AuthBaseComponent.prototype.form;
    /** @type {?} */
    AuthBaseComponent.prototype.error;
    /** @type {?} */
    AuthBaseComponent.prototype.appName;
    /** @type {?} */
    AuthBaseComponent.prototype.stateKey;
    /** @type {?} */
    AuthBaseComponent.prototype.userToken;
    /** @type {?} */
    AuthBaseComponent.prototype.password;
    /** @type {?} */
    AuthBaseComponent.prototype.rememberMe;
}
