// @er/core/esm2015/lib/bases/abstract/change-pwd.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormBuilder, Validators } from '@angular/forms';
import { ConfigUtils, DateUtils } from '@er/utils';
import { NgValidators } from '@er/validate';
import { AuthUtils } from '../../utils/auth/index';
import { DiUtils } from '../../utils/di/index';
import { RouteUtils } from '../../utils/nav/index';
import { AuthBaseComponent } from './auth.component';
export class AbstractChangePwdComponent extends AuthBaseComponent {
    /**
     * @return {?}
     */
    ngOnInit() {
        this.form = DiUtils.get(FormBuilder).group({
            [this.oldPasswordKey]: ['', Validators.required],
            [this.newPasswordKey]: ['', [NgValidators.required, NgValidators.minLength(6)]],
            reNewPassword: ['', [NgValidators.required, NgValidators.minLength(6), NgValidators.equalTo(this.newPasswordKey)]]
        });
    }
    /**
     * @return {?}
     */
    changePwd() {
        this.error = undefined;
        /** @type {?} */
        const data = this.form.value;
        delete data['reNewPassword'];
        data['userPwdSetDate'] = DateUtils.getFormattedDateTime();
        AuthUtils.changePwd(this.getSubmitData(this.form.value))
            .subscribe(response => {
            if (response.success === true) {
                RouteUtils.nav(ConfigUtils.getConfig().auth.changePwdSuccessRouter);
            }
            else {
                this.error = response.message || '密码修改失败';
            }
        });
    }
}
if (false) {
    /** @type {?} */
    AbstractChangePwdComponent.prototype.oldPasswordKey;
    /** @type {?} */
    AbstractChangePwdComponent.prototype.newPasswordKey;
}
