// @er/core/esm2015/lib/bases/abstract/register.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormBuilder, FormControl } from '@angular/forms';
import { ApiServerType } from '@er/types';
import { NgValidators } from '@er/validate';
import { DiUtils } from '../../utils/di/index';
import { NgAsyncValidator } from '../../validators/async';
import { AuthBaseComponent } from './auth.component';
export class AbstractRegisterComponent extends AuthBaseComponent {
    constructor() {
        super(...arguments);
        this.checkExists = false;
        this.inFocus = false;
        this.errorMessage = {
            duplicated: '该用户名已经被注册'
        };
    }
    /**
     * @return {?}
     */
    get isUserNameValidate() {
        return !this.inFocus && this.form.get(this.userToken).valid;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.form = DiUtils.get(FormBuilder).group({
            [this.userToken]: new FormControl('', {
                validators: NgValidators.required,
                asyncValidators: NgAsyncValidator(this.userExistsApiOptions(), 'duplicated'),
                updateOn: 'blur'
            }),
            [this.password]: ['', [NgValidators.required, NgValidators.minLength(6)]],
            rePassword: ['', [NgValidators.required, NgValidators.minLength(6), NgValidators.equalTo(this.password)]]
        });
    }
    /**
     * @return {?}
     */
    userExistsApiOptions() {
        return {
            requestOptions: {
                serverType: ApiServerType.ER
            },
            apiEntry: 'user/exists',
            payload: (val) => {
                return { userName: val };
            }
        };
    }
}
if (false) {
    /** @type {?} */
    AbstractRegisterComponent.prototype.checkExists;
    /** @type {?} */
    AbstractRegisterComponent.prototype.inFocus;
    /** @type {?} */
    AbstractRegisterComponent.prototype.errorMessage;
}
