// @er/formly-primeng/esm2015/lib/components/debug/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
export class PngFormDebugComponent extends BaseComponent {
}
PngFormDebugComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-form-debug',
                template: `
    <p-tabView>
      <p-tabPanel header="Form Data" leftIcon="fa fa-eye text-dark">
        <pre [erJsonViewer]="model"></pre>
      </p-tabPanel>
      <p-tabPanel header="Form Json" leftIcon="fa fa-gear text-info">
        <pre [erJsonViewer]="debugFormProps"></pre>
      </p-tabPanel>
      <p-tabPanel header="Runtime Form Json" leftIcon="fa fa-cogs text-danger">
        <pre [erJsonViewer]="runtimeFormProps"></pre>
      </p-tabPanel>
    </p-tabView>
  `,
                styles: ["/deep/ .ui-tabview-selected{border-bottom:5px solid red!important}/deep/ .ui-tabview-selected .ui-tabview-title{color:#000!important}"]
            }] }
];
PngFormDebugComponent.propDecorators = {
    model: [{ type: Input }],
    debugFormProps: [{ type: Input }],
    runtimeFormProps: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngFormDebugComponent.prototype.model;
    /** @type {?} */
    PngFormDebugComponent.prototype.debugFormProps;
    /** @type {?} */
    PngFormDebugComponent.prototype.runtimeFormProps;
}
