// @er/types/esm5/lib/api/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ApiParamsKeys() { }
if (false) {
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.id;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.ids;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.data;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.q;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.query;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.filter;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.globalFiler;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.page;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.size;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.sort;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.order;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.fields;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.from;
    /** @type {?|undefined} */
    ApiParamsKeys.prototype.to;
}
/**
 * @record
 */
export function ApiPayload() { }
if (false) {
    /** @type {?|undefined} */
    ApiPayload.prototype.id;
    /** @type {?|undefined} */
    ApiPayload.prototype.ids;
    /** @type {?|undefined} */
    ApiPayload.prototype.data;
    /** @type {?|undefined} */
    ApiPayload.prototype.query;
    /** @type {?|undefined} */
    ApiPayload.prototype.filter;
    /** @type {?|undefined} */
    ApiPayload.prototype.aggs;
    /** @type {?|undefined} */
    ApiPayload.prototype.page;
    /** @type {?|undefined} */
    ApiPayload.prototype.size;
    /** @type {?|undefined} */
    ApiPayload.prototype.from;
    /** @type {?|undefined} */
    ApiPayload.prototype.to;
    /** @type {?|undefined} */
    ApiPayload.prototype.sort;
    /** @type {?|undefined} */
    ApiPayload.prototype.fields;
    /** @type {?|undefined} */
    ApiPayload.prototype.script;
    /** @type {?|undefined} */
    ApiPayload.prototype.params;
    /* Skipping unhandled member: [key: string]: any;*/
}
/** @enum {string} */
var ApiServerType = {
    App: 'app',
    ES: 'es',
    ER: 'er',
    MOCK: 'mock',
    OTHER: 'other',
};
export { ApiServerType };
/**
 * @record
 */
export function ApiRequestOptions() { }
if (false) {
    /** @type {?|undefined} */
    ApiRequestOptions.prototype.paramsKeys;
    /** @type {?|undefined} */
    ApiRequestOptions.prototype.uriResolver;
    /** @type {?|undefined} */
    ApiRequestOptions.prototype.payloadResolver;
    /** @type {?|undefined} */
    ApiRequestOptions.prototype.actions;
}
