// @er/primeng/esm2015/lib/components/button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { WrapperComponent } from '@er/core';
import { DEFAULT_BUTTON_PROPS } from '@er/presets';
import { CommonsUtils } from '@er/utils';
export class PngButtonComponent extends WrapperComponent {
    /**
     * @return {?}
     */
    get isHidden() {
        if (this.$props && this.$props.hidden) {
            return CommonsUtils.getValue(this.$props.hidden, this.outputContext) || false;
        }
        return false;
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        /** @type {?} */
        const btnProps = CommonsUtils.copy(this.$props || {});
        if (btnProps.menu) {
            btnProps.menu.forEach(item => {
                if (item.command) {
                    /** @type {?} */
                    const command = item.command;
                    item.command = (event) => this.onMenuCommand(command, event);
                }
            });
        }
        this.$props = Object.assign({}, DEFAULT_BUTTON_PROPS, btnProps);
    }
    /**
     * @param {?} command
     * @param {?} event
     * @return {?}
     */
    onMenuCommand(command, event) {
        command.call(this, Object.assign({}, event || {}, this.outputContext || {}));
    }
    /**
     * @param {?} state
     * @return {?}
     */
    setState(state) {
        this.icon = state ? 'fa fa-cog fa-spin' : this.$props['icon'];
    }
}
PngButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-button',
                template: "<ng-container *ngIf=\"!isHidden\">\n  <p-splitButton *ngIf=\"$props?.menu\"\n                 erPropsBind\n                 [props]=\"$props\"\n                 [context]=\"outputContext\"\n  >\n  </p-splitButton>\n  <p-button *ngIf=\"!$props?.menu\" [title]=\"$props?.title || ''\"\n            erPropsBind\n            [props]=\"$props\"\n            [context]=\"outputContext\"\n  >\n  </p-button>\n</ng-container>\n"
            }] }
];
if (false) {
    /** @type {?} */
    PngButtonComponent.prototype.icon;
}
