// @er/primeng/esm2015/lib/work-space/menu/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { PngAdminWorkSpaceComponent } from '../component';
export class WorkSpaceMainMenuComponent {
    /**
     * @param {?} ws
     */
    constructor(ws) {
        this.ws = ws;
        this.theme = 'blue';
        this.layout = 'blue';
        this.version = 'v4';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.menus$ = StatesUtils.observe(StateNames.rootMenu)
            .pipe(map(menu => (menu && menu.menuItems) || []), distinctUntilChanged());
    }
    /**
     * @param {?} theme
     * @return {?}
     */
    changeTheme(theme) {
        /** @type {?} */
        const themeLink = (/** @type {?} */ (document.getElementById('theme-css')));
        if (this.version === 'v3') {
            themeLink.href = 'assets/theme/theme-' + theme + '.css';
        }
        else {
            themeLink.href = 'assets/theme/theme-' + theme + '-v4' + '.css';
        }
        this.theme = theme;
    }
    /**
     * @param {?} layout
     * @param {?=} special
     * @return {?}
     */
    changeLayout(layout, special) {
        /** @type {?} */
        const layoutLink = (/** @type {?} */ (document.getElementById('layout-css')));
        if (this.version === 'v3') {
            layoutLink.href = 'assets/layout/css/layout-' + layout + '.css';
        }
        else {
            layoutLink.href = 'assets/layout/css/layout-' + layout + '-v4' + '.css';
        }
        this.layout = layout;
        if (special) {
            this.ws.darkMenu = true;
        }
    }
    /**
     * @param {?} version
     * @return {?}
     */
    changeVersion(version) {
        /** @type {?} */
        const themeLink = (/** @type {?} */ (document.getElementById('theme-css')));
        /** @type {?} */
        const layoutLink = (/** @type {?} */ (document.getElementById('layout-css')));
        if (version === 'v3') {
            this.version = 'v3';
            themeLink.href = 'assets/theme/theme-' + this.theme + '.css';
            layoutLink.href = 'assets/layout/css/layout-' + this.layout + '.css';
        }
        else {
            themeLink.href = 'assets/theme/theme-' + this.theme + '-v4' + '.css';
            layoutLink.href = 'assets/layout/css/layout-' + this.layout + '-v4' + '.css';
            this.version = '-v4';
        }
    }
}
WorkSpaceMainMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-ws-main-menu',
                template: `
    <ul png-ws-sub-menu
        erAnimate="fadeInDown"
        class="layout-menu layout-main-menu clearfix"
        [erHttpState]="httpMenuTag"
        [item]="menus$|async" [root]="true"
        [reset]="reset"
        [visible]="true"
        [parentActive]="true"></ul>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
WorkSpaceMainMenuComponent.ctorParameters = () => [
    { type: PngAdminWorkSpaceComponent }
];
WorkSpaceMainMenuComponent.propDecorators = {
    reset: [{ type: Input }],
    httpMenuTag: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.reset;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.httpMenuTag;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.menus$;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.theme;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.layout;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.version;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.ws;
}
