// @er/core/esm5/lib/components/breadcrumb/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { StateNames } from '@er/types';
import { CommonsUtils, ConfigUtils, RegUtils, StatesUtils } from '@er/utils';
import { filter, map } from 'rxjs/operators';
import { BaseComponent } from '../../bases/component';
/** @type {?} */
var ROUTE_DATA_BREADCRUMB_KEY = 'breadcrumb';
var BreadcrumbsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BreadcrumbsComponent, _super);
    function BreadcrumbsComponent(router, route) {
        var _this = _super.call(this) || this;
        _this.router = router;
        _this.route = route;
        _this.outlet = 'primary';
        return _this;
    }
    /**
     * @return {?}
     */
    BreadcrumbsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.breadcrumbs$ = StatesUtils.observe(StateNames.nav)
            .pipe(filter(function (state) { return state && state.status === 'success'; }), map(function () { return _this.resolveRoute(); }));
    };
    /**
     * @param {?} breadcrumb
     * @return {?}
     */
    BreadcrumbsComponent.prototype.getRouterLink = /**
     * @param {?} breadcrumb
     * @return {?}
     */
    function (breadcrumb) {
        if (breadcrumb.params) {
            return [breadcrumb.url, breadcrumb.params];
        }
        else {
            return [breadcrumb.url];
        }
    };
    /**
     * @private
     * @return {?}
     */
    BreadcrumbsComponent.prototype.resolveRoute = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var breadcrumbs = [];
        /** @type {?} */
        var currentRoute = this.route.root;
        /** @type {?} */
        var menus = StatesUtils.getValue(StateNames.rootMenu);
        /** @type {?} */
        var url;
        /** @type {?} */
        var prefix = CommonsUtils.get(ConfigUtils.getConfig(), 'breadcrumb.prefix');
        /** @type {?} */
        var excludes = CommonsUtils.get(ConfigUtils.getConfig(), 'breadcrumb.excludes');
        /** @type {?} */
        var lastLabel = '';
        do {
            /** @type {?} */
            var childrenRoutes = currentRoute.children;
            currentRoute = null;
            childrenRoutes.forEach(function (route) {
                if (route.outlet === _this.outlet) {
                    /** @type {?} */
                    var routeSnapshot = route.snapshot;
                    /** @type {?} */
                    var label = null;
                    if (routeSnapshot.routeConfig && routeSnapshot.routeConfig.data && routeSnapshot.routeConfig.data[ROUTE_DATA_BREADCRUMB_KEY]) {
                        label = routeSnapshot.routeConfig.data[ROUTE_DATA_BREADCRUMB_KEY];
                    }
                    if (!label) {
                        /** @type {?} */
                        var currentUrl = routeSnapshot.url.join('/');
                        if (currentUrl && currentUrl.trim().length > 0) {
                            if (!url) {
                                url = currentUrl;
                            }
                            else {
                                url = url + '/' + currentUrl;
                            }
                        }
                        if (url && !RegUtils.test(url, excludes)) {
                            /** @type {?} */
                            var parts = url.split('/');
                            /** @type {?} */
                            var menu = CommonsUtils.deepFind(menus, { routerLink: url }, { routerLink: "/" + url }, { name: parts[parts.length - 1] });
                            if (menu && menu.length > 0) {
                                label = ((/** @type {?} */ (menu[0]))).label;
                            }
                        }
                    }
                    if (label && label.length !== 0 && label !== lastLabel) {
                        if (url.length === 0) {
                            routeSnapshot.params = {};
                        }
                        breadcrumbs.push((/** @type {?} */ ({
                            label: label,
                            url: url,
                            params: routeSnapshot.params
                        })));
                    }
                    currentRoute = route;
                    lastLabel = label;
                }
            });
        } while (currentRoute);
        if (prefix) {
            return prefix.concat(breadcrumbs);
        }
        else {
            return breadcrumbs;
        }
    };
    BreadcrumbsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-breadcrumbs',
                    template: "<ol class=\"breadcrumb\">\n    <ng-container *ngTemplateOutlet=\"templates?.before\"></ng-container>\n    <ng-container *ngFor=\"let breadcrumb of breadcrumbs$|async;let last=last\">\n        <li class=\"breadcrumb-item\" [ngClass]=\"{active: last}\">\n            <a *ngIf=\"!last\" [routerLink]=\"getRouterLink(breadcrumb)\">{{breadcrumb.label}}</a>\n            <span *ngIf=\"last\">{{breadcrumb.label}}</span>\n        </li>\n    </ng-container>\n    <ng-container *ngTemplateOutlet=\"templates?.after\"></ng-container>\n</ol>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    BreadcrumbsComponent.ctorParameters = function () { return [
        { type: Router },
        { type: ActivatedRoute }
    ]; };
    BreadcrumbsComponent.propDecorators = {
        outlet: [{ type: Input }]
    };
    return BreadcrumbsComponent;
}(BaseComponent));
export { BreadcrumbsComponent };
if (false) {
    /** @type {?} */
    BreadcrumbsComponent.prototype.outlet;
    /** @type {?} */
    BreadcrumbsComponent.prototype.breadcrumbs$;
    /**
     * @type {?}
     * @private
     */
    BreadcrumbsComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    BreadcrumbsComponent.prototype.route;
}
