// @er/formly-primeng/esm5/lib/components/debug/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
var PngFormDebugComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngFormDebugComponent, _super);
    function PngFormDebugComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PngFormDebugComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-form-debug',
                    template: "\n    <p-tabView>\n      <p-tabPanel header=\"Form Data\" leftIcon=\"fa fa-eye text-dark\">\n        <pre [erJsonViewer]=\"model\"></pre>\n      </p-tabPanel>\n      <p-tabPanel header=\"Form Json\" leftIcon=\"fa fa-gear text-info\">\n        <pre [erJsonViewer]=\"debugFormProps\"></pre>\n      </p-tabPanel>\n      <p-tabPanel header=\"Runtime Form Json\" leftIcon=\"fa fa-cogs text-danger\">\n        <pre [erJsonViewer]=\"runtimeFormProps\"></pre>\n      </p-tabPanel>\n    </p-tabView>\n  ",
                    styles: ["/deep/ .ui-tabview-selected{border-bottom:5px solid red!important}/deep/ .ui-tabview-selected .ui-tabview-title{color:#000!important}"]
                }] }
    ];
    PngFormDebugComponent.propDecorators = {
        model: [{ type: Input }],
        debugFormProps: [{ type: Input }],
        runtimeFormProps: [{ type: Input }]
    };
    return PngFormDebugComponent;
}(BaseComponent));
export { PngFormDebugComponent };
if (false) {
    /** @type {?} */
    PngFormDebugComponent.prototype.model;
    /** @type {?} */
    PngFormDebugComponent.prototype.debugFormProps;
    /** @type {?} */
    PngFormDebugComponent.prototype.runtimeFormProps;
}
