// @er/primeng/esm2015/lib/components/table/component/caption/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { OverlayPanel } from 'primeng/primeng';
import { PngTableComponent } from '../table/component';
export class PngTableCaptionComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     */
    constructor(pngTable) {
        super();
        this.pngTable = pngTable;
        this.onSimpleSearch = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    search(event) {
        this.onSimpleSearch.emit(event);
    }
    /**
     * @return {?}
     */
    openPanel() {
        StatesUtils.update(StateNames.popup, (/** @type {?} */ ((/** @type {?} */ ({
            header: '查询条件定义',
            content: this.searchBuilderPanel,
            visible: true,
            minWidth: 700,
            contentMaxHeight: 400
        })))));
    }
    /**
     * @return {?}
     */
    refresh() {
        this.pngTable.load();
    }
}
PngTableCaptionComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-table-caption',
                template: `
    <ng-container *ngIf="customerTpl;else default">
      <ng-container *ngTemplateOutlet="customerTpl;context:{$implicit:pngTable}"></ng-container>
    </ng-container>
    <ng-template #default>
      <div>
        <span class="pull-left table-caption" [innerHTML]="caption|erSafeHtml"></span>
        <span class="pull-right d-inline"><i class="fa fa-refresh text-primary" title="刷新表格" (click)="refresh()"></i></span>
      </div>
      <div class="clearfix"></div>
      <div *ngIf="hasSearchBox || hasAddAction || hasSearchInput">
        <div class="d-flex align-items-start">
          <div class="pl-3 py-3" *ngIf="hasAddAction">
            <png-button erPropsBind [props]="addButton" [outputContext]="{pngTable: pngTable}"></png-button>
          </div>
          <div class="ml-auto py-3" *ngIf="hasSearchInput && !hasSearchBox">
            <png-search-box title="综合搜索，输入回车键执行" (onSearch)="search($event)"></png-search-box>
          </div>
          <div class="ml-auto" *ngIf="hasSearchBox">
            <button pButton icon="fa fa-search" label="高级查询" (click)="openPanel()"></button>
            <ng-template #searchBuilderPanel>
              <png-table-query-builder></png-table-query-builder>
            </ng-template>
          </div>
        </div>
      </div>
    </ng-template>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngTableCaptionComponent.ctorParameters = () => [
    { type: PngTableComponent }
];
PngTableCaptionComponent.propDecorators = {
    caption: [{ type: Input }],
    hasAddAction: [{ type: Input }],
    hasSearchInput: [{ type: Input }],
    hasSearchBox: [{ type: Input }],
    addButton: [{ type: Input }],
    onSimpleSearch: [{ type: Output }],
    searchBuilderPanel: [{ type: ViewChild, args: ['searchBuilderPanel',] }]
};
if (false) {
    /** @type {?} */
    PngTableCaptionComponent.prototype.caption;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasAddAction;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasSearchInput;
    /** @type {?} */
    PngTableCaptionComponent.prototype.hasSearchBox;
    /** @type {?} */
    PngTableCaptionComponent.prototype.addButton;
    /** @type {?} */
    PngTableCaptionComponent.prototype.onSimpleSearch;
    /** @type {?} */
    PngTableCaptionComponent.prototype.searchBuilderPanel;
    /** @type {?} */
    PngTableCaptionComponent.prototype.pngTable;
}
