// @er/core/esm5/lib/directives/animate-if/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgIf } from '@angular/common';
import { Directive, EventEmitter, Input, Output, TemplateRef, ViewContainerRef } from '@angular/core';
import { Subject } from 'rxjs/index';
import { Observable } from 'rxjs/internal/Observable';
import { takeUntil } from 'rxjs/operators';
/**
 * @param {?} el
 * @param {?} maxWait
 * @return {?}
 */
function animationEndSafe(el, maxWait) {
    return Observable.create(function (observer) {
        el.addEventListener('animationend', observer);
        setTimeout(observer, maxWait);
    });
    // return new Promise((resolve, reject) => {
    //   el.addEventListener('animationend', resolve);
    //   setTimeout(resolve, maxWait);
    // });
}
var AnimatedIfDirective = /** @class */ (function (_super) {
    tslib_1.__extends(AnimatedIfDirective, _super);
    function AnimatedIfDirective(viewContainer, _templateRef) {
        var _this = _super.call(this, viewContainer, _templateRef) || this;
        _this.viewContainer = viewContainer;
        _this.animatedIfOnShow = new EventEmitter(false);
        _this.animatedIfOnHide = new EventEmitter(false);
        _this.visible = false;
        _this.destroyed$ = new Subject();
        return _this;
    }
    Object.defineProperty(AnimatedIfDirective.prototype, "erAnimatedIf", {
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (this.visible && !val) {
                this.hide();
            }
            else if (!this.visible && val) {
                this.show();
            }
            this.visible = val;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @return {?}
     */
    AnimatedIfDirective.prototype.getContainer = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.viewContainer.length > 0) {
            return ((/** @type {?} */ (this.viewContainer.get(0)))).rootNodes[0];
        }
        else {
            return null;
        }
    };
    /**
     * Hides the element. Animation is performed through css on the container.
     * This method will set the class to hiding and preserve the element for the animation duration.
     */
    /**
     * Hides the element. Animation is performed through css on the container.
     * This method will set the class to hiding and preserve the element for the animation duration.
     * @private
     * @return {?}
     */
    AnimatedIfDirective.prototype.hide = /**
     * Hides the element. Animation is performed through css on the container.
     * This method will set the class to hiding and preserve the element for the animation duration.
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var container = this.getContainer();
        if (!!container) {
            container.classList.remove('showing');
            container.classList.add('hiding');
            animationEndSafe(container, 1000)
                .pipe(takeUntil(this.destroyed$))
                .subscribe(function () {
                _this.ngIf = _this.visible;
                container.classList.remove('hiding');
            });
            this.animatedIfOnHide.emit();
        }
    };
    /**
     * Shows the element. Animation is performed through css on the container
     */
    /**
     * Shows the element. Animation is performed through css on the container
     * @private
     * @return {?}
     */
    AnimatedIfDirective.prototype.show = /**
     * Shows the element. Animation is performed through css on the container
     * @private
     * @return {?}
     */
    function () {
        this.ngIf = true;
        /** @type {?} */
        var container = this.getContainer();
        if (!!container) {
            container.classList.remove('hiding');
            container.classList.add('showing');
            animationEndSafe(container, 1000)
                .pipe(takeUntil(this.destroyed$))
                .subscribe(function () {
                container.classList.remove('showing');
            });
            this.animatedIfOnShow.emit();
        }
    };
    /**
     * @return {?}
     */
    AnimatedIfDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.destroyed$.next();
        this.destroyed$.complete();
    };
    AnimatedIfDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erAnimatedIf]'
                },] }
    ];
    /** @nocollapse */
    AnimatedIfDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: TemplateRef }
    ]; };
    AnimatedIfDirective.propDecorators = {
        animatedIfOnShow: [{ type: Output }],
        animatedIfOnHide: [{ type: Output }],
        erAnimatedIf: [{ type: Input }]
    };
    return AnimatedIfDirective;
}(NgIf));
export { AnimatedIfDirective };
if (false) {
    /** @type {?} */
    AnimatedIfDirective.prototype.animatedIfOnShow;
    /** @type {?} */
    AnimatedIfDirective.prototype.animatedIfOnHide;
    /**
     * @type {?}
     * @private
     */
    AnimatedIfDirective.prototype.visible;
    /**
     * @type {?}
     * @protected
     */
    AnimatedIfDirective.prototype.destroyed$;
    /**
     * @type {?}
     * @private
     */
    AnimatedIfDirective.prototype.viewContainer;
}
