// @er/core/esm2015/lib/utils/log/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LogService } from '../../services/log/service';
import { DiUtils } from '../di';
export class LogUtils {
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    static debug(from, message, ...context) {
        LogUtils.getLogService().debug(from, message, context);
    }
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    static info(from, message, ...context) {
        LogUtils.getLogService().info(from, message, context);
    }
    /**
     * @param {?} messages
     * @return {?}
     */
    static warn(messages) {
        LogUtils.getLogService().warn(messages);
    }
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    static error(from, message, ...context) {
        LogUtils.getLogService().error(from, message, context);
    }
    /**
     * @private
     * @return {?}
     */
    static getLogService() {
        if (!LogUtils._logService) {
            LogUtils._logService = DiUtils.get(LogService);
        }
        return LogUtils._logService;
    }
}
LogUtils._logService = null;
if (false) {
    /**
     * @type {?}
     * @private
     */
    LogUtils._logService;
}
