// @er/core/esm2015/lib/components/layout/left-right/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '../../../bases/component';
export class LeftRightContainerComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.leftStyleClass = 'col-3';
        this.rightStyleClass = 'col-9';
        this.switchable = true;
        this.onLeftHide = new EventEmitter();
        this.templateNames = {
            left: 'left',
            right: 'right'
        };
    }
    /**
     * @param {?} event
     * @return {?}
     */
    reLayout(event) {
        this.hideLeft = event;
        this.onLeftHide.emit(event);
    }
}
LeftRightContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-left-right',
                template: `
    <div class="d-flex">
      <div *ngIf="!hideLeft" class="w-25 shadow-lg">
        <ng-container *ngTemplateOutlet="template(templateNames.left)"></ng-container>
      </div>
      <div class="flex-grow-1">
        <div *ngIf="switchable">
          <i class="fa fa-2x  text-primary" [title]="hideLeft?'展开左侧内容':'收起左侧内容'"
             [ngClass]="hideLeft?'fa-arrow-circle-right':'fa-arrow-circle-left'"
             (click)="reLayout(!hideLeft)"></i>
        </div>
        <ng-container *ngTemplateOutlet="template(templateNames.right)"></ng-container>
      </div>
    </div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
LeftRightContainerComponent.propDecorators = {
    leftStyleClass: [{ type: Input }],
    rightStyleClass: [{ type: Input }],
    hideLeft: [{ type: Input }],
    switchable: [{ type: Input }],
    onLeftHide: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    LeftRightContainerComponent.prototype.leftStyleClass;
    /** @type {?} */
    LeftRightContainerComponent.prototype.rightStyleClass;
    /** @type {?} */
    LeftRightContainerComponent.prototype.hideLeft;
    /** @type {?} */
    LeftRightContainerComponent.prototype.switchable;
    /** @type {?} */
    LeftRightContainerComponent.prototype.onLeftHide;
    /** @type {?} */
    LeftRightContainerComponent.prototype.templateNames;
}
