// @er/primeng/esm5/lib/components/search-box/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '@er/core';
import { DialogUtils } from '@er/utils';
var PngSearchBoxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngSearchBoxComponent, _super);
    function PngSearchBoxComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.searched = false;
        _this.title = '按关键字搜索,输入回车键执行';
        _this.onSearch = new EventEmitter();
        _this.onSearchChange = new EventEmitter();
        return _this;
    }
    /**
     * @return {?}
     */
    PngSearchBoxComponent.prototype.search = /**
     * @return {?}
     */
    function () {
        this.searched = true;
        this.onSearch.emit(this.searchKey);
    };
    /**
     * @return {?}
     */
    PngSearchBoxComponent.prototype.keyChange = /**
     * @return {?}
     */
    function () {
        if (!this.searchKey || this.searchKey.length === 0) {
            this.clearSearch();
        }
        this.onSearchChange.emit(this.searchKey);
    };
    /**
     * @return {?}
     */
    PngSearchBoxComponent.prototype.clearSearch = /**
     * @return {?}
     */
    function () {
        this.searchKey = undefined;
        if (this.searched) {
            this.onSearch.emit(undefined);
        }
        this.searched = false;
    };
    /**
     * @return {?}
     */
    PngSearchBoxComponent.prototype.showHelp = /**
     * @return {?}
     */
    function () {
        DialogUtils.info('查询操作符', "\n    <div class=\"ml-3\">\n       <div><b class=\"font-weight-bold\">+</b> \u5E76\u4E14</div>\n       <div><b class=\"font-weight-bold\">|</b> \u6216\u8005</div>\n       <div><b class=\"font-weight-bold\">-</b> \u6392\u9664</div>\n       <div><b class=\"font-weight-bold\">*</b> \u5F00\u5934</div>\n       <div><b class=\"font-weight-bold\">( )</b> \u4F18\u5148</div>\n    </div>\n    ");
    };
    PngSearchBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-search-box',
                    template: "\n    <span class=\"inner-addon\" [ngClass]=\"styleClass\" [ngStyle]=\"style\">\n      <i class=\"fa fa-search text-primary left\"></i>\n      <input [title]=\"title\" pInputText placeholder=\"\u8F93\u5165\u641C\u7D22\u5173\u952E\u5B57\u6216\u62FC\u97F3\" [(ngModel)]=\"searchKey\" (keydown.enter)=\"search()\" (keyup)=\"keyChange()\"/>\n      <i *ngIf=\"searchKey\" class=\"fa fa-trash gray right\" [title]=\"title\" (click)=\"clearSearch()\"></i>\n   </span>\n    <span class=\"position-container\"><i title=\"\u591A\u503C\u67E5\u8BE2\u5E2E\u52A9\" class=\"fa fa-question-circle green position-absolute\" style=\"right:2px\"\n                                        (click)=\"showHelp()\"></i></span>\n\n  "
                }] }
    ];
    PngSearchBoxComponent.propDecorators = {
        styleClass: [{ type: Input }],
        style: [{ type: Input }],
        title: [{ type: Input }],
        onSearch: [{ type: Output }],
        onSearchChange: [{ type: Output }]
    };
    return PngSearchBoxComponent;
}(BaseComponent));
export { PngSearchBoxComponent };
if (false) {
    /** @type {?} */
    PngSearchBoxComponent.prototype.searchKey;
    /** @type {?} */
    PngSearchBoxComponent.prototype.searched;
    /** @type {?} */
    PngSearchBoxComponent.prototype.styleClass;
    /** @type {?} */
    PngSearchBoxComponent.prototype.style;
    /** @type {?} */
    PngSearchBoxComponent.prototype.title;
    /** @type {?} */
    PngSearchBoxComponent.prototype.onSearch;
    /** @type {?} */
    PngSearchBoxComponent.prototype.onSearchChange;
}
