// @er/primeng/esm2015/lib/components/slider/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { DEFAULT_PNG_SLIDER_PROPS } from './type';
export class PngSliderComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.$defaultProps = DEFAULT_PNG_SLIDER_PROPS;
    }
    /**
     * @return {?}
     */
    get current() {
        return CommonsUtils.join(this.value, '——');
    }
    /**
     * @param {?} event
     * @return {?}
     */
    initValue(event) {
        if (event.hasOwnProperty('min') && event.hasOwnProperty('max')) {
            this.min = event.min;
            this.max = event.max;
            this.value = [event.min, event.max];
        }
        else {
            this.value = event;
        }
    }
}
PngSliderComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-slider',
                template: `
    <small class="d-inline w-25">值：{{current}}</small>
    <p-slider class="d-inline"
              [erApiData]="apiDataProps"
              [(ngModel)]="value"
              erPropsBind
              [props]="$props"
              [min]="min"
              [max]="max"
              (onChange)="updateTouch()"
              (onDataReady)="initValue($event)"
    >
    </p-slider>

  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngSliderComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngSliderComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSliderComponent.prototype.min;
    /** @type {?} */
    PngSliderComponent.prototype.max;
}
