// @er/presets/esm5/lib/defaults/animation/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { transition, trigger, useAnimation } from '@angular/animations';
import { bounce, bounceIn, bounceInDown, bounceInLeft, bounceInRight, bounceInUp, bounceOut, bounceOutDown, bounceOutLeft, bounceOutRight, bounceOutUp, fadeIn, fadeInDown, fadeInLeft, fadeInRight, fadeInUp, fadeOut, fadeOutDown, fadeOutLeft, fadeOutRight, fadeOutUp, flash, flip, flipInX, flipInY, flipOutX, flipOutY, hinge, jackInTheBox, jello, lightSpeedIn, lightSpeedOut, pulse, rollIn, rollOut, rotateIn, rotateInDownLeft, rotateInDownRight, rotateInUpLeft, rotateInUpRight, rotateOut, rotateOutDownLeft, rotateOutDownRight, rotateOutUpLeft, rotateOutUpRight, rubberBand, shake, slideInDown, slideInLeft, slideInRight, slideInUp, slideOutDown, slideOutLeft, slideOutRight, slideOutUp, swing, tada, wobble, zoomIn, zoomInDown, zoomInLeft, zoomInRight, zoomInUp, zoomOut, zoomOutDown, zoomOutLeft, zoomOutRight, zoomOutUp } from 'ng-animate';
/** @type {?} */
export var ANIMATION_NAMES = {
    bounce: bounce,
    flash: flash,
    pulse: pulse,
    rubberBand: rubberBand,
    shake: shake,
    swing: swing,
    tada: tada,
    wobble: wobble,
    jello: jello,
    bounceInDown: bounceInDown,
    bounceInLeft: bounceInLeft,
    bounceInRight: bounceInRight,
    bounceInUp: bounceInUp,
    bounceOutDown: bounceOutDown,
    bounceOutLeft: bounceOutLeft,
    bounceOutRight: bounceOutRight,
    bounceOutUp: bounceOutUp,
    fadeIn: fadeIn,
    fadeInDown: fadeInDown,
    fadeInLeft: fadeInLeft,
    fadeInRight: fadeInRight,
    fadeInUp: fadeInUp,
    fadeOut: fadeOut,
    fadeOutDown: fadeOutDown,
    fadeOutLeft: fadeOutLeft,
    fadeOutRight: fadeOutRight,
    fadeOutUp: fadeOutUp,
    slideInDown: slideInDown,
    slideInLeft: slideInLeft,
    slideInRight: slideInRight,
    slideInUp: slideInUp,
    slideOutDown: slideOutDown,
    slideOutLeft: slideOutLeft,
    slideOutRight: slideOutRight,
    slideOutUp: slideOutUp,
    flip: flip,
    flipInX: flipInX,
    flipInY: flipInY,
    flipOutX: flipOutX,
    flipOutY: flipOutY,
    lightSpeedIn: lightSpeedIn,
    lightSpeedOut: lightSpeedOut,
    rotateIn: rotateIn,
    rotateInDownLeft: rotateInDownLeft,
    rotateInDownRight: rotateInDownRight,
    rotateInUpLeft: rotateInUpLeft,
    rotateInUpRight: rotateInUpRight,
    rotateOut: rotateOut,
    rotateOutDownLeft: rotateOutDownLeft,
    rotateOutDownRight: rotateOutDownRight,
    rotateOutUpLeft: rotateOutUpLeft,
    rotateOutUpRight: rotateOutUpRight,
    jackInTheBox: jackInTheBox,
    hinge: hinge,
    rollIn: rollIn,
    rollOut: rollOut,
    zoomIn: zoomIn,
    zoomOut: zoomOut,
    zoomInDown: zoomInDown,
    zoomInLeft: zoomInLeft,
    zoomInRight: zoomInRight,
    zoomInUp: zoomInUp,
    zoomOutDown: zoomOutDown,
    zoomOutLeft: zoomOutLeft,
    zoomOutRight: zoomOutRight,
    zoomOutUp: zoomOutUp
};
/** @type {?} */
export var ANIMATIONS = {
    bounce: trigger('bounce', [transition('* => *', useAnimation(bounce))]),
    flash: trigger('flash', [transition('* => *', useAnimation(flash))]),
    pulse: trigger('pulse', [transition('* => *', useAnimation(pulse))]),
    rubberBand: trigger('rubberBand', [transition('* => *', useAnimation(rubberBand))]),
    shake: trigger('shake', [transition('* => *', useAnimation(shake))]),
    swing: trigger('swing', [transition('* => *', useAnimation(swing))]),
    tada: trigger('tada', [transition('* => *', useAnimation(tada))]),
    wobble: trigger('wobble', [transition('* => *', useAnimation(wobble))]),
    jello: trigger('jello', [transition('* => *', useAnimation(jello))]),
    bounceIn: trigger('bounceIn', [transition('* => *', useAnimation(bounceIn))]),
    bounceInDown: trigger('bounceInDown', [transition('* => *', useAnimation(bounceInDown))]),
    bounceInLeft: trigger('bounceInLeft', [transition('* => *', useAnimation(bounceInLeft))]),
    bounceOut: trigger('bounceOut', [transition('* => *', useAnimation(bounceOut))]),
    bounceOutRight: trigger('bounceOutRight', [
        transition('* => *', useAnimation(bounceOutRight))
    ]),
    bounceOutUp: trigger('bounceOutUp', [transition('* => *', useAnimation(bounceOutUp))]),
    fadeIn: trigger('fadeIn', [transition('* => *', useAnimation(fadeIn))]),
    fadeInDown: trigger('fadeInDown', [transition('* => *', useAnimation(fadeInDown))]),
    fadeInLeft: trigger('fadeInLeft', [transition('* => *', useAnimation(fadeInLeft))]),
    fadeOut: trigger('fadeOut', [transition('* => *', useAnimation(fadeOut))]),
    fadeOutUp: trigger('fadeOutUp', [transition('* => *', useAnimation(fadeOutUp))]),
    fadeOutRight: trigger('fadeOutRight', [transition('* => *', useAnimation(fadeOutRight))]),
    slideInDown: trigger('slideInDown', [transition('* => *', useAnimation(slideInDown))]),
    slideInLeft: trigger('slideInLeft', [transition('* => *', useAnimation(slideInLeft))]),
    slideOutUp: trigger('slideOutUp', [transition('* => *', useAnimation(slideOutUp))]),
    slideOutRight: trigger('slideOutRight', [
        transition('* => *', useAnimation(slideOutRight))
    ]),
    flip: trigger('flip', [transition('* => *', useAnimation(flip))]),
    flipInX: trigger('flipInX', [transition('* => *', useAnimation(flipInX))]),
    flipInY: trigger('flipInY', [transition('* => *', useAnimation(flipInY))]),
    flipOutX: trigger('flipOutX', [transition('* => *', useAnimation(flipOutX))]),
    flipOutY: trigger('flipOutY', [transition('* => *', useAnimation(flipOutY))]),
    lightSpeedIn: trigger('lightSpeedIn', [transition('* => *', useAnimation(lightSpeedIn))]),
    lightSpeedOut: trigger('lightSpeedOut', [
        transition('* => *', useAnimation(lightSpeedOut))
    ]),
    rotateIn: trigger('rotateIn', [transition('* => *', useAnimation(rotateIn))]),
    rotateInDownLeft: trigger('rotateInDownLeft', [
        transition('* => *', useAnimation(rotateInDownLeft))
    ]),
    rotateInUpRight: trigger('rotateInUpRight', [
        transition('* => *', useAnimation(rotateInUpRight))
    ]),
    rotateOut: trigger('rotateOut', [transition('* => *', useAnimation(rotateOut))]),
    rotateOutUpLeft: trigger('rotateOutUpLeft', [
        transition('* => *', useAnimation(rotateOutUpLeft))
    ]),
    rotateOutDownRight: trigger('rotateOutDownRight', [
        transition('* => *', useAnimation(rotateOutDownRight))
    ]),
    hinge: trigger('hinge', [transition('* => *', useAnimation(hinge))]),
    jackInTheBox: trigger('jackInTheBox', [transition('* => *', useAnimation(jackInTheBox))]),
    rollIn: trigger('rollIn', [transition('* => *', useAnimation(rollIn))]),
    rollOut: trigger('rollOut', [transition('* => *', useAnimation(rollOut))]),
    zoomIn: trigger('zoomIn', [transition('* => *', useAnimation(zoomIn))]),
    zoomInLeft: trigger('zoomInLeft', [transition('* => *', useAnimation(zoomInLeft))]),
    zoomInDown: trigger('zoomInDown', [transition('* => *', useAnimation(zoomInDown))]),
    zoomOut: trigger('zoomOut', [transition('* => *', useAnimation(zoomOut))]),
    zoomOutUp: trigger('zoomOutUp', [transition('* => *', useAnimation(zoomOutUp))]),
    zoomOutRight: trigger('zoomOutRight', [transition('* => *', useAnimation(zoomOutRight))])
};
