// @er/types/esm5/lib/dialog/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var DialogType = {
    CONFIRM: 'confirm',
    POPUP: 'popup',
};
export { DialogType };
/**
 * @record
 */
export function DialogProp() { }
if (false) {
    /** @type {?|undefined} */
    DialogProp.prototype.header;
    /** @type {?|undefined} */
    DialogProp.prototype.width;
    /** @type {?|undefined} */
    DialogProp.prototype.height;
    /** @type {?|undefined} */
    DialogProp.prototype.closeOnEscape;
    /** @type {?|undefined} */
    DialogProp.prototype.context;
}
/**
 * @record
 */
export function ConfirmDialogProp() { }
if (false) {
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.content;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.icon;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.key;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rtl;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.closable;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.responsive;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.appendTo;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.acceptLabel;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.acceptIcon;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.acceptVisible;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rejectLabel;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rejectIcon;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rejectVisible;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.acceptButtonStyleClass;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rejectButtonStyleClass;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.onAccept;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.onReject;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.acceptEvent;
    /** @type {?|undefined} */
    ConfirmDialogProp.prototype.rejectEvent;
}
/**
 * @record
 */
export function PopupDialogProp() { }
if (false) {
    /** @type {?|undefined} */
    PopupDialogProp.prototype.content;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.header;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.draggable;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.resizable;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.minWidth;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.minHeight;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.contentMaxHeight;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.contentMaxWidth;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.contentWidth;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.contentHeight;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.contentStyle;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.visible;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.modal;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.dismissableMask;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.rtl;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.closable;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.responsive;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.breakpoint;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.appendTo;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.style;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.styleClass;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.showHeader;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.showFooter;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.positionLeft;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.positionTop;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.baseZIndex;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.autoZIndex;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.ariaLabelledBy;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.blockScroll;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.maximizable;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.closeLabel;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.closeIcon;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.onClose;
    /** @type {?|undefined} */
    PopupDialogProp.prototype.buttons;
}
