// @er/core/esm2015/lib/components/we-chat/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ConfigUtils } from '@er/utils';
export class WechatQrCodeComponent {
    constructor() {
        this.height = 300;
        this.width = 300;
        this.baseUrl = 'https://open.weixin.qq.com/connect/qrconnect';
    }
    /**
     * @return {?}
     */
    get qrCodeUrl() {
        /** @type {?} */
        const wechat = ConfigUtils.getConfig().wechat;
        /** @type {?} */
        const redirect = encodeURIComponent(wechat.redirect_uri);
        /** @type {?} */
        let url = `${this.baseUrl}?appid=${wechat.appid}&redirect_uri=${redirect}&scope=${wechat.scope}&response_type=code`;
        url += wechat.style ? '&style=' + wechat.style : '';
        url += wechat.self_redirect ? '&self_redirect=' + wechat.self_redirect : '';
        url += wechat.href ? '&href=' + wechat.href : '';
        return url;
    }
}
WechatQrCodeComponent.decorators = [
    { type: Component, args: [{
                selector: 'wechat-qrcode',
                template: `
    <iframe [src]="qrCodeUrl|erSafeHtml:'resource'" [width]="width" [height]="height" scrolling="no" [attr.frameborder]="0"></iframe>
  `
            }] }
];
WechatQrCodeComponent.propDecorators = {
    height: [{ type: Input }],
    width: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    WechatQrCodeComponent.prototype.height;
    /** @type {?} */
    WechatQrCodeComponent.prototype.width;
    /**
     * @type {?}
     * @private
     */
    WechatQrCodeComponent.prototype.baseUrl;
}
