// @er/core/esm2015/lib/components/button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output } from '@angular/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { BaseComponent } from '../../bases/component';
export class StateButtonComponent extends BaseComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super();
        this.cd = cd;
        this.type = 'button';
        this.hint = '';
        this.onClick = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.processIcon = 'fa fa-cog fa-spin';
        this.processing$ = new Subject();
    }
    /**
     * @param {?} processing
     * @return {?}
     */
    set processing(processing) {
        this.processing$.next(processing);
        this.cd.markForCheck(); // used for not dom event modify
    }
    /**
     * @return {?}
     */
    get btnLabel() {
        if (this.processing) {
            return this.stateLabel || this.label;
        }
        else {
            return this.label || '';
        }
    }
    /**
     * @return {?}
     */
    get btnTitle() {
        if (this.processing) {
            return this.stateHint || '执行中...';
        }
        else {
            return this.hint || '';
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.stateKey) {
            this.$subscriptions = StatesUtils.observe(StateNames.http, this.stateKey)
                .pipe(filter(state => !!state), map(state => state['start'])).subscribe(b => this.processing$.next(b));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    click(event) {
        this.onClick.emit(Object.assign({}, event, this.context || {}));
    }
}
StateButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-button',
                template: "<button *ngIf=\"!hidden\"\n        [attr.type]=\"type\"\n        [ngClass]=\"styleClass\"\n        [ngStyle]=\"style\"\n        [disabled]=\"disabled || (processing$|async) \"\n        [title]=\"hint\"\n        (click)=\"click($event)\"\n        (focus)=\"emitEvent('onFocus',$event)\"\n        (blur)=\"emitEvent('onBlur',$event)\"\n>\n  <i *ngIf=\"!iconRight && !(processing$|async)\" [ngClass]=\"icon\"></i> {{btnLabel}}\n  <ng-content></ng-content>\n  <i *ngIf=\"iconRight && !(processing$|async)\" [ngClass]=\"icon\"  ></i>\n  <i *ngIf=\"processing$|async\" [ngClass]=\"processIcon\" style=\"margin-left: 5px\"></i>\n</button>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
StateButtonComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
StateButtonComponent.propDecorators = {
    type: [{ type: Input }],
    hint: [{ type: Input }],
    label: [{ type: Input }],
    icon: [{ type: Input }],
    style: [{ type: Input }],
    hidden: [{ type: Input }],
    styleClass: [{ type: Input }],
    disabled: [{ type: Input }],
    iconRight: [{ type: Input }],
    stateKey: [{ type: Input }],
    stateLabel: [{ type: Input }],
    stateHint: [{ type: Input }],
    context: [{ type: Input }],
    onClick: [{ type: Output }],
    onFocus: [{ type: Output }],
    onBlur: [{ type: Output }],
    processing: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    StateButtonComponent.prototype.type;
    /** @type {?} */
    StateButtonComponent.prototype.hint;
    /** @type {?} */
    StateButtonComponent.prototype.label;
    /** @type {?} */
    StateButtonComponent.prototype.icon;
    /** @type {?} */
    StateButtonComponent.prototype.style;
    /** @type {?} */
    StateButtonComponent.prototype.hidden;
    /** @type {?} */
    StateButtonComponent.prototype.styleClass;
    /** @type {?} */
    StateButtonComponent.prototype.disabled;
    /** @type {?} */
    StateButtonComponent.prototype.iconRight;
    /** @type {?} */
    StateButtonComponent.prototype.stateKey;
    /** @type {?} */
    StateButtonComponent.prototype.stateLabel;
    /** @type {?} */
    StateButtonComponent.prototype.stateHint;
    /** @type {?} */
    StateButtonComponent.prototype.context;
    /** @type {?} */
    StateButtonComponent.prototype.onClick;
    /** @type {?} */
    StateButtonComponent.prototype.onFocus;
    /** @type {?} */
    StateButtonComponent.prototype.onBlur;
    /** @type {?} */
    StateButtonComponent.prototype.processIcon;
    /** @type {?} */
    StateButtonComponent.prototype.processing$;
    /**
     * @type {?}
     * @protected
     */
    StateButtonComponent.prototype.cd;
}
