// @er/core/esm5/lib/directives/api-data/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
import { ApiUtils } from '../../utils/api';
import { LogUtils } from '../../utils/log';
var ApiDataDirective = /** @class */ (function (_super) {
    tslib_1.__extends(ApiDataDirective, _super);
    function ApiDataDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onDataReady = new EventEmitter();
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ApiDataDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.apiDataProps
            && changes.apiDataProps.currentValue
            && (this.apiDataProps.autoFetch !== false || !changes.apiDataProps.isFirstChange())) {
            this.fetchData();
        }
    };
    /**
     * @return {?}
     */
    ApiDataDirective.prototype.fetchData = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (CommonsUtils.isEmpty(this.apiDataProps)) {
            LogUtils.debug('apiDataDirective', '没有定义apiDataProps');
            return;
        }
        ApiUtils.fetch(this.apiDataProps)
            .subscribe(function (data) {
            LogUtils.debug(_this, 'apiData待订阅数据', data);
            if (_this.apiDataProps.data$) {
                _this.apiDataProps.data$.next(data);
            }
            _this.onDataReady.emit(data);
        });
    };
    ApiDataDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erApiData]'
                },] }
    ];
    ApiDataDirective.propDecorators = {
        apiDataProps: [{ type: Input, args: ['erApiData',] }],
        onRequestData: [{ type: Output }],
        onDataReady: [{ type: Output }]
    };
    return ApiDataDirective;
}(BaseDirective));
export { ApiDataDirective };
if (false) {
    /** @type {?} */
    ApiDataDirective.prototype.apiDataProps;
    /** @type {?} */
    ApiDataDirective.prototype.onRequestData;
    /** @type {?} */
    ApiDataDirective.prototype.onDataReady;
}
