// @er/core/esm2015/lib/utils/nav/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Location } from '@angular/common';
import { ActivatedRoute, Router } from '@angular/router';
import { StateNames } from '@er/types';
import { CommonsUtils, StatesUtils } from '@er/utils';
import { DiUtils } from '../di';
export class RouteUtils {
    /**
     * @return {?}
     */
    static getActiveRoute() {
        /** @type {?} */
        let route = DiUtils.get(ActivatedRoute);
        while (route.firstChild) {
            route = route.firstChild;
        }
        return route;
    }
    /**
     * @return {?}
     */
    static getActiveRouteData() {
        /** @type {?} */
        const route = RouteUtils.getActiveRoute();
        return route && route.snapshot.data;
    }
    /**
     * @param {?=} outletName
     * @param {?=} isRelative
     * @return {?}
     */
    static clearOutlet(outletName = 'primary', isRelative = false) {
        /** @type {?} */
        const extra = {};
        if (!isRelative) {
            extra['relativeTo'] = RouteUtils.getActiveRoute();
        }
        DiUtils.get(Router).navigate([{ outlets: { [outletName]: null } }], extra);
    }
    /**
     * @param {?} commands
     * @param {?=} paramsOrRelative
     * @param {?=} extras
     * @return {?}
     */
    static nav(commands, paramsOrRelative, extras) {
        if (CommonsUtils.isString(commands)) {
            commands = [commands];
        }
        if (!extras) {
            extras = (/** @type {?} */ ({}));
        }
        if (paramsOrRelative && Object.keys(paramsOrRelative).length > 0) {
            CommonsUtils.set((/** @type {?} */ (extras)), 'queryParams', paramsOrRelative);
        }
        if ((CommonsUtils.isString(commands) && !((/** @type {?} */ (commands))).startsWith('/'))
            || (paramsOrRelative === true)) {
            CommonsUtils.set((/** @type {?} */ (extras)), 'relativeTo', RouteUtils.getActiveRoute());
        }
        DiUtils.get(Router).navigate((/** @type {?} */ (commands)), extras);
    }
    /**
     * @return {?}
     */
    static back() {
        DiUtils.get(Location).back();
    }
    /**
     * @return {?}
     */
    static forward() {
        DiUtils.get(Location).forward();
    }
    /**
     * @return {?}
     */
    static refresh() {
        RouteUtils.navigate(StatesUtils.getValue(StateNames.nav)['url']);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    static navigate(url) {
        if (url.startsWith('http')) {
            window.location.href = url;
        }
        else {
            DiUtils.get(Router).navigateByUrl(url);
        }
    }
}
