// @er/core/esm2015/lib/pipes/up-to-now/pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, NgZone, Pipe } from '@angular/core';
import { DateUtils } from '@er/utils';
export class UpToNowPipe {
    /**
     * @param {?} cd
     * @param {?} ngZone
     */
    constructor(cd, ngZone) {
        this.cd = cd;
        this.ngZone = ngZone;
    }
    /**
     * @param {?} value
     * @param {?=} due
     * @return {?}
     */
    transform(value, due) {
        /** @type {?} */
        const d = DateUtils.toDate(value);
        if (!d) {
            return '';
        }
        /** @type {?} */
        const span = DateUtils.getDateByNow(d, due);
        this.removeTimer();
        /** @type {?} */
        const timeToUpdate = (Number.isNaN(span.seconds)) ? 1000 : DateUtils.getUpdateSeconds(span.seconds) * 1000;
        this.timer = this.ngZone.runOutsideAngular(() => {
            if (typeof window !== 'undefined') {
                return window.setTimeout(() => {
                    this.ngZone.run(() => this.cd.detectChanges());
                }, timeToUpdate);
            }
            return null;
        });
        return span.message;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.removeTimer();
    }
    /**
     * @private
     * @return {?}
     */
    removeTimer() {
        if (this.timer) {
            window.clearTimeout(this.timer);
            this.timer = null;
        }
    }
}
UpToNowPipe.decorators = [
    { type: Pipe, args: [{
                name: 'erUpToNow',
                pure: false
            },] }
];
/** @nocollapse */
UpToNowPipe.ctorParameters = () => [
    { type: ChangeDetectorRef },
    { type: NgZone }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.timer;
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.ngZone;
}
