// @er/formly-primeng/esm5/lib/components/wizard/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Output } from '@angular/core';
import { FormArray, FormGroup } from '@angular/forms';
import { LogUtils } from '@er/core';
import { PngFormlyUtils } from '../../utils';
import { PngFormlyFormComponent } from '../form/component';
var PngFormlyWizardFormComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngFormlyWizardFormComponent, _super);
    function PngFormlyWizardFormComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.stepChange = new EventEmitter();
        return _this;
    }
    /**
     * @param {?} formProps
     * @return {?}
     */
    PngFormlyWizardFormComponent.prototype.resolveFormFields = /**
     * @param {?} formProps
     * @return {?}
     */
    function (formProps) {
        var _this = this;
        if (!formProps || !formProps['steps']) {
            LogUtils.error(this, '没有定义steps', formProps);
        }
        if (formProps.steps) {
            formProps.steps.forEach(function (step) {
                step.fields.forEach(function (field) {
                    PngFormlyUtils.resolveField(formProps, field, _this);
                });
            });
            this.formArray = new FormArray(formProps.steps.map(function () { return new FormGroup({}); }));
            this.stepFormlyOptions = formProps.steps.map(function (step) {
                return tslib_1.__assign({}, formProps.options || {}, step.options || {});
            });
        }
        this.currentStepIndex = formProps.activeIndex || 0;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngFormlyWizardFormComponent.prototype.onStepChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.currentStepIndex = event;
        this.stepChange.emit(event);
    };
    PngFormlyWizardFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-formly-wizard',
                    template: "\n    <png-panel [header]=\"formProps.header\" [footer]=\"formProps.footer\" [icon]=\"formProps.icon\" [toggleable]=\"formProps.collapsable\">\n      <png-wizard [(activeIndex)]=\"currentStepIndex\"\n                  (stepChange)=\"onStepChange($event)\"\n                  [styleClass]=\"formProps.styleClass\"\n                  [style]=\"formProps.style\"\n                  [stepStyleClass]=\"formProps['stepStyleClass']\"\n                  [stepStyle]=\"formProps['stepStyle']\"\n      >\n        <ng-container *ngFor=\"let step of formProps['steps']; let i = index;\">\n          <png-step\n            [label]=\"step.label\"\n            [styleClass]=\"step.styleClass\"\n            [style]=\"step.style\"\n          >\n            <formly-form\n              [form]=\"formArray.at(i)\"\n              [(model)]=\"formProps.model\"\n              [fields]=\"step.fields\"\n              [options]=\"stepFormlyOptions[i]\">\n            </formly-form>\n          </png-step>\n        </ng-container>\n        <ng-template erTypedTemplate=\"footer\">\n          <button pButton type=\"button\" *ngIf=\"currentStepIndex+1 === formProps['steps'].length\"\n                  class=\"pull-right\" icon=\"fa fa-save\" iconPos=\"right\" label=\"\u63D0\u4EA4\"></button>\n        </ng-template>\n      </png-wizard>\n    </png-panel>\n    <!--<png-form-debug [model]=\"formModel\"-->\n    <!--[debugFormProps]=\"debugFormProps\"-->\n    <!--[runtimeFormProps]=\"formProps\">-->\n    <!--</png-form-debug>-->\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PngFormlyWizardFormComponent.propDecorators = {
        stepChange: [{ type: Output }]
    };
    return PngFormlyWizardFormComponent;
}(PngFormlyFormComponent));
export { PngFormlyWizardFormComponent };
if (false) {
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.formArray;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.stepFormlyOptions;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.currentStepIndex;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.stepChange;
}
