// @er/utils/esm2015/lib/config/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ApiServerType, LogLevel, NotifyType } from '@er/types';
import { CommonsUtils } from '../comm';
/** @type {?} */
const DEFAULT_APP_CONFIG = {
    version: '0.0.1',
    stomp: {
        enabled: false,
        destination: '/message/push',
        headers: {},
        heartbeat_in: 0,
        // Typical value 0 - disabled
        heartbeat_out: 20000,
        // Typical value 20000 - every 20 seconds
        reconnect_delay: 5000,
        debug: false
    },
    auth: {
        storageKey: 'erCurrentUser',
        headerName: 'Authorization',
        authScheme: 'Bearer ',
        userTokenKey: 'userName',
        passwordKey: 'password',
        rememberMeKey: 'rememberMe',
        requiredUserProps: 'org',
        loginRouter: 'user/login',
        loginApiOptions: { serverType: ApiServerType.App },
        loginSuccessRouter: '/',
        logoutSuccessRouter: 'user/login'
    },
    alert: {},
    announce: {},
    notify: NotifyType.ALERT,
    onLoading: 'fa fa-cog fa-spin fa-3x indigo',
    enableLogger: true,
    dataFieldProps: {
        dataCode: {
            apiEntry: 'data_code',
            codeKey: 'code',
            nameKey: 'name',
            parentKey: 'parent',
            catKey: 'codeCat'
        }
    },
    wechat: {
        self_redirect: true,
        scope: 'snsapi_login',
        state: '',
        style: 'black',
        href: ''
    },
    log: {
        enableApp: true,
        enableRoute: false,
        level: LogLevel.DEBUG,
        httpPost: false,
        blackLList: ['PropsBindDirective']
    }
};
export class ConfigUtils {
    /**
     * @return {?}
     */
    static getConfig() {
        return ConfigUtils.moduleConfig || ConfigUtils.appConfig;
    }
    /**
     * @param {?} customer
     * @return {?}
     */
    static configApp(customer) {
        ConfigUtils.appConfig = CommonsUtils.merge({}, DEFAULT_APP_CONFIG, customer);
        return ConfigUtils.appConfig;
    }
    /**
     * @param {?} moduleConf
     * @param {?} url
     * @return {?}
     */
    static configModule(moduleConf, url) {
        /** @type {?} */
        let config = moduleConf;
        if (url && url === ConfigUtils.lastUrl) {
            config = CommonsUtils.merge(ConfigUtils.lastConfig, moduleConf);
        }
        else {
            ConfigUtils.lastUrl = url;
            ConfigUtils.lastConfig = moduleConf;
        }
        ConfigUtils.moduleConfig = CommonsUtils.merge({}, ConfigUtils.appConfig, config);
        return ConfigUtils.moduleConfig;
    }
    /**
     * @param {?} namePath
     * @param {?=} defaultValue
     * @return {?}
     */
    static get(namePath, defaultValue) {
        return CommonsUtils.get(ConfigUtils.getConfig(), namePath, defaultValue);
    }
}
ConfigUtils.lastUrl = undefined;
ConfigUtils.lastConfig = {};
ConfigUtils.appConfig = Object.assign({}, DEFAULT_APP_CONFIG);
ConfigUtils.moduleConfig = undefined;
if (false) {
    /** @type {?} */
    ConfigUtils.lastUrl;
    /** @type {?} */
    ConfigUtils.lastConfig;
    /** @type {?} */
    ConfigUtils.appConfig;
    /** @type {?} */
    ConfigUtils.moduleConfig;
}
