// @er/primeng/esm5/lib/components/check-box/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { DEFAULT_CHECK_BOX_PROPS } from './type';
var PngCheckBoxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngCheckBoxComponent, _super);
    function PngCheckBoxComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.valueParser = _this.parseValue;
        _this.modelFormatter = _this.formatModel;
        _this.$defaultProps = DEFAULT_CHECK_BOX_PROPS;
        return _this;
    }
    Object.defineProperty(PngCheckBoxComponent.prototype, "chkValue", {
        get: /**
         * @return {?}
         */
        function () {
            return this._chkValue;
        },
        set: /**
         * @param {?} values
         * @return {?}
         */
        function (values) {
            var _this = this;
            this._chkValue = values;
            /** @type {?} */
            var data = this.data$.value;
            if (CommonsUtils.isArray(data)) {
                /** @type {?} */
                var objectValues_1 = [];
                values.forEach(function (value) {
                    /** @type {?} */
                    var matched = ((/** @type {?} */ (data))).filter(function (v) { return v[_this.dataKey] === value; });
                    if (matched.length === 1) {
                        objectValues_1.push(DataItemUtils.getItemValue(matched[0], _this.valueKey));
                    }
                });
                this.value = objectValues_1.length > 0 ? objectValues_1 : null;
            }
            else {
                this.value = values;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngCheckBoxComponent.prototype, "dataKey", {
        get: /**
         * @return {?}
         */
        function () {
            return CommonsUtils.get(this.$props, '$ext.dataItemProps.idKey', 'value');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngCheckBoxComponent.prototype, "valueKey", {
        get: /**
         * @return {?}
         */
        function () {
            return CommonsUtils.get(this.$props, '$ext.dataItemProps.valueKey', 'value');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} model
     * @return {?}
     */
    PngCheckBoxComponent.prototype.formatModel = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        var _this = this;
        if (this.$props && !this.$props.label && model) {
            /** @type {?} */
            var models = CommonsUtils.getArrayValue(model);
            this._chkValue = [];
            models.forEach(function (m) {
                if (CommonsUtils.isJson(m)) {
                    _this._chkValue.push(DataItemUtils.getItemValue(m, _this.dataKey));
                }
                else {
                    _this._chkValue.push(m);
                }
            });
            return models;
        }
        else {
            return model;
        }
    };
    PngCheckBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-check-box',
                    template: "\n    <ng-container *ngIf=\"$props.label;else multiTpl\">\n      <p-checkbox\n        class=\"m-2\"\n        [(ngModel)]=\"value\"\n        erPropsBind\n        [props]=\"$props\"\n        [context]=\"ctx\"\n        binary=\"true\"\n        (onChange)=\"updateTouch()\"\n      >\n      </p-checkbox>\n    </ng-container>\n    <ng-template #multiTpl>\n      <div class=\"d-flex pb-2\"\n           [ngClass]=\"$props.$ext['inline']===false?'flex-column':'flex-wrap'\"\n           [erApiData]=\"apiDataProps\"\n      >\n        <ng-container *ngFor=\"let option of data$|async|erDataItems:$props.$ext.dataItemProps;let i=index\">\n          <span class=\"d-flex align-items-center\">\n             <p-checkbox\n               class=\"m-2\"\n               [(ngModel)]=\"chkValue\"\n               [attr.index]=\"i\"\n               erPropsBind\n               [props]=\"$props\"\n               [context]=\"ctx\"\n               [value]=\"option.data|erValue:dataKey\"\n               (onChange)=\"updateTouch()\"\n             >\n             </p-checkbox>\n             <span class=\"mr-4\" [innerHTML]=\"option.label|erSafeHtml\"></span>\n          </span>\n        </ng-container>\n      </div>\n    </ng-template>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngCheckBoxComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngCheckBoxComponent;
}(DataAwareControlWrapperComponent));
export { PngCheckBoxComponent };
if (false) {
    /** @type {?} */
    PngCheckBoxComponent.prototype.valueParser;
    /** @type {?} */
    PngCheckBoxComponent.prototype.modelFormatter;
    /** @type {?} */
    PngCheckBoxComponent.prototype.$defaultProps;
    /**
     * @type {?}
     * @private
     */
    PngCheckBoxComponent.prototype._chkValue;
}
