// @er/core/esm5/lib/components/animate-number/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { Subject, timer } from 'rxjs';
import { mapTo, scan, startWith, switchMap, takeWhile } from 'rxjs/operators';
import { BaseComponent } from '../../bases/component';
var AnimateNumberComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AnimateNumberComponent, _super);
    function AnimateNumberComponent(cd) {
        var _this = _super.call(this) || this;
        _this.cd = cd;
        _this.currentNumber = 0;
        _this.fontSize = '.8em';
        _this.counter$ = new Subject();
        _this.$subscriptions = _this.counter$
            .pipe(switchMap(function (endRange) {
            return timer(1000, 200)
                .pipe(mapTo(_this.positiveOrNegative(endRange, _this.currentNumber)), startWith(_this.currentNumber), scan(function (acc, curr) { return acc + curr; }), takeWhile(_this.takeUntil(endRange, _this.currentNumber)));
        }))
            .subscribe(function (val) {
            _this.currentNumber = val;
            if (_this.currentNumber === _this.end) {
                _this.fontSize = '1em';
            }
            _this.cd.detectChanges();
        });
        return _this;
    }
    Object.defineProperty(AnimateNumberComponent.prototype, "end", {
        get: /**
         * @return {?}
         */
        function () {
            return this._end;
        },
        set: /**
         * @param {?} endRange
         * @return {?}
         */
        function (endRange) {
            this._end = endRange;
            this.currentNumber = endRange - CommonsUtils.getRandomNumber(20);
            this.counter$.next(endRange);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    AnimateNumberComponent.prototype.positiveOrNegative = /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    function (endRange, currentNumber) {
        return endRange > currentNumber ? 1 : -1;
    };
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    AnimateNumberComponent.prototype.takeUntil = /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    function (endRange, currentNumber) {
        return endRange > currentNumber
            ? function (val) { return val <= endRange; }
            : function (val) { return val >= endRange; };
    };
    AnimateNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'animate-number',
                    template: "\n    <span [ngStyle]=\"{'font-size': fontSize, 'font-weight': 'bold'}\">\n       {{currentNumber}}\n    </span>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    AnimateNumberComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    AnimateNumberComponent.propDecorators = {
        end: [{ type: Input }]
    };
    return AnimateNumberComponent;
}(BaseComponent));
export { AnimateNumberComponent };
if (false) {
    /** @type {?} */
    AnimateNumberComponent.prototype.currentNumber;
    /** @type {?} */
    AnimateNumberComponent.prototype.fontSize;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype.counter$;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype._end;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype.cd;
}
