// @er/primeng/esm2015/lib/components/select/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, forwardRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { Dropdown } from 'primeng/primeng';
import { DEFAULT_PNG_SELECT_PROPS } from './type';
export class PngSelectComponent extends DataAwareControlWrapperComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super();
        this.cd = cd;
        this.$defaultProps = DEFAULT_PNG_SELECT_PROPS;
        this.templateNames = {
            group: 'group',
            item: 'item'
        };
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        super.onPropsInit();
        if (!this.$props.group && this.$props.$ext && this.$props.$ext.dataItemProps && this.$props.$ext.dataItemProps.groupKey) {
            this.$props.group = true;
        }
    }
}
PngSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-select',
                template: `
    <p-dropdown
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      [erApiData]="apiDataProps"
      [options]="data$|async|erDataItems:$props?.$ext?.dataItemProps"
      (onBlur)="updateTouch()"
    >
      <ng-template let-group pTemplate="group">
        <ng-container *ngTemplateOutlet="template(templateNames.group,groupTpl);context:{$implicit:group}"></ng-container>
      </ng-template>
      <ng-template let-group #groupTpl>
        <span *ngIf="group" style="font-size: 1em"><i class="fa fa-group text-primary"></i> {{group.label}}</span>
      </ng-template>
      <ng-template let-item pTemplate="item">
        <er-data-item [customerTpl]="template(templateNames.item)"
                      [item]="item.data"
                      [props]="$props.$ext.dataItemProps">
        </er-data-item>
      </ng-template>
    </p-dropdown>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngSelectComponent),
                        multi: true
                    }],
                styles: ["/deep/ .ui-dropdown-item{border-bottom:1px solid #a8a8a8!important;font-size:.8em;color:#000!important}/deep/ .ui-dropdown-item:hover{background-color:transparent!important;font-weight:700;color:red!important;font-size:1em}"]
            }] }
];
/** @nocollapse */
PngSelectComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PngSelectComponent.propDecorators = {
    dropDown: [{ type: ViewChild, args: [Dropdown,] }]
};
if (false) {
    /** @type {?} */
    PngSelectComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSelectComponent.prototype.templateNames;
    /** @type {?} */
    PngSelectComponent.prototype.dropDown;
    /**
     * @type {?}
     * @private
     */
    PngSelectComponent.prototype.cd;
}
