// @er/formly/esm5/lib/base/field-type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { QueryList, ViewChild, ViewChildren, ViewContainerRef } from '@angular/core';
import { ComponentUtils } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { ValidationMessageDirective } from '@er/validate';
import { FieldType } from '@ngx-formly/core';
import { FormlyPropsBindDirective } from '../directives/props/directive';
/**
 * @abstract
 */
var ErFormlyFieldType = /** @class */ (function (_super) {
    tslib_1.__extends(ErFormlyFieldType, _super);
    function ErFormlyFieldType() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ErFormlyFieldType.prototype, "attrs", {
        get: /**
         * @return {?}
         */
        function () {
            return this.field['attrs'] || {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ErFormlyFieldType.prototype, "props", {
        get: /**
         * @return {?}
         */
        function () {
            return this.field['props'] || {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ErFormlyFieldType.prototype, "context", {
        get: /**
         * @return {?}
         */
        function () {
            return this.field['context'] || {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ErFormlyFieldType.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.formControl.value;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            this.formControl.setValue(val);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} key
     * @return {?}
     */
    ErFormlyFieldType.prototype.component = /**
     * @param {?=} key
     * @return {?}
     */
    function (key) {
        return this.form['components'] &&
            this.form['components'][key || this.field.key];
    };
    /**
     * @param {?=} key
     * @return {?}
     */
    ErFormlyFieldType.prototype.data$ = /**
     * @param {?=} key
     * @return {?}
     */
    function (key) {
        return this.form['components'] &&
            this.form['components'][key || this.field.key] &&
            this.form['components'][key || this.field.key].data$;
    };
    /**
     * @param {?=} key
     * @return {?}
     */
    ErFormlyFieldType.prototype.control = /**
     * @param {?=} key
     * @return {?}
     */
    function (key) {
        return this.form.get(key);
    };
    /**
     * @param {?} key
     * @return {?}
     */
    ErFormlyFieldType.prototype.fieldProps = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        /** @type {?} */
        var props = this.field['formProps'].fields.filter(function (prop) { return prop.key === key; });
        return (props && props.length === 1) ? props[0] : undefined;
    };
    /**
     * @return {?}
     */
    ErFormlyFieldType.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var _widgets = this.widgets.toArray();
        /** @type {?} */
        var component;
        if (_widgets.length === 1) {
            component = ComponentUtils.getHostComponentInst(_widgets[0]);
        }
        else if (_widgets.length > 1) {
            component = _widgets.map(function (widget) { return ComponentUtils.getHostComponentInst(widget); });
        }
        if (this.validateMessage) {
            this.validateMessage.showError = this.showError || true;
        }
        if (!this.form['components']) {
            this.form['components'] = {};
        }
        this.form['components'][this.field.key] = component;
    };
    /**
     * @return {?}
     */
    ErFormlyFieldType.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.field['props'] = this.field['props'] || {};
        if (this.defaultProps) {
            this.field['props'] = CommonsUtils.defaults({}, this.field['props'], this.defaultProps);
        }
        this.field['formlyField'] = this;
    };
    ErFormlyFieldType.propDecorators = {
        widgets: [{ type: ViewChildren, args: [FormlyPropsBindDirective, { read: ViewContainerRef },] }],
        validateMessage: [{ type: ViewChild, args: [ValidationMessageDirective,] }]
    };
    return ErFormlyFieldType;
}(FieldType));
export { ErFormlyFieldType };
if (false) {
    /** @type {?} */
    ErFormlyFieldType.prototype.defaultProps;
    /** @type {?} */
    ErFormlyFieldType.prototype.widgets;
    /** @type {?} */
    ErFormlyFieldType.prototype.validateMessage;
}
