// @er/share/esm2015/lib/image-editor/module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ErCoreModule } from '@er/core';
import { Ng2ImgMaxModule, Ng2ImgMaxService } from 'ng2-img-max';
import { ImageEditorComponent } from './component';
import { ImageToolsService } from './services/image-tools.service';
import { ImgCropService } from './services/img-crop.service';
import { ImgResizeExactService } from './services/img-resize-exact.service';
export { ImageToolsService } from './services/image-tools.service';
export { ImgCropService } from './services/img-crop.service';
export { ImgResizeExactService } from './services/img-resize-exact.service';
export class ErImageEditorModule {
}
ErImageEditorModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    ErCoreModule,
                    Ng2ImgMaxModule
                ],
                declarations: [
                    ImageEditorComponent
                ],
                providers: [
                    { provide: ImgResizeExactService, useClass: ImgResizeExactService },
                    { provide: ImgCropService, useClass: ImgCropService },
                    { provide: ImageToolsService, useClass: ImageToolsService },
                    { provide: Ng2ImgMaxService, useClass: Ng2ImgMaxService }
                ],
                exports: [
                    ImageEditorComponent
                ]
            },] }
];
