// @er/core/esm5/lib/directives/high-light/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input } from '@angular/core';
import { BaseDirective } from '../../bases/directive';
var HighlightDirective = /** @class */ (function (_super) {
    tslib_1.__extends(HighlightDirective, _super);
    function HighlightDirective(el) {
        var _this = _super.call(this) || this;
        _this.el = el;
        _this.searchTerm = undefined;
        _this.caseSensitive = true;
        _this.viewRendered = false;
        _this.WRAPPER_TOKEN = '==--==##';
        return _this;
    }
    Object.defineProperty(HighlightDirective.prototype, "caseSensitivity", {
        get: /**
         * @private
         * @return {?}
         */
        function () {
            return this.caseSensitive ? '' : 'i';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    HighlightDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.highlightSearchTerm();
    };
    /**
     * @return {?}
     */
    HighlightDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.highlightSearchTerm();
    };
    /**
     * @return {?}
     */
    HighlightDirective.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        this.viewRendered = true;
    };
    /**
     * @return {?}
     */
    HighlightDirective.prototype.highlightSearchTerm = /**
     * @return {?}
     */
    function () {
        if (!this.searchTerm) {
            if (this.viewRendered) {
                this.removePreviouslyMarkedTextInNode();
            }
            return;
        }
        if (this.el.nativeElement) {
            this.removePreviouslyMarkedTextInNode();
            this.markMatchedTextTokens(this.el.nativeElement);
        }
    };
    /**
     * @private
     * @param {?} htmlNode
     * @return {?}
     */
    HighlightDirective.prototype.markMatchedTextTokens = /**
     * @private
     * @param {?} htmlNode
     * @return {?}
     */
    function (htmlNode) {
        var _this = this;
        /** @type {?} */
        var _searchTerm = this.getSearchTerm();
        /** @type {?} */
        var searchRegex = new RegExp(_searchTerm, 'gmi');
        /** @type {?} */
        var _searchTermUniqueTokens = this.getUniqueTokenWrappedSearchTerm();
        /** @type {?} */
        var searchRegexUniqueTokens = new RegExp(_searchTermUniqueTokens, 'gmi');
        this.traverseHtmlElementsTree(htmlNode, function (e) {
            _this.traverseNodesInElement(htmlNode.childNodes, function (node) { return _this.wrapUniqueTokensAroundMatchedText(node, searchRegex); });
        });
        this.markMatchedTextAndRemoveUniqueTokens(htmlNode, searchRegexUniqueTokens);
    };
    /**
     * @private
     * @param {?} htmlNode
     * @param {?} searchRegex
     * @return {?}
     */
    HighlightDirective.prototype.markMatchedTextAndRemoveUniqueTokens = /**
     * @private
     * @param {?} htmlNode
     * @param {?} searchRegex
     * @return {?}
     */
    function (htmlNode, searchRegex) {
        var _this = this;
        if (htmlNode.innerHTML) {
            /** @type {?} */
            var innerHtml = htmlNode.innerHTML;
            /** @type {?} */
            var newHtml = innerHtml.replace(searchRegex, function (match) {
                /** @type {?} */
                var wrapperLength = _this.WRAPPER_TOKEN.length;
                /** @type {?} */
                var markedStr = match.substr(wrapperLength, match.length - (wrapperLength * 2));
                return "<mark>" + markedStr + "</mark>";
            });
            htmlNode.innerHTML = newHtml;
        }
    };
    /**
     * @private
     * @param {?} nodes
     * @param {?} visitCallback
     * @return {?}
     */
    HighlightDirective.prototype.traverseNodesInElement = /**
     * @private
     * @param {?} nodes
     * @param {?} visitCallback
     * @return {?}
     */
    function (nodes, visitCallback) {
        for (var i = 0; i < nodes.length; i++) {
            /** @type {?} */
            var node = nodes[i];
            if (node.nodeType === 3) {
                visitCallback(node);
            }
        }
    };
    /**
     * @private
     * @param {?} htmlNode
     * @param {?} searchRegex
     * @return {?}
     */
    HighlightDirective.prototype.wrapUniqueTokensAroundMatchedText = /**
     * @private
     * @param {?} htmlNode
     * @param {?} searchRegex
     * @return {?}
     */
    function (htmlNode, searchRegex) {
        /** @type {?} */
        var innerText = htmlNode.nodeValue;
        /** @type {?} */
        var newText = innerText.replace(searchRegex, this.WRAPPER_TOKEN + "$&" + this.WRAPPER_TOKEN);
        htmlNode.nodeValue = newText;
    };
    /**
     * @private
     * @param {?} currentNode
     * @param {?} visitCallback
     * @return {?}
     */
    HighlightDirective.prototype.traverseHtmlElementsTree = /**
     * @private
     * @param {?} currentNode
     * @param {?} visitCallback
     * @return {?}
     */
    function (currentNode, visitCallback) {
        if (currentNode) {
            visitCallback(currentNode);
        }
        for (var i = 0; i < currentNode.children.length; i++) {
            /** @type {?} */
            var childNode = currentNode.children[i];
            this.markMatchedTextTokens((/** @type {?} */ (childNode)));
        }
    };
    /**
     * @private
     * @return {?}
     */
    HighlightDirective.prototype.removePreviouslyMarkedTextInNode = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var node = this.el.nativeElement;
        /** @type {?} */
        var markingPattern = new RegExp('<mark>|<\/mark>', 'g');
        /** @type {?} */
        var cleanText = node.innerHTML.replace(markingPattern, '');
        node.innerHTML = cleanText;
    };
    // private highlightedNewTextInNode() {
    //   const node = this.el.nativeElement;
    //   const htmlNegativeLookaheadPattern = '(?![^<>]*>)';
    //   const searchTermPattern = this.getSearchTerm();
    //   const searchRegex = new RegExp(searchTermPattern + htmlNegativeLookaheadPattern, 'gmi');
    //   const markedText = node.innerHTML.replace(searchRegex, '<mark>$&</mark>');
    //   node.innerHTML = markedText;
    // }
    // private highlightedNewTextInNode() {
    //   const node = this.el.nativeElement;
    //   const htmlNegativeLookaheadPattern = '(?![^<>]*>)';
    //   const searchTermPattern = this.getSearchTerm();
    //   const searchRegex = new RegExp(searchTermPattern + htmlNegativeLookaheadPattern, 'gmi');
    //   const markedText = node.innerHTML.replace(searchRegex, '<mark>$&</mark>');
    //   node.innerHTML = markedText;
    // }
    /**
     * @private
     * @return {?}
     */
    HighlightDirective.prototype.getSearchTerm = 
    // private highlightedNewTextInNode() {
    //   const node = this.el.nativeElement;
    //   const htmlNegativeLookaheadPattern = '(?![^<>]*>)';
    //   const searchTermPattern = this.getSearchTerm();
    //   const searchRegex = new RegExp(searchTermPattern + htmlNegativeLookaheadPattern, 'gmi');
    //   const markedText = node.innerHTML.replace(searchRegex, '<mark>$&</mark>');
    //   node.innerHTML = markedText;
    // }
    /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var escapedSearchTerm = "" + this.escapeRegExp(this.searchTerm);
        /** @type {?} */
        var spaceToMultiMatchRegex = new RegExp(' ', 'gm');
        escapedSearchTerm = escapedSearchTerm.replace(spaceToMultiMatchRegex, '|');
        return escapedSearchTerm;
    };
    /**
     * @private
     * @return {?}
     */
    HighlightDirective.prototype.getUniqueTokenWrappedSearchTerm = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var escapedSearchTerm = this.escapeRegExp(this.searchTerm);
        /** @type {?} */
        var spaceToMultiMatchRegex = new RegExp(' ', 'gm');
        escapedSearchTerm = escapedSearchTerm.replace(spaceToMultiMatchRegex, this.WRAPPER_TOKEN + "|" + this.WRAPPER_TOKEN);
        escapedSearchTerm = "" + this.WRAPPER_TOKEN + escapedSearchTerm + this.WRAPPER_TOKEN;
        return escapedSearchTerm;
    };
    /**
     * @private
     * @param {?} str
     * @return {?}
     */
    HighlightDirective.prototype.escapeRegExp = /**
     * @private
     * @param {?} str
     * @return {?}
     */
    function (str) {
        return str.replace('/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g', '\\$&');
    };
    HighlightDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erHighlight]'
                },] }
    ];
    /** @nocollapse */
    HighlightDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    HighlightDirective.propDecorators = {
        searchTerm: [{ type: Input, args: ['erHighlight',] }],
        caseSensitive: [{ type: Input }]
    };
    return HighlightDirective;
}(BaseDirective));
export { HighlightDirective };
if (false) {
    /** @type {?} */
    HighlightDirective.prototype.searchTerm;
    /** @type {?} */
    HighlightDirective.prototype.caseSensitive;
    /**
     * @type {?}
     * @private
     */
    HighlightDirective.prototype.viewRendered;
    /**
     * @type {?}
     * @private
     */
    HighlightDirective.prototype.WRAPPER_TOKEN;
    /**
     * @type {?}
     * @private
     */
    HighlightDirective.prototype.el;
}
