// @er/core/esm5/lib/components/data-item/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { BaseComponent } from '../../bases/component';
var DataItemComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DataItemComponent, _super);
    function DataItemComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(DataItemComponent.prototype, "labels", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var _label = DataItemUtils.getItemLabel(this.item, this.labelKeys);
            if (CommonsUtils.isArray(_label)) {
                return _label;
            }
            else {
                return [_label];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DataItemComponent.prototype, "labelKeys", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var keys = this.props.dropDownOptionKey || this.props.labelKey;
            if (!keys) {
                if (this.item.hasOwnProperty('label')) {
                    keys = 'label';
                }
                else {
                    keys = Object.keys(this.item).filter(function (key) { return !key.startsWith('_') && key !== 'id'; });
                }
            }
            return keys;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DataItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.props) {
            this.props = {};
        }
    };
    DataItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-data-item',
                    template: "\n    <ng-container *ngIf=\"props.template;else default\">\n      <er-dyna-content [content]=\"props.template\" [context]=\"item\"></er-dyna-content>\n    </ng-container>\n    <ng-template #default>\n      <div class=\"d-flex data-item\" [ngClass]=\"props.itemStyleClass || 'ui-clearfix'\">\n        <span class=\"label mr-3 border-right\" *ngFor=\"let label of labels\" [ngClass]=\"props.labelStyleClass\" [innerHTML]=\"label\"></span>\n        <span class=\"pull-right count\" *ngIf=\"props.showCounter && item['value']\">\n          <span [ngClass]=\"props.counterStyleClass || 'badge badge-pill badge-info'\">{{item['value']}}</span>\n        </span>\n      </div>\n    </ng-template>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["/deep/ .ui-autocomplete-list-item{border-bottom:1px solid #9f9f9f!important;font-size:.8em!important;color:#000!important}/deep/ .ui-autocomplete-list-item:hover{background-color:transparent!important;font-weight:700!important;color:red!important;font-size:1em!important}"]
                }] }
    ];
    DataItemComponent.propDecorators = {
        item: [{ type: Input }],
        props: [{ type: Input }]
    };
    return DataItemComponent;
}(BaseComponent));
export { DataItemComponent };
if (false) {
    /** @type {?} */
    DataItemComponent.prototype.item;
    /** @type {?} */
    DataItemComponent.prototype.props;
}
