// @er/primeng/esm2015/lib/components/table/component/column-selector/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '@er/core';
export class PngTableColumnSelectorComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.onColumnsChange = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changeColumns(event) {
        event.value.forEach(column => column.hidden = false);
        this.onColumnsChange.emit(event);
    }
}
PngTableColumnSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-table-column-selector',
                template: `

    <p-multiSelect [style]="{height: '30px', minWidth: '150px'}"
                   [ngClass]="styleClass"
                   [ngStyle]="style"
                   [options]="optionColumns"
                   [(ngModel)]="columns"
                   optionLabel="label"
                   [selectedItemsLabel]="'当前显示{0}('+optionColumns?.length+')个列'"
                   defaultLabel="选择显示字段"
                   (onChange)="changeColumns($event)"
    >
    </p-multiSelect>

  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PngTableColumnSelectorComponent.propDecorators = {
    columns: [{ type: Input }],
    optionColumns: [{ type: Input }],
    styleClass: [{ type: Input }],
    style: [{ type: Input }],
    onColumnsChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.columns;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.optionColumns;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.styleClass;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.style;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.onColumnsChange;
}
