// @er/core/esm2015/lib/utils/dyna/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Compiler, Component, ComponentFactoryResolver, NgModule } from '@angular/core';
import { CacheUtils } from '@er/utils';
import { DiUtils } from '../di';
/** @type {?} */
const cacheKey = '$component-factory';
export class DynaComponentUtils {
    /**
     * @param {?} hostViewContainerRef
     * @param {?} componentType
     * @return {?}
     */
    static createComponentByType(hostViewContainerRef, componentType) {
        if (!componentType) {
            return;
        }
        /** @type {?} */
        let componentFactory = CacheUtils.get(cacheKey, componentType);
        if (!componentFactory) {
            componentFactory = DynaComponentUtils.getComponentFactoryResolver().resolveComponentFactory(componentType);
            CacheUtils.put(cacheKey, componentType, componentFactory);
        }
        return DynaComponentUtils.createComponent(hostViewContainerRef, componentFactory);
    }
    /**
     * @param {?} hostViewContainerRef
     * @param {?} template
     * @return {?}
     */
    static createComponentByTemplate(hostViewContainerRef, template) {
        if (!template) {
            return;
        }
        /** @type {?} */
        let componentFactory = CacheUtils.get(cacheKey, template);
        if (!componentFactory) {
            /** @type {?} */
            const component = DynaComponentUtils.createDecoratedComponent(null, template);
            /** @type {?} */
            const module = DynaComponentUtils.createNgModule(component);
            componentFactory = module.componentFactories.find(factory => factory.componentType === component);
            CacheUtils.put(cacheKey, template, componentFactory);
        }
        return DynaComponentUtils.createComponent(hostViewContainerRef, componentFactory);
    }
    /**
     * @private
     * @return {?}
     */
    static getComponentFactoryResolver() {
        if (!DynaComponentUtils._componentFactoryResolver) {
            DynaComponentUtils._componentFactoryResolver = DiUtils.get(ComponentFactoryResolver);
        }
        return DynaComponentUtils._componentFactoryResolver;
    }
    /**
     * @private
     * @param {?} useClass
     * @param {?} template
     * @return {?}
     */
    static createDecoratedComponent(useClass, template) {
        /** @type {?} */
        const componentClass = useClass || class {
        };
        return Component({
            template: template
        })(componentClass);
    }
    /**
     * @private
     * @param {?} component
     * @return {?}
     */
    static createNgModule(component) {
        /** @type {?} */
        const ngModule = NgModule({ declarations: [component] })(class {
        });
        return DiUtils.get(Compiler).compileModuleAndAllComponentsSync(ngModule);
    }
    /**
     * @private
     * @param {?} hostViewContainerRef
     * @param {?} componentFactory
     * @return {?}
     */
    static createComponent(hostViewContainerRef, componentFactory) {
        hostViewContainerRef.clear();
        /** @type {?} */
        const componentRef = hostViewContainerRef.createComponent(componentFactory);
        return componentRef;
    }
}
DynaComponentUtils._componentFactoryResolver = null;
if (false) {
    /**
     * @type {?}
     * @private
     */
    DynaComponentUtils._componentFactoryResolver;
}
