// @er/primeng/esm2015/lib/components/popup/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '@er/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { of } from 'rxjs';
import { filter, map, takeUntil } from 'rxjs/operators';
import { DEFAULT_POPUP_DIALOG_PROPS } from './defaults';
export class PngPopupDialogComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.onClose = new EventEmitter();
        this.onAction = new EventEmitter();
        this.templateNames = {
            header: 'header',
            content: 'content',
            footer: 'footer'
        };
    }
    /**
     * @param {?} options
     * @return {?}
     */
    set options(options) {
        this.dlgOptions$ = of(this.resolveDlgOptions(options));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.dlgOptions$ = StatesUtils.observe(StateNames.popup)
            .pipe(filter(state => !!state), map(options => this.resolveDlgOptions(options)), takeUntil(this.destroyed$));
    }
    /**
     * @param {?} options
     * @return {?}
     */
    resolveDlgOptions(options) {
        /** @type {?} */
        const opt = Object.assign({}, DEFAULT_POPUP_DIALOG_PROPS, { header: this.header, content: this.content }, options);
        if (options.contentMaxHeight || options.contentMaxWidth) {
            opt.contentStyle = Object.assign({}, opt.contentStyle || {}, { 'max-height': `${options.contentMaxHeight || 3000}px`, 'max-width': `${options.contentMaxWidth || 3000}px` });
        }
        return opt;
    }
    /**
     * @param {?} dlg
     * @param {?} options
     * @return {?}
     */
    showEvent(dlg, options) {
        super.emitEvent('onShow', {
            scope: options,
            event: dlg,
            context: options.context
        });
    }
    /**
     * @param {?} dlg
     * @param {?} options
     * @return {?}
     */
    closeDlg(dlg, options) {
        if (this.canClose && !this.canClose({
            scope: options,
            event: dlg,
            context: options.context
        })) {
            return;
        }
        StatesUtils.create(StateNames.popup, { visible: false });
    }
}
PngPopupDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-popup-dialog',
                template: "<ng-container *ngIf=\"dlgOptions$|async as dlgOptions\">\n  <p-dialog #pngDialog\n            [(visible)]=\"dlgOptions.visible\"\n            [minWidth]=\"dlgOptions.minWidth\"\n            [minHeight]=\"dlgOptions.minHeight\"\n            [width]=\"dlgOptions.width\"\n            [height]=\"dlgOptions.height\"\n            [modal]=\"dlgOptions.modal\"\n            [draggable]=\"dlgOptions.draggable\"\n            [resizable]=\"dlgOptions.resizable\"\n            [closeOnEscape]=\"dlgOptions.closeOnEscape\"\n            [contentStyle]=\"dlgOptions.contentStyle\"\n            [dismissableMask]=\"dlgOptions.dismissableMask\"\n            [rtl]=\"dlgOptions.rtl\"\n            [closable]=\"dlgOptions.closable\"\n            [responsive]=\"dlgOptions.responsive\"\n            [breakpoint]=\"dlgOptions.breakpoint\"\n            [appendTo]=\"dlgOptions.appendTo\"\n            [maximizable]=\"dlgOptions.maximizable\"\n            [blockScroll]=\"dlgOptions.blockScroll\"\n            [style]=\"dlgOptions.style\"\n            [styleClass]=\"dlgOptions.styleClass\"\n            [showHeader]=\"dlgOptions.showHeader\"\n            [positionLeft]=\"dlgOptions.positionLeft\"\n            [positionTop]=\"dlgOptions.positionTop\"\n            (onShow)=\"showEvent(pngDialog,dlgOptions)\"\n  >\n    <p-header>\n      <ng-container *ngIf=\"templates[templateNames.header]\">\n        <ng-container *ngTemplateOutlet=\"templates[templateNames.header]\"></ng-container>\n      </ng-container>\n      <span [innerHTML]=\"dlgOptions.header\"></span>\n    </p-header>\n\n    <ng-container *ngTemplateOutlet=\"templates[templateNames.content]\"></ng-container>\n\n    <ng-container *ngIf=\"dlgOptions.content\">\n      <er-dyna-content [content]=\"dlgOptions.content\"></er-dyna-content>\n    </ng-container>\n\n    <ng-content></ng-content>\n\n    <p-footer *ngIf=\"dlgOptions.showFooter\">\n      <ng-container *ngIf=\"templates[templateNames.footer]\">\n        <ng-container *ngTemplateOutlet=\"templates[templateNames.footer]\"></ng-container>\n      </ng-container>\n      <ng-container *ngIf=\"!templates[templateNames.footer]\">\n        <png-buttons [btns]=\"dlgOptions.buttons\" [context]=\"dlgOptions.context\"></png-buttons>\n        <button pButton type=\"button\" class=\"ui-button-rounded ui-button-danger\" iconPos=\"left\"\n                (click)=\"closeDlg(pngDialog,dlgOptions)\"\n                [icon]=\"dlgOptions.closeIcon\" [label]=\"dlgOptions.closeLabel\"></button>\n      </ng-container>\n    </p-footer>\n  </p-dialog>\n</ng-container>\n\n\n\n\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host::ng-deep .ui-dialog-titlebar{background-color:#0d7a8b!important;color:#fff!important}"]
            }] }
];
PngPopupDialogComponent.propDecorators = {
    header: [{ type: Input }],
    content: [{ type: Input }],
    canClose: [{ type: Input }],
    onClose: [{ type: Output }],
    onAction: [{ type: Output }],
    options: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngPopupDialogComponent.prototype.header;
    /** @type {?} */
    PngPopupDialogComponent.prototype.content;
    /** @type {?} */
    PngPopupDialogComponent.prototype.canClose;
    /** @type {?} */
    PngPopupDialogComponent.prototype.onClose;
    /** @type {?} */
    PngPopupDialogComponent.prototype.onAction;
    /** @type {?} */
    PngPopupDialogComponent.prototype.dlgOptions$;
    /** @type {?} */
    PngPopupDialogComponent.prototype.templateNames;
}
