// @er/core/esm2015/lib/interceptors/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpAppConfiguredInterceptor } from './customer/interceptor';
import { HttpErrorHandlerInterceptor } from './error/interceptor';
import { HttpJwtAuthInterceptor } from './jwt/interceptor';
import { HttpLogInterceptor } from './log/interceptor';
import { HttpProgressInterceptor } from './progress/interceptor';
import { HttpTimeoutInterceptor } from './timeout/interceptor';
export { HttpAppConfiguredInterceptor } from './customer/interceptor';
export { HttpErrorHandlerInterceptor } from './error/interceptor';
export { HttpJwtAuthInterceptor } from './jwt/interceptor';
export { HttpProgressInterceptor } from './progress/interceptor';
export { HttpLogInterceptor } from './log/interceptor';
export { HttpTimeoutInterceptor } from './timeout/interceptor';
export { esBoundInterceptor, credentialHttpRequestInterceptor } from './customer/preset/all';
/** @type {?} */
export const CoreInterceptors = [
    { provide: HTTP_INTERCEPTORS, useClass: HttpErrorHandlerInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: HttpProgressInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: HttpLogInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: HttpTimeoutInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: HttpJwtAuthInterceptor, multi: true },
    { provide: HTTP_INTERCEPTORS, useClass: HttpAppConfiguredInterceptor, multi: true }
];
