// @er/primeng/esm5/lib/components/table/component/body-cell/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, HostListener, Input, TemplateRef } from '@angular/core';
import { BaseComponent } from '@er/core';
import { PngTableComponent } from '../table/component';
var PngTableBodyCellComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableBodyCellComponent, _super);
    function PngTableBodyCellComponent(pngTable) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        return _this;
    }
    Object.defineProperty(PngTableBodyCellComponent.prototype, "width", {
        get: /**
         * @return {?}
         */
        function () {
            return this.column.width;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngTableBodyCellComponent.prototype, "height", {
        get: /**
         * @return {?}
         */
        function () {
            return this.column.height;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} event
     * @return {?}
     */
    PngTableBodyCellComponent.prototype.cellContext = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        return {
            $event: event,
            $id: this.pngTable.$props.$id,
            column: this.column,
            row: this.row,
            rowIndex: this.rowIndex,
            pngTable: this.pngTable
        };
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngTableBodyCellComponent.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.column.onMouseEnter) {
            this.column.onMouseEnter.call(this.pngTable, this.cellContext(event));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngTableBodyCellComponent.prototype.onMouseOver = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.column.onMouseOver) {
            this.column.onMouseOver.call(this.pngTable, this.cellContext(event));
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngTableBodyCellComponent.prototype.onMouseOut = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.column.onMouseOut) {
            this.column.onMouseOut.call(this.pngTable, this.cellContext(event));
        }
    };
    PngTableBodyCellComponent.decorators = [
        { type: Component, args: [{
                    selector: '[png-table-body-cell]',
                    template: "\n    <ng-container *ngIf=\"customerTpl;else default\">\n      <ng-container *ngTemplateOutlet=\"customerTpl;context:cellContext\"></ng-container>\n    </ng-container>\n    <ng-template #default>\n      <png-table-body-cell-content\n        [customerTpl]=\"dataCellTpl\"\n        [column]=\"column\"\n        [row]=\"row\"\n        [rowIndex]=\"rowIndex\"\n      ></png-table-body-cell-content>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngTableBodyCellComponent.ctorParameters = function () { return [
        { type: PngTableComponent }
    ]; };
    PngTableBodyCellComponent.propDecorators = {
        column: [{ type: Input }],
        row: [{ type: Input }],
        rowIndex: [{ type: Input }],
        dataCellTpl: [{ type: Input }],
        width: [{ type: HostBinding, args: ['style.width',] }],
        height: [{ type: HostBinding, args: ['style.height',] }],
        onMouseEnter: [{ type: HostListener, args: ['mouseenter', ['$event'],] }],
        onMouseOver: [{ type: HostListener, args: ['mouseover', ['$event'],] }],
        onMouseOut: [{ type: HostListener, args: ['mouseout', ['$event'],] }]
    };
    return PngTableBodyCellComponent;
}(BaseComponent));
export { PngTableBodyCellComponent };
if (false) {
    /** @type {?} */
    PngTableBodyCellComponent.prototype.column;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.row;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.rowIndex;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.dataCellTpl;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.pngTable;
}
