// @er/types/esm2015/lib/life-cycle/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const LifeCycleNames = {
    onInit: 'onInit',
    onChanges: 'onChanges',
    doCheck: 'doCheck',
    afterContentInit: 'afterContentInit',
    afterContentChecked: 'afterContentChecked',
    afterViewInit: 'afterViewInit',
    afterViewChecked: 'afterViewChecked',
    onDestroy: 'onDestroy',
};
export { LifeCycleNames };
/**
 * @record
 */
export function LifeCycles() { }
if (false) {
    /** @type {?|undefined} */
    LifeCycles.prototype.onInit;
    /** @type {?|undefined} */
    LifeCycles.prototype.onChanges;
    /** @type {?|undefined} */
    LifeCycles.prototype.doCheck;
    /** @type {?|undefined} */
    LifeCycles.prototype.afterContentInit;
    /** @type {?|undefined} */
    LifeCycles.prototype.afterContentChecked;
    /** @type {?|undefined} */
    LifeCycles.prototype.afterViewInit;
    /** @type {?|undefined} */
    LifeCycles.prototype.afterViewChecked;
    /** @type {?|undefined} */
    LifeCycles.prototype.onDestroy;
}
