// @er/core/esm2015/lib/directives/focus-high-light/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { BaseDirective } from '../../bases/directive';
export class FocusHighlightDirective extends BaseDirective {
    /**
     * @param {?} el
     */
    constructor(el) {
        super();
        this.el = el;
        this.defaultBgColor = 'yellow';
    }
    /**
     * @return {?}
     */
    onMouseEnter() {
        this.highlight(this.bgColor || this.defaultBgColor);
    }
    /**
     * @return {?}
     */
    onMouseLeave() {
        this.highlight(null);
    }
    /**
     * @private
     * @param {?} color
     * @return {?}
     */
    highlight(color) {
        this.el.nativeElement.style.backgroundColor = color;
    }
}
FocusHighlightDirective.decorators = [
    { type: Directive, args: [{ selector: '[erFocusHighlight]' },] }
];
/** @nocollapse */
FocusHighlightDirective.ctorParameters = () => [
    { type: ElementRef }
];
FocusHighlightDirective.propDecorators = {
    bgColor: [{ type: Input }],
    onMouseEnter: [{ type: HostListener, args: ['mouseenter',] }],
    onMouseLeave: [{ type: HostListener, args: ['mouseleave',] }]
};
if (false) {
    /** @type {?} */
    FocusHighlightDirective.prototype.bgColor;
    /** @type {?} */
    FocusHighlightDirective.prototype.defaultBgColor;
    /**
     * @type {?}
     * @private
     */
    FocusHighlightDirective.prototype.el;
}
