// @er/formly/esm5/lib/utils/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DataUtils } from '@er/core';
import { StatesUtils, TreeUtils } from '@er/utils';
import { distinctUntilChanged, map, takeUntil } from 'rxjs/operators';
var FormlyUtils = /** @class */ (function () {
    function FormlyUtils() {
    }
    /**
     * @param {?} ctx
     * @param {?} entity
     * @param {?=} nextSeqProps
     * @return {?}
     */
    FormlyUtils.resolveDistinctFieldValue = /**
     * @param {?} ctx
     * @param {?} entity
     * @param {?=} nextSeqProps
     * @return {?}
     */
    function (ctx, entity, nextSeqProps) {
        console.log(ctx);
        /** @type {?} */
        var fieldName = 'code';
        if (nextSeqProps) {
            fieldName = nextSeqProps.seqKey || 'code';
        }
        /** @type {?} */
        var currentValue = ctx.model[fieldName];
        if (!currentValue) {
            return DataUtils.getNextValue(tslib_1.__assign({}, nextSeqProps || {}, { apiEntry: entity.apiEntry })).pipe(map(function (result) {
                ctx.model[fieldName] = result;
                return true;
            }));
        }
        else {
            return DataUtils.checkDistinct({
                apiProps: {
                    apiEntry: entity.apiEntry,
                    query: nextSeqProps && nextSeqProps.query,
                    searchFields: fieldName,
                    queryValue: currentValue
                },
                id: ctx.model['id']
            });
        }
    };
    /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @param {?=} valueResolver
     * @return {?}
     */
    FormlyUtils.initFieldValueFromSelectedTreeNode = /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @param {?=} valueResolver
     * @return {?}
     */
    function (formlyForm, fieldName, valueResolver) {
        var _a;
        /** @type {?} */
        var fieldValue = formlyForm.model && formlyForm.model[fieldName];
        if (!fieldValue) {
            /** @type {?} */
            var node = TreeUtils.getSelectedNode(formlyForm && formlyForm['tableId']);
            if (node) {
                /** @type {?} */
                var val = void 0;
                if (valueResolver) {
                    val = valueResolver(node);
                }
                else {
                    val = { code: node['code'], name: node['name'] };
                }
                formlyForm.patchModel((_a = {}, _a[fieldName] = val, _a));
            }
        }
    };
    /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @return {?}
     */
    FormlyUtils.onFieldChange = /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @return {?}
     */
    function (formlyForm, fieldName) {
        return formlyForm.formControl(fieldName).valueChanges
            .pipe(takeUntil(formlyForm.destroyed$), distinctUntilChanged());
    };
    /**
     * @param {?} ctx
     * @return {?}
     */
    FormlyUtils.refreshContainerNavTree = /**
     * @param {?} ctx
     * @return {?}
     */
    function (ctx) {
        /** @type {?} */
        var tableId = ctx.formlyForm.tableId;
        if (tableId) {
            /** @type {?} */
            var table = tableId && StatesUtils.getValue(tableId);
            if (table && table.component) {
                /** @type {?} */
                var container = table.component.$container;
                container.refreshTree = true;
            }
        }
    };
    return FormlyUtils;
}());
export { FormlyUtils };
