// @er/core/esm5/lib/directives/json-viewer/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import JSONFormatter from 'json-formatter-js';
import { BehaviorSubject, Observable } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { BaseDirective } from '../../bases/directive';
import { ObjectDiff } from './diff';
export { ObjectDiff } from './diff';
var JsonViewerDirective = /** @class */ (function (_super) {
    tslib_1.__extends(JsonViewerDirective, _super);
    function JsonViewerDirective(el, cd) {
        var _this = _super.call(this) || this;
        _this.el = el;
        _this.cd = cd;
        _this.showCurrent = true;
        _this.openLevel = 2;
        _this.showDiff = false;
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    JsonViewerDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (changes.model) {
            this.model = changes.model.currentValue;
            if (this.model) {
                Object.keys(this.model).forEach(function (key) {
                    if (_this.model[key] instanceof BehaviorSubject) {
                        _this.model[key] = _this.model[key].value;
                    }
                    else if (_this.model[key] instanceof Observable) {
                        ((/** @type {?} */ (_this.model[key])))
                            .pipe(takeUntil(_this.destroyed$))
                            .subscribe(function (v) { return _this.model[key] = v; });
                    }
                    else if (_this.showDiff) {
                        try {
                            _this.diff = ObjectDiff.diff(changes.model.previousValue, changes.model.currentValue);
                        }
                        catch (e) {
                        }
                    }
                });
            }
            this.showJson();
        }
    };
    /**
     * @return {?}
     */
    JsonViewerDirective.prototype.showJson = /**
     * @return {?}
     */
    function () {
        this.el.nativeElement.innerHTML = '';
        if (this.showCurrent) {
            /** @type {?} */
            var formatter = new JSONFormatter(this.model || {}, this.openLevel, {
                hoverPreviewEnabled: true
            });
            try {
                /** @type {?} */
                var divEle = formatter.render();
                divEle.classList.add('json-viewer');
                this.el.nativeElement.appendChild(divEle);
            }
            catch (e) {
            }
        }
        if (this.showDiff) {
            /** @type {?} */
            var divEle = document.createElement('div');
            divEle.classList.add('diff-viewer');
            divEle.innerHTML = this.diff;
            this.el.nativeElement.appendChild(divEle);
        }
    };
    JsonViewerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erJsonViewer]'
                },] }
    ];
    /** @nocollapse */
    JsonViewerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ChangeDetectorRef }
    ]; };
    JsonViewerDirective.propDecorators = {
        model: [{ type: Input, args: ['erJsonViewer',] }],
        showCurrent: [{ type: Input, args: ['current',] }],
        openLevel: [{ type: Input }],
        showDiff: [{ type: Input, args: ['diff',] }]
    };
    return JsonViewerDirective;
}(BaseDirective));
export { JsonViewerDirective };
if (false) {
    /** @type {?} */
    JsonViewerDirective.prototype.model;
    /** @type {?} */
    JsonViewerDirective.prototype.showCurrent;
    /** @type {?} */
    JsonViewerDirective.prototype.openLevel;
    /** @type {?} */
    JsonViewerDirective.prototype.showDiff;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.diff;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.cd;
}
