// @er/formly/esm2015/lib/components/form/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, Input } from '@angular/core';
import { BaseFormComponent } from '@er/core';
import { Subject } from 'rxjs';
export class FormlyFormComponent extends BaseFormComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super(cd);
        this.cd = cd;
        this.options = {};
        this.valueChanges$ = new Subject();
    }
    /**
     * @return {?}
     */
    get formlyOptions() {
        return Object.assign({ fieldChanges: this.valueChanges$ }, this.options);
    }
    /**
     * @return {?}
     */
    get submitTips() {
        if (!this.form.valid) {
            /** @type {?} */
            const fields = this.getFormErrorFields();
            /** @type {?} */
            const labels = [];
            fields.forEach(field => {
                /** @type {?} */
                const formlyField = this.formlyField(field);
                if (formlyField) {
                    /** @type {?} */
                    const label = formlyField.templateOptions && formlyField.templateOptions.label || field;
                    labels.push(label.replace(':', ''));
                }
            });
            return `请正确写以下信息️：${labels.join(',')}`;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        /** @type {?} */
        const valueChanges$ = this.options.fieldChanges || this.valueChanges$;
        if (this.valueChanges) {
            this.$subscriptions = valueChanges$.subscribe(event => {
                this.valueChanges(Object.assign({}, event, { formlyForm: this }));
            });
        }
    }
    /**
     * @param {?=} other
     * @return {?}
     */
    getEventContext(other) {
        return super.getEventContext({
            fields: this.fields,
            formOptions: this.options,
            formProps: this.formProps,
            formlyForm: this
        });
    }
    /**
     * @param {?} key
     * @return {?}
     */
    formlyField(key) {
        /** @type {?} */
        const props = this.fields.filter(prop => prop.key === key);
        return (props && props.length === 1) ? props[0] : undefined;
    }
}
FormlyFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-formly-form',
                template: "<div [ngClass]=\"styleClass\" [ngStyle]=\"style\">\n  <ng-container *ngIf=\"header\">\n    <er-dyna-content [content]=\"header\"></er-dyna-content>\n  </ng-container>\n  <ng-container *ngIf=\"template(templateNames.header)\">\n    <ng-container *ngTemplateOutlet=\"template(templateNames.header);\n                                     context: getEventContext()\"></ng-container>\n  </ng-container>\n\n  <form [formGroup]=\"form\" [ngClass]=\"formStyleClass || 'mt-2'\" [ngStyle]=\"formStyle\" novalidate>\n    <formly-form [form]=\"form\"\n                 [options]=\"formlyOptions\"\n                 [(model)]=\"model\"\n                 [fields]=\"fields\"\n                 class=\"formly-form\">\n      <ng-content></ng-content>\n      <ng-container *ngTemplateOutlet=\"template(templateNames.buttons,default)\"></ng-container>\n    </formly-form>\n  </form>\n  <ng-container *ngIf=\"footer\">\n    <er-dyna-content [content]=\"footer\"></er-dyna-content>\n  </ng-container>\n  <ng-container *ngIf=\"template(templateNames.footer)\">\n    <ng-container *ngTemplateOutlet=\"template(templateNames.footer);\n                                     context: getEventContext()\"></ng-container>\n  </ng-container>\n</div>\n\n<ng-template #default>\n  <div [ngClass]=\"buttonsStyleClass\">\n    <er-button [stateKey]=\"formId\"\n               icon=\"fa fa-save\"\n               label=\"\u63D0&nbsp;&nbsp;&nbsp;\u4EA4\"\n               title=\"\u4FDD\u7559\u6570\u636E\u5230\u670D\u52A1\u7AEF\"\n               class=\"mr-2\"\n               type=\"button\"\n               [pTooltip]=\"submitTips\"\n               tooltipPosition=\"top\"\n               [tooltipDisabled]=\"form.valid\"\n               [disabled]=\"!form.valid\"\n               (onClick)=\"submit()\"\n               [styleClass]=\"submitBtnStyleClass\"\n    ></er-button>\n    <er-button icon=\"fa fa-circle-o\"\n               class=\"mr-2\"\n               label=\"\u91CD&nbsp;&nbsp;&nbsp;\u586B\"\n               title=\"\u6062\u590D\u5230\u4FEE\u6539\u524D\u7684\u72B6\u6001\"\n               [styleClass]=\"resetBtnStyleClass\"\n               (onClick)=\"resetModel()\">\n    </er-button>\n\n    <ng-container *ngFor=\"let btn of buttons\">\n      <er-button class=\"mr-2\" erPropsBind [props]=\"btn\" [context]=\"getEventContext(btn.context)\"></er-button>\n    </ng-container>\n\n  </div>\n</ng-template>\n"
            }] }
];
/** @nocollapse */
FormlyFormComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
FormlyFormComponent.propDecorators = {
    fields: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    FormlyFormComponent.prototype.fields;
    /** @type {?} */
    FormlyFormComponent.prototype.options;
    /** @type {?} */
    FormlyFormComponent.prototype.formProps;
    /** @type {?} */
    FormlyFormComponent.prototype.valueChanges$;
    /**
     * @type {?}
     * @protected
     */
    FormlyFormComponent.prototype.cd;
}
