// @er/core/esm5/lib/directives/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AnimatedIfDirective } from './animate-if/directive';
import { AnimateDirective } from './animate/directive';
import { ApiDataDirective } from './api-data/directive';
import { DynaTemplateDirective } from './dyna-template/directive';
import { FocusHighlightDirective } from './focus-high-light/directive';
import { HighlightDirective } from './high-light/directive';
import { JsonViewerDirective } from './json-viewer/directive';
import { MovableHandleDirective } from './movable/movable-handle.directive';
import { MovableDirective } from './movable/movable.directive';
import { PositionDirective } from './position/directive';
import { PropsBindDirective } from './props-bind/directive';
import { RemoveMeDirective } from './remove-me/directive';
import { SelectorDirective } from './selector/directive';
import { HttpSpinnerDirective } from './state-spinner/http/directive';
import { NotifySpinnerDirective } from './state-spinner/notify/directive';
import { TagDirective } from './text-tag/directive';
import { TypedTemplateDirective } from './typed-template/directive';
export { AnimateDirective } from './animate/directive';
export { ApiDataDirective } from './api-data/directive';
export { TypedTemplateDirective } from './typed-template/directive';
export { DynaTemplateDirective } from './dyna-template/directive';
export { PropsBindDirective } from './props-bind/directive';
export { TagDirective } from './text-tag/directive';
export { SelectorDirective } from './selector/directive';
export { PositionDirective } from './position/directive';
export { RemoveMeDirective } from './remove-me/directive';
export { HttpSpinnerDirective } from './state-spinner/http/directive';
export { NotifySpinnerDirective } from './state-spinner/notify/directive';
export { HighlightDirective } from './high-light/directive';
export { FocusHighlightDirective } from './focus-high-light/directive';
export { MovableHandleDirective } from './movable/movable-handle.directive';
export { Position, MovableDirective } from './movable/movable.directive';
export { JsonViewerDirective, ObjectDiff } from './json-viewer/directive';
/** @type {?} */
export var CoreDirectives = [
    AnimateDirective,
    AnimatedIfDirective,
    ApiDataDirective,
    TypedTemplateDirective,
    DynaTemplateDirective,
    TagDirective,
    SelectorDirective,
    PositionDirective,
    RemoveMeDirective,
    PropsBindDirective,
    MovableDirective,
    MovableHandleDirective,
    JsonViewerDirective,
    HttpSpinnerDirective,
    NotifySpinnerDirective,
    HighlightDirective,
    FocusHighlightDirective
];
