// @er/formly/esm2015/lib/utils/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataUtils } from '@er/core';
import { StatesUtils, TreeUtils } from '@er/utils';
import { distinctUntilChanged, map, takeUntil } from 'rxjs/operators';
export class FormlyUtils {
    /**
     * @param {?} ctx
     * @param {?} entity
     * @param {?=} nextSeqProps
     * @return {?}
     */
    static resolveDistinctFieldValue(ctx, entity, nextSeqProps) {
        console.log(ctx);
        /** @type {?} */
        let fieldName = 'code';
        if (nextSeqProps) {
            fieldName = nextSeqProps.seqKey || 'code';
        }
        /** @type {?} */
        const currentValue = ctx.model[fieldName];
        if (!currentValue) {
            return DataUtils.getNextValue(Object.assign({}, nextSeqProps || {}, { apiEntry: entity.apiEntry })).pipe(map(result => {
                ctx.model[fieldName] = result;
                return true;
            }));
        }
        else {
            return DataUtils.checkDistinct({
                apiProps: {
                    apiEntry: entity.apiEntry,
                    query: nextSeqProps && nextSeqProps.query,
                    searchFields: fieldName,
                    queryValue: currentValue
                },
                id: ctx.model['id']
            });
        }
    }
    /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @param {?=} valueResolver
     * @return {?}
     */
    static initFieldValueFromSelectedTreeNode(formlyForm, fieldName, valueResolver) {
        /** @type {?} */
        const fieldValue = formlyForm.model && formlyForm.model[fieldName];
        if (!fieldValue) {
            /** @type {?} */
            const node = TreeUtils.getSelectedNode(formlyForm && formlyForm['tableId']);
            if (node) {
                /** @type {?} */
                let val;
                if (valueResolver) {
                    val = valueResolver(node);
                }
                else {
                    val = { code: node['code'], name: node['name'] };
                }
                formlyForm.patchModel({ [fieldName]: val });
            }
        }
    }
    /**
     * @param {?} formlyForm
     * @param {?} fieldName
     * @return {?}
     */
    static onFieldChange(formlyForm, fieldName) {
        return formlyForm.formControl(fieldName).valueChanges
            .pipe(takeUntil(formlyForm.destroyed$), distinctUntilChanged());
    }
    /**
     * @param {?} ctx
     * @return {?}
     */
    static refreshContainerNavTree(ctx) {
        /** @type {?} */
        const tableId = ctx.formlyForm.tableId;
        if (tableId) {
            /** @type {?} */
            const table = tableId && StatesUtils.getValue(tableId);
            if (table && table.component) {
                /** @type {?} */
                const container = table.component.$container;
                container.refreshTree = true;
            }
        }
    }
}
