// @er/primeng/esm5/lib/components/notify.announce/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { BaseComponent } from '@er/core';
import { NotifyLevel, NotifyType, StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { MessageService } from 'primeng/api';
import { filter } from 'rxjs/operators';
/** @type {?} */
export var DEFAULT_ANNOUNCE_PROPS = {
    level: NotifyLevel.INFO,
    animation: true,
    closable: true,
    baseZIndex: 2000
};
var PngAnnounceComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngAnnounceComponent, _super);
    function PngAnnounceComponent(messageService, cd) {
        var _this = _super.call(this) || this;
        _this.messageService = messageService;
        _this.cd = cd;
        _this.announce = tslib_1.__assign({}, DEFAULT_ANNOUNCE_PROPS, ConfigUtils.getConfig().announce || {});
        _this.defaultNotifyType = ConfigUtils.getConfig().notify;
        return _this;
    }
    /**
     * @return {?}
     */
    PngAnnounceComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$subscriptions = StatesUtils.observe(StateNames.notify)
            .pipe(filter(function (notify) { return notify && notify.message; }), filter(function (notify) { return notify.type === NotifyType.ANNOUNCE || (!notify.type && _this.defaultNotifyType === NotifyType.ANNOUNCE); })).subscribe(function (notify) {
            /** @type {?} */
            var model = _this.getNotifyModel(notify);
            _this.messageService.add(model);
            _this.cd.detectChanges();
        });
    };
    /**
     * @param {?} notify
     * @return {?}
     */
    PngAnnounceComponent.prototype.getNotifyModel = /**
     * @param {?} notify
     * @return {?}
     */
    function (notify) {
        return {
            key: notify.key,
            severity: notify.level || NotifyLevel.INFO,
            summary: notify.title,
            detail: notify.message,
            closable: notify.closable,
            data: notify.data
        };
    };
    PngAnnounceComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-announce',
                    template: "\n    <p-messages [styleClass]=\"announce.styleClass\"\n                [style]=\"announce.style\"\n                [closable]=\"announce.closable\"\n    ></p-messages>\n  ",
                    providers: [MessageService],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngAnnounceComponent.ctorParameters = function () { return [
        { type: MessageService },
        { type: ChangeDetectorRef }
    ]; };
    return PngAnnounceComponent;
}(BaseComponent));
export { PngAnnounceComponent };
if (false) {
    /** @type {?} */
    PngAnnounceComponent.prototype.announce;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.defaultNotifyType;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    PngAnnounceComponent.prototype.cd;
}
