// @er/core/esm2015/lib/services/log/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LogLevel } from '@er/types';
import { CommonsUtils, ConfigUtils, DateUtils, DialogUtils } from '@er/utils';
import { HttpUtils } from '../../utils/http';
import * as i0 from "@angular/core";
export class LogService {
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    debug(from, message, ...context) {
        if (!ConfigUtils.getConfig().isProduction
            && ConfigUtils.getConfig().log.enableApp
            && ConfigUtils.getConfig().log.level <= LogLevel.DEBUG) {
            if ((typeof from) !== 'string' && from.constructor) {
                from = from.constructor.name;
            }
            if (ConfigUtils.getConfig().log.blackLList.indexOf(from) < 0) {
                this.getLogger().log(`[${DateUtils.getFormattedTime()}] %c DEBUG `, 'color: white; background: green; font-size: 18px;', this.getLogData([from, message, ...context]));
            }
        }
    }
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    info(from, message, ...context) {
        if (ConfigUtils.getConfig().log.level <= LogLevel.INFO) {
            if (typeof from !== 'string' && from.constructor) {
                from = from.constructor.name;
            }
            if (ConfigUtils.getConfig().log.blackLList.indexOf(from) < 0) {
                this.getLogger().log(`[${DateUtils.getFormattedTime()}] %c INFO `, 'color: white; background: blue; font-size: 18px;', this.getLogData([from, message, ...context]));
            }
        }
    }
    /**
     * @param {?} messages
     * @return {?}
     */
    warn(messages) {
        if (ConfigUtils.getConfig().log.level <= LogLevel.WARN) {
            this.getLogger().log(`[${DateUtils.getFormattedTime()}] %c WARN `, 'color: white; background: blue; font-size: 18px;', this.getLogData(messages));
        }
    }
    /**
     * @param {?} from
     * @param {?} message
     * @param {...?} context
     * @return {?}
     */
    error(from, message, ...context) {
        DialogUtils.error('错误', message);
        if (ConfigUtils.getConfig().log.level <= LogLevel.ERROR) {
            this.getLogger().log(`[${DateUtils.getFormattedTime()}] %c ERROR `, 'color: white; background: red; font-size: 18px;', this.getLogData([from, message, ...context]));
            if (ConfigUtils.getConfig().log.httpPost) {
                if (CommonsUtils.isFunction(from)) {
                    from = from.constructor.name;
                }
                this.getHttpLogger().log(DateUtils.getFormattedDateTime(), this.getLogData([from, message]));
            }
        }
    }
    /**
     * @protected
     * @return {?}
     */
    getLogger() {
        return console;
    }
    /**
     * @protected
     * @return {?}
     */
    getHttpLogger() {
        return (/** @type {?} */ ({
            log: (message, ...optionalParams) => {
                HttpUtils.request(Object.assign({}, ConfigUtils.getConfig().logHttpOptions, { body: Object.assign({ message }, optionalParams) }));
            }
        }));
    }
    /**
     * @private
     * @param {?} messages
     * @return {?}
     */
    getLogData(messages) {
        return Object.assign({}, messages);
    }
}
LogService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */ LogService.ngInjectableDef = i0.defineInjectable({ factory: function LogService_Factory() { return new LogService(); }, token: LogService, providedIn: "root" });
