// @er/utils/esm2015/lib/form/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormGroup } from '@angular/forms';
export class FormUtils {
    /**
     * @param {?} form
     * @param {...?} path
     * @return {?}
     */
    static getFormControl(form, ...path) {
        /** @type {?} */
        let _form = form;
        /** @type {?} */
        let control = null;
        if (path.length > 0) {
            path.forEach(p => {
                control = _form.controls[p];
                if (control instanceof FormGroup) {
                    _form = control;
                }
            });
        }
        return (/** @type {?} */ (control));
    }
    /**
     * @param {?} control
     * @param {?} value
     * @return {?}
     */
    static setValue(control, value) {
        control.setValue(value, {
            emitEvent: false,
            emitModelToViewChange: false,
            emitViewToModelChange: false
        });
    }
}
