// @er/presets/esm5/lib/api-servers/er.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ApiServerType, HttpMethod, HttpRequestPayloadType, Order } from '@er/types';
/**
 * @record
 */
export function ErHttpResponse() { }
if (false) {
    /** @type {?|undefined} */
    ErHttpResponse.prototype.success;
    /** @type {?|undefined} */
    ErHttpResponse.prototype.message;
    /** @type {?|undefined} */
    ErHttpResponse.prototype.error;
    /** @type {?|undefined} */
    ErHttpResponse.prototype.content;
    /** @type {?|undefined} */
    ErHttpResponse.prototype.contents;
}
/** @type {?} */
var paramKeys = {
    sort: '_sort',
    order: '_order',
    from: 'from',
    size: 'size',
    page: '_page',
    to: '_end'
};
/** @type {?} */
var responseResolver = function (response) { return response; };
var ɵ0 = responseResolver;
/** @type {?} */
var updateResponseResolver = function (response) {
    return response && response.hasOwnProperty('id')
        ? { success: true, result: response.id }
        : { failure: true, message: response };
};
var ɵ1 = updateResponseResolver;
/** @type {?} */
var deleteResponseResolver = function (response) {
    return response && Object.keys(response).length === 0
        ? { success: true, result: {} }
        : { failure: true, message: response };
};
var ɵ2 = deleteResponseResolver;
/** @type {?} */
export var ErServerApiRequestOptions = {
    serverType: ApiServerType.ER,
    paramsKeys: paramKeys,
    actions: {
        login: {
            uriResolver: function () { return 'login'; },
            method: HttpMethod.POST,
            notifySuccess: true,
            successMessage: '登录成功',
            notifyFailure: true,
            onResponse: responseResolver,
            onSuccess: (function (resp) {
                /** @type {?} */
                var content = resp.content;
                return content || {};
            })
        },
        logout: {
            uriResolver: function () { return 'logout'; },
            notifySuccess: false,
            onResponse: responseResolver
        },
        changePwd: {
            uriResolver: function () { return 'user/updatePwd'; },
            notifySuccess: false,
            onResponse: responseResolver
        },
        register: {
            uriResolver: function () { return 'user/reg'; },
            method: HttpMethod.POST,
            notifySuccess: true,
            successMessage: '注册成功'
        },
        getById: {
            uriResolver: function (apiEntry, payload) { return apiEntry + "/" + payload; },
            onResponse: function (response) {
                return tslib_1.__assign({ id: response['_id'] }, response['_source']);
            }
        },
        getByIds: {
            method: HttpMethod.GET,
            payloadType: HttpRequestPayloadType.PARAMS,
            payloadResolver: function (apiEntry, payload, options) {
                var _a;
                /** @type {?} */
                var _ids = payload;
                if (_ids.length === 0) {
                    throw new Error('id值为空');
                }
                /** @type {?} */
                var keys = options.paramsKeys;
                return _a = {},
                    _a[keys.ids] = _ids,
                    _a[keys.fields] = payload.fields,
                    _a;
            }
        },
        getByQuery: {
            method: HttpMethod.GET,
            payloadType: HttpRequestPayloadType.PARAMS,
            payloadResolver: function (apiEntry, payload, options) {
                /** @type {?} */
                var qParams = '';
                /** @type {?} */
                var requestPayload = payload || {};
                /** @type {?} */
                var query = requestPayload.query;
                if (query) {
                    if (query instanceof Array) {
                        // (query as Query[]).forEach(q => qParams += `${q.field}=${q.value}&`);
                    }
                    else {
                        /** @type {?} */
                        var q = (/** @type {?} */ (query));
                        // qParams = `${q.field}=${q.value}`;
                    }
                }
                var size = requestPayload.size, page = requestPayload.page;
                if (size || page) {
                    if (!qParams.endsWith('&')) {
                        qParams += '&';
                    }
                    qParams += options.paramsKeys.size + "=" + (size || 10) + "&" + options.paramsKeys.page + "=" + (page || 1);
                }
                /** @type {?} */
                var sort = requestPayload.sort;
                if (sort) {
                    if (!qParams.endsWith('&')) {
                        qParams += '&';
                    }
                    /** @type {?} */
                    var sortFields_1 = '';
                    /** @type {?} */
                    var sortOrders_1 = '';
                    if (typeof sort === 'string') {
                        qParams += options.paramsKeys.sort + "=" + sort + "&" + options.paramsKeys.order + "=" + Order.ASC;
                    }
                    else {
                        Object.keys(sort).map(function (field) {
                            sortFields_1 += field + ",";
                            sortOrders_1 += sort[field] + ",";
                        });
                        qParams += options.paramsKeys.sort + "=" + sortFields_1 + "&" + options.paramsKeys.order + "=" + sortOrders_1;
                    }
                }
                return qParams;
            },
            onResponse: function (response) { return response; }
        },
        create: {
            method: HttpMethod.POST,
            onResponse: updateResponseResolver,
            notifySuccess: false,
            alertSuccess: true,
            successMessage: '数据保存成功'
        },
        updateById: {
            method: HttpMethod.PUT,
            payloadType: HttpRequestPayloadType.BODY,
            uriResolver: function (apiEntry, payload) { return apiEntry + "/" + payload.id; },
            payloadResolver: function (apiEntry, payload) { return payload.data; },
            onResponse: updateResponseResolver,
            notifySuccess: false,
            alertSuccess: true,
            successMessage: '数据更新成功'
        },
        patchById: {
            method: HttpMethod.PATCH,
            payloadType: HttpRequestPayloadType.BODY,
            uriResolver: function (apiEntry, payload) { return apiEntry + "/" + payload.id; },
            payloadResolver: function (apiEntry, payload) { return payload; },
            onResponse: updateResponseResolver,
            notifySuccess: true,
            successMessage: '数据更新成功'
        },
        deleteById: {
            method: HttpMethod.DELETE,
            uriResolver: function (apiEntry, payload) { return apiEntry + "/" + payload; },
            onResponse: deleteResponseResolver,
            notifySuccess: true,
            successMessage: '数据删除成功'
        }
    }
};
export { ɵ0, ɵ1, ɵ2 };
