// @er/presets/esm5/lib/api-servers/mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var paramKeys = {
    sort: '_sort',
    order: '_order',
    size: '_limit',
    page: '_page',
    from: '_start',
    to: '_end'
};
/** @type {?} */
var getResponseResolver = function (response) {
    if (response) {
        if (response instanceof Array) {
            if (response.length === 1) {
                response = response[0];
            }
        }
        return { success: true, result: response };
    }
    else {
        return { success: false };
    }
};
var ɵ0 = getResponseResolver;
/** @type {?} */
var updateResponseResolver = function (response) {
    return response && response.hasOwnProperty('id')
        ? { success: true, result: response.id }
        : { failure: true, message: response };
};
var ɵ1 = updateResponseResolver;
/** @type {?} */
var deleteResponseResolver = function (response) {
    return response && Object.keys(response).length === 0
        ? { success: true, result: {} }
        : { failure: true, message: response };
};
var ɵ2 = deleteResponseResolver;
/** @type {?} */
export var MockServerApiRequestOptions = {};
export { ɵ0, ɵ1, ɵ2 };
