// @er/types/esm2015/lib/chart/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const ChartType = {
    PIE: 'pie',
    LINE: 'line',
    BAR: 'bar',
    RADAR: 'radar',
    AREA: 'polarArea',
    DOUGHNUT: 'doughnut',
};
export { ChartType };
/**
 * @record
 */
export function ChartDimension() { }
if (false) {
    /** @type {?|undefined} */
    ChartDimension.prototype.title;
    /** @type {?|undefined} */
    ChartDimension.prototype.dataSrcPath;
    /** @type {?|undefined} */
    ChartDimension.prototype.dataItemProps;
}
/**
 * @record
 */
export function ChartProps() { }
if (false) {
    /** @type {?|undefined} */
    ChartProps.prototype.type;
    /** @type {?|undefined} */
    ChartProps.prototype.data;
    /** @type {?|undefined} */
    ChartProps.prototype.rawData;
    /** @type {?|undefined} */
    ChartProps.prototype.options;
    /** @type {?|undefined} */
    ChartProps.prototype.caption;
    /** @type {?|undefined} */
    ChartProps.prototype.captionIcon;
    /** @type {?|undefined} */
    ChartProps.prototype.width;
    /** @type {?|undefined} */
    ChartProps.prototype.height;
    /** @type {?|undefined} */
    ChartProps.prototype.switchable;
    /** @type {?|undefined} */
    ChartProps.prototype.dataSource;
    /** @type {?|undefined} */
    ChartProps.prototype.dimensions;
    /** @type {?|undefined} */
    ChartProps.prototype.dataItemProps;
    /** @type {?|undefined} */
    ChartProps.prototype.resolveDataSource;
    /** @type {?|undefined} */
    ChartProps.prototype.resolveDimension;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function ChartDataProps() { }
if (false) {
    /** @type {?|undefined} */
    ChartDataProps.prototype.labels;
    /** @type {?|undefined} */
    ChartDataProps.prototype.datasets;
}
/**
 * @record
 */
export function ChartDataSetsProps() { }
if (false) {
    /** @type {?|undefined} */
    ChartDataSetsProps.prototype.label;
    /** @type {?|undefined} */
    ChartDataSetsProps.prototype.data;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function LineChartOptions() { }
if (false) {
    /** @type {?|undefined} */
    LineChartOptions.prototype.showLines;
    /** @type {?|undefined} */
    LineChartOptions.prototype.spanGaps;
}
/**
 * @record
 */
export function LineChartDataSetProps() { }
if (false) {
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.xAxisID;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.yAxisID;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderWidth;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderDash;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderDashOffset;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderCapStyle;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.borderJoinStyle;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.cubicInterpolationMode;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.fill;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.lineTension;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointBackgroundColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointBorderColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointBorderWidth;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointRadius;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointStyle;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointRotation;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointHitRadius;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointHoverBackgroundColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointHoverBorderColor;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointHoverBorderWidth;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.pointHoverRadius;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.showLine;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.spanGaps;
    /** @type {?|undefined} */
    LineChartDataSetProps.prototype.steppedLine;
}
/**
 * @record
 */
export function BarChartOptions() { }
if (false) {
    /** @type {?|undefined} */
    BarChartOptions.prototype.barPercentage;
    /** @type {?|undefined} */
    BarChartOptions.prototype.categoryPercentage;
    /** @type {?|undefined} */
    BarChartOptions.prototype.barThickness;
    /** @type {?|undefined} */
    BarChartOptions.prototype.maxBarThickness;
}
/**
 * @record
 */
export function BarChartDataSetProps() { }
if (false) {
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.xAxisID;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.yAxisID;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.borderColor;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.borderWidth;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.borderSkipped;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.hoverBackgroundColo;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.hoverBorderColor;
    /** @type {?|undefined} */
    BarChartDataSetProps.prototype.hoverBorderWidth;
}
/**
 * @record
 */
export function RadarChartDataSetProps() { }
if (false) {
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderWidth;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderDash;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderDashOffset;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderCapStyle;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.borderJoinStyle;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.fill;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.lineTension;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointBackgroundColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointBorderColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointBorderWidth;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointRadius;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointRotation;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointStyle;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointHitRadius;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointHoverBackgroundColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointHoverBorderColor;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointHoverBorderWidth;
    /** @type {?|undefined} */
    RadarChartDataSetProps.prototype.pointHoverRadius;
}
/**
 * @record
 */
export function PieChartOptions() { }
if (false) {
    /** @type {?|undefined} */
    PieChartOptions.prototype.cutoutPercentage;
    /** @type {?|undefined} */
    PieChartOptions.prototype.rotation;
    /** @type {?|undefined} */
    PieChartOptions.prototype.circumference;
}
/**
 * @record
 */
export function PieChartDataSetProps() { }
if (false) {
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.borderColor;
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.borderWidth;
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.hoverBackgroundColor;
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.hoverBorderColor;
    /** @type {?|undefined} */
    PieChartDataSetProps.prototype.hoverBorderWidth;
}
/**
 * @record
 */
export function PolarAreaDataSetProps() { }
if (false) {
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.backgroundColor;
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.borderColor;
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.borderWidth;
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.hoverBackgroundColor;
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.hoverBorderColor;
    /** @type {?|undefined} */
    PolarAreaDataSetProps.prototype.hoverBorderWidth;
}
/**
 * @record
 */
export function PolarAreaOptions() { }
if (false) {
    /** @type {?|undefined} */
    PolarAreaOptions.prototype.startAngle;
}
