// @er/types/esm2015/lib/notify/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function NotifyProp() { }
if (false) {
    /** @type {?|undefined} */
    NotifyProp.prototype.type;
    /** @type {?|undefined} */
    NotifyProp.prototype.key;
    /** @type {?|undefined} */
    NotifyProp.prototype.level;
    /** @type {?|undefined} */
    NotifyProp.prototype.animation;
    /** @type {?|undefined} */
    NotifyProp.prototype.title;
    /** @type {?|undefined} */
    NotifyProp.prototype.icon;
    /** @type {?|undefined} */
    NotifyProp.prototype.message;
    /** @type {?|undefined} */
    NotifyProp.prototype.style;
    /** @type {?|undefined} */
    NotifyProp.prototype.styleClass;
    /** @type {?|undefined} */
    NotifyProp.prototype.sticky;
    /** @type {?|undefined} */
    NotifyProp.prototype.duration;
    /** @type {?|undefined} */
    NotifyProp.prototype.modal;
    /** @type {?|undefined} */
    NotifyProp.prototype.position;
    /** @type {?|undefined} */
    NotifyProp.prototype.immutable;
    /** @type {?|undefined} */
    NotifyProp.prototype.closable;
    /** @type {?|undefined} */
    NotifyProp.prototype.baseZIndex;
    /** @type {?|undefined} */
    NotifyProp.prototype.autoZIndex;
    /** @type {?|undefined} */
    NotifyProp.prototype.data;
    /* Skipping unhandled member: [key: string]: any;*/
}
/** @enum {string} */
const NotifyType = {
    ALERT: 'alert',
    ANNOUNCE: 'announce',
};
export { NotifyType };
/** @enum {string} */
const NotifyLevel = {
    INFO: 'info',
    ERROR: 'error',
    SUCCESS: 'success',
    WARN: 'warn',
    QUESTION: 'question',
};
export { NotifyLevel };
/** @enum {string} */
const NotifyMethod = {
    APPEND: 'append',
    REPLACE: 'replace',
};
export { NotifyMethod };
