// @er/core/esm2015/lib/directives/dyna-template/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2, ViewContainerRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
import { DynaComponentUtils } from '../../utils/dyna';
export class DynaTemplateDirective extends BaseDirective {
    /**
     * @param {?} viewContainerRef
     * @param {?} el
     * @param {?} renderer
     */
    constructor(viewContainerRef, el, renderer) {
        super();
        this.viewContainerRef = viewContainerRef;
        this.el = el;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (CommonsUtils.isEmptyString(this.template)) {
            this.template = this.el.nativeElement.innerHTML;
        }
        if (CommonsUtils.isEmptyString(this.template)) {
            return;
        }
        this.template = this.template.replace(/\[\[/g, '{{');
        this.template = this.template.replace(/]]/g, '}}');
        this.componentRef = DynaComponentUtils.createComponentByTemplate(this.viewContainerRef, this.template);
        /** @type {?} */
        const parent = this.el.nativeElement.parentNode;
        this.renderer.removeChild(parent, this.el.nativeElement);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        super.ngOnDestroy();
        if (this.componentRef) {
            this.componentRef.destroy();
            this.componentRef = null;
        }
    }
}
DynaTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erDynaTemplate]',
            },] }
];
/** @nocollapse */
DynaTemplateDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ElementRef },
    { type: Renderer2 }
];
DynaTemplateDirective.propDecorators = {
    template: [{ type: Input, args: ['erDynaTemplate',] }],
    context: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DynaTemplateDirective.prototype.template;
    /** @type {?} */
    DynaTemplateDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.renderer;
}
