// @er/types/esm2015/lib/dyna-content/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Context() { }
/**
 * @record
 */
export function DynaDirective() { }
if (false) {
    /** @type {?|undefined} */
    DynaDirective.prototype.type;
    /** @type {?|undefined} */
    DynaDirective.prototype.inputs;
    /** @type {?|undefined} */
    DynaDirective.prototype.outputs;
}
/**
 * @record
 */
export function DynaComponent() { }
if (false) {
    /** @type {?|undefined} */
    DynaComponent.prototype.type;
    /** @type {?|undefined} */
    DynaComponent.prototype.props;
    /** @type {?|undefined} */
    DynaComponent.prototype.directives;
    /** @type {?|undefined} */
    DynaComponent.prototype.attrs;
}
/**
 * @record
 */
export function ContextedContent() { }
if (false) {
    /** @type {?|undefined} */
    ContextedContent.prototype.content;
    /** @type {?|undefined} */
    ContextedContent.prototype.context;
    /** @type {?|undefined} */
    ContextedContent.prototype.styleClass;
    /** @type {?|undefined} */
    ContextedContent.prototype.style;
}
/** @enum {string} */
const ContentType = {
    STRING: 'string',
    TEMPLATE_STR: 'templateStr',
    DATA: 'data',
    FUNCTION: 'function',
    TEMPLATE: 'template',
    COMPONENT: 'component',
};
export { ContentType };
