// @er/primeng/esm2015/lib/components/file-upload/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngFileUploadProps() { }
if (false) {
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.name;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.url;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.method;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.multiple;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.accept;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.auto;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.withCredentials;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.maxFileSize;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.invalidFileSizeMessageSummary;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.invalidFileSizeMessageDetail;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.invalidFileTypeMessageSummary;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.invalidFileTypeMessageDetail;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.previewWidth;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.chooseLabel;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.uploadLabel;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.cancelLabel;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.showUploadButton;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.showCancelButton;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.mode;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.customUpload;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.files;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onBeforeUpload;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onBeforeSend;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onUpload;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onError;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onClear;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onRemove;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onSelect;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.onProgress;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.uploadHandler;
    /** @type {?|undefined} */
    PngFileUploadProps.prototype.$ext;
}
/** @type {?} */
export const DEFAULT_PNG_FILE_UPLOAD_PROPS = (/** @type {?} */ ({
    accept: 'image/*',
    name: 'files',
    files: [],
    multiple: false,
    previewWidth: 50,
    invalidFileSizeMessageSummary: '"{0}: 文件大小无效, "',
    invalidFileSizeMessageDetail: '"最大上传文件大小 {0}."',
    invalidFileTypeMessageSummary: '"{0}: 无效的文件类型, "',
    invalidFileTypeMessageDetail: '"可上传的文件类型: {0}."',
    chooseLabel: '选择',
    uploadLabel: '上传',
    cancelLabel: '取消',
    $ext: {
        attrsName: 'attrs',
        waterMark: '公司名称',
        maxSize: 0.5,
        maxHeight: 200,
        maxWidth: 300,
        smallSize: 50,
        storeTo: 'db'
    }
}));
