// @er/primeng/esm2015/lib/components/table/component/summary/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
import { AggMethodsProps } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { PngTableComponent } from '../table/component';
export class PngTableSummaryComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     * @param {?} cd
     */
    constructor(pngTable, cd) {
        super();
        this.pngTable = pngTable;
        this.cd = cd;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.aggValues) {
            if (changes.aggValues.currentValue) {
                this.colAggValues = {};
                Object.keys(changes.aggValues.currentValue)
                    .forEach(key => {
                    /** @type {?} */
                    let name = key;
                    if (key.indexOf('-') > 0) {
                        name = key.substr(0, key.indexOf('-'));
                    }
                    /** @type {?} */
                    const type = key.substr(key.indexOf('-') + 1);
                    /** @type {?} */
                    let values = this.colAggValues[name];
                    if (!values) {
                        values = [];
                        this.colAggValues[name] = values;
                    }
                    /** @type {?} */
                    const agg = {};
                    agg['type'] = AggMethodsProps[type] && AggMethodsProps[type].label;
                    if (CommonsUtils.isArray(this.aggValues[key])) {
                        agg['values'] = this.aggValues[key];
                    }
                    else {
                        agg['value'] = this.aggValues[key];
                    }
                    values.push(agg);
                });
            }
            else {
                this.colAggValues = undefined;
            }
            this.cd.detectChanges();
        }
    }
    /**
     * @param {?} column
     * @return {?}
     */
    cellClass(column) {
        return CommonsUtils.getValue(column.aggCellClass, column);
    }
    /**
     * @param {?} column
     * @return {?}
     */
    cellStyle(column) {
        return CommonsUtils.getValue(column.aggCellStyle, column);
    }
}
PngTableSummaryComponent.decorators = [
    { type: Component, args: [{
                selector: '[png-table-summary]',
                template: `
    <ng-container *ngIf="customerTpl;else default">
      <ng-container *ngTemplateOutlet="customerTpl;context:{$implicit:aggValues,columns:columns,context:pngTable}"></ng-container>
    </ng-container>
    <ng-template #default>
      <th *ngIf="pngTable.$props.$ext.rowExpandable" class="col-expandable col-summary"></th>
      <th *ngIf="pngTable.$props.$ext.hasRowCheckbox" class="col-checkbox col-summary"></th>
      <th *ngIf="pngTable.$props.$ext.hasRowRadioButton" class="col-radio col-summary"></th>
      <ng-container *ngIf="colAggValues">
        <th *ngFor="let col of columns" [width]="col.width" class="text-right col-summary col-summary-{{col.key}}">
          <ng-container *ngFor="let agg of colAggValues[col.key]">
            <ng-container *ngIf="agg.value">
              <div [ngClass]="cellClass(col)" style="font-size: 0.8em"
                   [ngStyle]="cellStyle(col)">{{agg.type}}: {{agg.value | floor: 2}}
              </div>
            </ng-container>
            <ng-container *ngIf="agg.values">
              <div [ngClass]="cellClass(col)" style="font-size: 0.8em"
                   [ngStyle]="cellStyle(col)"> {{agg.type}}:
              </div>
              <ul>
                <ng-container *ngFor="let data of agg.values">
                  <li class="border-bottom"><span class="">{{data.label}}:</span><span class="badge badge-pill btn-info">{{data.value}}</span></li>
                </ng-container>
              </ul>
            </ng-container>
          </ng-container>
        </th>
      </ng-container>
    </ng-template>

  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngTableSummaryComponent.ctorParameters = () => [
    { type: PngTableComponent },
    { type: ChangeDetectorRef }
];
PngTableSummaryComponent.propDecorators = {
    columns: [{ type: Input }],
    aggValues: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngTableSummaryComponent.prototype.columns;
    /** @type {?} */
    PngTableSummaryComponent.prototype.aggValues;
    /** @type {?} */
    PngTableSummaryComponent.prototype.colAggValues;
    /** @type {?} */
    PngTableSummaryComponent.prototype.pngTable;
    /**
     * @type {?}
     * @protected
     */
    PngTableSummaryComponent.prototype.cd;
}
