// @er/core/esm2015/lib/components/dyna-component/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, EventEmitter, Input, Output, Renderer2, ViewContainerRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { ComponentUtils } from '../../utils/component';
import { DynaComponentUtils } from '../../utils/dyna';
export class DynaCompComponent {
    /**
     * @param {?} viewContainerRef
     * @param {?} el
     * @param {?} renderer
     */
    constructor(viewContainerRef, el, renderer) {
        this.viewContainerRef = viewContainerRef;
        this.el = el;
        this.renderer = renderer;
        this.onCreated = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.type) {
            this.createComponent();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.component) {
            this.component.destroy();
            this.component = null;
        }
    }
    /**
     * @private
     * @return {?}
     */
    createComponent() {
        if (this.type) {
            if (CommonsUtils.isString(this.type)) {
                this.component = DynaComponentUtils.createComponentByTemplate(this.viewContainerRef, (/** @type {?} */ (this.type)));
            }
            else {
                this.component = DynaComponentUtils.createComponentByType(this.viewContainerRef, (/** @type {?} */ (this.type)));
            }
            ComponentUtils.bind(this.component.instance, this.props);
            this.onCreated.emit(this.component);
            /** @type {?} */
            const parent = this.el.nativeElement.parentNode;
            this.renderer.removeChild(parent, this.el.nativeElement);
        }
    }
}
DynaCompComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-dyna-component',
                template: '',
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynaCompComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ElementRef },
    { type: Renderer2 }
];
DynaCompComponent.propDecorators = {
    type: [{ type: Input }],
    props: [{ type: Input }],
    onCreated: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynaCompComponent.prototype.type;
    /** @type {?} */
    DynaCompComponent.prototype.props;
    /** @type {?} */
    DynaCompComponent.prototype.onCreated;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.component;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DynaCompComponent.prototype.renderer;
}
