// @er/validate/esm5/lib/components/form/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChild, Input, Optional, TemplateRef } from '@angular/core';
import { ControlContainer, FormGroup } from '@angular/forms';
import { CommonsUtils } from '@er/utils';
import { Subject } from 'rxjs';
import { ValidateUtils } from '../../utils';
var FormValidateMessagesComponent = /** @class */ (function () {
    function FormValidateMessagesComponent(controlContainer) {
        this.controlContainer = controlContainer;
        this.errors$ = new Subject();
    }
    /**
     * @return {?}
     */
    FormValidateMessagesComponent.prototype.checkFormValidate = /**
     * @return {?}
     */
    function () {
        if (this.form.invalid) {
            alert();
            this.errors$.next(ValidateUtils.getFormErrors(this.form, this.messages, this.fields));
        }
        else {
            this.errors$.next(undefined);
        }
    };
    /**
     * @return {?}
     */
    FormValidateMessagesComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.form && this.controlContainer) {
            this.form = this.controlContainer.formDirective ? ((/** @type {?} */ (this.controlContainer.formDirective))).form : null;
        }
        if (this.form) {
            this.subscription = this.form.statusChanges
                .subscribe(function () {
                _this.checkFormValidate();
            });
        }
    };
    /**
     * @param {?} errors
     * @return {?}
     */
    FormValidateMessagesComponent.prototype.errorKeys = /**
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        return Object.keys(errors);
    };
    /**
     * @param {?} error
     * @return {?}
     */
    FormValidateMessagesComponent.prototype.errorMsg = /**
     * @param {?} error
     * @return {?}
     */
    function (error) {
        return CommonsUtils.getText(error);
    };
    /**
     * @return {?}
     */
    FormValidateMessagesComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    };
    FormValidateMessagesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-form-validate-messages',
                    template: "<div *ngIf=\"errors$|async as errors\">\n  <ng-container *ngTemplateOutlet=\"errorMsgTpl?errorMsgTpl:defaultTpl;context:{$implicit:errors}\"></ng-container>\n</div>\n\n<ng-template #defaultTpl let-errors>\n  <div class=\"text-danger\" [ngClass]=\"styleClass\">\n    <ul *ngFor=\"let key of errorKeys(errors)\" class=\"d-flex flex-column justify-content-start\">\n     <li class=\"my-2\">{{key}}</li>\n    </ul>\n  </div>\n</ng-template>",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    FormValidateMessagesComponent.ctorParameters = function () { return [
        { type: ControlContainer, decorators: [{ type: Optional }] }
    ]; };
    FormValidateMessagesComponent.propDecorators = {
        form: [{ type: Input }],
        messages: [{ type: Input }],
        fields: [{ type: Input }],
        styleClass: [{ type: Input }],
        errorMsgTpl: [{ type: ContentChild, args: [TemplateRef,] }]
    };
    return FormValidateMessagesComponent;
}());
export { FormValidateMessagesComponent };
if (false) {
    /** @type {?} */
    FormValidateMessagesComponent.prototype.form;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.messages;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.fields;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.styleClass;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.errorMsgTpl;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.errors$;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    FormValidateMessagesComponent.prototype.controlContainer;
}
