// @er/formly/esm2015/lib/er-formly.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ANALYZE_FOR_ENTRY_COMPONENTS, Inject, InjectionToken, NgModule, Optional } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { ErCoreModule } from '@er/core';
import { ErValidateModule } from '@er/validate';
import { FormlyConfig, FormlyModule } from '@ngx-formly/core';
import { TooltipModule } from 'primeng/primeng';
import { Components } from './components/all';
import { ER_FORMLY_CONFIG } from './config/all';
import { Directives } from './directives/all';
import { FormlyFieldValidatorExtension } from './extensions/field-validation';
import { FORMLY_WRAPPER_COMPONENTS } from './field-wrappers/all';
import { Pipes } from './pipes/all';
/** @type {?} */
export const Er_FORMLY_CONFIG = new InjectionToken('Er_FORMLY_CONFIG');
export class ErFormlyModule {
}
ErFormlyModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    ReactiveFormsModule,
                    FormlyModule,
                    ErCoreModule,
                    ErValidateModule,
                    TooltipModule
                ],
                declarations: [
                    ...Components,
                    ...Directives,
                    ...Pipes
                ],
                exports: [
                    ...Components,
                    ...Directives,
                    ...Pipes
                ]
            },] }
];
export class ErFormlyRootModule {
    /**
     * @param {?} formlyConfig
     * @param {?} erFormlyConfigOption
     */
    constructor(formlyConfig, erFormlyConfigOption) {
        if (erFormlyConfigOption) {
            formlyConfig.addConfig(erFormlyConfigOption);
        }
        if (formlyConfig.extensions.hasOwnProperty('field-validation')) {
            // delete formlyConfig.extensions['field-validation'];
        }
        if (formlyConfig.extensions.hasOwnProperty('field-expression')) {
            // delete formlyConfig.extensions['field-expression'];
        }
        formlyConfig.addConfig({
            extensions: [
                { name: 'validator', extension: new FormlyFieldValidatorExtension }
            ]
        });
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    static forRoot(config = {}) {
        return {
            ngModule: ErFormlyRootModule,
            providers: [
                { provide: Er_FORMLY_CONFIG, useValue: config, multi: true },
                { provide: ANALYZE_FOR_ENTRY_COMPONENTS, useValue: config, multi: true }
            ]
        };
    }
}
ErFormlyRootModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    ReactiveFormsModule,
                    ErCoreModule,
                    ErValidateModule,
                    FormlyModule.forRoot(ER_FORMLY_CONFIG)
                ],
                declarations: [
                    ...FORMLY_WRAPPER_COMPONENTS
                ]
            },] }
];
/** @nocollapse */
ErFormlyRootModule.ctorParameters = () => [
    { type: FormlyConfig },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [Er_FORMLY_CONFIG,] }] }
];
