// @er/primeng/esm2015/lib/components/table/component/body-row/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, TemplateRef, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { TableCheckbox } from 'primeng/table';
import { PngTableComponent } from '../table/component';
export class PngTableBodyRowComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     * @param {?} cd
     */
    constructor(pngTable, cd) {
        super();
        this.pngTable = pngTable;
        this.cd = cd;
    }
    /**
     * @return {?}
     */
    get rowContext() {
        return {
            columns: this.columns,
            row: this.row,
            rowIndex: this.rowIndex,
            expanded: this.expanded,
            context: this.pngTable
        };
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.chk) {
            this.$subscriptions = this.chk.tableService.selectionSource$.subscribe(() => {
                this.cd.detectChanges();
            });
        }
    }
    /**
     * @param {?} col
     * @return {?}
     */
    colContext(col) {
        return Object.assign({}, this.rowContext, col);
    }
    /**
     * @param {?} col
     * @return {?}
     */
    getColStyleClass(col) {
        return CommonsUtils.getValue(col.styleClass, this.colContext(col));
    }
    /**
     * @param {?} col
     * @return {?}
     */
    getColStyle(col) {
        return CommonsUtils.getValue(col.style, this.colContext(col));
    }
}
PngTableBodyRowComponent.decorators = [
    { type: Component, args: [{
                selector: '[png-table-body-row]',
                template: `
    <ng-container *ngIf="customerTpl;else default">
      <ng-container *ngTemplateOutlet="customerTpl;context:rowContext"></ng-container>
    </ng-container>
    <ng-template #default>
      <tr *ngIf="rowHeaderTpl">
        <ng-container *ngTemplateOutlet="rowHeaderTpl;context:rowContext"></ng-container>
      </tr>
      <td *ngIf="pngTable.$props.$ext.rowExpandable" class="col-expandable">
        <a href="#" [pRowToggler]="row">
          <i [ngClass]="expanded ? 'pi pi-chevron-down' : 'pi pi-chevron-right'"></i>
        </a>
      </td>
      <td *ngIf="pngTable.$props.$ext.hasRowCheckbox" class="col-checkbox">
        <p-tableCheckbox #chk [value]="row"></p-tableCheckbox>
      </td>
      <td *ngIf="pngTable.$props.$ext.hasRowRadioButton" class="col-radio">
        <p-tableRadioButton [value]="row"></p-tableRadioButton>
      </td>
      <ng-container *ngFor="let col of columns">
        <td png-table-body-cell class="col-data col-data-{{col.key}}"
            [ngClass]="getColStyleClass(col)"
            [ngStyle]="getColStyle(col)"
            [customerTpl]="cellTpl"
            [column]="col"
            [row]="row"
            [rowIndex]="rowIndex"
            [dataCellTpl]="dataCellTpl"
        >
        </td>
      </ng-container>
      <tr *ngIf="rowFooterTpl">
        <ng-container *ngTemplateOutlet="rowFooterTpl;context:rowContext"></ng-container>
      </tr>
    </ng-template>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngTableBodyRowComponent.ctorParameters = () => [
    { type: PngTableComponent },
    { type: ChangeDetectorRef }
];
PngTableBodyRowComponent.propDecorators = {
    columns: [{ type: Input }],
    row: [{ type: Input }],
    rowIndex: [{ type: Input }],
    expanded: [{ type: Input }],
    rowHeaderTpl: [{ type: Input }],
    rowFooterTpl: [{ type: Input }],
    cellTpl: [{ type: Input }],
    dataCellTpl: [{ type: Input }],
    chk: [{ type: ViewChild, args: ['chk',] }]
};
if (false) {
    /** @type {?} */
    PngTableBodyRowComponent.prototype.columns;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.row;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowIndex;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.expanded;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowHeaderTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.rowFooterTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.cellTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.dataCellTpl;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.chk;
    /** @type {?} */
    PngTableBodyRowComponent.prototype.pngTable;
    /**
     * @type {?}
     * @private
     */
    PngTableBodyRowComponent.prototype.cd;
}
