// @er/bootstrap/bundles/er-bootstrap.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@er/types'), require('@er/utils'), require('@angular/core'), require('@er/core')) :
    typeof define === 'function' && define.amd ? define('@er/bootstrap', ['exports', '@er/types', '@er/utils', '@angular/core', '@er/core'], factory) :
    (factory((global.er = global.er || {}, global.er.bootstrap = {}),global.types,global.utils,global.ng.core,global.core$1));
}(this, (function (exports,types,utils,core,core$1) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BsNavPageComponent = /** @class */ (function (_super) {
        __extends(BsNavPageComponent, _super);
        function BsNavPageComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @return {?}
         */
        BsNavPageComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                this.title = utils.StatesUtils.getValue(types.StateNames.appTitle);
            };
        BsNavPageComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'bs-nav-page',
                        template: "<div class=\"container h-100\">\n  <div class=\"row\">\n    <div class=\"col-12 col-md-3 col-xl-3 py-md-4 bd-sidebar\">\n      <bs-side-nav></bs-side-nav>\n    </div>\n    <main class=\"col-12 col-md-9 col-xl-8 py-md-4 pl-md-4 bd-content\">\n      <div class=\"d-flex\">\n        <div style=\"flex:1;\">\n          <h2 class=\"bd-title\">{{title}}</h2>\n        </div>\n        <div style=\"align-self:center;\">\n        </div>\n      </div>\n\n      <div class=\"card\">\n        <div class=\"card-body\">\n          <router-outlet></router-outlet>\n        </div>\n      </div>\n    </main>\n  </div>\n</div>",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        return BsNavPageComponent;
    }(core$1.BaseComponent));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BsSideNavComponent = /** @class */ (function () {
        function BsSideNavComponent() {
            this.items = [];
            this.items = this.routes.filter(function (route) { return route.component; })
                .map(function (route) {
                return ({
                    title: route.data.title,
                    url: "/" + route.path,
                    data: route.data
                });
            });
        }
        BsSideNavComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'bs-side-nav',
                        template: "\n    <ul class=\"nav nav-pills flex-column\">\n      <li class=\"nav-item\" routerLinkActive=\"active\" *ngFor=\"let item of items\">\n        <a class=\"nav-link\" routerLink=\"{{item.url}}\" routerLinkActive=\"active\">{{item.title}}</a>\n      </li>\n    </ul>\n  ",
                        styles: [".section-nav{padding-left:0;border-left:1px solid #eee}.section-nav ul{padding-left:1rem}.section-nav ul ul{display:none}.toc-entry{display:block}.toc-entry a{display:block;padding:.125rem 1.5rem;color:#99979c}.toc-entry a:hover{color:#00f;text-decoration:none}.bd-sidebar{order:0;background-color:#fff;height:100%}.bd-sidebar .nav{margin-top:10px}.bd-sidebar .nav>li>a{display:block;color:rgba(0,0,0,.65);font-weight:500}.bd-sidebar .nav>li>a:hover{color:rgba(0,0,0,.85);text-decoration:none;background-color:transparent}.bd-sidebar .nav>.active:hover>a,.bd-sidebar .nav>.active>a{font-weight:500;color:rgba(0,0,0,.85);background-color:transparent}.bd-sidebar .nav>.active>a.active{border:0 solid #343a40;border-left-width:4px;background-color:rgba(248,248,248,.62);border-radius:0}.bd-links{padding-top:1rem;padding-bottom:1rem;margin-right:-15px;margin-left:-15px}.bd-search-docs-toggle{line-height:1;color:grey}.bd-sidenav{display:none}.bd-toc-link{display:block;padding:.25rem 1.5rem;font-weight:500;color:rgba(0,0,0,.65)}.bd-toc-link:hover{color:rgba(0,0,0,.85);text-decoration:none}.bd-toc-item.active{margin-bottom:1rem}.bd-toc-item.active:not(:first-child){margin-top:1rem}.bd-toc-item.active>.bd-toc-link{color:rgba(0,0,0,.85)}.bd-toc-item.active>.bd-toc-link:hover{background-color:transparent}.bd-toc-item.active>.bd-sidenav{display:block}"]
                    }] }
        ];
        /** @nocollapse */
        BsSideNavComponent.ctorParameters = function () { return []; };
        BsSideNavComponent.propDecorators = {
            routes: [{ type: core.Input }]
        };
        return BsSideNavComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /** @type {?} */
    var BsComponents = [
        BsSideNavComponent,
        BsNavPageComponent
    ];

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var ErBootstrapModule = /** @class */ (function () {
        function ErBootstrapModule() {
        }
        ErBootstrapModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            core$1.ErCoreModule,
                        ],
                        declarations: __spread(BsComponents),
                        exports: __spread(BsComponents)
                    },] }
        ];
        return ErBootstrapModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.ErBootstrapModule = ErBootstrapModule;
    exports.BsNavPageComponent = BsNavPageComponent;
    exports.BsSideNavComponent = BsSideNavComponent;
    exports.BsComponents = BsComponents;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
