// @er/core/esm2015/lib/services/error/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { LocationStrategy, PathLocationStrategy } from '@angular/common';
import { Injectable, Injector, NgZone } from '@angular/core';
import { DomUtils } from '@er/utils';
import { LogService } from '../log/service';
import * as i0 from "@angular/core";
export class GlobalErrorHandler {
    /**
     * @param {?} injector
     * @param {?} zone
     */
    constructor(injector, zone) {
        this.injector = injector;
        this.zone = zone;
        this.logService = injector.get(LogService);
        this.location = injector.get(LocationStrategy);
    }
    /**
     * @param {?} error
     * @return {?}
     */
    handleError(error) {
        /** @type {?} */
        const message = error.message ? error.message : error.toString();
        /** @type {?} */
        const url = this.location instanceof PathLocationStrategy
            ? this.location.path() : '';
        this.logService.error(this, message, url, DomUtils.resolveUserAgent());
    }
}
GlobalErrorHandler.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
GlobalErrorHandler.ctorParameters = () => [
    { type: Injector },
    { type: NgZone }
];
/** @nocollapse */ GlobalErrorHandler.ngInjectableDef = i0.defineInjectable({ factory: function GlobalErrorHandler_Factory() { return new GlobalErrorHandler(i0.inject(i0.INJECTOR), i0.inject(i0.NgZone)); }, token: GlobalErrorHandler, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    GlobalErrorHandler.prototype.logService;
    /**
     * @type {?}
     * @private
     */
    GlobalErrorHandler.prototype.location;
    /**
     * @type {?}
     * @private
     */
    GlobalErrorHandler.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    GlobalErrorHandler.prototype.zone;
}
