// @er/primeng/esm2015/lib/components/table/component/query/field-filter/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent, DataFieldUtils } from '@er/core';
import { DataAttr, DataType } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { PngTableComponent } from '../../table/component';
export class PngTableFieldDataFilterComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     */
    constructor(pngTable) {
        super();
        this.pngTable = pngTable;
        this.onFilterChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get column() {
        return this._column;
    }
    /**
     * @param {?} column
     * @return {?}
     */
    set column(column) {
        if (!this.queryKey && column) {
            this.queryKey = DataFieldUtils.getDefaultFieldName(column.dataField) || column.key;
        }
        if (!this.queryDataType && column) {
            this.queryDataType = column.filterDataType || column.dataType;
        }
        if (!this.queryDataAttr && column && column.dataAttr) {
            this.queryDataAttr = CommonsUtils.getArrayValue(column.dataAttr)[0];
        }
        this._column = column;
    }
    /**
     * @return {?}
     */
    get dataField() {
        return this._dataField;
    }
    /**
     * @param {?} dataField
     * @return {?}
     */
    set dataField(dataField) {
        if (!this.queryKey && dataField) {
            this.queryKey = DataFieldUtils.getDefaultFieldName(dataField);
        }
        if (!this.queryDataType && dataField) {
            this.queryDataType = dataField.dataType;
        }
        this._dataField = dataField;
    }
    /**
     * @return {?}
     */
    onFilter() {
        this.onFilterChange.emit({
            key: this.queryKey,
            column: this.column,
            value: this.queryValue
        });
    }
    /**
     * @return {?}
     */
    clearFilter() {
        this.queryValue = undefined;
        this.onFilter();
    }
}
PngTableFieldDataFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-field-data-filter',
                template: `
    <div class="d-flex justify-content-end align-items-end">
      <png-query-input [queryKey]="queryKey" [queryDataType]="queryDataType" [queryDataAttr]="queryDataAttr"
                       [(queryValue)]="queryValue"></png-query-input>
      <ng-container *ngIf="queryValue">
      <span class="ml-2">
       <p-button
         type="button"
         icon="fa fa-search"
         label="查询"
         (click)="onFilter()"
       ></p-button>
    </span>
        <span class="ml-2">
    <p-button
      type="button"
      icon="fa fa-trash"
      label="清除"
      (click)="clearFilter()"
    ></p-button>
    </span>
      </ng-container>
    </div>
  `
            }] }
];
/** @nocollapse */
PngTableFieldDataFilterComponent.ctorParameters = () => [
    { type: PngTableComponent }
];
PngTableFieldDataFilterComponent.propDecorators = {
    queryKey: [{ type: Input }],
    queryDataType: [{ type: Input }],
    queryDataAttr: [{ type: Input }],
    queryValue: [{ type: Input }],
    onFilterChange: [{ type: Output }],
    column: [{ type: Input }],
    dataField: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryKey;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryDataType;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryDataAttr;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryValue;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.onFilterChange;
    /**
     * @type {?}
     * @private
     */
    PngTableFieldDataFilterComponent.prototype._column;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype._dataField;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.pngTable;
}
