// @er/formly-primeng/esm2015/lib/components/wizard/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Output } from '@angular/core';
import { FormArray, FormGroup } from '@angular/forms';
import { LogUtils } from '@er/core';
import { PngFormlyUtils } from '../../utils';
import { PngFormlyFormComponent } from '../form/component';
export class PngFormlyWizardFormComponent extends PngFormlyFormComponent {
    constructor() {
        super(...arguments);
        this.stepChange = new EventEmitter();
    }
    /**
     * @param {?} formProps
     * @return {?}
     */
    resolveFormFields(formProps) {
        if (!formProps || !formProps['steps']) {
            LogUtils.error(this, '没有定义steps', formProps);
        }
        if (formProps.steps) {
            formProps.steps.forEach(step => {
                step.fields.forEach(field => {
                    PngFormlyUtils.resolveField(formProps, field, this);
                });
            });
            this.formArray = new FormArray(formProps.steps.map(() => new FormGroup({})));
            this.stepFormlyOptions = formProps.steps.map((step) => {
                return Object.assign({}, formProps.options || {}, step.options || {});
            });
        }
        this.currentStepIndex = formProps.activeIndex || 0;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onStepChange(event) {
        this.currentStepIndex = event;
        this.stepChange.emit(event);
    }
}
PngFormlyWizardFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-formly-wizard',
                template: `
    <png-panel [header]="formProps.header" [footer]="formProps.footer" [icon]="formProps.icon" [toggleable]="formProps.collapsable">
      <png-wizard [(activeIndex)]="currentStepIndex"
                  (stepChange)="onStepChange($event)"
                  [styleClass]="formProps.styleClass"
                  [style]="formProps.style"
                  [stepStyleClass]="formProps['stepStyleClass']"
                  [stepStyle]="formProps['stepStyle']"
      >
        <ng-container *ngFor="let step of formProps['steps']; let i = index;">
          <png-step
            [label]="step.label"
            [styleClass]="step.styleClass"
            [style]="step.style"
          >
            <formly-form
              [form]="formArray.at(i)"
              [(model)]="formProps.model"
              [fields]="step.fields"
              [options]="stepFormlyOptions[i]">
            </formly-form>
          </png-step>
        </ng-container>
        <ng-template erTypedTemplate="footer">
          <button pButton type="button" *ngIf="currentStepIndex+1 === formProps['steps'].length"
                  class="pull-right" icon="fa fa-save" iconPos="right" label="提交"></button>
        </ng-template>
      </png-wizard>
    </png-panel>
    <!--<png-form-debug [model]="formModel"-->
    <!--[debugFormProps]="debugFormProps"-->
    <!--[runtimeFormProps]="formProps">-->
    <!--</png-form-debug>-->
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PngFormlyWizardFormComponent.propDecorators = {
    stepChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.formArray;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.stepFormlyOptions;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.currentStepIndex;
    /** @type {?} */
    PngFormlyWizardFormComponent.prototype.stepChange;
}
