// @er/core/esm2015/lib/directives/position/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
export class PositionDirective {
    /**
     * @param {?} el
     * @param {?} render
     */
    constructor(el, render) {
        this.el = el;
        this.render = render;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.render.addClass(this.el.nativeElement.parentNode, 'd-flex');
        this.render.addClass(this.el.nativeElement.parentNode, 'align-items-center');
        this.render.addClass(this.el.nativeElement.parentNode, 'justify-content-center');
        this.render.addClass(this.el.nativeElement, this.positionClass || 'text-center');
    }
}
PositionDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erPosition]'
            },] }
];
/** @nocollapse */
PositionDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
PositionDirective.propDecorators = {
    positionClass: [{ type: Input, args: ['erPosition',] }]
};
if (false) {
    /** @type {?} */
    PositionDirective.prototype.positionClass;
    /**
     * @type {?}
     * @private
     */
    PositionDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    PositionDirective.prototype.render;
}
