// @er/utils/esm2015/lib/date/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateNames } from '@er/types';
import * as momentNs from 'moment';
import { CommonsUtils } from '../comm';
import { StatesUtils } from '../states';
/** @type {?} */
const moment = momentNs;
export class DateUtils {
    /**
     * @param {?} data
     * @return {?}
     */
    static getMoment(data) {
        if (moment.isMoment(data)) {
            return data;
        }
        /** @type {?} */
        const date = DateUtils.toDate(data);
        if (date) {
            return moment(data);
        }
        return null;
    }
    /**
     * @param {?} v
     * @return {?}
     */
    static isDate(v) {
        return !!DateUtils.toDate(v);
    }
    /**
     * @param {?} v
     * @return {?}
     */
    static toDate(v) {
        if (CommonsUtils.isString(v)) {
            /** @type {?} */
            const date = Date.parse(v);
            if (!isNaN(date)) {
                return new Date(date);
            }
        }
        else {
            if (v && typeof v.getMonth === 'function') {
                return v;
            }
        }
        return undefined;
    }
    /**
     * @param {?=} date
     * @param {?=} format
     * @return {?}
     */
    static getFormattedDate(date, format = 'YYYY-MM-DD') {
        return DateUtils.getFormattedDateTime(format, date);
    }
    /**
     * @param {?=} date
     * @param {?=} format
     * @return {?}
     */
    static getFormattedTime(date, format = 'HH:mm:ss') {
        return DateUtils.getFormattedDateTime(format, date);
    }
    /**
     * @param {?=} format
     * @param {?=} date
     * @return {?}
     */
    static getFormattedDateTime(format = 'YYYY-MM-DD HH:mm:ss', date) {
        if (!date) {
            date = DateUtils.getDate();
        }
        else {
            date = DateUtils.toDate(date);
        }
        /** @type {?} */
        const momentData = DateUtils.getMoment(date);
        if (momentData) {
            return momentData.format(format);
        }
        return '';
    }
    /**
     * @return {?}
     */
    static getDate() {
        /** @type {?} */
        let date;
        /** @type {?} */
        const dateState = StatesUtils.getValue(StateNames.serverTime);
        if (dateState) {
            date = dateState.date;
        }
        else {
            date = new Date();
        }
        return date;
    }
    /**
     * @return {?}
     */
    static getStamp() {
        return DateUtils.getDate().getTime();
    }
    /**
     * @param {?} present
     * @param {?=} due
     * @return {?}
     */
    static getDateByNow(present, due) {
        if (!present) {
            return { seconds: 0, message: '' };
        }
        if (!DateUtils.isDate(present)) {
            try {
                present = new Date(present);
            }
            catch (e) {
                present = null;
            }
        }
        if (!present) {
            return { seconds: 0, message: '' };
        }
        /** @type {?} */
        let _due = due;
        if (!_due) {
            _due = new Date();
        }
        if (!DateUtils.isDate(_due)) {
            _due = new Date(_due);
        }
        /** @type {?} */
        const seconds = Math.round(Math.abs((_due.getTime() - present.getTime()) / 1000));
        /** @type {?} */
        const minutes = Math.round(Math.abs(seconds / 60));
        /** @type {?} */
        const hours = Math.round(Math.abs(minutes / 60));
        /** @type {?} */
        const days = Math.round(Math.abs(hours / 24));
        /** @type {?} */
        const months = Math.round(Math.abs(days / 30.416));
        /** @type {?} */
        const years = Math.round(Math.abs(days / 365));
        /** @type {?} */
        let message = '';
        if (seconds <= 10) {
            message = (_due > present) ? '刚刚' : '马上';
        }
        else if (seconds <= 30) {
            message = '30秒';
        }
        else if (seconds <= 90) {
            message = '1分钟';
        }
        else if (minutes <= 45) {
            message = minutes + '分钟';
        }
        else if (minutes <= 90) {
            message = '1小时';
        }
        else if (hours <= 22) {
            message = hours + '小时';
        }
        else if (hours <= 36) {
            message = '1天';
        }
        else if (days <= 25) {
            message = days + '天';
        }
        else if (days <= 45) {
            message = '1个月';
        }
        else if (days <= 345) {
            message = months + '月';
        }
        else if (days <= 545) {
            message = '1年';
        }
        else { // (days > 545)
            message = years + '年';
        }
        if (seconds > 10) {
            message += (_due > present) ? '前' : '后';
        }
        return {
            seconds,
            message
        };
    }
    /**
     * @param {?} seconds
     * @return {?}
     */
    static getUpdateSeconds(seconds) {
        /** @type {?} */
        const min = 60;
        /** @type {?} */
        const hr = min * 60;
        /** @type {?} */
        const day = hr * 24;
        if (seconds < min) { // less than 1 min, update every 2 secs
            return 2;
        }
        else if (seconds < hr) { // less than an hour, update every 30 secs
            return 30;
        }
        else if (seconds < day) { // less then a day, update every 5 mins
            return 300;
        }
        else { // update every hour
            return 3600;
        }
    }
    /**
     * @param {?=} date
     * @return {?}
     */
    static getYear(date) {
        return (date || DateUtils.getDate()).getFullYear();
    }
    /**
     * @param {?=} date
     * @return {?}
     */
    static getMonth(date) {
        return (date || DateUtils.getDate()).getMonth() + 1;
    }
    /**
     * @param {?=} date
     * @return {?}
     */
    static getDay(date) {
        return (date || DateUtils.getDate()).getDate();
    }
    /**
     * @return {?}
     */
    static getYesterday() {
        return DateUtils.getDay(new Date(DateUtils.getDate().getTime() - 24 * 60 * 60 * 1000));
    }
    /**
     * @return {?}
     */
    static getTomorrow() {
        return DateUtils.getDay(new Date(DateUtils.getDate().getTime() + 24 * 60 * 60 * 1000));
    }
    /**
     * @return {?}
     */
    static getLastMonth() {
        /** @type {?} */
        const month = DateUtils.getMonth() - 1;
        return month < 1 ? 12 : month;
    }
    /**
     * @return {?}
     */
    static getNextMonth() {
        /** @type {?} */
        const month = DateUtils.getMonth() + 1;
        return month > 12 ? 1 : month;
    }
}
