// @er/core/esm2015/lib/directives/animate-if/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgIf } from '@angular/common';
import { Directive, EventEmitter, Input, Output, TemplateRef, ViewContainerRef } from '@angular/core';
import { Subject } from 'rxjs/index';
import { Observable } from 'rxjs/internal/Observable';
import { takeUntil } from 'rxjs/operators';
/**
 * @param {?} el
 * @param {?} maxWait
 * @return {?}
 */
function animationEndSafe(el, maxWait) {
    return Observable.create((observer) => {
        el.addEventListener('animationend', observer);
        setTimeout(observer, maxWait);
    });
    // return new Promise((resolve, reject) => {
    //   el.addEventListener('animationend', resolve);
    //   setTimeout(resolve, maxWait);
    // });
}
export class AnimatedIfDirective extends NgIf {
    /**
     * @param {?} viewContainer
     * @param {?} _templateRef
     */
    constructor(viewContainer, _templateRef) {
        super(viewContainer, _templateRef);
        this.viewContainer = viewContainer;
        this.animatedIfOnShow = new EventEmitter(false);
        this.animatedIfOnHide = new EventEmitter(false);
        this.visible = false;
        this.destroyed$ = new Subject();
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set erAnimatedIf(val) {
        if (this.visible && !val) {
            this.hide();
        }
        else if (!this.visible && val) {
            this.show();
        }
        this.visible = val;
    }
    /**
     * @private
     * @return {?}
     */
    getContainer() {
        if (this.viewContainer.length > 0) {
            return ((/** @type {?} */ (this.viewContainer.get(0)))).rootNodes[0];
        }
        else {
            return null;
        }
    }
    /**
     * Hides the element. Animation is performed through css on the container.
     * This method will set the class to hiding and preserve the element for the animation duration.
     * @private
     * @return {?}
     */
    hide() {
        /** @type {?} */
        const container = this.getContainer();
        if (!!container) {
            container.classList.remove('showing');
            container.classList.add('hiding');
            animationEndSafe(container, 1000)
                .pipe(takeUntil(this.destroyed$))
                .subscribe(() => {
                this.ngIf = this.visible;
                container.classList.remove('hiding');
            });
            this.animatedIfOnHide.emit();
        }
    }
    /**
     * Shows the element. Animation is performed through css on the container
     * @private
     * @return {?}
     */
    show() {
        this.ngIf = true;
        /** @type {?} */
        const container = this.getContainer();
        if (!!container) {
            container.classList.remove('hiding');
            container.classList.add('showing');
            animationEndSafe(container, 1000)
                .pipe(takeUntil(this.destroyed$))
                .subscribe(() => {
                container.classList.remove('showing');
            });
            this.animatedIfOnShow.emit();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroyed$.next();
        this.destroyed$.complete();
    }
}
AnimatedIfDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erAnimatedIf]'
            },] }
];
/** @nocollapse */
AnimatedIfDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef }
];
AnimatedIfDirective.propDecorators = {
    animatedIfOnShow: [{ type: Output }],
    animatedIfOnHide: [{ type: Output }],
    erAnimatedIf: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AnimatedIfDirective.prototype.animatedIfOnShow;
    /** @type {?} */
    AnimatedIfDirective.prototype.animatedIfOnHide;
    /**
     * @type {?}
     * @private
     */
    AnimatedIfDirective.prototype.visible;
    /**
     * @type {?}
     * @protected
     */
    AnimatedIfDirective.prototype.destroyed$;
    /**
     * @type {?}
     * @private
     */
    AnimatedIfDirective.prototype.viewContainer;
}
