// @er/primeng/esm5/lib/components/tree/types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngTreeNodeProps() { }
/**
 * @record
 */
export function PngTreeProps() { }
if (false) {
    /** @type {?|undefined} */
    PngTreeProps.prototype.nodes;
    /** @type {?|undefined} */
    PngTreeProps.prototype.selectionMode;
    /** @type {?|undefined} */
    PngTreeProps.prototype.selection;
    /** @type {?|undefined} */
    PngTreeProps.prototype.contextMenu;
    /** @type {?|undefined} */
    PngTreeProps.prototype.layout;
    /** @type {?|undefined} */
    PngTreeProps.prototype.draggableScope;
    /** @type {?|undefined} */
    PngTreeProps.prototype.droppableScope;
    /** @type {?|undefined} */
    PngTreeProps.prototype.draggableNodes;
    /** @type {?|undefined} */
    PngTreeProps.prototype.droppableNodes;
    /** @type {?|undefined} */
    PngTreeProps.prototype.metaKeySelection;
    /** @type {?|undefined} */
    PngTreeProps.prototype.propagateSelectionUp;
    /** @type {?|undefined} */
    PngTreeProps.prototype.propagateSelectionDown;
    /** @type {?|undefined} */
    PngTreeProps.prototype.loadingIcon;
    /** @type {?|undefined} */
    PngTreeProps.prototype.emptyMessage;
    /** @type {?|undefined} */
    PngTreeProps.prototype.selectionChange;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeSelect;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeUnselect;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeExpand;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeCollapse;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeContextMenuSelect;
    /** @type {?|undefined} */
    PngTreeProps.prototype.onNodeDrop;
    /** @type {?|undefined} */
    PngTreeProps.prototype.$container;
    /** @type {?|undefined} */
    PngTreeProps.prototype.$ext;
}
/** @type {?} */
export var PNG_TREE_NODE_KEYS = {
    idKey: 'id',
    labelKey: 'label',
    parentKey: 'parent',
    childrenKey: 'children',
    leafKey: 'leaf',
    dataKey: 'data',
    expandedKey: 'expanded'
};
/** @type {?} */
export var DEFAULT_PNG_TREE_NODE_PROPS = {};
/** @type {?} */
export var DEFAULT_PNG_TREE_PROPS = {
    style: { 'max-height.px': 500, 'max-width.px': 350, 'overflow': 'auto' },
    emptyMessage: '没有节点信息',
    $ext: {
        apiDataProps: {
            dataPath: 'items',
            size: 9999
        },
        dataItemProps: {
            idKey: 'id',
            parentKey: 'parent',
            labelKey: 'name',
            leafKey: 'leaf'
        },
        nodesFormat: 'plainArray',
        lazy: false,
        hasSearchBox: true,
        autoExpandLevel: 1,
        expandedIcon: 'fa fa-folder-open',
        collapsedIcon: 'fa fa-folder',
        icon: 'fa fa-leaf'
    }
};
