// @er/formly-bootstrap/esm5/lib/field-wrappers/addons.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, ViewContainerRef } from '@angular/core';
import { FieldWrapper } from '@ngx-formly/core';
var BsAddonsWrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BsAddonsWrapperComponent, _super);
    function BsAddonsWrapperComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    BsAddonsWrapperComponent.prototype.addonRightClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.to.addonRight.onClick) {
            this.to.addonRight.onClick(this.to, this, $event);
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    BsAddonsWrapperComponent.prototype.addonLeftClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.to.addonLeft.onClick) {
            this.to.addonLeft.onClick(this.to, this, $event);
        }
    };
    BsAddonsWrapperComponent.decorators = [
        { type: Component, args: [{
                    template: "\n    <div class=\"input-group\">\n      <div class=\"input-group-prepend\"\n        *ngIf=\"to['addonLeft']\"\n        [ngStyle]=\"{cursor: to.addonLeft.onClick ? 'pointer' : 'inherit'}\"\n        (click)=\"addonLeftClick($event)\">\n        <i class=\"input-group-text\" [ngClass]=\"to['addonLeft'].class\" *ngIf=\"to['addonLeft'].class\"></i>\n        <span *ngIf=\"to['addonLeft'].text\" class=\"input-group-text\">{{ to['addonLeft'].text }}</span>\n      </div>\n      <ng-template #fieldComponent></ng-template>\n      <div class=\"input-group-append\"\n        *ngIf=\"to['addonRight']\"\n        [ngStyle]=\"{cursor: to['addonRight'].onClick ? 'pointer' : 'inherit'}\"\n        (click)=\"addonRightClick($event)\">\n        <i class=\"input-group-text\" [ngClass]=\"to['addonRight'].class\" *ngIf=\"to['addonRight'].class\"></i>\n        <span *ngIf=\"to['addonRight'].text\" class=\"input-group-text\">{{ to['addonRight'].text }}</span>\n      </div>\n    </div>\n  ",
                    styles: ["\n    :host /deep/ .input-group>:not(:first-child)> .form-control {\n      border-top-left-radius: 0;\n      border-bottom-left-radius: 0;\n    }\n    :host /deep/ .input-group>:not(:last-child)> .form-control {\n      border-top-right-radius: 0;\n      border-bottom-right-radius: 0;\n    }\n  "]
                }] }
    ];
    BsAddonsWrapperComponent.propDecorators = {
        fieldComponent: [{ type: ViewChild, args: ['fieldComponent', { read: ViewContainerRef },] }]
    };
    return BsAddonsWrapperComponent;
}(FieldWrapper));
export { BsAddonsWrapperComponent };
if (false) {
    /** @type {?} */
    BsAddonsWrapperComponent.prototype.fieldComponent;
}
