// @er/formly-primeng/esm5/lib/components/tab/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Output } from '@angular/core';
import { FormArray, FormGroup } from '@angular/forms';
import { LogUtils } from '@er/core';
import { PngFormlyUtils } from '../../utils';
import { PngFormlyFormComponent } from '../form/component';
var PngFormlyTabsFormComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngFormlyTabsFormComponent, _super);
    function PngFormlyTabsFormComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onTabChange = new EventEmitter();
        _this.onTabClose = new EventEmitter();
        return _this;
    }
    /**
     * @param {?} formProps
     * @return {?}
     */
    PngFormlyTabsFormComponent.prototype.resolveFormFields = /**
     * @param {?} formProps
     * @return {?}
     */
    function (formProps) {
        var _this = this;
        if (!formProps || !formProps['tabs']) {
            LogUtils.error(this, '没有定义tabs', formProps);
        }
        if (formProps.tabs) {
            formProps.tabs.forEach(function (tab) {
                tab.fields.forEach(function (field) {
                    PngFormlyUtils.resolveField(_this._formProps, field, _this);
                });
            });
            this.formArray = new FormArray(formProps.tabs.map(function () { return new FormGroup({}); }));
            this.tabFormlyOptions = formProps.tabs.map(function (tab) {
                return tslib_1.__assign({}, formProps.options || {}, tab.options || {});
            });
        }
        this.currentTabIndex = formProps.activeIndex || 0;
    };
    /**
     * @return {?}
     */
    PngFormlyTabsFormComponent.prototype.afterPropsInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngFormlyTabsFormComponent.prototype.tabChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.currentTabIndex = event;
        this.onTabChange.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngFormlyTabsFormComponent.prototype.tabClose = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onTabClose.emit(event);
    };
    PngFormlyTabsFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-formly-tab',
                    template: "\n    <png-panel [header]=\"formProps.caption\" [footer]=\"formProps.footer\" [icon]=\"formProps.icon\" [toggleable]=\"formProps.collapsable\">\n      <p-tabView [activeIndex]=\"formProps['activeIndex']\"\n                 [styleClass]=\"formProps.styleClass\"\n                 [style]=\"formProps.style\"\n                 [orientation]=\"formProps['tabOrientation']\"\n                 (onChange)=\"tabChange($event)\"\n                 (onClose)=\"tabClose($event)\"\n      >\n        <ng-container>\n          <p-tabPanel *ngFor=\"let tab of formProps['tabs']; let i = index\"\n                      [header]=\"tab.caption\"\n                      [headerStyle]=\"tab.style\"\n                      [headerStyleClass]=\"tab.styleClass\"\n                      [leftIcon]=\"tab.leftIcon\"\n                      [rightIcon]=\"tab.rightIcon\"\n          >\n            <formly-form\n              [form]=\"formArray.at(i)\"\n              [(model)]=\"formProps.model\"\n              [fields]=\"tab.fields\"\n              [options]=\"tabFormlyOptions[i]\">\n            </formly-form>\n          </p-tabPanel>\n        </ng-container>\n      </p-tabView>\n    </png-panel>\n\n    <div [ngClass]=\"formProps.buttonsStyleClass\">\n      <er-button [stateKey]=\"formProps.$id\"\n                 icon=\"fa fa-save\"\n                 label=\"\u63D0&nbsp;&nbsp;&nbsp;\u4EA4\"\n                 title=\"\u4FDD\u7559\u6570\u636E\u5230\u670D\u52A1\u7AEF\"\n                 class=\"mr-2\"\n                 type=\"button\"\n                 (onClick)=\"formlyForm.submit()\"\n                 [styleClass]=\"formProps.submitBtnStyleClass\"\n      ></er-button>\n      <er-button icon=\"fa fa-circle-o\"\n                 class=\"mr-2\"\n                 label=\"\u91CD&nbsp;&nbsp;&nbsp;\u586B\"\n                 title=\"\u6062\u590D\u5230\u4FEE\u6539\u524D\u7684\u72B6\u6001\"\n                 [styleClass]=\"formProps.resetBtnStyleClass\"\n                 (onClick)=\"formlyForm.resetModel()\">\n      </er-button>\n\n      <ng-container *ngFor=\"let btn of formProps.buttons\">\n        <er-button erPropsBind [props]=\"btn\" [context]=\"btn.context\"></er-button>\n      </ng-container>\n    </div>\n\n    <!--<png-form-debug [model]=\"formModel\"-->\n    <!--[debugFormProps]=\"props\"-->\n    <!--[runtimeFormProps]=\"formProps\">-->\n    <!--</png-form-debug>-->\n  ",
                    styles: [":host::ng-deep .ui-tabview .ui-state-active{background-color:transparent!important}:host::ng-deep .ui-tabview .ui-tabview-panel{margin:10px}:host::ng-deep .ui-tabview .ui-tabview-selected{border-left:5px solid #3bdc06}"]
                }] }
    ];
    PngFormlyTabsFormComponent.propDecorators = {
        onTabChange: [{ type: Output }],
        onTabClose: [{ type: Output }]
    };
    return PngFormlyTabsFormComponent;
}(PngFormlyFormComponent));
export { PngFormlyTabsFormComponent };
if (false) {
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.formArray;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.tabFormlyOptions;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.currentTabIndex;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.onTabChange;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.onTabClose;
}
