// @er/primeng/esm2015/lib/components/option-item/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
export class PngOptionItemComponent extends BaseComponent {
    /**
     * @return {?}
     */
    get labels() {
        /** @type {?} */
        const _label = DataItemUtils.getItemLabel(this.item, this.props.dropDownOptionKey || this.props.labelKey || 'label');
        if (CommonsUtils.isArray(_label)) {
            return _label;
        }
        else {
            return [_label];
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.props) {
            this.props = {};
        }
    }
}
PngOptionItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-option-item',
                template: `
    <ng-container *ngIf="customerTpl;else default">
      <ng-container *ngTemplateOutlet="customerTpl;context:{$implicit:item}"></ng-container>
    </ng-container>
    <ng-template #default>
      <div class="d-flex option-item" [ngClass]="props.itemStyleClass || 'ui-helper-clearfix'">
          <span class="label mr-3" *ngFor="let label of labels" [ngClass]="props.labelStyleClass" [innerHTML]="label"></span>
          <span class="pull-right count" *ngIf="props.showCounter && item['count']">
          <span [ngClass]="props.counterStyleClass || 'badge badge-pill badge-info'">{{item['count']}}</span>
        </span>
      </div>
    </ng-template>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".option-item{border-bottom:1px solid #d5d5d5;font-size:.8em}.option-item:hover{font-weight:700;font-size:1em}"]
            }] }
];
PngOptionItemComponent.propDecorators = {
    item: [{ type: Input }],
    props: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngOptionItemComponent.prototype.item;
    /** @type {?} */
    PngOptionItemComponent.prototype.props;
}
