// @er/primeng/esm5/lib/work-space/menu/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { PngAdminWorkSpaceComponent } from '../component';
var WorkSpaceMainMenuComponent = /** @class */ (function () {
    function WorkSpaceMainMenuComponent(ws) {
        this.ws = ws;
        this.theme = 'blue';
        this.layout = 'blue';
        this.version = 'v4';
    }
    /**
     * @return {?}
     */
    WorkSpaceMainMenuComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.menus$ = StatesUtils.observe(StateNames.rootMenu)
            .pipe(map(function (menu) { return (menu && menu.menuItems) || []; }), distinctUntilChanged());
    };
    /**
     * @param {?} theme
     * @return {?}
     */
    WorkSpaceMainMenuComponent.prototype.changeTheme = /**
     * @param {?} theme
     * @return {?}
     */
    function (theme) {
        /** @type {?} */
        var themeLink = (/** @type {?} */ (document.getElementById('theme-css')));
        if (this.version === 'v3') {
            themeLink.href = 'assets/theme/theme-' + theme + '.css';
        }
        else {
            themeLink.href = 'assets/theme/theme-' + theme + '-v4' + '.css';
        }
        this.theme = theme;
    };
    /**
     * @param {?} layout
     * @param {?=} special
     * @return {?}
     */
    WorkSpaceMainMenuComponent.prototype.changeLayout = /**
     * @param {?} layout
     * @param {?=} special
     * @return {?}
     */
    function (layout, special) {
        /** @type {?} */
        var layoutLink = (/** @type {?} */ (document.getElementById('layout-css')));
        if (this.version === 'v3') {
            layoutLink.href = 'assets/layout/css/layout-' + layout + '.css';
        }
        else {
            layoutLink.href = 'assets/layout/css/layout-' + layout + '-v4' + '.css';
        }
        this.layout = layout;
        if (special) {
            this.ws.darkMenu = true;
        }
    };
    /**
     * @param {?} version
     * @return {?}
     */
    WorkSpaceMainMenuComponent.prototype.changeVersion = /**
     * @param {?} version
     * @return {?}
     */
    function (version) {
        /** @type {?} */
        var themeLink = (/** @type {?} */ (document.getElementById('theme-css')));
        /** @type {?} */
        var layoutLink = (/** @type {?} */ (document.getElementById('layout-css')));
        if (version === 'v3') {
            this.version = 'v3';
            themeLink.href = 'assets/theme/theme-' + this.theme + '.css';
            layoutLink.href = 'assets/layout/css/layout-' + this.layout + '.css';
        }
        else {
            themeLink.href = 'assets/theme/theme-' + this.theme + '-v4' + '.css';
            layoutLink.href = 'assets/layout/css/layout-' + this.layout + '-v4' + '.css';
            this.version = '-v4';
        }
    };
    WorkSpaceMainMenuComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-ws-main-menu',
                    template: "\n    <ul png-ws-sub-menu\n        erAnimate=\"fadeInDown\"\n        class=\"layout-menu layout-main-menu clearfix\"\n        [erHttpState]=\"httpMenuTag\"\n        [item]=\"menus$|async\" [root]=\"true\"\n        [reset]=\"reset\"\n        [visible]=\"true\"\n        [parentActive]=\"true\"></ul>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    WorkSpaceMainMenuComponent.ctorParameters = function () { return [
        { type: PngAdminWorkSpaceComponent }
    ]; };
    WorkSpaceMainMenuComponent.propDecorators = {
        reset: [{ type: Input }],
        httpMenuTag: [{ type: Input }]
    };
    return WorkSpaceMainMenuComponent;
}());
export { WorkSpaceMainMenuComponent };
if (false) {
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.reset;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.httpMenuTag;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.menus$;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.theme;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.layout;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.version;
    /** @type {?} */
    WorkSpaceMainMenuComponent.prototype.ws;
}
