// @er/utils/esm5/lib/chart/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChartType } from '@er/types';
import { ColorUtils } from '../color';
import { CommonsUtils } from '../comm';
import { DataItemUtils } from '../data-item';
var ChartUtils = /** @class */ (function () {
    function ChartUtils() {
    }
    /**
     * @param {?} rawData
     * @param {?} chartOptions
     * @param {?=} dimensionsIndex
     * @return {?}
     */
    ChartUtils.buildChartData = /**
     * @param {?} rawData
     * @param {?} chartOptions
     * @param {?=} dimensionsIndex
     * @return {?}
     */
    function (rawData, chartOptions, dimensionsIndex) {
        if (dimensionsIndex === void 0) { dimensionsIndex = 0; }
        /** @type {?} */
        var type = chartOptions.type;
        /** @type {?} */
        var data = rawData;
        /** @type {?} */
        var dimension = chartOptions.dimension || {};
        if (dimension.dataSrcPath) {
            data = CommonsUtils.get(rawData, dimension.dataSrcPath);
        }
        /** @type {?} */
        var dataItemProps = dimension.dataItemProps || chartOptions.dataItemProps;
        /** @type {?} */
        var label = dimension.title || '';
        switch (type || ChartType.BAR) {
            case ChartType.BAR:
                data = ChartUtils.buildBarChart(data, dataItemProps, label, dimensionsIndex);
                break;
            case ChartType.LINE:
                data = ChartUtils.buildLineChart(data, dataItemProps, label, dimensionsIndex);
                break;
            case ChartType.PIE:
                data = ChartUtils.buildPieChart(data, dataItemProps, label, dimensionsIndex);
                break;
            case ChartType.AREA:
                data = ChartUtils.buildAreaChart(data, dataItemProps, label, dimensionsIndex);
                break;
            case ChartType.RADAR:
                data = ChartUtils.buildRadarChart(data, dataItemProps, label, dimensionsIndex);
                break;
            case ChartType.DOUGHNUT:
                data = ChartUtils.buildPieChart(data, dataItemProps, label, dimensionsIndex);
                break;
        }
        return data;
    };
    /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    ChartUtils.buildBarChart = /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    function (dataItems, config, label, colorIndex) {
        if (label === void 0) { label = ''; }
        if (colorIndex === void 0) { colorIndex = 0; }
        /** @type {?} */
        var _data = DataItemUtils.transItems(dataItems, config);
        /** @type {?} */
        var labels = [];
        /** @type {?} */
        var datasets = [];
        /** @type {?} */
        var data = [];
        /** @type {?} */
        var backgroundColor = [];
        /** @type {?} */
        var hoverBackgroundColor = [];
        /** @type {?} */
        var items = _data;
        items.forEach(function (item) {
            if (item['group']) {
                datasets.push({
                    label: items[''].label,
                    data: items[''].items,
                    backgroundColor: ColorUtils.getColor(items[''].label)
                });
                labels.push('');
            }
            else {
                /** @type {?} */
                var lb = item['key_as_string'] || item['label'];
                labels.push(lb);
                data.push(item['value']);
                backgroundColor.push(ColorUtils.getColor(lb, colorIndex));
                hoverBackgroundColor.push(ColorUtils.getColor(lb, colorIndex + 10));
            }
        });
        datasets.push({
            label: label,
            backgroundColor: backgroundColor,
            hoverBackgroundColor: hoverBackgroundColor,
            data: data
        });
        return {
            labels: labels,
            datasets: datasets
        };
    };
    /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    ChartUtils.buildPieChart = /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    function (dataItems, config, label, colorIndex) {
        if (label === void 0) { label = ''; }
        if (colorIndex === void 0) { colorIndex = 0; }
        /** @type {?} */
        var _data = DataItemUtils.transItems(dataItems, config);
        /** @type {?} */
        var labels = [];
        /** @type {?} */
        var datasets = [];
        /** @type {?} */
        var data = [];
        /** @type {?} */
        var backgroundColor = [];
        /** @type {?} */
        var hoverBackgroundColor = [];
        /** @type {?} */
        var items = _data;
        items.forEach(function (item) {
            if (item['group']) {
                datasets.push({
                    label: items[''].label,
                    data: items[''].items,
                    backgroundColor: ColorUtils.getColor(items[''].label)
                });
                labels.push('');
            }
            else {
                /** @type {?} */
                var lb = item['key_as_string'] || item['label'];
                labels.push(lb);
                data.push(item['value']);
                backgroundColor.push(ColorUtils.getColor(lb, colorIndex));
                hoverBackgroundColor.push(ColorUtils.getColor(lb, colorIndex + 10));
            }
        });
        datasets.push({
            label: label || '',
            backgroundColor: backgroundColor,
            hoverBackgroundColor: hoverBackgroundColor,
            data: data
        });
        return {
            labels: labels,
            datasets: datasets
        };
    };
    /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    ChartUtils.buildAreaChart = /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    function (dataItems, config, label, colorIndex) {
        if (label === void 0) { label = ''; }
        if (colorIndex === void 0) { colorIndex = 0; }
        /** @type {?} */
        var _data = DataItemUtils.transItems(dataItems, config);
        /** @type {?} */
        var labels = [];
        /** @type {?} */
        var datasets = [];
        /** @type {?} */
        var backgroundColor = [];
        /** @type {?} */
        var data = [];
        /** @type {?} */
        var items = _data;
        items.forEach(function (item) {
            if (item['group']) {
                datasets.push({
                    label: items[''].label,
                    data: items[''].items,
                    backgroundColor: ColorUtils.getColor(items[''].label)
                });
                labels.push('');
            }
            else {
                /** @type {?} */
                var lb = item['key_as_string'] || item['label'];
                labels.push(lb);
                data.push(item['value']);
                backgroundColor.push(ColorUtils.getColor(lb, colorIndex));
            }
        });
        datasets.push({
            label: label || '',
            data: data,
            backgroundColor: backgroundColor
        });
        return {
            labels: labels,
            datasets: datasets
        };
    };
    /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    ChartUtils.buildLineChart = /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    function (dataItems, config, label, colorIndex) {
        if (label === void 0) { label = ''; }
        if (colorIndex === void 0) { colorIndex = 0; }
        /** @type {?} */
        var _data = DataItemUtils.transItems(dataItems, config);
        /** @type {?} */
        var labels = [];
        /** @type {?} */
        var datasets = [];
        /** @type {?} */
        var data = [];
        /** @type {?} */
        var items = _data;
        items.forEach(function (item) {
            if (item['group']) {
                datasets.push({
                    label: items[''].label,
                    data: items[''].items,
                    backgroundColor: ColorUtils.getColor(items[''].label)
                });
                labels.push('');
            }
            else {
                /** @type {?} */
                var lb = item['key_as_string'] || item['label'];
                labels.push(lb);
                data.push(item['value']);
            }
        });
        datasets.push({
            label: label || '',
            data: data,
            fill: false,
            borderColor: ColorUtils.getColor(colorIndex)
        });
        return {
            labels: labels,
            datasets: datasets
        };
    };
    /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    ChartUtils.buildRadarChart = /**
     * @param {?} dataItems
     * @param {?} config
     * @param {?=} label
     * @param {?=} colorIndex
     * @return {?}
     */
    function (dataItems, config, label, colorIndex) {
        if (label === void 0) { label = ''; }
        if (colorIndex === void 0) { colorIndex = 0; }
        /** @type {?} */
        var _data = DataItemUtils.transItems(dataItems, config);
        /** @type {?} */
        var labels = [];
        /** @type {?} */
        var datasets = [];
        /** @type {?} */
        var backgroundColor = [];
        /** @type {?} */
        var borderColor = [];
        /** @type {?} */
        var pointBackgroundColor = [];
        /** @type {?} */
        var pointHoverBorderColor = [];
        /** @type {?} */
        var data = [];
        /** @type {?} */
        var items = _data;
        items.forEach(function (item) {
            if (item['group']) {
                datasets.push({
                    label: items[''].label,
                    data: items[''].items,
                    backgroundColor: ColorUtils.getColor(items[''].label)
                });
                labels.push('');
            }
            else {
                /** @type {?} */
                var lb = item['key_as_string'] || item['label'];
                labels.push(lb);
                data.push(item['value']);
                backgroundColor.push(ColorUtils.getColor(lb, colorIndex));
                borderColor.push(ColorUtils.getColor(lb, colorIndex + 5));
                pointBackgroundColor.push(ColorUtils.getColor(lb, colorIndex + 5));
                pointHoverBorderColor.push(ColorUtils.getColor(lb, colorIndex + 5));
            }
        });
        datasets.push({
            label: label || '',
            data: data,
            backgroundColor: backgroundColor,
            borderColor: borderColor,
            pointBackgroundColor: pointBackgroundColor,
            pointHoverBorderColor: pointHoverBorderColor,
            pointBorderColor: '#fff',
            pointHoverBackgroundColor: '#fff'
        });
        return {
            labels: labels,
            datasets: datasets
        };
    };
    return ChartUtils;
}());
export { ChartUtils };
