// @er/primeng/esm5/lib/components/input/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ControlWrapperComponent } from '@er/core';
var PngInputComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngInputComponent, _super);
    function PngInputComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PngInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-input',
                    template: "\n    <div class=\"ui-inputgroup\">\n      <ng-container *ngIf=\"attrs?.$ext?.preIcon || attrs?.$ext?.preText\">\n        <span class=\"ui-inputgroup-addon\"><i [ngClass]=\"attrs?.$ext?.preIcon\"></i>{{attrs?.$ext?.preText}}</span>\n      </ng-container>\n      <png-buttons *ngIf=\"attrs?.$ext?.preButtons\" [btns]=\"attrs.$ext.preButtons\" [context]=\"ctx\"></png-buttons>\n      <ng-container *ngIf=\"attrs?.$ext?.floatLabel;else inputTpl\">\n       <span class=\"ui-float-label\">\n         <ng-container *ngTemplateOutlet=\"inputTpl\"></ng-container>\n         <label style=\"font-weight: bold;\">{{attrs?.$ext.label}}</label>\n       </span>\n      </ng-container>\n      <ng-template #inputTpl>\n        <span class=\"w-100 inner-addon\">\n          <input pInputText\n                 [(ngModel)]=\"value\"\n                 type=\"{{attrs?.type || 'text'}}\"\n                 erPropsBind\n                 [attrs]=\"attrs\"\n                 [context]=\"ctx\"\n                 [pKeyFilter]=\"attrs?.$ext?.inputRule || 'alphanum'\"\n                 (blur)=\"updateTouch()\"/>\n          <i *ngIf=\"value && !disabled && resettable\" title=\"\u6E05\u9664\u8F93\u5165\u5185\u5BB9\" class=\"fa fa-trash gray right\" (click)=\"reset()\"></i>\n       </span>\n      </ng-template>\n      <ng-container *ngIf=\"attrs?.$ext?.postIcon || attrs?.$ext?.postText\">\n        <span class=\"ui-inputgroup-addon\"><i [ngClass]=\"attrs?.$ext?.postIcon\"></i>{{attrs?.$ext?.postText}}</span>\n      </ng-container>\n      <png-buttons *ngIf=\"attrs?.$ext?.postButtons\" [btns]=\"attrs.$ext.postButtons\" [context]=\"ctx\"></png-buttons>\n    </div>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngInputComponent; }),
                            multi: true
                        }],
                    styles: ["\n    .ui-float-label {\n      width: 100%\n    }\n\n    .ui-float-label > label {\n      font-size: 1em;\n      color: inherit\n    }\n  "]
                }] }
    ];
    PngInputComponent.propDecorators = {
        attrs: [{ type: Input }]
    };
    return PngInputComponent;
}(ControlWrapperComponent));
export { PngInputComponent };
if (false) {
    /** @type {?} */
    PngInputComponent.prototype.attrs;
}
