// @er/core/esm5/lib/components/we-chat/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ConfigUtils } from '@er/utils';
var WechatQrCodeComponent = /** @class */ (function () {
    function WechatQrCodeComponent() {
        this.height = 300;
        this.width = 300;
        this.baseUrl = 'https://open.weixin.qq.com/connect/qrconnect';
    }
    Object.defineProperty(WechatQrCodeComponent.prototype, "qrCodeUrl", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var wechat = ConfigUtils.getConfig().wechat;
            /** @type {?} */
            var redirect = encodeURIComponent(wechat.redirect_uri);
            /** @type {?} */
            var url = this.baseUrl + "?appid=" + wechat.appid + "&redirect_uri=" + redirect + "&scope=" + wechat.scope + "&response_type=code";
            url += wechat.style ? '&style=' + wechat.style : '';
            url += wechat.self_redirect ? '&self_redirect=' + wechat.self_redirect : '';
            url += wechat.href ? '&href=' + wechat.href : '';
            return url;
        },
        enumerable: true,
        configurable: true
    });
    WechatQrCodeComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wechat-qrcode',
                    template: "\n    <iframe [src]=\"qrCodeUrl|erSafeHtml:'resource'\" [width]=\"width\" [height]=\"height\" scrolling=\"no\" [attr.frameborder]=\"0\"></iframe>\n  "
                }] }
    ];
    WechatQrCodeComponent.propDecorators = {
        height: [{ type: Input }],
        width: [{ type: Input }]
    };
    return WechatQrCodeComponent;
}());
export { WechatQrCodeComponent };
if (false) {
    /** @type {?} */
    WechatQrCodeComponent.prototype.height;
    /** @type {?} */
    WechatQrCodeComponent.prototype.width;
    /**
     * @type {?}
     * @private
     */
    WechatQrCodeComponent.prototype.baseUrl;
}
