// @er/utils/esm2015/lib/pager/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { range } from 'lodash';
export class PagerUtil {
    /**
     * @param {?} totalItems
     * @param {?=} currentPage
     * @param {?=} pageSize
     * @return {?}
     */
    static getPagerModel(totalItems, currentPage = 1, pageSize = 10) {
        /** @type {?} */
        const totalPages = Math.ceil(totalItems / pageSize);
        /** @type {?} */
        let startPage;
        /** @type {?} */
        let endPage;
        if (totalPages <= 10) {
            startPage = 1;
            endPage = totalPages;
        }
        else {
            if (currentPage <= 6) {
                startPage = 1;
                endPage = 10;
            }
            else if (currentPage + 4 >= totalPages) {
                startPage = totalPages - 9;
                endPage = totalPages;
            }
            else {
                startPage = currentPage - 5;
                endPage = currentPage + 4;
            }
        }
        /** @type {?} */
        const startIndex = (currentPage - 1) * pageSize;
        /** @type {?} */
        const endIndex = Math.min(startIndex + pageSize - 1, totalItems - 1);
        /** @type {?} */
        const pages = range(startPage, endPage + 1);
        return {
            totalItems: totalItems,
            currentPage: currentPage,
            pageSize: pageSize,
            totalPages: totalPages,
            startPage: startPage,
            endPage: endPage,
            startIndex: startIndex,
            endIndex: endIndex,
            pages: pages
        };
    }
}
