// @er/primeng/esm2015/lib/components/radio/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { DEFAULT_RADIO_PROPS } from './type';
export class PngRadioComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.valueParser = this.parseValue;
        this.modelFormatter = this.formatModel;
        this.$defaultProps = DEFAULT_RADIO_PROPS;
    }
    /**
     * @return {?}
     */
    get rbValue() {
        return this._rbValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set rbValue(value) {
        this._rbValue = value;
        /** @type {?} */
        const values = this.data$.value;
        if (CommonsUtils.isArray(values)) {
            /** @type {?} */
            const matched = ((/** @type {?} */ (values))).filter(v => v[this.dataKey] === value);
            if (matched.length === 1) {
                this.value = DataItemUtils.getItemValue(matched[0], this.valueKey);
            }
            else {
                this.value = null;
            }
        }
        else {
            this.value = value;
        }
    }
    /**
     * @return {?}
     */
    get dataKey() {
        return CommonsUtils.get(this.$props, '$ext.dataItemProps.idKey', 'value');
    }
    /**
     * @return {?}
     */
    get valueKey() {
        return CommonsUtils.get(this.$props, '$ext.dataItemProps.valueKey', 'value');
    }
    /**
     * @param {?} model
     * @return {?}
     */
    formatModel(model) {
        if (CommonsUtils.isJson(model)) {
            this._rbValue = DataItemUtils.getItemValue(model, this.dataKey);
        }
        return model;
    }
}
PngRadioComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-radio',
                template: `
    <div class="d-flex pb-2" [erApiData]="apiDataProps" [ngClass]="$props.$ext['inline']===false?'flex-column':'flex-wrap'">
      <ng-container *ngFor="let option of data$|async|erDataItems:$props.$ext.dataItemProps;let i=index">
       <span class="d-flex align-items-center">
        <p-radioButton
          class="m-2"
          [(ngModel)]="rbValue"
          erPropsBind
          [props]="$props"
          [context]="ctx"
          [inputId]="i"
          [value]="option.data|erValue:dataKey"
          (onClick)="updateTouch()"
        >
        </p-radioButton>
        <span class="mr-4" [innerHTML]="option.label|erSafeHtml"></span>
       </span>
      </ng-container>
    </div>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngRadioComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngRadioComponent.prototype.valueParser;
    /** @type {?} */
    PngRadioComponent.prototype.modelFormatter;
    /** @type {?} */
    PngRadioComponent.prototype.$defaultProps;
    /** @type {?} */
    PngRadioComponent.prototype._rbValue;
}
