// @er/formly-primeng/esm5/lib/config/formly.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PngFormlyAutoCompleteComponent } from '../field-types/auto-complete/component';
import { PngFormlyButtonComponent } from '../field-types/button/component';
import { PngFormlyCalendarComponent } from '../field-types/calendar/component';
import { PngFormlyCheckboxComponent } from '../field-types/checkbox/component';
import { PngFormlyChipsComponent } from '../field-types/chips/component';
import { PngFormlyFieldDynaComponent } from '../field-types/dyna/component';
import { PngFormlyEditorComponent } from '../field-types/editor/component';
import { PngFormlyFileUploadComponent } from '../field-types/file-upload/component';
import { PngFormlyInputComponent } from '../field-types/input/component';
import { PngFormlyKeyFilterComponent } from '../field-types/key-filter/component';
import { PngFormlyListBoxComponent } from '../field-types/list-box/component';
import { PngFormlyMaskComponent } from '../field-types/mask/component';
import { PngFormlyMultiSelectComponent } from '../field-types/multi-select/component';
import { PngFormlyRadioComponent } from '../field-types/radio/component';
import { PngFormlyRatingComponent } from '../field-types/rating/component';
import { PngFormlySelectButtonComponent } from '../field-types/select-button/component';
import { PngFormlySelectComponent } from '../field-types/select/component';
import { PngFormlySliderComponent } from '../field-types/slider/component';
import { PngFormlySpinnerComponent } from '../field-types/spinner/component';
import { PngFormlySwitchComponent } from '../field-types/switch/component';
import { PngFormlyTableComponent } from '../field-types/table/component';
import { PngFormlyTextAreaComponent } from '../field-types/text-area/component';
import { PngFormlyTreeComponent } from '../field-types/tree/component';
import { PngFormlyFieldValueComponent } from '../field-types/value/component';
import { PngCardWrapperComponent } from '../field-wrappers/card/wrapper';
import { PngFieldSetWrapperComponent } from '../field-wrappers/field-set/wrapper';
import { PngFormlyHorizontalWrapperComponent } from '../field-wrappers/horizontal/wrapper';
import { PngPanelWrapperComponent } from '../field-wrappers/panel/wrapper';
import { PngFormlyResponsiveWrapperComponent } from '../field-wrappers/responsive/wrapper';
import { PngScrollPanelWrapperComponent } from '../field-wrappers/scrol-panel/wrapper';
import { PngFormlyVerticalWrapperComponent } from '../field-wrappers/vertical/wrapper';
import { TemplatePngField } from '../run/field';
import { PngFormlyTypes, PngFormlyWrappers } from '../types/all';
/** @type {?} */
export var PNG_FORMLY_CONFIG = {
    types: [
        {
            name: PngFormlyTypes.input,
            component: PngFormlyInputComponent
        },
        {
            name: PngFormlyTypes.checkbox,
            component: PngFormlyCheckboxComponent
        },
        {
            name: PngFormlyTypes.textArea,
            component: PngFormlyTextAreaComponent
        },
        {
            name: PngFormlyTypes.radio,
            component: PngFormlyRadioComponent
        },
        {
            name: PngFormlyTypes.select,
            component: PngFormlySelectComponent
        },
        {
            name: PngFormlyTypes.button,
            component: PngFormlyButtonComponent
        },
        {
            name: PngFormlyTypes.autoComplete,
            component: PngFormlyAutoCompleteComponent
        },
        {
            name: PngFormlyTypes.calendar,
            component: PngFormlyCalendarComponent
        },
        {
            name: PngFormlyTypes.chips,
            component: PngFormlyChipsComponent
        },
        {
            name: PngFormlyTypes.editor,
            component: PngFormlyEditorComponent
        },
        {
            name: PngFormlyTypes.fileUpload,
            component: PngFormlyFileUploadComponent
        },
        {
            name: PngFormlyTypes.keyFilter,
            component: PngFormlyKeyFilterComponent
        },
        {
            name: PngFormlyTypes.listBox,
            component: PngFormlyListBoxComponent
        },
        {
            name: PngFormlyTypes.mask,
            component: PngFormlyMaskComponent
        },
        {
            name: PngFormlyTypes.switch,
            component: PngFormlySwitchComponent
        },
        {
            name: PngFormlyTypes.multiSelect,
            component: PngFormlyMultiSelectComponent
        },
        {
            name: PngFormlyTypes.rating,
            component: PngFormlyRatingComponent
        },
        {
            name: PngFormlyTypes.slider,
            component: PngFormlySliderComponent
        },
        {
            name: PngFormlyTypes.spinner,
            component: PngFormlySpinnerComponent
        },
        {
            name: PngFormlyTypes.selectButton,
            component: PngFormlySelectButtonComponent
        },
        {
            name: PngFormlyTypes.table,
            component: PngFormlyTableComponent
        },
        {
            name: PngFormlyTypes.tree,
            component: PngFormlyTreeComponent
        },
        {
            name: PngFormlyTypes.value,
            component: PngFormlyFieldValueComponent
        },
        {
            name: PngFormlyTypes.dyna,
            component: PngFormlyFieldDynaComponent
        }
    ],
    wrappers: [
        { name: PngFormlyWrappers.pngVerticalWrapper, component: PngFormlyVerticalWrapperComponent },
        { name: PngFormlyWrappers.pngHorizontalWrapper, component: PngFormlyHorizontalWrapperComponent },
        { name: PngFormlyWrappers.pngResponsiveWrapper, component: PngFormlyResponsiveWrapperComponent },
        { name: PngFormlyWrappers.pngFieldSetWrapper, component: PngFieldSetWrapperComponent },
        { name: PngFormlyWrappers.pngCardWrapper, component: PngCardWrapperComponent },
        { name: PngFormlyWrappers.pngPanelWrapper, component: PngPanelWrapperComponent },
        { name: PngFormlyWrappers.pngScrollPanelWrapper, component: PngScrollPanelWrapperComponent }
    ],
    manipulators: [
        { class: TemplatePngField, method: 'run' }
    ]
};
