// @er/core/esm2015/lib/directives/props-bind/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, Input, KeyValueDiffers, Renderer2, SimpleChange, ViewContainerRef } from '@angular/core';
import { CommonsUtils, IdUtils } from '@er/utils';
import { Observable } from 'rxjs';
import { BaseDirective } from '../../bases/directive';
import { WrapperComponent } from '../../bases/wrappe';
import { ComponentUtils } from '../../utils/component';
import { LogUtils } from '../../utils/log';
import { changesFromRecord } from './util';
/** @type {?} */
const recordToChanges = changesFromRecord({ isFirstChanges: true });
/** @type {?} */
const recordToNewChanges = changesFromRecord({ onlyNewChanges: true });
export class PropsBindDirective extends BaseDirective {
    /**
     * @param {?} differs
     * @param {?} elementRef
     * @param {?} renderer
     * @param {?} viewContainerRef
     */
    constructor(differs, elementRef, renderer, viewContainerRef) {
        super();
        this.differs = differs;
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.viewContainerRef = viewContainerRef;
        this.autoBind = true;
        this.differ = this.differs.find({}).create();
        this._hostComponentInst = undefined;
        this._viewComponentRef = undefined;
    }
    /**
     * @return {?}
     */
    get hostComponentInst() {
        if (this._hostComponentInst === undefined) {
            this._hostComponentInst = ComponentUtils.getHostComponentInst(this.viewContainerRef);
        }
        return this._hostComponentInst;
    }
    /**
     * @return {?}
     */
    get viewComponentRef() {
        if (this._viewComponentRef === undefined) {
            this._viewComponentRef = ComponentUtils.getViewComponentRef(this.viewContainerRef);
        }
        return this._viewComponentRef;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.autoBind) {
            if (changes.props && !CommonsUtils.isEmpty(this.props)) {
                this.resolveProps();
                this.differs.find({}).create();
                LogUtils.debug(this, '属性绑定', this.props, this.inputs, this.outputs, this.hooks);
                if (this.methods) {
                    this.bindMethods();
                }
                if (this.inputs) {
                    /** @type {?} */
                    const inputsChanges = this.differ.diff(this.inputs);
                    this.expInputs = this.getExpressionInputs(this.inputs);
                    if (inputsChanges) {
                        this.updateInputsChanges(inputsChanges);
                        this.updateInputs(!this.lastInputChanges);
                    }
                }
                if (this.outputs) {
                    /** @type {?} */
                    const outputsChanges = this.differ.diff(this.outputs);
                    if (outputsChanges) {
                        this.bindOutputs();
                    }
                }
                if (this.hooks) {
                    this.bindLifeCycles();
                }
                if (changes.props.isFirstChange()) {
                    if (this.hostComponentInst._propsBind) {
                        this.hostComponentInst._propsBind();
                    }
                }
            }
            if (changes.attrs && this.attrs) {
                this.resolveAttrs();
                /** @type {?} */
                const attrChanges = this.differ.diff(this.attrs);
                if (attrChanges) {
                    this.lastAttrChanges = this.changesToAttrChanges(attrChanges);
                    this.updateAttributes(this.lastAttrChanges);
                }
                // this.expAttrs = this.getExpressionInputs(this.attrs);
            }
        }
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        if (this.attrs && this.autoBind) {
            if (this.expAttrs && Object.keys(this.expAttrs).length > 0) {
                Object.keys(this.expAttrs).forEach(key => {
                    this.attrs[key] = this.expAttrs[key].call(this, this.getEventContext());
                });
            }
            /** @type {?} */
            const attrChanges = this.differ.diff(this.attrs);
            if (attrChanges) {
                this.lastAttrChanges = this.changesToAttrChanges(attrChanges);
                this.updateAttributes(this.lastAttrChanges);
            }
        }
        if (this.inputs && this.autoBind) {
            if (this.expInputs && Object.keys(this.expInputs).length > 0) {
                Object.keys(this.expInputs).forEach(key => {
                    /** @type {?} */
                    const result = this.expInputs[key](this.getEventContext());
                    if (result instanceof Observable) {
                        this.$subscriptions = ((/** @type {?} */ (result))).subscribe(data => {
                            this.inputs[key] = data;
                        });
                    }
                    else {
                        this.inputs[key] = result;
                    }
                });
            }
            /** @type {?} */
            const inputsChanges = this.differ.diff(this.inputs);
            if (inputsChanges) {
                /** @type {?} */
                const isNotFirstChange = !!this.lastInputChanges;
                this.updateInputsChanges(inputsChanges);
                if (isNotFirstChange) {
                    this.updateInputs();
                }
            }
        }
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    getEventContext(event) {
        return Object.assign({}, event, { $from: this.hostComponentInst }, this.context || {});
    }
    /**
     * @private
     * @param {?=} isFirstChange
     * @return {?}
     */
    updateInputs(isFirstChange = false) {
        if (!this.hostComponentInst || !this.inputs) {
            return;
        }
        Object.keys(this.inputs)
            .forEach(p => (this.hostComponentInst[p] = this.inputs[p]));
        this.notifyOnInputChanges(this.lastInputChanges, isFirstChange);
    }
    /**
     * @private
     * @return {?}
     */
    bindOutputs() {
        this.stopSubscribe();
        if (!this.outputs) {
            return;
        }
        Object.keys(this.outputs)
            .filter(p => this.hostComponentInst[p])
            .forEach(p => this.$subscriptions = ((/** @type {?} */ (this.hostComponentInst[p])))
            .subscribe(event => {
            try {
                this.outputs[p](this.getEventContext(event));
            }
            catch (e) {
                LogUtils.error(this.hostComponentInst, '操作失败', e);
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    bindLifeCycles() {
        /** @type {?} */
        const constructor = ComponentUtils.getConstructor(this.hostComponentInst);
        /** @type {?} */
        const destructor = constructor.prototype['ngOnDestroy'];
        ComponentUtils.bindLifeCycles(this.hostComponentInst, this.hooks, constructor, destructor);
    }
    /**
     * @private
     * @return {?}
     */
    bindMethods() {
        if (this.methods) {
            Object.keys(this.methods).forEach(method => {
                this.hostComponentInst[method] = this.methods[method].bind(this.hostComponentInst);
            });
        }
    }
    /**
     * @private
     * @param {?=} changes
     * @param {?=} forceFirstChanges
     * @return {?}
     */
    notifyOnInputChanges(changes = {}, forceFirstChanges) {
        if (!this.hostComponentInst.ngOnChanges) {
            return;
        }
        if (forceFirstChanges) {
            changes = this.collectFirstChanges();
        }
        this.hostComponentInst.ngOnChanges(changes);
        if (this.hostComponentInst.cd && this.hostComponentInst.cd.detectChanges) {
            this.hostComponentInst.cd.detectChanges();
        }
    }
    /**
     * @private
     * @param {?} props
     * @return {?}
     */
    getExpressionInputs(props) {
        if (props) {
            /** @type {?} */
            const expProps = {};
            Object.keys(props).forEach(key => {
                if (props[key] instanceof Function) {
                    expProps[key] = props[key];
                }
            });
            return expProps;
        }
        return undefined;
    }
    /**
     * @private
     * @param {?} differ
     * @return {?}
     */
    updateInputsChanges(differ) {
        this.lastInputChanges = this.collectChangesFromDiffer(differ);
    }
    /**
     * @private
     * @return {?}
     */
    collectFirstChanges() {
        /** @type {?} */
        const changes = (/** @type {?} */ ({}));
        /** @type {?} */
        const inputs = this.inputs;
        Object.keys(inputs)
            .forEach(prop => (changes[prop] = new SimpleChange(undefined, inputs[prop], true)));
        return changes;
    }
    /**
     * @private
     * @param {?} differ
     * @return {?}
     */
    collectChangesFromDiffer(differ) {
        /** @type {?} */
        const changes = (/** @type {?} */ ({}));
        differ.forEachAddedItem(recordToChanges(changes));
        differ.forEachItem(recordToNewChanges(changes));
        return changes;
    }
    /**
     * @private
     * @param {?} changes
     * @return {?}
     */
    changesToAttrChanges(changes) {
        /** @type {?} */
        const attrChanges = {
            set: {},
            remove: []
        };
        changes.forEachAddedItem(r => (attrChanges.set[r.key] = r.currentValue));
        changes.forEachChangedItem(r => (attrChanges.set[r.key] = r.currentValue));
        changes.forEachRemovedItem(r => attrChanges.remove.push(r.key));
        return attrChanges;
    }
    /**
     * @private
     * @param {?} actions
     * @return {?}
     */
    updateAttributes(actions) {
        Object.keys(actions.set).forEach(key => this.setAttribute(key, actions.set[key]));
        actions.remove.forEach(key => this.removeAttribute(key));
    }
    /**
     * @private
     * @param {?} name
     * @param {?} value
     * @param {?=} namespace
     * @return {?}
     */
    setAttribute(name, value, namespace) {
        this.renderer.setAttribute(this.elementRef.nativeElement, name, value, namespace);
    }
    /**
     * @private
     * @param {?} name
     * @param {?=} namespace
     * @return {?}
     */
    removeAttribute(name, namespace) {
        this.renderer.removeAttribute(this.elementRef.nativeElement, name, namespace);
    }
    /**
     * @private
     * @return {?}
     */
    resolveProps() {
        if (this.hostComponentInst && CommonsUtils.isJson(this.props)) {
            this.inputs = {};
            this.outputs = {};
            if (this.hostComponentInst._initProps) {
                this.hostComponentInst.$props = CommonsUtils.merge({}, this.props);
                this.hostComponentInst._initProps();
                this.props = CommonsUtils.merge({}, this.hostComponentInst.$props);
            }
            if (!this.props['$id']) {
                this.props['$id'] = IdUtils.getRandom();
            }
            if (this.props['methods']) {
                this.methods = this.props['methods'];
                delete this.props['methods'];
            }
            if (this.props['hooks']) {
                this.hooks = this.props['hooks'];
                delete this.props['hooks'];
            }
            // else if (this.props['$ext'] && this.props['$ext'].hooks) {
            //   this.hooks = this.props['$ext'].hooks;
            //   delete this.props['$ext'].hooks;
            // }
            if (this.hostComponentInst instanceof WrapperComponent) {
                this.props = CommonsUtils.get(this.props, '$ext', {});
                if (this.props['hooks']) {
                    this.hooks = this.props['hooks'];
                    delete this.props['hooks'];
                }
            }
            else {
                this.props = CommonsUtils.omit(this.props, '$ext');
            }
            Object.keys(this.props).forEach(key => {
                if (this.hostComponentInst[key] instanceof EventEmitter) {
                    this.outputs[key] = this.props[key];
                }
                else {
                    this.inputs[key] = this.props[key];
                }
            });
            if (Object.keys(this.inputs).length === 0) {
                this.inputs = undefined;
            }
            if (Object.keys(this.outputs).length === 0) {
                this.outputs = undefined;
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    resolveAttrs() {
        if (this.hostComponentInst instanceof WrapperComponent) {
            this.attrs = CommonsUtils.get(this.attrs, '$ext', {});
        }
        else {
            this.attrs = CommonsUtils.omit(this.attrs, '$ext');
        }
    }
}
PropsBindDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erPropsBind]'
            },] }
];
/** @nocollapse */
PropsBindDirective.ctorParameters = () => [
    { type: KeyValueDiffers },
    { type: ElementRef },
    { type: Renderer2 },
    { type: ViewContainerRef }
];
PropsBindDirective.propDecorators = {
    props: [{ type: Input }],
    context: [{ type: Input }],
    attrs: [{ type: Input }],
    autoBind: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PropsBindDirective.prototype.props;
    /** @type {?} */
    PropsBindDirective.prototype.context;
    /** @type {?} */
    PropsBindDirective.prototype.attrs;
    /** @type {?} */
    PropsBindDirective.prototype.autoBind;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.methods;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.inputs;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.outputs;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.expInputs;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.expAttrs;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.hooks;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.lastInputChanges;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.lastAttrChanges;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.differ;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype._hostComponentInst;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype._viewComponentRef;
    /**
     * @type {?}
     * @private
     */
    PropsBindDirective.prototype.differs;
    /**
     * @type {?}
     * @protected
     */
    PropsBindDirective.prototype.elementRef;
    /**
     * @type {?}
     * @protected
     */
    PropsBindDirective.prototype.renderer;
    /**
     * @type {?}
     * @protected
     */
    PropsBindDirective.prototype.viewContainerRef;
}
