// @er/validate/esm2015/lib/components/form/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ContentChild, Input, Optional, TemplateRef } from '@angular/core';
import { ControlContainer, FormGroup } from '@angular/forms';
import { CommonsUtils } from '@er/utils';
import { Subject } from 'rxjs';
import { ValidateUtils } from '../../utils';
export class FormValidateMessagesComponent {
    /**
     * @param {?} controlContainer
     */
    constructor(controlContainer) {
        this.controlContainer = controlContainer;
        this.errors$ = new Subject();
    }
    /**
     * @return {?}
     */
    checkFormValidate() {
        if (this.form.invalid) {
            alert();
            this.errors$.next(ValidateUtils.getFormErrors(this.form, this.messages, this.fields));
        }
        else {
            this.errors$.next(undefined);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.form && this.controlContainer) {
            this.form = this.controlContainer.formDirective ? ((/** @type {?} */ (this.controlContainer.formDirective))).form : null;
        }
        if (this.form) {
            this.subscription = this.form.statusChanges
                .subscribe(() => {
                this.checkFormValidate();
            });
        }
    }
    /**
     * @param {?} errors
     * @return {?}
     */
    errorKeys(errors) {
        return Object.keys(errors);
    }
    /**
     * @param {?} error
     * @return {?}
     */
    errorMsg(error) {
        return CommonsUtils.getText(error);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
}
FormValidateMessagesComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-form-validate-messages',
                template: "<div *ngIf=\"errors$|async as errors\">\n  <ng-container *ngTemplateOutlet=\"errorMsgTpl?errorMsgTpl:defaultTpl;context:{$implicit:errors}\"></ng-container>\n</div>\n\n<ng-template #defaultTpl let-errors>\n  <div class=\"text-danger\" [ngClass]=\"styleClass\">\n    <ul *ngFor=\"let key of errorKeys(errors)\" class=\"d-flex flex-column justify-content-start\">\n     <li class=\"my-2\">{{key}}</li>\n    </ul>\n  </div>\n</ng-template>",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
FormValidateMessagesComponent.ctorParameters = () => [
    { type: ControlContainer, decorators: [{ type: Optional }] }
];
FormValidateMessagesComponent.propDecorators = {
    form: [{ type: Input }],
    messages: [{ type: Input }],
    fields: [{ type: Input }],
    styleClass: [{ type: Input }],
    errorMsgTpl: [{ type: ContentChild, args: [TemplateRef,] }]
};
if (false) {
    /** @type {?} */
    FormValidateMessagesComponent.prototype.form;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.messages;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.fields;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.styleClass;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.errorMsgTpl;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.errors$;
    /** @type {?} */
    FormValidateMessagesComponent.prototype.subscription;
    /**
     * @type {?}
     * @private
     */
    FormValidateMessagesComponent.prototype.controlContainer;
}
