// @er/primeng/esm5/lib/components/table/component/query/field-filter/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent, DataFieldUtils } from '@er/core';
import { DataAttr, DataType } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { PngTableComponent } from '../../table/component';
var PngTableFieldDataFilterComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableFieldDataFilterComponent, _super);
    function PngTableFieldDataFilterComponent(pngTable) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        _this.onFilterChange = new EventEmitter();
        return _this;
    }
    Object.defineProperty(PngTableFieldDataFilterComponent.prototype, "column", {
        get: /**
         * @return {?}
         */
        function () {
            return this._column;
        },
        set: /**
         * @param {?} column
         * @return {?}
         */
        function (column) {
            if (!this.queryKey && column) {
                this.queryKey = DataFieldUtils.getDefaultFieldName(column.dataField) || column.key;
            }
            if (!this.queryDataType && column) {
                this.queryDataType = column.filterDataType || column.dataType;
            }
            if (!this.queryDataAttr && column && column.dataAttr) {
                this.queryDataAttr = CommonsUtils.getArrayValue(column.dataAttr)[0];
            }
            this._column = column;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PngTableFieldDataFilterComponent.prototype, "dataField", {
        get: /**
         * @return {?}
         */
        function () {
            return this._dataField;
        },
        set: /**
         * @param {?} dataField
         * @return {?}
         */
        function (dataField) {
            if (!this.queryKey && dataField) {
                this.queryKey = DataFieldUtils.getDefaultFieldName(dataField);
            }
            if (!this.queryDataType && dataField) {
                this.queryDataType = dataField.dataType;
            }
            this._dataField = dataField;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PngTableFieldDataFilterComponent.prototype.onFilter = /**
     * @return {?}
     */
    function () {
        this.onFilterChange.emit({
            key: this.queryKey,
            column: this.column,
            value: this.queryValue
        });
    };
    /**
     * @return {?}
     */
    PngTableFieldDataFilterComponent.prototype.clearFilter = /**
     * @return {?}
     */
    function () {
        this.queryValue = undefined;
        this.onFilter();
    };
    PngTableFieldDataFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-field-data-filter',
                    template: "\n    <div class=\"d-flex justify-content-end align-items-end\">\n      <png-query-input [queryKey]=\"queryKey\" [queryDataType]=\"queryDataType\" [queryDataAttr]=\"queryDataAttr\"\n                       [(queryValue)]=\"queryValue\"></png-query-input>\n      <ng-container *ngIf=\"queryValue\">\n      <span class=\"ml-2\">\n       <p-button\n         type=\"button\"\n         icon=\"fa fa-search\"\n         label=\"\u67E5\u8BE2\"\n         (click)=\"onFilter()\"\n       ></p-button>\n    </span>\n        <span class=\"ml-2\">\n    <p-button\n      type=\"button\"\n      icon=\"fa fa-trash\"\n      label=\"\u6E05\u9664\"\n      (click)=\"clearFilter()\"\n    ></p-button>\n    </span>\n      </ng-container>\n    </div>\n  "
                }] }
    ];
    /** @nocollapse */
    PngTableFieldDataFilterComponent.ctorParameters = function () { return [
        { type: PngTableComponent }
    ]; };
    PngTableFieldDataFilterComponent.propDecorators = {
        queryKey: [{ type: Input }],
        queryDataType: [{ type: Input }],
        queryDataAttr: [{ type: Input }],
        queryValue: [{ type: Input }],
        onFilterChange: [{ type: Output }],
        column: [{ type: Input }],
        dataField: [{ type: Input }]
    };
    return PngTableFieldDataFilterComponent;
}(BaseComponent));
export { PngTableFieldDataFilterComponent };
if (false) {
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryKey;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryDataType;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryDataAttr;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.queryValue;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.onFilterChange;
    /**
     * @type {?}
     * @private
     */
    PngTableFieldDataFilterComponent.prototype._column;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype._dataField;
    /** @type {?} */
    PngTableFieldDataFilterComponent.prototype.pngTable;
}
