// @er/primeng/esm5/lib/components/chart/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { ApiUtils, BaseComponent } from '@er/core';
import { ChartType } from '@er/types';
import { ChartUtils, CommonsUtils, IdUtils } from '@er/utils';
import { BehaviorSubject } from 'rxjs';
var PngChartComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngChartComponent, _super);
    function PngChartComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = ChartType.BAR;
        _this.plugins = [];
        _this.responsive = true;
        _this.switchable = true;
        _this.templateNames = {
            header: 'header'
        };
        _this.onDataSelect = new EventEmitter();
        _this.data$ = new BehaviorSubject(undefined);
        _this.chartTypes = [
            { label: '柱状图', value: ChartType.BAR, icon: 'fa fa-bar-chart' },
            { label: '圆饼图', value: ChartType.PIE, icon: 'fa fa-pie-chart' },
            { label: '曲线图', value: ChartType.LINE, icon: 'fa fa-line-chart' },
            { label: '雷达图', value: ChartType.RADAR, icon: 'fa fa-bar' },
            { label: '区域图', value: ChartType.AREA, icon: 'fa fa-area-chart' },
            { label: '环形图', value: ChartType.DOUGHNUT, icon: 'fa fa-circle' }
        ];
        return _this;
    }
    Object.defineProperty(PngChartComponent.prototype, "data", {
        get: /**
         * @return {?}
         */
        function () {
            return this.data$.value;
        },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this.data$.next(data || {});
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    PngChartComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.rawData = undefined;
        this.buildChart();
    };
    /**
     * @return {?}
     */
    PngChartComponent.prototype.changeType = /**
     * @return {?}
     */
    function () {
        this.buildChart();
    };
    /**
     * @return {?}
     */
    PngChartComponent.prototype.reBuildChart = /**
     * @return {?}
     */
    function () {
        this.rawData = undefined;
        this.buildChart();
    };
    /**
     * @return {?}
     */
    PngChartComponent.prototype.buildChart = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.$id) {
            this.$id = IdUtils.getRandom();
        }
        if (this.rawData) {
            this.resolvedRawData(this.rawData);
        }
        else if (this.dataSource) {
            /** @type {?} */
            var ds = this.dataSource;
            if (!ds.hasOwnProperty('apiDataProps')) {
                /** @type {?} */
                var apiDataProps = ds;
                ds = {};
                ds['apiDataProps'] = apiDataProps;
            }
            if (this.resolveDataSource) {
                ds = this.resolveDataSource.method((/** @type {?} */ (ds)));
            }
            ApiUtils.batchFetch(ds, this.$id)
                .subscribe(function (data) {
                _this.rawData = data;
                _this.resolvedRawData(_this.rawData);
            });
        }
    };
    /**
     * @param {?} rawData
     * @return {?}
     */
    PngChartComponent.prototype.resolvedRawData = /**
     * @param {?} rawData
     * @return {?}
     */
    function (rawData) {
        var _this = this;
        if (!CommonsUtils.isEmpty(rawData)) {
            /** @type {?} */
            var data = void 0;
            /** @type {?} */
            var datasets_1 = [];
            /** @type {?} */
            var labels_1 = [];
            if (CommonsUtils.isFunction(this.dimensions)) {
                data = ((/** @type {?} */ (this.dimensions)))(rawData);
            }
            else {
                if (this.dimensions) {
                    /** @type {?} */
                    var dimensions = CommonsUtils.getArrayValue(this.dimensions);
                    dimensions.forEach(function (dimension, index) {
                        if (_this.resolveDimension) {
                            dimension = _this.resolveDimension.method(dimension);
                        }
                        /** @type {?} */
                        var chartData = ChartUtils.buildChartData(rawData, {
                            type: _this.type,
                            dimension: dimension,
                            dataItemProps: _this.dataItemProps
                        }, index);
                        datasets_1.push.apply(datasets_1, tslib_1.__spread(chartData.datasets));
                        if (labels_1.length === 0) {
                            labels_1 = chartData.labels;
                        }
                    });
                    data = { labels: labels_1, datasets: datasets_1 };
                }
                else {
                    data = ChartUtils.buildChartData(rawData, {
                        type: this.type,
                        dataItemProps: this.dataItemProps
                    });
                }
            }
            this.data = data;
        }
        else {
            this.data = {};
        }
    };
    PngChartComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-chart',
                    template: "\n    <p-panel [toggleable]=\"true\" [showHeader]=\"!!caption\">\n      <p-header>\n        <i [ngClass]=\"captionIcon || 'fa fa-bar-chart'\"></i>\n        <span [innerHTML]=\"caption|erSafeHtml\"></span>\n        <ng-container *ngTemplateOutlet=\"templates[templateNames.header]\"></ng-container>\n      </p-header>\n      <er-http-status class=\"mr-2\" [tag]=\"$id\"></er-http-status>\n      <div class=\"p-3\" *ngIf=\"switchable\">\n        <p-selectButton erPosition=\"center\" [options]=\"chartTypes\" [(ngModel)]=\"type\" (onChange)=\"changeType()\"></p-selectButton>\n      </div>\n      <p-chart\n        [type]=\"type\"\n        [options]=\"options\"\n        [data]=\"data$|async\"\n        [plugins]=\"plugins\"\n        [width]=\"width \"\n        [height]=\"height\"\n        [responsive]=\"responsive\"\n        (onDataSelect)=\"onDataSelect.emit($event)\"\n      >\n      </p-chart>\n    </p-panel>\n\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PngChartComponent.propDecorators = {
        $id: [{ type: Input }],
        type: [{ type: Input }],
        rawData: [{ type: Input }],
        options: [{ type: Input }],
        caption: [{ type: Input }],
        captionIcon: [{ type: Input }],
        query: [{ type: Input }],
        plugins: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }],
        responsive: [{ type: Input }],
        switchable: [{ type: Input }],
        dataSource: [{ type: Input }],
        dataItemProps: [{ type: Input }],
        dimensions: [{ type: Input }],
        resolveDataSource: [{ type: Input }],
        resolveDimension: [{ type: Input }],
        onDataSelect: [{ type: Output }],
        data: [{ type: Input }]
    };
    return PngChartComponent;
}(BaseComponent));
export { PngChartComponent };
if (false) {
    /** @type {?} */
    PngChartComponent.prototype.$id;
    /** @type {?} */
    PngChartComponent.prototype.type;
    /** @type {?} */
    PngChartComponent.prototype.rawData;
    /** @type {?} */
    PngChartComponent.prototype.options;
    /** @type {?} */
    PngChartComponent.prototype.caption;
    /** @type {?} */
    PngChartComponent.prototype.captionIcon;
    /** @type {?} */
    PngChartComponent.prototype.query;
    /** @type {?} */
    PngChartComponent.prototype.plugins;
    /** @type {?} */
    PngChartComponent.prototype.width;
    /** @type {?} */
    PngChartComponent.prototype.height;
    /** @type {?} */
    PngChartComponent.prototype.responsive;
    /** @type {?} */
    PngChartComponent.prototype.switchable;
    /** @type {?} */
    PngChartComponent.prototype.dataSource;
    /** @type {?} */
    PngChartComponent.prototype.dataItemProps;
    /** @type {?} */
    PngChartComponent.prototype.dimensions;
    /** @type {?} */
    PngChartComponent.prototype.resolveDataSource;
    /** @type {?} */
    PngChartComponent.prototype.resolveDimension;
    /** @type {?} */
    PngChartComponent.prototype.templateNames;
    /** @type {?} */
    PngChartComponent.prototype.onDataSelect;
    /** @type {?} */
    PngChartComponent.prototype.data$;
    /** @type {?} */
    PngChartComponent.prototype.chartTypes;
}
