// @er/utils/esm2015/lib/regex/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { RegPatterns } from '@er/types';
import { CommonsUtils } from '../comm';
export class RegUtils {
    /**
     * @param {?} regexStr
     * @return {?}
     */
    static escapeRegexp(regexStr) {
        /** @type {?} */
        const result = regexStr.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');
        return result;
    }
    /**
     * @param {?} pattern
     * @return {?}
     */
    static getRegexps(pattern) {
        if (!pattern) {
            return [];
        }
        /** @type {?} */
        const exps = [];
        /** @type {?} */
        const patterns = [];
        if (CommonsUtils.isString(pattern)) {
            patterns.push(pattern);
        }
        else {
            patterns.concat(pattern);
        }
        for (const p of patterns) {
            if (p) {
                if (p.indexOf(',') > 0) {
                    exps.concat(RegUtils.getRegexps(p.split(',')));
                }
                else {
                    exps.push(RegUtils.getRegexp(p));
                }
            }
        }
        return exps;
    }
    /**
     * @param {?} src
     * @param {?} pattern
     * @return {?}
     */
    static test(src, pattern) {
        /** @type {?} */
        let regex;
        if (typeof pattern === 'string') {
            regex = new RegExp(pattern);
        }
        else {
            regex = pattern;
        }
        if (regex instanceof RegExp) {
            return regex.test(src);
        }
        else if (regex && Array.isArray(regex)) {
            return regex.every((item) => {
                return RegUtils.test(src, item);
            });
        }
        else if (regex && Array.isArray(regex.and)) {
            return regex.and.every((item) => {
                return RegUtils.test(src, item);
            });
        }
        else if (regex && Array.isArray(regex.or)) {
            return regex.or.some((item) => {
                return RegUtils.test(src, item);
            });
        }
        else if (regex && regex.not) {
            return !RegUtils.test(src, regex.not);
        }
        else {
            return false;
        }
    }
    /**
     * @param {?} string
     * @param {?} regex
     * @return {?}
     */
    static exec(string, regex) {
        /** @type {?} */
        const self = this;
        if (typeof regex === 'string') {
            regex = new RegExp(regex);
        }
        if (regex instanceof RegExp) {
            return regex.exec(string);
        }
        else if (regex && Array.isArray(regex)) {
            return regex.reduce(function (res, item) {
                return (!!res) ? res : self.exec(string, item);
            }, null);
        }
        else {
            return null;
        }
    }
    /**
     * @private
     * @param {?} pattern
     * @return {?}
     */
    static getRegexp(pattern) {
        /** @type {?} */
        let exp = null;
        if (!pattern) {
            return null;
        }
        if (CommonsUtils.isString(pattern)) {
            if (pattern === 'letter') {
                exp = RegPatterns.letter;
            }
            else if (pattern === 'digit') {
                exp = RegPatterns.digit;
            }
            else if (pattern === 'character') {
                exp = RegPatterns.character;
            }
            else if (pattern === 'integer') {
                exp = RegPatterns.integer;
            }
            else if (pattern === 'numeric') {
                exp = RegPatterns.numeric;
            }
            else {
                exp = new RegExp((/** @type {?} */ (pattern)));
            }
        }
        else {
            exp = (/** @type {?} */ (pattern));
        }
        return exp;
    }
}
