// @er/core/esm2015/lib/utils/date-sync/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { BehaviorSubject } from 'rxjs';
import { first } from 'rxjs/operators';
import { HttpUtils } from '../http';
import { LogUtils } from '../log';
export class DateSyncUtils {
    /**
     * @return {?}
     */
    static start() {
        /** @type {?} */
        const timeServer = ConfigUtils.getConfig().api.servers.timer;
        if (timeServer) {
            HttpUtils.request({
                uri: timeServer.uri,
                observe: 'events'
            })
                .pipe(first())
                .subscribe(response => {
                /** @type {?} */
                const dates = response.headers && response.headers.getAll('Date');
                if (dates) {
                    LogUtils.info('DateSyncUtils', '同步服务器时间', dates);
                    if (dates && dates.length > 0) {
                        DateSyncUtils.updateLocalDate(dates[0]);
                    }
                }
            });
        }
    }
    /**
     * @param {?} date
     * @return {?}
     */
    static updateLocalDate(date) {
        /** @type {?} */
        const d = new Date(date);
        DateSyncUtils.timeServer$.next(d);
        DateSyncUtils.stopSyncSchedule();
        DateSyncUtils.startRefreshSchedule();
        DateSyncUtils.startSyncSchedule();
    }
    /**
     * @private
     * @return {?}
     */
    static stopSyncSchedule() {
        if (DateSyncUtils.interval) {
            clearInterval(DateSyncUtils.interval);
        }
    }
    /**
     * @private
     * @return {?}
     */
    static startRefreshSchedule() {
        DateSyncUtils.interval = setInterval(() => {
            /** @type {?} */
            const d = DateSyncUtils.timeServer$.value;
            d.setSeconds(DateSyncUtils.timeServer$.value.getSeconds() + 1);
            DateSyncUtils.timeServer$.next(d);
            StatesUtils.create(StateNames.serverTime, { date: d }, true);
        }, 1000);
    }
    /**
     * @private
     * @return {?}
     */
    static startSyncSchedule() {
        setInterval(() => {
            DateSyncUtils.start();
        }, 1000 * 60 * 15);
    }
}
DateSyncUtils.timeServer$ = new BehaviorSubject(new Date());
DateSyncUtils.interval = undefined;
if (false) {
    /** @type {?} */
    DateSyncUtils.timeServer$;
    /**
     * @type {?}
     * @private
     */
    DateSyncUtils.interval;
}
