// @er/primeng/esm2015/lib/components/table/component/query/fields-builder/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { BaseComponent } from '@er/core';
import { EsQueryMethods } from '@er/types';
import { PngTableComponent } from '../../table/component';
export class PngTableQueryBuilderComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     */
    constructor(pngTable) {
        super();
        this.pngTable = pngTable;
        this.qbConfig = {};
        this.switches = [
            { label: '并且', value: 'and', icon: 'fa fa-fw fa-cc-paypal' },
            { label: '或者', value: 'or', icon: 'fa fa-fw fa-cc-visa' },
            { label: '非', value: 'not', icon: 'fa fa-fw fa-cc-mastercard' }
        ];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.qbConfig.fields = this.pngTable.$props.columns
            .filter(column => column.dataField)
            .map(column => column.dataField);
        this.qbConfig.operators = (field) => Object.keys(EsQueryMethods).map(key => {
            return {
                label: EsQueryMethods[key].label,
                value: EsQueryMethods[key].key
            };
        });
    }
}
PngTableQueryBuilderComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-table-query-builder',
                template: `
    <er-query-builder [config]="qbConfig">
      <!--<ng-template erTypedTemplate="switchGroup" let-qb="qb">-->
      <!--<p-selectButton [options]="switches" [ngModel]="qb.value.condition"></p-selectButton>-->
      <!--</ng-template>-->
      <!--<ng-template erTypedTemplate="buttonGroup" let-qb="qb">-->
      <!--<p-button label="条件" icon="fa fa-plus" (onClick)="qb.addRule()"></p-button>-->
      <!--<p-button label="组" icon="fa fa-group" (onClick)="qb.addRuleSet()"></p-button>-->
      <!--<ng-container *ngIf="!!parentRuleSet && qb.config.allowRuleSet">-->
      <!--<p-button icon="fa fa-minus" (onClick)="qb.removeRuleSet()"></p-button>-->
      <!--</ng-container>-->
      <!--</ng-template>-->
      <ng-template erTypedTemplate="input" let-rule let-qb="qb">
        <png-query-input [options]="pngTable.$props.$ext"></png-query-input>
      </ng-template>
    </er-query-builder>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngTableQueryBuilderComponent),
                        multi: true
                    }],
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngTableQueryBuilderComponent.ctorParameters = () => [
    { type: PngTableComponent }
];
if (false) {
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.qbConfig;
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.switches;
    /** @type {?} */
    PngTableQueryBuilderComponent.prototype.pngTable;
}
