// @er/types/esm2015/lib/query/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const DEFAULT_QUERY_CONDITION = 'AND';
/**
 * @record
 */
export function QueryOperator() { }
if (false) {
    /** @type {?|undefined} */
    QueryOperator.prototype.label;
    /** @type {?|undefined} */
    QueryOperator.prototype.key;
    /** @type {?|undefined} */
    QueryOperator.prototype.inputs;
    /** @type {?|undefined} */
    QueryOperator.prototype.multiple;
    /** @type {?|undefined} */
    QueryOperator.prototype.dataTypes;
    /** @type {?|undefined} */
    QueryOperator.prototype.operator;
    /** @type {?|undefined} */
    QueryOperator.prototype.getValue;
}
/**
 * @record
 */
export function QueryOperators() { }
/**
 * @record
 */
export function Query() { }
if (false) {
    /** @type {?|undefined} */
    Query.prototype.$and;
    /** @type {?|undefined} */
    Query.prototype.$or;
    /** @type {?|undefined} */
    Query.prototype.$not;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function QueryRule() { }
if (false) {
    /** @type {?|undefined} */
    QueryRule.prototype.entity;
    /** @type {?|undefined} */
    QueryRule.prototype.field;
    /** @type {?|undefined} */
    QueryRule.prototype.value;
    /** @type {?|undefined} */
    QueryRule.prototype.operator;
    /** @type {?|undefined} */
    QueryRule.prototype.data;
}
/**
 * @record
 */
export function QueryRuleSet() { }
if (false) {
    /** @type {?|undefined} */
    QueryRuleSet.prototype.condition;
    /** @type {?|undefined} */
    QueryRuleSet.prototype.rules;
}
