// @er/formly-primeng/esm2015/lib/field-wrappers/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PngCardWrapperComponent } from './card/wrapper';
import { PngFieldSetWrapperComponent } from './field-set/wrapper';
import { PngFormlyHorizontalWrapperComponent } from './horizontal/wrapper';
import { PngPanelWrapperComponent } from './panel/wrapper';
import { PngFormlyResponsiveWrapperComponent } from './responsive/wrapper';
import { PngScrollPanelWrapperComponent } from './scrol-panel/wrapper';
import { PngFormlyVerticalWrapperComponent } from './vertical/wrapper';
export { PngCardWrapperComponent } from './card/wrapper';
export { PngFormlyHorizontalWrapperComponent } from './horizontal/wrapper';
export { PngFieldSetWrapperComponent } from './field-set/wrapper';
export { PngPanelWrapperComponent } from './panel/wrapper';
export { PngFormlyVerticalWrapperComponent } from './vertical/wrapper';
export { PngScrollPanelWrapperComponent } from './scrol-panel/wrapper';
export { PngFormlyResponsiveWrapperComponent } from './responsive/wrapper';
/** @type {?} */
export const PNG_FORMLY_WRAPPERS = [
    PngFormlyVerticalWrapperComponent,
    PngFormlyHorizontalWrapperComponent,
    PngFormlyResponsiveWrapperComponent,
    PngPanelWrapperComponent,
    PngCardWrapperComponent,
    PngFieldSetWrapperComponent,
    PngScrollPanelWrapperComponent
];
