// @er/core/esm2015/lib/pipes/tracker/pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Pipe } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BehaviorSubject, timer } from 'rxjs';
import { mapTo, scan, startWith, switchMap, tap } from 'rxjs/operators';
export class TrackerPipe {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        this.cd = cd;
        this.currentNumber = 0;
        this.counter$ = new BehaviorSubject(0);
        //   .subscribe();
    }
    /**
     * @param {?} number
     * @param {?=} offset
     * @return {?}
     */
    transform(number, offset = 20) {
        this.currentNumber = number - CommonsUtils.getRandomNumber(offset);
        this.counter$.next(this.currentNumber);
        this.counter$
            .pipe(switchMap(endRange => {
            return timer(1000, 200)
                .pipe(mapTo(this.positiveOrNegative(endRange, this.currentNumber)), startWith(this.currentNumber), scan((acc, curr) => acc + curr), tap(n => console.log(n))
            //  takeWhile(this.takeUntil(endRange, this.currentNumber))
            );
        })).subscribe(() => this.cd.detectChanges());
        return this.counter$;
    }
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    positiveOrNegative(endRange, currentNumber) {
        return endRange > currentNumber ? 1 : -1;
    }
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    takeUntil(endRange, currentNumber) {
        return endRange > currentNumber
            ? val => val <= endRange
            : val => val >= endRange;
    }
}
TrackerPipe.decorators = [
    { type: Pipe, args: [{
                name: 'erTracker'
            },] }
];
/** @nocollapse */
TrackerPipe.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
if (false) {
    /** @type {?} */
    TrackerPipe.prototype.currentNumber;
    /** @type {?} */
    TrackerPipe.prototype.counter$;
    /**
     * @type {?}
     * @private
     */
    TrackerPipe.prototype.cd;
}
