// @er/validate/esm2015/lib/directives/control-on-blur/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Optional } from '@angular/core';
import { NgControl } from '@angular/forms';
export class ValidateOnBlurMessageDirective {
    /**
     * @param {?} model
     */
    constructor(model) {
        this.model = model;
    }
    /**
     * @return {?}
     */
    get control() {
        if (this.model && this.model.control) {
            return this.model.control;
        }
        else {
            return undefined;
        }
    }
    /**
     * @return {?}
     */
    onFocus() {
        if (this.control) {
            this.validators = this.control.validator;
            this.asyncValidators = this.control.asyncValidator;
            this.control.clearAsyncValidators();
            this.control.clearValidators();
        }
    }
    /**
     * @return {?}
     */
    onBlur() {
        if (this.control) {
            this.control.setAsyncValidators(this.asyncValidators);
            this.control.setValidators(this.validators);
            this.control.updateValueAndValidity();
        }
    }
}
ValidateOnBlurMessageDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erValidateOnBlur]'
            },] }
];
/** @nocollapse */
ValidateOnBlurMessageDirective.ctorParameters = () => [
    { type: NgControl, decorators: [{ type: Optional }] }
];
ValidateOnBlurMessageDirective.propDecorators = {
    onFocus: [{ type: HostListener, args: ['focus',] }],
    onBlur: [{ type: HostListener, args: ['blur',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.validators;
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.asyncValidators;
    /**
     * @type {?}
     * @private
     */
    ValidateOnBlurMessageDirective.prototype.model;
}
