// @er/primeng/esm2015/lib/components/auto-complete/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { DEFAULT_PNG_AUTO_COMPLETE_PROPS } from './type';
export class PngAutoCompleteComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.valueParser = this.parseValue;
        this.$defaultProps = DEFAULT_PNG_AUTO_COMPLETE_PROPS;
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        super.onPropsInit();
        /** @type {?} */
        let placeholder = '请输入';
        if (this.$props.$ext.apiDataProps) {
            if (!this.$props.$ext.apiDataProps.withPrefix &&
                !this.$props.$ext.apiDataProps.withTerm &&
                !this.$props.$ext.apiDataProps.withWildcard) {
                this.$props.$ext.apiDataProps.withPy = this.$props.$ext.apiDataProps.withPy || true;
                this.$props.$ext.apiDataProps.withZh = this.$props.$ext.apiDataProps.withZh || true;
            }
            if (!this.$props.placeholder) {
                if (this.$props.$ext.apiDataProps.withPy) {
                    placeholder += '拼音首字母，';
                }
                if (this.$props.$ext.apiDataProps.withZh) {
                    placeholder += '汉字关键字';
                }
                placeholder += '筛选';
                this.$props.placeholder = placeholder;
            }
        }
        if (this.$props.$ext.dataItemProps) {
            if (!this.$props.field && CommonsUtils.isString(this.$props.$ext.dataItemProps.labelKey)) {
                this.$props.field = (/** @type {?} */ (this.$props.$ext.dataItemProps.labelKey));
            }
            if (this.$props.$ext.dataItemProps.showCounter) {
                if (!this.$props.$ext.apiDataProps.aggProps) {
                    /** @type {?} */
                    const fields = this.$props.$ext.apiDataProps.returnFields;
                    this.$props.$ext.apiDataProps.aggProps = fields;
                }
            }
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    complete(event) {
        this.apiDataProps = Object.assign({}, this.apiDataProps, { queryValue: event.query });
    }
}
PngAutoCompleteComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-auto-complete',
                template: `
    <div class="inner-addon">
      <p-autoComplete
        [(ngModel)]="value"
        erPropsBind
        [props]="$props"
        [context]="ctx"
        [erApiData]="apiDataProps"
        (completeMethod)="complete($event)"
        [suggestions]="data$|async"
        (onBlur)="updateTouch()"
      >
        <ng-template let-item pTemplate="item">
          <er-data-item [customerTpl]="template('_')"
                        [item]="item"
                        [props]="$props.$ext.dataItemProps"
          >
          </er-data-item>
        </ng-template>
      </p-autoComplete>
      <i *ngIf="value && resettable && !disabled" title="清除输入内容" class="fa fa-trash text-muted right" [ngClass]="$props.dropdown? 'clear1': 'clear2'" (click)="reset()"></i>
    </div>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngAutoCompleteComponent),
                        multi: true
                    }],
                styles: [".clear1{right:22px!important}.clear2{right:2px!important}"]
            }] }
];
if (false) {
    /** @type {?} */
    PngAutoCompleteComponent.prototype.valueParser;
    /** @type {?} */
    PngAutoCompleteComponent.prototype.$defaultProps;
}
