// @er/primeng/esm2015/lib/components/auto-complete/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngAutoCompleteProps() { }
if (false) {
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.field;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.scrollHeight;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.dropdown;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.multiple;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.minLength;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.delay;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.inputStyle;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.inputStyleClass;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.inputId;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.placeholder;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.readonly;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.maxlength;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.size;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.appendTo;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.tabindex;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.dataKey;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.autoHighlight;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.type;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.emptyMessage;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.immutable;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.required;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.forceSelection;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.dropdownMode;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.completeMethod;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onFocus;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onBlur;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onKeyUp;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onSelect;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onUnselect;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onDropdownClick;
    /** @type {?|undefined} */
    PngAutoCompleteProps.prototype.onClear;
}
/** @type {?} */
export const DEFAULT_PNG_AUTO_COMPLETE_PROPS = {
    minLength: 0,
    size: 10,
    dropdown: true,
    forceSelection: true,
    $ext: {
        apiDataProps: {
            dataPath: 'items',
            autoFetch: false
        }
    }
};
