// @er/core/esm5/lib/components/http-status/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BaseComponent } from '../../bases/component';
import { HTTP_STATE$, HTTP_TAG_STATE$ } from '../../states/all';
var HttpStatusComponent = /** @class */ (function (_super) {
    tslib_1.__extends(HttpStatusComponent, _super);
    function HttpStatusComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.loadingClass = 'loading-circle';
        return _this;
    }
    Object.defineProperty(HttpStatusComponent.prototype, "httpState$", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var subject;
            if (!this.tag) {
                subject = HTTP_STATE$;
            }
            else {
                subject = HTTP_TAG_STATE$(this.tag);
            }
            return subject;
        },
        enumerable: true,
        configurable: true
    });
    HttpStatusComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-http-status',
                    template: "    \n      <ng-container *ngIf=\"httpState$|async\">\n        <ng-container *ngTemplateOutlet=\"template('_',default)\"></ng-container>\n        <ng-template #default>\n          <div [ngClass]=\"loadingClass\"></div>\n        </ng-template>\n      </ng-container>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    HttpStatusComponent.propDecorators = {
        tag: [{ type: Input }],
        loadingClass: [{ type: Input }]
    };
    return HttpStatusComponent;
}(BaseComponent));
export { HttpStatusComponent };
if (false) {
    /** @type {?} */
    HttpStatusComponent.prototype.tag;
    /** @type {?} */
    HttpStatusComponent.prototype.loadingClass;
}
