// @er/primeng/esm2015/lib/components/table/component/body-cell/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, HostBinding, HostListener, Input, TemplateRef } from '@angular/core';
import { BaseComponent } from '@er/core';
import { PngTableComponent } from '../table/component';
export class PngTableBodyCellComponent extends BaseComponent {
    /**
     * @param {?} pngTable
     */
    constructor(pngTable) {
        super();
        this.pngTable = pngTable;
    }
    /**
     * @return {?}
     */
    get width() {
        return this.column.width;
    }
    /**
     * @return {?}
     */
    get height() {
        return this.column.height;
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    cellContext(event) {
        return {
            $event: event,
            $id: this.pngTable.$props.$id,
            column: this.column,
            row: this.row,
            rowIndex: this.rowIndex,
            pngTable: this.pngTable
        };
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseEnter(event) {
        if (this.column.onMouseEnter) {
            this.column.onMouseEnter.call(this.pngTable, this.cellContext(event));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseOver(event) {
        if (this.column.onMouseOver) {
            this.column.onMouseOver.call(this.pngTable, this.cellContext(event));
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onMouseOut(event) {
        if (this.column.onMouseOut) {
            this.column.onMouseOut.call(this.pngTable, this.cellContext(event));
        }
    }
}
PngTableBodyCellComponent.decorators = [
    { type: Component, args: [{
                selector: '[png-table-body-cell]',
                template: `
    <ng-container *ngIf="customerTpl;else default">
      <ng-container *ngTemplateOutlet="customerTpl;context:cellContext"></ng-container>
    </ng-container>
    <ng-template #default>
      <png-table-body-cell-content
        [customerTpl]="dataCellTpl"
        [column]="column"
        [row]="row"
        [rowIndex]="rowIndex"
      ></png-table-body-cell-content>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
PngTableBodyCellComponent.ctorParameters = () => [
    { type: PngTableComponent }
];
PngTableBodyCellComponent.propDecorators = {
    column: [{ type: Input }],
    row: [{ type: Input }],
    rowIndex: [{ type: Input }],
    dataCellTpl: [{ type: Input }],
    width: [{ type: HostBinding, args: ['style.width',] }],
    height: [{ type: HostBinding, args: ['style.height',] }],
    onMouseEnter: [{ type: HostListener, args: ['mouseenter', ['$event'],] }],
    onMouseOver: [{ type: HostListener, args: ['mouseover', ['$event'],] }],
    onMouseOut: [{ type: HostListener, args: ['mouseout', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    PngTableBodyCellComponent.prototype.column;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.row;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.rowIndex;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.dataCellTpl;
    /** @type {?} */
    PngTableBodyCellComponent.prototype.pngTable;
}
