// @er/types/esm2015/lib/overlay/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const OverlayType = {
    TABLE: 'table',
    CHART: 'chart',
    IMAGE: 'image',
    PANEL: 'panel',
    FRAME: 'frame',
    MENU: 'menu',
    TREE: 'tree',
    COMPONENT: 'component',
};
export { OverlayType };
/**
 * @record
 */
export function Overlay() { }
if (false) {
    /** @type {?|undefined} */
    Overlay.prototype.type;
    /** @type {?|undefined} */
    Overlay.prototype.icon;
    /** @type {?|undefined} */
    Overlay.prototype.content;
    /** @type {?|undefined} */
    Overlay.prototype.context;
    /** @type {?|undefined} */
    Overlay.prototype.styleClass;
    /** @type {?|undefined} */
    Overlay.prototype.style;
    /** @type {?|undefined} */
    Overlay.prototype.overlayStyleClass;
    /** @type {?|undefined} */
    Overlay.prototype.overlayStyle;
    /** @type {?|undefined} */
    Overlay.prototype.from;
}
/**
 * @record
 */
export function OverlayContext() { }
if (false) {
    /** @type {?|undefined} */
    OverlayContext.prototype.$event;
    /** @type {?|undefined} */
    OverlayContext.prototype.type;
    /** @type {?|undefined} */
    OverlayContext.prototype.icon;
    /** @type {?|undefined} */
    OverlayContext.prototype.dismissable;
    /** @type {?|undefined} */
    OverlayContext.prototype.showCloseIcon;
    /** @type {?|undefined} */
    OverlayContext.prototype.content;
    /** @type {?|undefined} */
    OverlayContext.prototype.context;
    /** @type {?|undefined} */
    OverlayContext.prototype.styleClass;
    /** @type {?|undefined} */
    OverlayContext.prototype.style;
    /** @type {?|undefined} */
    OverlayContext.prototype.overlayStyleClass;
    /** @type {?|undefined} */
    OverlayContext.prototype.overlayStyle;
    /** @type {?|undefined} */
    OverlayContext.prototype.id;
    /* Skipping unhandled member: [key: string]: any;*/
}
