// @er/primeng/esm5/lib/components/auto-complete/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { DEFAULT_PNG_AUTO_COMPLETE_PROPS } from './type';
var PngAutoCompleteComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngAutoCompleteComponent, _super);
    function PngAutoCompleteComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.valueParser = _this.parseValue;
        _this.$defaultProps = DEFAULT_PNG_AUTO_COMPLETE_PROPS;
        return _this;
    }
    /**
     * @return {?}
     */
    PngAutoCompleteComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.onPropsInit.call(this);
        /** @type {?} */
        var placeholder = '请输入';
        if (this.$props.$ext.apiDataProps) {
            if (!this.$props.$ext.apiDataProps.withPrefix &&
                !this.$props.$ext.apiDataProps.withTerm &&
                !this.$props.$ext.apiDataProps.withWildcard) {
                this.$props.$ext.apiDataProps.withPy = this.$props.$ext.apiDataProps.withPy || true;
                this.$props.$ext.apiDataProps.withZh = this.$props.$ext.apiDataProps.withZh || true;
            }
            if (!this.$props.placeholder) {
                if (this.$props.$ext.apiDataProps.withPy) {
                    placeholder += '拼音首字母，';
                }
                if (this.$props.$ext.apiDataProps.withZh) {
                    placeholder += '汉字关键字';
                }
                placeholder += '筛选';
                this.$props.placeholder = placeholder;
            }
        }
        if (this.$props.$ext.dataItemProps) {
            if (!this.$props.field && CommonsUtils.isString(this.$props.$ext.dataItemProps.labelKey)) {
                this.$props.field = (/** @type {?} */ (this.$props.$ext.dataItemProps.labelKey));
            }
            if (this.$props.$ext.dataItemProps.showCounter) {
                if (!this.$props.$ext.apiDataProps.aggProps) {
                    /** @type {?} */
                    var fields = this.$props.$ext.apiDataProps.returnFields;
                    this.$props.$ext.apiDataProps.aggProps = fields;
                }
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngAutoCompleteComponent.prototype.complete = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.apiDataProps = tslib_1.__assign({}, this.apiDataProps, { queryValue: event.query });
    };
    PngAutoCompleteComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-auto-complete',
                    template: "\n    <div class=\"inner-addon\">\n      <p-autoComplete\n        [(ngModel)]=\"value\"\n        erPropsBind\n        [props]=\"$props\"\n        [context]=\"ctx\"\n        [erApiData]=\"apiDataProps\"\n        (completeMethod)=\"complete($event)\"\n        [suggestions]=\"data$|async\"\n        (onBlur)=\"updateTouch()\"\n      >\n        <ng-template let-item pTemplate=\"item\">\n          <er-data-item [customerTpl]=\"template('_')\"\n                        [item]=\"item\"\n                        [props]=\"$props.$ext.dataItemProps\"\n          >\n          </er-data-item>\n        </ng-template>\n      </p-autoComplete>\n      <i *ngIf=\"value && resettable && !disabled\" title=\"\u6E05\u9664\u8F93\u5165\u5185\u5BB9\" class=\"fa fa-trash text-muted right\" [ngClass]=\"$props.dropdown? 'clear1': 'clear2'\" (click)=\"reset()\"></i>\n    </div>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngAutoCompleteComponent; }),
                            multi: true
                        }],
                    styles: [".clear1{right:22px!important}.clear2{right:2px!important}"]
                }] }
    ];
    return PngAutoCompleteComponent;
}(DataAwareControlWrapperComponent));
export { PngAutoCompleteComponent };
if (false) {
    /** @type {?} */
    PngAutoCompleteComponent.prototype.valueParser;
    /** @type {?} */
    PngAutoCompleteComponent.prototype.$defaultProps;
}
