// @er/bootstrap/fesm5/er-bootstrap.js
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { __spread, __extends } from 'tslib';
import { ChangeDetectionStrategy, Component, Input, NgModule } from '@angular/core';
import { BaseComponent, ErCoreModule } from '@er/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BsNavPageComponent = /** @class */ (function (_super) {
    __extends(BsNavPageComponent, _super);
    function BsNavPageComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    BsNavPageComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.title = StatesUtils.getValue(StateNames.appTitle);
    };
    BsNavPageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-nav-page',
                    template: "<div class=\"container h-100\">\n  <div class=\"row\">\n    <div class=\"col-12 col-md-3 col-xl-3 py-md-4 bd-sidebar\">\n      <bs-side-nav></bs-side-nav>\n    </div>\n    <main class=\"col-12 col-md-9 col-xl-8 py-md-4 pl-md-4 bd-content\">\n      <div class=\"d-flex\">\n        <div style=\"flex:1;\">\n          <h2 class=\"bd-title\">{{title}}</h2>\n        </div>\n        <div style=\"align-self:center;\">\n        </div>\n      </div>\n\n      <div class=\"card\">\n        <div class=\"card-body\">\n          <router-outlet></router-outlet>\n        </div>\n      </div>\n    </main>\n  </div>\n</div>",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    return BsNavPageComponent;
}(BaseComponent));

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BsSideNavComponent = /** @class */ (function () {
    function BsSideNavComponent() {
        this.items = [];
        this.items = this.routes.filter(function (route) { return route.component; })
            .map(function (route) { return ({
            title: route.data.title,
            url: "/" + route.path,
            data: route.data
        }); });
    }
    BsSideNavComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bs-side-nav',
                    template: "\n    <ul class=\"nav nav-pills flex-column\">\n      <li class=\"nav-item\" routerLinkActive=\"active\" *ngFor=\"let item of items\">\n        <a class=\"nav-link\" routerLink=\"{{item.url}}\" routerLinkActive=\"active\">{{item.title}}</a>\n      </li>\n    </ul>\n  ",
                    styles: [".section-nav{padding-left:0;border-left:1px solid #eee}.section-nav ul{padding-left:1rem}.section-nav ul ul{display:none}.toc-entry{display:block}.toc-entry a{display:block;padding:.125rem 1.5rem;color:#99979c}.toc-entry a:hover{color:#00f;text-decoration:none}.bd-sidebar{order:0;background-color:#fff;height:100%}.bd-sidebar .nav{margin-top:10px}.bd-sidebar .nav>li>a{display:block;color:rgba(0,0,0,.65);font-weight:500}.bd-sidebar .nav>li>a:hover{color:rgba(0,0,0,.85);text-decoration:none;background-color:transparent}.bd-sidebar .nav>.active:hover>a,.bd-sidebar .nav>.active>a{font-weight:500;color:rgba(0,0,0,.85);background-color:transparent}.bd-sidebar .nav>.active>a.active{border:0 solid #343a40;border-left-width:4px;background-color:rgba(248,248,248,.62);border-radius:0}.bd-links{padding-top:1rem;padding-bottom:1rem;margin-right:-15px;margin-left:-15px}.bd-search-docs-toggle{line-height:1;color:grey}.bd-sidenav{display:none}.bd-toc-link{display:block;padding:.25rem 1.5rem;font-weight:500;color:rgba(0,0,0,.65)}.bd-toc-link:hover{color:rgba(0,0,0,.85);text-decoration:none}.bd-toc-item.active{margin-bottom:1rem}.bd-toc-item.active:not(:first-child){margin-top:1rem}.bd-toc-item.active>.bd-toc-link{color:rgba(0,0,0,.85)}.bd-toc-item.active>.bd-toc-link:hover{background-color:transparent}.bd-toc-item.active>.bd-sidenav{display:block}"]
                }] }
    ];
    /** @nocollapse */
    BsSideNavComponent.ctorParameters = function () { return []; };
    BsSideNavComponent.propDecorators = {
        routes: [{ type: Input }]
    };
    return BsSideNavComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var BsComponents = [
    BsSideNavComponent,
    BsNavPageComponent
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ErBootstrapModule = /** @class */ (function () {
    function ErBootstrapModule() {
    }
    ErBootstrapModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        ErCoreModule,
                    ],
                    declarations: __spread(BsComponents),
                    exports: __spread(BsComponents)
                },] }
    ];
    return ErBootstrapModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { ErBootstrapModule, BsNavPageComponent, BsSideNavComponent, BsComponents };
