// @er/validate/esm2015/lib/directives/control/validate-state.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, HostBinding, Input, Optional, ViewContainerRef } from '@angular/core';
import { NgControl } from '@angular/forms';
/* tslint:disable:directive-selector */
export class ValidateStateDirective {
    /**
     * @param {?} componentFactoryResolver
     * @param {?} viewContainerRef
     * @param {?} ngControl
     */
    constructor(componentFactoryResolver, viewContainerRef, ngControl) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.viewContainerRef = viewContainerRef;
        this.ngControl = ngControl;
        this.showError = true;
        this.enableValidation = true;
    }
    /**
     * @return {?}
     */
    get control() {
        if (this.ngControl && this.ngControl.control) {
            return this.ngControl.control;
        }
        else {
            return undefined;
        }
    }
    /**
     * @return {?}
     */
    get isValid() {
        if (!this.showError || (!this.ngControl && this.ngControl.control == null)) {
            return false;
        }
        return this.ngControl && this.ngControl.control && this.ngControl.control.valid;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this.ngControl || !this.enableValidation) {
            return;
        }
        /** @type {?} */
        const attrs = this.viewContainerRef.element.nativeElement.attributes;
        if (attrs) {
            /** @type {?} */
            const tag = !!attrs.getNamedItem('noerror');
            if (this.showError && tag) {
                this.showError = false;
                return;
            }
        }
    }
}
ValidateStateDirective.decorators = [
    { type: Directive, args: [{
                selector: '[formControlName],[formControl],[ngModel]'
            },] }
];
/** @nocollapse */
ValidateStateDirective.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ViewContainerRef },
    { type: NgControl, decorators: [{ type: Optional }] }
];
ValidateStateDirective.propDecorators = {
    messages: [{ type: Input }],
    showError: [{ type: Input }],
    enableValidation: [{ type: Input }],
    isValid: [{ type: HostBinding, args: ['class.er-valid',] }]
};
if (false) {
    /** @type {?} */
    ValidateStateDirective.prototype.messages;
    /** @type {?} */
    ValidateStateDirective.prototype.showError;
    /** @type {?} */
    ValidateStateDirective.prototype.enableValidation;
    /** @type {?} */
    ValidateStateDirective.prototype.currentState;
    /**
     * @type {?}
     * @private
     */
    ValidateStateDirective.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    ValidateStateDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    ValidateStateDirective.prototype.ngControl;
}
