// @er/validate/esm2015/lib/directives/control/validate-message.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, HostBinding, Input, Optional, ViewContainerRef } from '@angular/core';
import { NgControl } from '@angular/forms';
import { CacheUtils } from '@er/utils';
import { ValidateMessagesComponent } from '../../components/control/component';
/* tslint:disable:directive-selector */
export class ValidationMessageDirective {
    /**
     * @param {?} componentFactoryResolver
     * @param {?} viewContainerRef
     * @param {?} ngControl
     */
    constructor(componentFactoryResolver, viewContainerRef, ngControl) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.viewContainerRef = viewContainerRef;
        this.ngControl = ngControl;
        this.showError = true;
        this.enableValidation = true;
    }
    /**
     * @return {?}
     */
    get control() {
        if (this.ngControl && this.ngControl.control) {
            return this.ngControl.control;
        }
        else {
            return undefined;
        }
    }
    /**
     * @return {?}
     */
    get isValid() {
        if (!this.showError || (!this.ngControl && this.ngControl.control == null)) {
            return false;
        }
        return this.ngControl && this.ngControl.control && this.ngControl.control.valid;
    }
    /**
     * @return {?}
     */
    get isInvalid() {
        if (!this.showError || (!this.ngControl && this.ngControl.control == null)) {
            return false;
        }
        return this.ngControl && this.ngControl.control && this.ngControl.control.invalid;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this.ngControl || !this.enableValidation) {
            return;
        }
        /** @type {?} */
        const attrs = this.viewContainerRef.element.nativeElement.attributes;
        if (attrs) {
            /** @type {?} */
            const tag = !!attrs.getNamedItem('noerror');
            if (this.showError && tag) {
                this.showError = false;
                return;
            }
        }
        this.componentFactory = CacheUtils.get('$componentFactory', 'ValidateMessagesComponentFactory');
        if (!this.componentFactory) {
            this.componentFactory = this.componentFactoryResolver.resolveComponentFactory(ValidateMessagesComponent);
            CacheUtils.put('$componentFactory', 'validationMessagesComponentFactory', this.componentFactory);
        }
        /** @type {?} */
        const className = this.viewContainerRef.element.nativeElement.parentElement.className;
        if (className.indexOf('input-group') >= 0) {
        }
        this.componentRef = this.viewContainerRef.createComponent(this.componentFactory);
        ((/** @type {?} */ (this.componentRef.instance))).messages = this.messages;
        ((/** @type {?} */ (this.componentRef.instance))).control = (/** @type {?} */ (this.control));
    }
}
ValidationMessageDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erValidate]'
            },] }
];
/** @nocollapse */
ValidationMessageDirective.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ViewContainerRef },
    { type: NgControl, decorators: [{ type: Optional }] }
];
ValidationMessageDirective.propDecorators = {
    messages: [{ type: Input }],
    showError: [{ type: Input }],
    enableValidation: [{ type: Input }],
    isValid: [{ type: HostBinding, args: ['class.er-valid',] }],
    isInvalid: [{ type: HostBinding, args: ['class.er-invalid',] }]
};
if (false) {
    /** @type {?} */
    ValidationMessageDirective.prototype.messages;
    /** @type {?} */
    ValidationMessageDirective.prototype.showError;
    /** @type {?} */
    ValidationMessageDirective.prototype.enableValidation;
    /** @type {?} */
    ValidationMessageDirective.prototype.currentState;
    /**
     * @type {?}
     * @private
     */
    ValidationMessageDirective.prototype.componentFactory;
    /**
     * @type {?}
     * @private
     */
    ValidationMessageDirective.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    ValidationMessageDirective.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    ValidationMessageDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    ValidationMessageDirective.prototype.ngControl;
}
