// @er/core/esm2015/lib/services/task/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { StateNames } from '@er/types';
import { CommonsUtils, ConfigUtils, StatesUtils } from '@er/utils';
import { timer } from 'rxjs';
import { distinctUntilChanged, filter, map } from 'rxjs/operators';
import { DiUtils } from '../../utils/di';
import * as i0 from "@angular/core";
/** @type {?} */
export const BACKGROUND_TASK_TOKEN = new InjectionToken('BACKGROUND_TASK_TOKEN');
/**
 * used to run background task globally
 */
export class TaskService {
    /**
     * @param {?=} tasks
     */
    constructor(tasks = []) {
        this.subs = [];
        StatesUtils.create(StateNames.appTitle, ConfigUtils.getConfig().appTitle, true);
        setTimeout(() => {
            this.cleanStates();
            /** @type {?} */
            const _tasks = CommonsUtils.flatten(tasks);
            if (_tasks && _tasks.length > 0) {
                _tasks.forEach(task => this.startTask((/** @type {?} */ (task))));
            }
        }, 2000);
    }
    /**
     * @return {?}
     */
    updateTitle() {
        this.subs.push(StatesUtils.observe(StateNames.appTitle)
            .pipe(filter(state => state && state.hasOwnProperty('title')), map(state => state.title), distinctUntilChanged())
            .subscribe(title => {
            DiUtils.get(Title).setTitle(title);
        }));
    }
    /**
     * @return {?}
     */
    cleanStates() {
        this.subs.push(timer(30000, 30000)
            .subscribe(() => {
            /** @type {?} */
            let store = StatesUtils.getStore().value;
            Object.keys(store).forEach(key => {
                if (store[key]) {
                    if (!store[key].preserved && (store[key].state && store[key].state.observers.length === 0)) {
                        store = CommonsUtils.omit(store, key);
                    }
                }
            });
            StatesUtils.getStore().next(store);
        }));
    }
    /**
     * @param {?} task
     * @return {?}
     */
    startTask(task) {
        if (CommonsUtils.isFunction(task)) {
            ((/** @type {?} */ (task)))();
        }
        else {
            /** @type {?} */
            const schedule = (/** @type {?} */ (task));
            /** @type {?} */
            let context = schedule.context;
            if (CommonsUtils.isFunction(schedule.context)) {
                context = schedule.context();
            }
            if (schedule.interval) {
                this.subs.push(timer(0, schedule.interval)
                    .subscribe(() => {
                    schedule.run(context);
                }));
            }
            else {
                schedule.run(context);
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.subs.forEach(sub => {
            sub.unsubscribe();
        });
    }
}
TaskService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TaskService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [BACKGROUND_TASK_TOKEN,] }] }
];
/** @nocollapse */ TaskService.ngInjectableDef = i0.defineInjectable({ factory: function TaskService_Factory() { return new TaskService(i0.inject(BACKGROUND_TASK_TOKEN, 8)); }, token: TaskService, providedIn: "root" });
if (false) {
    /** @type {?} */
    TaskService.prototype.subs;
}
