// @er/types/esm2015/lib/aggs/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DataType } from '../data-type';
/** @enum {string} */
const AggMethods = {
    AVG: 'avg',
    SUM: 'sum',
    MAX: 'max',
    MIN: 'min',
    COUNT: 'count',
    DISTINCT: 'distinct',
    TERMS: 'terms',
    HISTOGRAM: 'histogram',
    DATE_HISTOGRAM: 'dateHistogram',
    CUSTOMER: 'customer',
};
export { AggMethods };
/** @type {?} */
export const AggMethodsProps = {
    avg: { label: '平均值', key: 'avg', dataTypes: [DataType.NUMBER, DataType.DATE] },
    sum: { label: '合计值', key: 'sum', dataTypes: [DataType.NUMBER] },
    max: { label: '最大值', key: 'max', dataTypes: [DataType.NUMBER, DataType.DATE] },
    min: { label: '最小值', key: 'min', dataTypes: [DataType.NUMBER, DataType.DATE] },
    distinct: { label: '唯一值', key: 'distinct' },
    count: { label: '总数', key: 'count' },
    terms: { label: '值分布', key: 'terms' }
};
/** @enum {string} */
const AggInterval = {
    YEAR: 'year',
    MONTH: 'month',
    MILL_SECOND: 'millSecond',
    SECOND: 'second',
    MINUTE: 'minute',
    HOUR: 'hour',
    DAY: 'day',
    WEEK: 'week',
    QUARTER: 'quarter',
};
export { AggInterval };
/** @type {?} */
export const AggIntervalProps = {
    year: { label: '年', key: 'year', format: 'yy' },
    month: { label: '月', key: 'month', format: 'MM' },
    day: { label: '日', key: 'day', format: 'MM-dd' },
    week: { label: '星期', key: 'week', format: 'MM-dd' },
    quarter: { label: '季度', key: 'quarter', format: 'MM' },
    hour: { label: '小时', key: 'hour', format: 'HH' },
    minute: { label: '分钟', key: 'minute', format: 'HH:mm' },
    second: { label: '秒', key: 'second', format: 'ss' },
    millSecond: { label: '毫秒', key: 'millSecond', format: 'ms' }
};
/**
 * @record
 */
export function AggProps() { }
if (false) {
    /** @type {?|undefined} */
    AggProps.prototype.name;
    /** @type {?} */
    AggProps.prototype.field;
    /** @type {?|undefined} */
    AggProps.prototype.method;
    /** @type {?|undefined} */
    AggProps.prototype.interval;
    /** @type {?|undefined} */
    AggProps.prototype.size;
    /** @type {?|undefined} */
    AggProps.prototype.format;
    /** @type {?|undefined} */
    AggProps.prototype.minCount;
    /** @type {?|undefined} */
    AggProps.prototype.missing;
    /** @type {?|undefined} */
    AggProps.prototype.aggs;
    /** @type {?|undefined} */
    AggProps.prototype.sort;
}
