// @er/core/esm2015/lib/er-core.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { InjectionToken, Injector, NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { RouteReuseStrategy, RouterModule } from '@angular/router';
import { NotifyLevel, NotifyType, StateNames, StompStates } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { StompConfig, StompService } from '@stomp/ng2-stompjs';
import { DynamicModule } from 'ng-dynamic-component';
import { NgPipesModule } from 'ngx-pipes';
import { distinctUntilChanged, tap } from 'rxjs/operators';
import { CoreComponents } from './components/all';
import { CoreDirectives } from './directives/all';
import { CoreInterceptors } from './interceptors/all';
import { CorePipes } from './pipes/all';
import { BACKGROUND_TASK_TOKEN } from './services/all';
import { RouterStateService } from './services/router/service';
import { CachedRouteReuseStrategy } from './services/strategy/service';
import { TaskService } from './services/task/service';
import { CORE_TASKS } from './tasks/all';
import { DateSyncUtils } from './utils/date-sync';
import { DiUtils } from './utils/di';
/** @type {?} */
export const APP_CONFIG = new InjectionToken('APP_CONFIG');
/**
 * @param {?} appConfig
 * @return {?}
 */
export function initialize(appConfig) {
    return function () {
        // ConfigUtils.configApp(appConfig);
    };
}
export class ErCoreModule {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        if (!DiUtils.injector) {
            DiUtils.injector = injector;
            // load services
            injector.get(RouterStateService);
            injector.get(TaskService);
            try {
                /** @type {?} */
                const stompService = injector.get(StompService);
                if (stompService && ConfigUtils.getConfig().stomp.enabled) {
                    stompService.state
                        .pipe(distinctUntilChanged(), tap((state) => {
                        StatesUtils.update(StateNames.socket, { state: StompStates[state] });
                        StatesUtils.update(StateNames.notify, {
                            type: state === 0 ? NotifyType.ANNOUNCE : NotifyType.ALERT,
                            message: `消息服务器状态：${StompStates[state]}`,
                            level: state === 0 ? NotifyLevel.ERROR : NotifyLevel.INFO
                        });
                    })).subscribe();
                }
            }
            catch (e) {
            }
        }
    }
    /**
     * @param {?} appConfig
     * @return {?}
     */
    static forRoot(appConfig) {
        return {
            ngModule: ErCoreModule,
            providers: [
                { provide: APP_CONFIG, useValue: appConfig },
                // {
                //   provide: APP_INITIALIZER,
                //   useFactory: initialize,
                //   deps: [APP_CONFIG],
                //   multi: true
                // },
                ...CoreInterceptors,
                appConfig.stomp.enabled ? StompService : [],
                appConfig.stomp.enabled ? { provide: StompConfig, useValue: appConfig.stomp } : [],
                { provide: BACKGROUND_TASK_TOKEN, useValue: DateSyncUtils.start, multi: true },
                { provide: BACKGROUND_TASK_TOKEN, useValue: CORE_TASKS, multi: true },
                {
                    provide: RouteReuseStrategy,
                    useClass: CachedRouteReuseStrategy
                }
            ]
        };
    }
}
ErCoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    RouterModule,
                    HttpClientModule,
                    NgPipesModule,
                    DynamicModule.withComponents(CoreComponents)
                ],
                declarations: [
                    ...CoreComponents,
                    ...CoreDirectives,
                    ...CorePipes
                ],
                exports: [
                    CommonModule,
                    RouterModule,
                    HttpClientModule,
                    NgPipesModule,
                    ...CoreComponents,
                    ...CoreDirectives,
                    ...CorePipes
                ]
            },] }
];
/** @nocollapse */
ErCoreModule.ctorParameters = () => [
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ErCoreModule.prototype.injector;
}
