// @er/formly-primeng/esm2015/lib/components/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { PngFormlyAuditFormComponent } from './audit/component';
import { PngFormDebugComponent } from './debug/component';
import { PngFormlyFormComponent } from './form/component';
import { PngFormlyTabsFormComponent } from './tab/component';
import { PngFormlyWizardFormComponent } from './wizard/component';
export { PngFormlyAuditFormComponent } from './audit/component';
export { PngFormlyFormComponent } from './form/component';
export { PngFormlyWizardFormComponent } from './wizard/component';
export { PngFormlyTabsFormComponent } from './tab/component';
export { PngFormDebugComponent } from './debug/component';
export {} from './tab/type';
export {} from './wizard/type';
/** @type {?} */
export const Components = [
    PngFormlyFormComponent,
    PngFormlyAuditFormComponent,
    PngFormlyWizardFormComponent,
    PngFormlyTabsFormComponent,
    PngFormDebugComponent
];
