// @er/core/esm2015/lib/utils/changes/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonsUtils } from '@er/utils';
export class ChangesUtils {
    /**
     * @param {?} diff
     * @return {?}
     */
    static getChangeByDiff(diff) {
        if (!diff) {
            return {};
        }
        /** @type {?} */
        const changes = {
            changed: [],
            added: [],
            deleted: []
        };
        if (diff) {
            diff.forEachChangedItem(r => {
                changes.changed.push({
                    key: r.key,
                    value: r.currentValue,
                    from: r.previousValue
                });
            });
            diff.forEachAddedItem(r => {
                changes.added.push({
                    key: r.key,
                    value: r.currentValue
                });
            });
            diff.forEachRemovedItem(r => {
                changes.deleted.push({
                    key: r.key,
                    value: r.previousValue
                });
            });
        }
        return CommonsUtils.omitBy(changes, CommonsUtils.lodash().isEmpty);
    }
}
