// @er/primeng/esm5/lib/components/panel/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
import { BaseComponent, DiUtils } from '@er/core';
import { DomUtils } from '@er/utils';
var PngPanelComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngPanelComponent, _super);
    function PngPanelComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.collapsed = false;
        _this.toggleable = false;
        _this.showHeader = true;
        _this.templateNames = {
            header: 'header',
            footer: 'footer'
        };
        return _this;
    }
    /**
     * @return {?}
     */
    PngPanelComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterViewInit.call(this);
        /** @type {?} */
        var titleBar = this.panelEle.nativeElement.querySelector('.ui-panel-titlebar');
        if (titleBar) {
            if (this.barStyleClass) {
                DomUtils.addClasses(titleBar, this.barStyleClass);
            }
            if (this.barStyle) {
                Object.keys(this.barStyle).forEach(function (name) {
                    DiUtils.get(Renderer2).setStyle(titleBar, name, _this.barStyle[name]);
                });
            }
        }
    };
    PngPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-panel',
                    template: "<div class=\"ui-g\">\n  <div class=\"ui-g-12\">\n    <p-panel #pngPanel [styleClass]=\"styleClass\" [style]=\"style\" [toggleable]=\"toggleable\" [showHeader]=\"showHeader\"\n             [collapsed]=\"collapsed\">\n      <p-header>\n        <ng-container *ngIf=\"header\">\n          <i *ngIf=\"icon\" [class]=\"icon\"></i>\n          <er-dyna-content [content]=\"header\" class=\"d-inline-block\"></er-dyna-content>\n        </ng-container>\n        <ng-container *ngIf=\"template(templateNames.header)\">\n          <ng-container *ngTemplateOutlet=\"template(templateNames.header)\"></ng-container>\n        </ng-container>\n      </p-header>\n      <ng-content></ng-content>\n      <ng-container *ngIf=\"footer || template(templateNames.footer)\">\n        <p-footer>\n          <ng-container *ngIf=\"footer\">\n            <er-dyna-content [content]=\"footer\"></er-dyna-content>\n          </ng-container>\n          <ng-container *ngIf=\"template(templateNames.footer)\">\n            <ng-container *ngTemplateOutlet=\"template(templateNames.footer)\"></ng-container>\n          </ng-container>\n        </p-footer>\n      </ng-container>\n    </p-panel>\n  </div>\n</div>\n\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PngPanelComponent.propDecorators = {
        barStyle: [{ type: Input }],
        barStyleClass: [{ type: Input }],
        style: [{ type: Input }],
        styleClass: [{ type: Input }],
        header: [{ type: Input }],
        footer: [{ type: Input }],
        icon: [{ type: Input }],
        collapsed: [{ type: Input }],
        toggleable: [{ type: Input }],
        showHeader: [{ type: Input }],
        panelEle: [{ type: ViewChild, args: ['pngPanel', { read: ElementRef },] }]
    };
    return PngPanelComponent;
}(BaseComponent));
export { PngPanelComponent };
if (false) {
    /** @type {?} */
    PngPanelComponent.prototype.barStyle;
    /** @type {?} */
    PngPanelComponent.prototype.barStyleClass;
    /** @type {?} */
    PngPanelComponent.prototype.style;
    /** @type {?} */
    PngPanelComponent.prototype.styleClass;
    /** @type {?} */
    PngPanelComponent.prototype.header;
    /** @type {?} */
    PngPanelComponent.prototype.footer;
    /** @type {?} */
    PngPanelComponent.prototype.icon;
    /** @type {?} */
    PngPanelComponent.prototype.collapsed;
    /** @type {?} */
    PngPanelComponent.prototype.toggleable;
    /** @type {?} */
    PngPanelComponent.prototype.showHeader;
    /** @type {?} */
    PngPanelComponent.prototype.panelEle;
    /** @type {?} */
    PngPanelComponent.prototype.templateNames;
}
