// @er/primeng/esm2015/lib/work-space/footer/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { distinctUntilChanged, map } from 'rxjs/operators';
import { PngAdminWorkSpaceComponent } from '../component';
export class WorkSpaceFooterComponent {
    /**
     * @param {?} ws
     */
    constructor(ws) {
        this.ws = ws;
    }
    /**
     * @return {?}
     */
    get elapseTime$() {
        return StatesUtils.observe(StateNames.httpTimer)
            .pipe(distinctUntilChanged(), map(v => v && v['elapsed']));
    }
}
WorkSpaceFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-ws-footer',
                template: "<div class=\"layout-footer\">\n  <span class=\"footer-text-left\">\n     <img [src]=\"ws.appLogo\"/>\n  </span>\n  <span class=\"footer-text-right\">\n    <span class=\"text-muted small\" *ngIf=\"elapseTime$|async as elapseTime\">\u7F51\u7EDC\u8BBF\u95EE\u8017\u65F6\uFF1A{{elapseTime}}\u6BEB\u79D2</span>\n  </span>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
WorkSpaceFooterComponent.ctorParameters = () => [
    { type: PngAdminWorkSpaceComponent }
];
if (false) {
    /** @type {?} */
    WorkSpaceFooterComponent.prototype.ws;
}
