// @er/primeng/esm2015/lib/components/search-box/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '@er/core';
import { DialogUtils } from '@er/utils';
export class PngSearchBoxComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.searched = false;
        this.title = '按关键字搜索,输入回车键执行';
        this.onSearch = new EventEmitter();
        this.onSearchChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    search() {
        this.searched = true;
        this.onSearch.emit(this.searchKey);
    }
    /**
     * @return {?}
     */
    keyChange() {
        if (!this.searchKey || this.searchKey.length === 0) {
            this.clearSearch();
        }
        this.onSearchChange.emit(this.searchKey);
    }
    /**
     * @return {?}
     */
    clearSearch() {
        this.searchKey = undefined;
        if (this.searched) {
            this.onSearch.emit(undefined);
        }
        this.searched = false;
    }
    /**
     * @return {?}
     */
    showHelp() {
        DialogUtils.info('查询操作符', `
    <div class="ml-3">
       <div><b class="font-weight-bold">+</b> 并且</div>
       <div><b class="font-weight-bold">|</b> 或者</div>
       <div><b class="font-weight-bold">-</b> 排除</div>
       <div><b class="font-weight-bold">*</b> 开头</div>
       <div><b class="font-weight-bold">( )</b> 优先</div>
    </div>
    `);
    }
}
PngSearchBoxComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-search-box',
                template: `
    <span class="inner-addon" [ngClass]="styleClass" [ngStyle]="style">
      <i class="fa fa-search text-primary left"></i>
      <input [title]="title" pInputText placeholder="输入搜索关键字或拼音" [(ngModel)]="searchKey" (keydown.enter)="search()" (keyup)="keyChange()"/>
      <i *ngIf="searchKey" class="fa fa-trash gray right" [title]="title" (click)="clearSearch()"></i>
   </span>
    <span class="position-container"><i title="多值查询帮助" class="fa fa-question-circle green position-absolute" style="right:2px"
                                        (click)="showHelp()"></i></span>

  `
            }] }
];
PngSearchBoxComponent.propDecorators = {
    styleClass: [{ type: Input }],
    style: [{ type: Input }],
    title: [{ type: Input }],
    onSearch: [{ type: Output }],
    onSearchChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PngSearchBoxComponent.prototype.searchKey;
    /** @type {?} */
    PngSearchBoxComponent.prototype.searched;
    /** @type {?} */
    PngSearchBoxComponent.prototype.styleClass;
    /** @type {?} */
    PngSearchBoxComponent.prototype.style;
    /** @type {?} */
    PngSearchBoxComponent.prototype.title;
    /** @type {?} */
    PngSearchBoxComponent.prototype.onSearch;
    /** @type {?} */
    PngSearchBoxComponent.prototype.onSearchChange;
}
