// @er/primeng/esm5/lib/components/list/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_LIST_BOX_PROPS } from './type';
var PngListBoxComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngListBoxComponent, _super);
    function PngListBoxComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.templateNames = {
            header: 'header',
            footer: 'footer',
            item: 'item'
        };
        _this.$defaultProps = DEFAULT_PNG_LIST_BOX_PROPS;
        return _this;
    }
    PngListBoxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-list-box',
                    template: "\n    <p-listbox\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      [erApiData]=\"apiDataProps\"\n      [options]=\"data$|async\"\n      (onChange)=\"updateTouch()\"\n    >\n      <p-header *ngIf=\"template(templateNames.header)\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.header)\"></ng-container>\n      </p-header>\n\n      <p-header *ngIf=\"header\">\n        <div [innerHTML]=\"header\"></div>\n      </p-header>\n\n      <p-footer *ngIf=\"template(templateNames.footer)\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.footer)\"></ng-container>\n      </p-footer>\n\n      <p-footer *ngIf=\"$props.$ext['showValue'] || footer\">\n        <ng-container *ngIf=\"$props.$ext['showValue']\">\n          \u9009\u4E2D\u503C\uFF1A{{value | erValue: $props.$ext['field']}}\n        </ng-container>\n        <ng-container *ngIf=\"footer\">\n          <div [innerHTML]=\"footer\"></div>\n        </ng-container>\n      </p-footer>\n\n      <ng-template let-item let-i=\"index\" pTemplate=\"item\">\n        <er-data-item [customerTpl]=\"template(templateNames.item)\"\n                      [item]=\"item.value\"\n                      [props]=\"$props.$ext.dataItemProps\">\n        </er-data-item>\n      </ng-template>\n    </p-listbox>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngListBoxComponent; }),
                            multi: true
                        }]
                }] }
    ];
    PngListBoxComponent.propDecorators = {
        header: [{ type: Input }],
        footer: [{ type: Input }]
    };
    return PngListBoxComponent;
}(DataAwareControlWrapperComponent));
export { PngListBoxComponent };
if (false) {
    /** @type {?} */
    PngListBoxComponent.prototype.header;
    /** @type {?} */
    PngListBoxComponent.prototype.footer;
    /** @type {?} */
    PngListBoxComponent.prototype.templateNames;
    /** @type {?} */
    PngListBoxComponent.prototype.$defaultProps;
}
