// @er/formly-primeng/esm2015/lib/components/tab/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Output } from '@angular/core';
import { FormArray, FormGroup } from '@angular/forms';
import { LogUtils } from '@er/core';
import { PngFormlyUtils } from '../../utils';
import { PngFormlyFormComponent } from '../form/component';
export class PngFormlyTabsFormComponent extends PngFormlyFormComponent {
    constructor() {
        super(...arguments);
        this.onTabChange = new EventEmitter();
        this.onTabClose = new EventEmitter();
    }
    /**
     * @param {?} formProps
     * @return {?}
     */
    resolveFormFields(formProps) {
        if (!formProps || !formProps['tabs']) {
            LogUtils.error(this, '没有定义tabs', formProps);
        }
        if (formProps.tabs) {
            formProps.tabs.forEach(tab => {
                tab.fields.forEach(field => {
                    PngFormlyUtils.resolveField(this._formProps, field, this);
                });
            });
            this.formArray = new FormArray(formProps.tabs.map(() => new FormGroup({})));
            this.tabFormlyOptions = formProps.tabs.map((tab) => {
                return Object.assign({}, formProps.options || {}, tab.options || {});
            });
        }
        this.currentTabIndex = formProps.activeIndex || 0;
    }
    /**
     * @return {?}
     */
    afterPropsInit() {
    }
    /**
     * @param {?} event
     * @return {?}
     */
    tabChange(event) {
        this.currentTabIndex = event;
        this.onTabChange.emit(event);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    tabClose(event) {
        this.onTabClose.emit(event);
    }
}
PngFormlyTabsFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-formly-tab',
                template: `
    <png-panel [header]="formProps.caption" [footer]="formProps.footer" [icon]="formProps.icon" [toggleable]="formProps.collapsable">
      <p-tabView [activeIndex]="formProps['activeIndex']"
                 [styleClass]="formProps.styleClass"
                 [style]="formProps.style"
                 [orientation]="formProps['tabOrientation']"
                 (onChange)="tabChange($event)"
                 (onClose)="tabClose($event)"
      >
        <ng-container>
          <p-tabPanel *ngFor="let tab of formProps['tabs']; let i = index"
                      [header]="tab.caption"
                      [headerStyle]="tab.style"
                      [headerStyleClass]="tab.styleClass"
                      [leftIcon]="tab.leftIcon"
                      [rightIcon]="tab.rightIcon"
          >
            <formly-form
              [form]="formArray.at(i)"
              [(model)]="formProps.model"
              [fields]="tab.fields"
              [options]="tabFormlyOptions[i]">
            </formly-form>
          </p-tabPanel>
        </ng-container>
      </p-tabView>
    </png-panel>

    <div [ngClass]="formProps.buttonsStyleClass">
      <er-button [stateKey]="formProps.$id"
                 icon="fa fa-save"
                 label="提&nbsp;&nbsp;&nbsp;交"
                 title="保留数据到服务端"
                 class="mr-2"
                 type="button"
                 (onClick)="formlyForm.submit()"
                 [styleClass]="formProps.submitBtnStyleClass"
      ></er-button>
      <er-button icon="fa fa-circle-o"
                 class="mr-2"
                 label="重&nbsp;&nbsp;&nbsp;填"
                 title="恢复到修改前的状态"
                 [styleClass]="formProps.resetBtnStyleClass"
                 (onClick)="formlyForm.resetModel()">
      </er-button>

      <ng-container *ngFor="let btn of formProps.buttons">
        <er-button erPropsBind [props]="btn" [context]="btn.context"></er-button>
      </ng-container>
    </div>

    <!--<png-form-debug [model]="formModel"-->
    <!--[debugFormProps]="props"-->
    <!--[runtimeFormProps]="formProps">-->
    <!--</png-form-debug>-->
  `,
                styles: [":host::ng-deep .ui-tabview .ui-state-active{background-color:transparent!important}:host::ng-deep .ui-tabview .ui-tabview-panel{margin:10px}:host::ng-deep .ui-tabview .ui-tabview-selected{border-left:5px solid #3bdc06}"]
            }] }
];
PngFormlyTabsFormComponent.propDecorators = {
    onTabChange: [{ type: Output }],
    onTabClose: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.formArray;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.tabFormlyOptions;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.currentTabIndex;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.onTabChange;
    /** @type {?} */
    PngFormlyTabsFormComponent.prototype.onTabClose;
}
