// @er/primeng/esm2015/lib/components/wizard/step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
export class PngStepComponent {
    constructor() {
        this.active = false;
    }
}
PngStepComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-step',
                template: `
    <div *ngIf="active" [ngClass]="styleClass" [ngStyle]="style" class="ui-widget-content ui-corner-all step-container">
      <ng-content></ng-content>
    </div>
  `,
                styles: ['.step-container {padding: 45px 25px 45px 25px; margin-bottom: 20px;}']
            }] }
];
PngStepComponent.propDecorators = {
    styleClass: [{ type: Input }],
    style: [{ type: Input }],
    label: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngStepComponent.prototype.styleClass;
    /** @type {?} */
    PngStepComponent.prototype.style;
    /** @type {?} */
    PngStepComponent.prototype.label;
    /** @type {?} */
    PngStepComponent.prototype.active;
}
