// @er/core/esm5/lib/pipes/up-to-now/pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, NgZone, Pipe } from '@angular/core';
import { DateUtils } from '@er/utils';
var UpToNowPipe = /** @class */ (function () {
    function UpToNowPipe(cd, ngZone) {
        this.cd = cd;
        this.ngZone = ngZone;
    }
    /**
     * @param {?} value
     * @param {?=} due
     * @return {?}
     */
    UpToNowPipe.prototype.transform = /**
     * @param {?} value
     * @param {?=} due
     * @return {?}
     */
    function (value, due) {
        var _this = this;
        /** @type {?} */
        var d = DateUtils.toDate(value);
        if (!d) {
            return '';
        }
        /** @type {?} */
        var span = DateUtils.getDateByNow(d, due);
        this.removeTimer();
        /** @type {?} */
        var timeToUpdate = (Number.isNaN(span.seconds)) ? 1000 : DateUtils.getUpdateSeconds(span.seconds) * 1000;
        this.timer = this.ngZone.runOutsideAngular(function () {
            if (typeof window !== 'undefined') {
                return window.setTimeout(function () {
                    _this.ngZone.run(function () { return _this.cd.detectChanges(); });
                }, timeToUpdate);
            }
            return null;
        });
        return span.message;
    };
    /**
     * @return {?}
     */
    UpToNowPipe.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeTimer();
    };
    /**
     * @private
     * @return {?}
     */
    UpToNowPipe.prototype.removeTimer = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.timer) {
            window.clearTimeout(this.timer);
            this.timer = null;
        }
    };
    UpToNowPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'erUpToNow',
                    pure: false
                },] }
    ];
    /** @nocollapse */
    UpToNowPipe.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: NgZone }
    ]; };
    return UpToNowPipe;
}());
export { UpToNowPipe };
if (false) {
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.timer;
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    UpToNowPipe.prototype.ngZone;
}
