// @er/core/esm2015/lib/components/data-item/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { BaseComponent } from '../../bases/component';
export class DataItemComponent extends BaseComponent {
    /**
     * @return {?}
     */
    get labels() {
        /** @type {?} */
        const _label = DataItemUtils.getItemLabel(this.item, this.labelKeys);
        if (CommonsUtils.isArray(_label)) {
            return _label;
        }
        else {
            return [_label];
        }
    }
    /**
     * @return {?}
     */
    get labelKeys() {
        /** @type {?} */
        let keys = this.props.dropDownOptionKey || this.props.labelKey;
        if (!keys) {
            if (this.item.hasOwnProperty('label')) {
                keys = 'label';
            }
            else {
                keys = Object.keys(this.item).filter(key => !key.startsWith('_') && key !== 'id');
            }
        }
        return keys;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this.props) {
            this.props = {};
        }
    }
}
DataItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-data-item',
                template: `
    <ng-container *ngIf="props.template;else default">
      <er-dyna-content [content]="props.template" [context]="item"></er-dyna-content>
    </ng-container>
    <ng-template #default>
      <div class="d-flex data-item" [ngClass]="props.itemStyleClass || 'ui-clearfix'">
        <span class="label mr-3 border-right" *ngFor="let label of labels" [ngClass]="props.labelStyleClass" [innerHTML]="label"></span>
        <span class="pull-right count" *ngIf="props.showCounter && item['value']">
          <span [ngClass]="props.counterStyleClass || 'badge badge-pill badge-info'">{{item['value']}}</span>
        </span>
      </div>
    </ng-template>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["/deep/ .ui-autocomplete-list-item{border-bottom:1px solid #9f9f9f!important;font-size:.8em!important;color:#000!important}/deep/ .ui-autocomplete-list-item:hover{background-color:transparent!important;font-weight:700!important;color:red!important;font-size:1em!important}"]
            }] }
];
DataItemComponent.propDecorators = {
    item: [{ type: Input }],
    props: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataItemComponent.prototype.item;
    /** @type {?} */
    DataItemComponent.prototype.props;
}
