// @er/primeng/esm2015/lib/components/panel/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ElementRef, Input, Renderer2, ViewChild } from '@angular/core';
import { BaseComponent, DiUtils } from '@er/core';
import { DomUtils } from '@er/utils';
export class PngPanelComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.collapsed = false;
        this.toggleable = false;
        this.showHeader = true;
        this.templateNames = {
            header: 'header',
            footer: 'footer'
        };
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        super.ngAfterViewInit();
        /** @type {?} */
        const titleBar = this.panelEle.nativeElement.querySelector('.ui-panel-titlebar');
        if (titleBar) {
            if (this.barStyleClass) {
                DomUtils.addClasses(titleBar, this.barStyleClass);
            }
            if (this.barStyle) {
                Object.keys(this.barStyle).forEach(name => {
                    DiUtils.get(Renderer2).setStyle(titleBar, name, this.barStyle[name]);
                });
            }
        }
    }
}
PngPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-panel',
                template: "<div class=\"ui-g\">\n  <div class=\"ui-g-12\">\n    <p-panel #pngPanel [styleClass]=\"styleClass\" [style]=\"style\" [toggleable]=\"toggleable\" [showHeader]=\"showHeader\"\n             [collapsed]=\"collapsed\">\n      <p-header>\n        <ng-container *ngIf=\"header\">\n          <i *ngIf=\"icon\" [class]=\"icon\"></i>\n          <er-dyna-content [content]=\"header\" class=\"d-inline-block\"></er-dyna-content>\n        </ng-container>\n        <ng-container *ngIf=\"template(templateNames.header)\">\n          <ng-container *ngTemplateOutlet=\"template(templateNames.header)\"></ng-container>\n        </ng-container>\n      </p-header>\n      <ng-content></ng-content>\n      <ng-container *ngIf=\"footer || template(templateNames.footer)\">\n        <p-footer>\n          <ng-container *ngIf=\"footer\">\n            <er-dyna-content [content]=\"footer\"></er-dyna-content>\n          </ng-container>\n          <ng-container *ngIf=\"template(templateNames.footer)\">\n            <ng-container *ngTemplateOutlet=\"template(templateNames.footer)\"></ng-container>\n          </ng-container>\n        </p-footer>\n      </ng-container>\n    </p-panel>\n  </div>\n</div>\n\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
PngPanelComponent.propDecorators = {
    barStyle: [{ type: Input }],
    barStyleClass: [{ type: Input }],
    style: [{ type: Input }],
    styleClass: [{ type: Input }],
    header: [{ type: Input }],
    footer: [{ type: Input }],
    icon: [{ type: Input }],
    collapsed: [{ type: Input }],
    toggleable: [{ type: Input }],
    showHeader: [{ type: Input }],
    panelEle: [{ type: ViewChild, args: ['pngPanel', { read: ElementRef },] }]
};
if (false) {
    /** @type {?} */
    PngPanelComponent.prototype.barStyle;
    /** @type {?} */
    PngPanelComponent.prototype.barStyleClass;
    /** @type {?} */
    PngPanelComponent.prototype.style;
    /** @type {?} */
    PngPanelComponent.prototype.styleClass;
    /** @type {?} */
    PngPanelComponent.prototype.header;
    /** @type {?} */
    PngPanelComponent.prototype.footer;
    /** @type {?} */
    PngPanelComponent.prototype.icon;
    /** @type {?} */
    PngPanelComponent.prototype.collapsed;
    /** @type {?} */
    PngPanelComponent.prototype.toggleable;
    /** @type {?} */
    PngPanelComponent.prototype.showHeader;
    /** @type {?} */
    PngPanelComponent.prototype.panelEle;
    /** @type {?} */
    PngPanelComponent.prototype.templateNames;
}
