// @er/primeng/esm2015/lib/components/rating/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_RATING_PROPS } from './type';
export class PngRatingComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.$defaultProps = DEFAULT_PNG_RATING_PROPS;
    }
}
PngRatingComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-rating',
                template: `
    <p-rating
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      (onRate)="updateTouch()"
    >
    </p-rating>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngRatingComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngRatingComponent.prototype.$defaultProps;
}
