// @er/core/esm5/lib/bases/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ContentChildren, Input, QueryList, TemplateRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { TypedTemplateDirective } from '../directives/typed-template/directive';
import { BaseDirective } from './directive';
/**
 * provide template & dynamic query support
 * @abstract
 */
var BaseComponent = /** @class */ (function (_super) {
    tslib_1.__extends(BaseComponent, _super);
    function BaseComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.templates = {};
        return _this;
    }
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.typedTemplates && this.typedTemplates.length > 0) {
            this.typedTemplates.forEach(function (item) {
                if (item.type) {
                    CommonsUtils.set(_this.templates, item.type, item.template);
                }
                else {
                    _this.templates['_'] = item.template;
                }
            });
        }
    };
    /**
     * @param {?} outputs
     * @param {?} eventName
     * @param {?} defaultHandler
     * @param {?} event
     * @return {?}
     */
    BaseComponent.prototype.eventHandler = /**
     * @param {?} outputs
     * @param {?} eventName
     * @param {?} defaultHandler
     * @param {?} event
     * @return {?}
     */
    function (outputs, eventName, defaultHandler, event) {
        if (outputs && outputs[eventName]) {
            return outputs[eventName].call(this, event);
        }
        else {
            return defaultHandler(event);
        }
    };
    /**
     * @param {?=} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    BaseComponent.prototype.template = /**
     * @param {?=} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    function (name) {
        var defaultTpls = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            defaultTpls[_i - 1] = arguments[_i];
        }
        if (!name) {
            name = '_';
        }
        if (this.templates[name]) {
            return this.templates[name];
        }
        else if (defaultTpls.length > 0) {
            /** @type {?} */
            var tpls = CommonsUtils.filter(defaultTpls, function (defaultTpl) { return !!defaultTpl; });
            if (tpls.length > 0) {
                return tpls[0];
            }
        }
        return undefined;
    };
    /**
     * @template C
     * @param {?} id
     * @return {?}
     */
    BaseComponent.prototype.getComponentBySelectorId = /**
     * @template C
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (this.selectors) {
            /** @type {?} */
            var matched = this.selectors.toArray().filter(function (comp) { return comp.id === id; });
            if (matched.length === 1) {
                return matched[0].hostComponent;
            }
        }
        return undefined;
    };
    /**
     * @param {?} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    BaseComponent.prototype.hasTemplate = /**
     * @param {?} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    function (name) {
        var defaultTpls = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            defaultTpls[_i - 1] = arguments[_i];
        }
        return this.template.apply(this, tslib_1.__spread([name], defaultTpls)) !== undefined;
    };
    /**
     * @param {?} templates
     * @return {?}
     */
    BaseComponent.prototype.hasChildren = /**
     * @param {?} templates
     * @return {?}
     */
    function (templates) {
        return CommonsUtils.isJson(templates);
    };
    // empty hooks for dyna bind
    // empty hooks for dyna bind
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngAfterContentChecked = 
    // empty hooks for dyna bind
    /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this['__ngAfterViewInit'] = true;
    };
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this['__ngOnInit'] = true;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    BaseComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @return {?}
     */
    BaseComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    BaseComponent.propDecorators = {
        customerTpl: [{ type: Input }],
        typedTemplates: [{ type: ContentChildren, args: [TypedTemplateDirective,] }]
    };
    return BaseComponent;
}(BaseDirective));
export { BaseComponent };
if (false) {
    /** @type {?} */
    BaseComponent.prototype.customerTpl;
    /** @type {?} */
    BaseComponent.prototype.typedTemplates;
    /** @type {?} */
    BaseComponent.prototype.selectors;
    /** @type {?} */
    BaseComponent.prototype.templateNames;
    /** @type {?} */
    BaseComponent.prototype.templates;
}
