// @er/core/esm5/lib/bases/control.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HostBinding, Input } from '@angular/core';
import { NO_OP } from '@er/types';
import { BehaviorSubject } from 'rxjs';
import { BaseComponent } from './component';
var ControlComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ControlComponent, _super);
    function ControlComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.disabled = false;
        _this.disabled$ = new BehaviorSubject(false);
        _this.onTouched = NO_OP;
        _this.onChange = NO_OP;
        return _this;
    }
    Object.defineProperty(ControlComponent.prototype, "opacity", {
        get: /**
         * @return {?}
         */
        function () {
            return this.disabled ? 0.25 : 1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ControlComponent.prototype, "cursor", {
        get: /**
         * @return {?}
         */
        function () {
            return this.disabled ? 'not-allowed' : 'default';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ControlComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.innerValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!value || value !== this.innerValue) {
                /** @type {?} */
                var v = this.parser ? this.parser(value) : value;
                this.innerValue = v;
                this.onChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ControlComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.defaultValue) {
            this.innerValue = this.defaultValue;
        }
    };
    /**
     * @return {?}
     */
    ControlComponent.prototype.updateTouch = /**
     * @return {?}
     */
    function () {
        this.onTouched();
    };
    /**
     * @param {?} model
     * @return {?}
     */
    ControlComponent.prototype.writeValue = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        /** @type {?} */
        var v = this.formatter ? this.formatter(model) : model;
        if (v !== this.innerValue) {
            this.innerValue = v;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ControlComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ControlComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    ControlComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
        this.disabled$.next(isDisabled);
    };
    ControlComponent.propDecorators = {
        parser: [{ type: Input }],
        formatter: [{ type: Input }],
        disabled: [{ type: Input }],
        opacity: [{ type: HostBinding, args: ['style.opacity',] }],
        cursor: [{ type: HostBinding, args: ['style.cursor',] }]
    };
    return ControlComponent;
}(BaseComponent));
export { ControlComponent };
if (false) {
    /** @type {?} */
    ControlComponent.prototype.parser;
    /** @type {?} */
    ControlComponent.prototype.formatter;
    /** @type {?} */
    ControlComponent.prototype.disabled;
    /** @type {?} */
    ControlComponent.prototype.defaultValue;
    /** @type {?} */
    ControlComponent.prototype.disabled$;
    /** @type {?} */
    ControlComponent.prototype.onTouched;
    /** @type {?} */
    ControlComponent.prototype.onChange;
    /**
     * @type {?}
     * @protected
     */
    ControlComponent.prototype.innerValue;
}
