// @er/core/esm5/lib/bases/control-wrapper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HostBinding, Input } from '@angular/core';
import { NO_OP } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { WrapperComponent } from './wrappe';
/**
 * @template T
 */
var ControlWrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ControlWrapperComponent, _super);
    function ControlWrapperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.resettable = true;
        _this.disabled = false;
        _this.onTouched = NO_OP;
        _this.onChange = NO_OP;
        return _this;
    }
    Object.defineProperty(ControlWrapperComponent.prototype, "opacity", {
        get: /**
         * @return {?}
         */
        function () {
            return this.disabled ? 0.25 : 1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ControlWrapperComponent.prototype, "cursor", {
        get: /**
         * @return {?}
         */
        function () {
            return this.disabled ? 'not-allowed' : 'default';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ControlWrapperComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.innerValue;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            /** @type {?} */
            var v = this.valueParser ? this.valueParser(value) : value;
            if (!CommonsUtils.isEqual(v, this.innerValue)) {
                this.innerValue = v;
                this.onChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ControlWrapperComponent.prototype.updateTouch = /**
     * @return {?}
     */
    function () {
        this.onTouched();
    };
    /**
     * @param {?} model
     * @return {?}
     */
    ControlWrapperComponent.prototype.writeValue = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        /** @type {?} */
        var v = this.modelFormatter ? this.modelFormatter(model) : model;
        if (v !== this.innerValue) {
            this.innerValue = v;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ControlWrapperComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    ControlWrapperComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    ControlWrapperComponent.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    /**
     * @return {?}
     */
    ControlWrapperComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () {
        if (this.disabled) {
            this.$props['disabled'] = this.disabled;
        }
    };
    /**
     * @return {?}
     */
    ControlWrapperComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.value = null;
    };
    ControlWrapperComponent.propDecorators = {
        valueParser: [{ type: Input }],
        modelFormatter: [{ type: Input }],
        resettable: [{ type: Input }],
        disabled: [{ type: Input }],
        opacity: [{ type: HostBinding, args: ['style.opacity',] }],
        cursor: [{ type: HostBinding, args: ['style.cursor',] }]
    };
    return ControlWrapperComponent;
}(WrapperComponent));
export { ControlWrapperComponent };
if (false) {
    /** @type {?} */
    ControlWrapperComponent.prototype.valueParser;
    /** @type {?} */
    ControlWrapperComponent.prototype.modelFormatter;
    /** @type {?} */
    ControlWrapperComponent.prototype.resettable;
    /** @type {?} */
    ControlWrapperComponent.prototype.disabled;
    /** @type {?} */
    ControlWrapperComponent.prototype.onTouched;
    /** @type {?} */
    ControlWrapperComponent.prototype.onChange;
    /**
     * @type {?}
     * @protected
     */
    ControlWrapperComponent.prototype.innerValue;
}
