// @er/core/esm5/lib/bases/data-aware-control-wrapper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input, ViewChild } from '@angular/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { BehaviorSubject } from 'rxjs';
import { ApiDataDirective } from '../directives/api-data/directive';
import { LogUtils } from '../utils/log';
import { ControlWrapperComponent } from './control-wrapper';
/**
 * @template T
 */
var DataAwareControlWrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DataAwareControlWrapperComponent, _super);
    function DataAwareControlWrapperComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.data$ = new BehaviorSubject(undefined);
        return _this;
    }
    Object.defineProperty(DataAwareControlWrapperComponent.prototype, "apiDataProps", {
        get: /**
         * @return {?}
         */
        function () {
            return this._apiDataProps || {};
        },
        set: /**
         * @param {?} props
         * @return {?}
         */
        function (props) {
            this._apiDataProps = props;
            if (this._apiDataProps && !this._apiDataProps.data$) {
                this._apiDataProps.data$ = this.data$;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DataAwareControlWrapperComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.onPropsInit.call(this);
        this.initLocalDataItem();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    DataAwareControlWrapperComponent.prototype.parseValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        try {
            if (value && this.$props.$ext.dataItemProps && !CommonsUtils.isString(value)) {
                return DataItemUtils.getItemValue(value, this.$props.$ext.dataItemProps.valueKey);
            }
            return value;
        }
        catch (e) {
            LogUtils.error(this, '数据解析错误', e);
            return undefined;
        }
    };
    /**
     * @param {?} model
     * @return {?}
     */
    DataAwareControlWrapperComponent.prototype.formatModel = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        try {
            if (CommonsUtils.isJson(model) && this.$props.$ext.dataItemProps) {
                return DataItemUtils.getItemLabel(model, this.$props.$ext.dataItemProps.labelKey);
            }
            return model;
        }
        catch (e) {
            LogUtils.error(this, '数据解析错误', e);
            return undefined;
        }
    };
    /**
     * @return {?}
     */
    DataAwareControlWrapperComponent.prototype.initLocalDataItem = /**
     * @return {?}
     */
    function () {
        if (this.$props.$ext && this.$props.$ext.dataItems) {
            /** @type {?} */
            var data = this.$props.$ext.dataItems;
            data = data.map(function (d) { return d['data'] = d; });
            this.data$.next(data);
        }
    };
    DataAwareControlWrapperComponent.propDecorators = {
        apiData: [{ type: ViewChild, args: [ApiDataDirective,] }],
        apiDataProps: [{ type: Input }]
    };
    return DataAwareControlWrapperComponent;
}(ControlWrapperComponent));
export { DataAwareControlWrapperComponent };
if (false) {
    /** @type {?} */
    DataAwareControlWrapperComponent.prototype.data$;
    /** @type {?} */
    DataAwareControlWrapperComponent.prototype.apiData;
    /**
     * @type {?}
     * @private
     */
    DataAwareControlWrapperComponent.prototype._apiDataProps;
}
