// @er/core/esm5/lib/bases/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter } from '@angular/core';
import { Unsubscribe } from './unsubscribe';
/**
 * provide event emit support
 * @abstract
 */
var /**
 * provide event emit support
 * @abstract
 */
BaseDirective = /** @class */ (function (_super) {
    tslib_1.__extends(BaseDirective, _super);
    function BaseDirective() {
        var _this = _super.call(this) || this;
        _this.animations = {};
        _this.ctx = {
            $from: (/** @type {?} */ (_this))
        };
        return _this;
    }
    /**
     * @return {?}
     */
    BaseDirective.prototype.getContext = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var id;
        if (!this['$id'] && this['$props'] && this['$props'].$id) {
            id = this['$props'].$id;
            this['$id'] = id;
        }
        /** @type {?} */
        var context = {
            $from: (/** @type {?} */ (this))
        };
        return id ? tslib_1.__assign({ $id: id }, context) : context;
    };
    /**
     * @param {?=} event
     * @return {?}
     */
    BaseDirective.prototype.getEventContext = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        return tslib_1.__assign({}, this.getContext(), event || {});
    };
    /**
     * @param {?} eventName
     * @param {?} event
     * @return {?}
     */
    BaseDirective.prototype.emitEvent = /**
     * @param {?} eventName
     * @param {?} event
     * @return {?}
     */
    function (eventName, event) {
        /** @type {?} */
        var eventEmitter = this[eventName];
        if (!eventEmitter) {
            eventEmitter = new EventEmitter();
            this[eventName] = eventEmitter;
        }
        if (eventEmitter instanceof EventEmitter) {
            return eventEmitter.emit(this.getEventContext(event));
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    BaseDirective.prototype.isFirstChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        return Object.keys(changes).every(function (prop) { return changes[prop].isFirstChange(); });
    };
    /**
     * @param {?=} name
     * @return {?}
     */
    BaseDirective.prototype.animate = /**
     * @param {?=} name
     * @return {?}
     */
    function (name) {
        return !(this.animations[name] || false);
    };
    return BaseDirective;
}(Unsubscribe));
/**
 * provide event emit support
 * @abstract
 */
export { BaseDirective };
if (false) {
    /** @type {?} */
    BaseDirective.prototype.ctx;
    /** @type {?} */
    BaseDirective.prototype.animations;
}
