// @er/core/esm5/lib/bases/roue.bind.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CommonsUtils } from '@er/utils';
import { ComponentUtils } from '../utils/component';
import { LogUtils } from '../utils/log';
import { RouteUtils } from '../utils/nav';
import { BaseComponent } from './component';
var RouteBindComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RouteBindComponent, _super);
    function RouteBindComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    RouteBindComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$subscriptions = RouteUtils.getActiveRoute().url.subscribe(function (url) {
            /** @type {?} */
            var activeRoute = RouteUtils.getActiveRoute();
            if (!activeRoute) {
                return;
            }
            /** @type {?} */
            var data = CommonsUtils.copy(activeRoute.snapshot.data);
            /** @type {?} */
            var params = tslib_1.__assign({}, activeRoute.snapshot.params || {}, activeRoute.snapshot.queryParams || {});
            while (activeRoute && !activeRoute.snapshot.data['$props']) {
                activeRoute = activeRoute.parent;
            }
            /** @type {?} */
            var routeData = tslib_1.__assign({}, activeRoute.snapshot.data, data);
            /** @type {?} */
            var props = CommonsUtils.copy(routeData['$resolvedProps'] || routeData['$props']);
            if (data['$propsKey']) {
                props = CommonsUtils.get(props, data['$propsKey']);
            }
            Object.keys(data).forEach(function (key) {
                if (CommonsUtils.isFunction(data[key])) {
                    props[key] = ((/** @type {?} */ (data[key])))(props, params, data);
                }
                else if (CommonsUtils.isString(data[key]) && data[key].startsWith(':')) {
                    /** @type {?} */
                    var param = ((/** @type {?} */ (data[key]))).substring(1);
                    if (params[param]) {
                        props[key] = params[param];
                    }
                }
                else if (key !== '$propsKey' && key !== '$propsWrapperKey' && key !== '$props') {
                    props[key] = data[key];
                }
            });
            props['$routeParams'] = params;
            LogUtils.debug(_this, '路由数据', routeData);
            props = _this.beforeRoutePropsBind(props);
            ComponentUtils.bind(_this, props);
            _this.afterRoutePropsBind();
        });
    };
    /**
     * @param {?} props
     * @return {?}
     */
    RouteBindComponent.prototype.beforeRoutePropsBind = /**
     * @param {?} props
     * @return {?}
     */
    function (props) {
        return props;
    };
    /**
     * @return {?}
     */
    RouteBindComponent.prototype.afterRoutePropsBind = /**
     * @return {?}
     */
    function () {
    };
    return RouteBindComponent;
}(BaseComponent));
export { RouteBindComponent };
