// @er/core/esm5/lib/bases/wrappe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { CommonsUtils, IdUtils } from '@er/utils';
import { BaseComponent } from './component';
/**
 * @abstract
 * @template T
 */
var WrapperComponent = /** @class */ (function (_super) {
    tslib_1.__extends(WrapperComponent, _super);
    function WrapperComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    WrapperComponent.prototype._initProps = /**
     * @return {?}
     */
    function () {
        this.$props = CommonsUtils.defaults({}, this.$props || {}, this.$defaultProps || {});
        if (!this.$props.$id) {
            this.$props.$id = IdUtils.getRandom();
        }
        if (!this.$props.$ext) {
            this.$props.$ext = {};
        }
        if (this.$props.$ext['apiDataProps']) {
            if (!this.$props.$ext['apiDataProps'].requestOptions) {
                this.$props.$ext['apiDataProps'].requestOptions = {};
            }
            CommonsUtils.defaults(this.$props.$ext['apiDataProps'].requestOptions, {
                context: this,
                tag: this.$props.$id
            });
        }
        this.onPropsInit();
    };
    /**
     * @return {?}
     */
    WrapperComponent.prototype._propsBind = /**
     * @return {?}
     */
    function () {
        this.onPropsBind();
    };
    /**
     * @return {?}
     */
    WrapperComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    WrapperComponent.prototype.onPropsBind = /**
     * @return {?}
     */
    function () {
    };
    WrapperComponent.propDecorators = {
        outputContext: [{ type: Input }],
        attrs: [{ type: Input }]
    };
    return WrapperComponent;
}(BaseComponent));
export { WrapperComponent };
if (false) {
    /** @type {?} */
    WrapperComponent.prototype.$props;
    /** @type {?} */
    WrapperComponent.prototype.outputContext;
    /** @type {?} */
    WrapperComponent.prototype.attrs;
    /**
     * @type {?}
     * @protected
     */
    WrapperComponent.prototype.$defaultProps;
}
