// @er/core/esm2015/lib/components/animate-number/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { Subject, timer } from 'rxjs';
import { mapTo, scan, startWith, switchMap, takeWhile } from 'rxjs/operators';
import { BaseComponent } from '../../bases/component';
export class AnimateNumberComponent extends BaseComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super();
        this.cd = cd;
        this.currentNumber = 0;
        this.fontSize = '.8em';
        this.counter$ = new Subject();
        this.$subscriptions = this.counter$
            .pipe(switchMap(endRange => {
            return timer(1000, 200)
                .pipe(mapTo(this.positiveOrNegative(endRange, this.currentNumber)), startWith(this.currentNumber), scan((acc, curr) => acc + curr), takeWhile(this.takeUntil(endRange, this.currentNumber)));
        }))
            .subscribe(val => {
            this.currentNumber = val;
            if (this.currentNumber === this.end) {
                this.fontSize = '1em';
            }
            this.cd.detectChanges();
        });
    }
    /**
     * @return {?}
     */
    get end() {
        return this._end;
    }
    /**
     * @param {?} endRange
     * @return {?}
     */
    set end(endRange) {
        this._end = endRange;
        this.currentNumber = endRange - CommonsUtils.getRandomNumber(20);
        this.counter$.next(endRange);
    }
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    positiveOrNegative(endRange, currentNumber) {
        return endRange > currentNumber ? 1 : -1;
    }
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    takeUntil(endRange, currentNumber) {
        return endRange > currentNumber
            ? val => val <= endRange
            : val => val >= endRange;
    }
}
AnimateNumberComponent.decorators = [
    { type: Component, args: [{
                selector: 'animate-number',
                template: `
    <span [ngStyle]="{'font-size': fontSize, 'font-weight': 'bold'}">
       {{currentNumber}}
    </span>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
AnimateNumberComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
AnimateNumberComponent.propDecorators = {
    end: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AnimateNumberComponent.prototype.currentNumber;
    /** @type {?} */
    AnimateNumberComponent.prototype.fontSize;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype.counter$;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype._end;
    /**
     * @type {?}
     * @private
     */
    AnimateNumberComponent.prototype.cd;
}
