// @er/core/esm2015/lib/pipes/all.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { KeyValuePipe } from './as-list/pipe';
import { Base64DecodePipe } from './base64/pipe';
import { CommaSeparatedNumberPipe } from './comma-separate/pipe';
import { DataItemsPipe } from './data-items/pipe';
import { DefaultPipe } from './default/pipe';
import { EmojifyPipe } from './emojify/pipe';
import { FilterPipe } from './filter/pipe';
import { HighLightPipe } from './high-light/pipe';
import { IdCardPipe } from './id-card/pipe';
import { JoinPipe } from './join/pipe';
import { LinkifyPipe } from './linky/pipe';
import { KeysPipe } from './map-keys/pipe';
import { NumberToTextPipe } from './number-text/pipe';
import { ObservablePipe } from './observable/pipe';
import { OrderByPipe } from './order/pipe';
import { PercentPipe } from './percent/pipe';
import { SafeHtmlPipe } from './safe/pipe';
import { SplitPipe } from './split/pipe';
import { StylePipe } from './style/pipe';
import { SubPipe } from './sub/pipe';
import { TextPipe } from './text/pipe';
import { TimerPipe } from './timer/pipe';
import { TrackerPipe } from './tracker/pipe';
import { TreeNodesPipe } from './tree-nodes/pipe';
import { UniquePipe } from './unique/pipe';
import { UnitNumberPipe } from './unit-number/pipe';
import { UpToNowPipe } from './up-to-now/pipe';
import { ValuePipe } from './value/pipe';
export { FilterPipe } from './filter/pipe';
export { HighLightPipe } from './high-light/pipe';
export { KeysPipe } from './map-keys/pipe';
export { SafeHtmlPipe } from './safe/pipe';
export { StylePipe } from './style/pipe';
export { KeyValuePipe } from './as-list/pipe';
export { JoinPipe } from './join/pipe';
export { ObservablePipe } from './observable/pipe';
export { OrderByPipe } from './order/pipe';
export { IdCardPipe } from './id-card/pipe';
export { SplitPipe } from './split/pipe';
export { UniquePipe } from './unique/pipe';
export { Base64DecodePipe } from './base64/pipe';
export { CommaSeparatedNumberPipe } from './comma-separate/pipe';
export { LinkifyPipe } from './linky/pipe';
export { NumberToTextPipe } from './number-text/pipe';
export { UnitNumberPipe } from './unit-number/pipe';
export { DataItemsPipe } from './data-items/pipe';
export { TreeNodesPipe } from './tree-nodes/pipe';
export { PercentPipe } from './percent/pipe';
export { SubPipe } from './sub/pipe';
export { TextPipe } from './text/pipe';
export { TimerPipe } from './timer/pipe';
export { UpToNowPipe } from './up-to-now/pipe';
export { ValuePipe } from './value/pipe';
export { DefaultPipe } from './default/pipe';
export { TrackerPipe } from './tracker/pipe';
export { EmojifyPipe } from './emojify/pipe';
/** @type {?} */
export const CorePipes = [
    UpToNowPipe,
    Base64DecodePipe,
    CommaSeparatedNumberPipe,
    LinkifyPipe,
    NumberToTextPipe,
    UnitNumberPipe,
    SubPipe,
    EmojifyPipe,
    IdCardPipe,
    ObservablePipe,
    StylePipe,
    SafeHtmlPipe,
    HighLightPipe,
    FilterPipe,
    TrackerPipe,
    KeyValuePipe,
    KeysPipe,
    DataItemsPipe,
    TreeNodesPipe,
    JoinPipe,
    OrderByPipe,
    SplitPipe,
    UniquePipe,
    StylePipe,
    TextPipe,
    ValuePipe,
    DefaultPipe,
    TimerPipe,
    PercentPipe
];
