// @er/core/esm2015/lib/directives/api-data/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
import { ApiUtils } from '../../utils/api';
import { LogUtils } from '../../utils/log';
export class ApiDataDirective extends BaseDirective {
    constructor() {
        super(...arguments);
        this.onDataReady = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.apiDataProps
            && changes.apiDataProps.currentValue
            && (this.apiDataProps.autoFetch !== false || !changes.apiDataProps.isFirstChange())) {
            this.fetchData();
        }
    }
    /**
     * @return {?}
     */
    fetchData() {
        if (CommonsUtils.isEmpty(this.apiDataProps)) {
            LogUtils.debug('apiDataDirective', '没有定义apiDataProps');
            return;
        }
        ApiUtils.fetch(this.apiDataProps)
            .subscribe(data => {
            LogUtils.debug(this, 'apiData待订阅数据', data);
            if (this.apiDataProps.data$) {
                this.apiDataProps.data$.next(data);
            }
            this.onDataReady.emit(data);
        });
    }
}
ApiDataDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erApiData]'
            },] }
];
ApiDataDirective.propDecorators = {
    apiDataProps: [{ type: Input, args: ['erApiData',] }],
    onRequestData: [{ type: Output }],
    onDataReady: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ApiDataDirective.prototype.apiDataProps;
    /** @type {?} */
    ApiDataDirective.prototype.onRequestData;
    /** @type {?} */
    ApiDataDirective.prototype.onDataReady;
}
