// @er/share/esm2015/lib/image-editor/services/img-crop.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { forwardRef, Inject, Injectable } from '@angular/core';
import { Ng2ImgMaxService } from 'ng2-img-max';
import { Subject } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "ng2-img-max/dist/src/ng2-img-max.service";
export class ImgCropService {
    /**
     * @param {?} ng2ImgMaxService
     */
    constructor(ng2ImgMaxService) {
        this.ng2ImgMaxService = ng2ImgMaxService;
    }
    /**
     * @param {?} file
     * @param {?} toWidth
     * @param {?} toHeight
     * @param {?=} startX
     * @param {?=} startY
     * @return {?}
     */
    cropImage(file, toWidth, toHeight, startX = 0, startY = 0) {
        /** @type {?} */
        const croppedImageSubject = new Subject();
        if (file.type !== 'image/jpeg' && file.type !== 'image/png') {
            // END OF CROPPING
            setTimeout(() => {
                croppedImageSubject.error({
                    croppedFile: file,
                    reason: 'File provided is neither of type jpg nor of type png.',
                    error: 'INVALID_EXTENSION'
                });
            }, 0);
            return croppedImageSubject.asObservable();
        }
        /** @type {?} */
        const cvs = document.createElement('canvas');
        /** @type {?} */
        let ctx = cvs.getContext('2d');
        /** @type {?} */
        const img = new Image();
        img.onload = () => {
            this.ng2ImgMaxService.getEXIFOrientedImage(img).then(orientedImg => {
                window.URL.revokeObjectURL(img.src);
                cvs.width = toWidth;
                cvs.height = toHeight;
                ctx.drawImage(orientedImg, startX, startY, toWidth, toHeight, 0, 0, toWidth, toHeight);
                /** @type {?} */
                const imageData = ctx.getImageData(0, 0, orientedImg.width, orientedImg.height);
                /** @type {?} */
                let useAlpha = true;
                if (file.type === 'image/jpeg' || (file.type === 'image/png' && !this.isImgUsingAlpha(imageData))) {
                    // image without alpha
                    useAlpha = false;
                    ctx = cvs.getContext('2d', { 'alpha': false });
                    ctx.drawImage(orientedImg, startX, startY, toWidth, toHeight, 0, 0, toWidth, toHeight);
                }
                cvs.toBlob((blob) => {
                    /** @type {?} */
                    const newFile = this.generateResultFile(blob, file.name, file.type, new Date().getTime());
                    // END OF CROPPING
                    croppedImageSubject.next(newFile);
                }, useAlpha ? 'image/png' : 'image/jpeg');
            });
        };
        img.src = window.URL.createObjectURL(file);
        return croppedImageSubject.asObservable();
    }
    /**
     * @private
     * @param {?} imageData
     * @return {?}
     */
    isImgUsingAlpha(imageData) {
        for (let i = 0; i < imageData.data.length; i += 4) {
            if (imageData.data[i + 3] !== 255) {
                return true;
            }
        }
        return false;
    }
    /**
     * @private
     * @param {?} blob
     * @param {?} name
     * @param {?} type
     * @param {?} lastModified
     * @return {?}
     */
    generateResultFile(blob, name, type, lastModified) {
        /** @type {?} */
        const resultFile = new Blob([blob], { type: type });
        return this.blobToFile(resultFile, name, lastModified);
    }
    /**
     * @private
     * @param {?} blob
     * @param {?} name
     * @param {?} lastModified
     * @return {?}
     */
    blobToFile(blob, name, lastModified) {
        /** @type {?} */
        const file = blob;
        file.name = name;
        file.lastModified = lastModified;
        // Cast to a File() type
        return (/** @type {?} */ (file));
    }
}
ImgCropService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
/** @nocollapse */
ImgCropService.ctorParameters = () => [
    { type: Ng2ImgMaxService, decorators: [{ type: Inject, args: [forwardRef(() => Ng2ImgMaxService),] }] }
];
/** @nocollapse */ ImgCropService.ngInjectableDef = i0.defineInjectable({ factory: function ImgCropService_Factory() { return new ImgCropService(i0.inject(i1.Ng2ImgMaxService)); }, token: ImgCropService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImgCropService.prototype.ng2ImgMaxService;
}
