// @er/core/esm5/lib/directives/dyna-template/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2, ViewContainerRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
import { DynaComponentUtils } from '../../utils/dyna';
var DynaTemplateDirective = /** @class */ (function (_super) {
    tslib_1.__extends(DynaTemplateDirective, _super);
    function DynaTemplateDirective(viewContainerRef, el, renderer) {
        var _this = _super.call(this) || this;
        _this.viewContainerRef = viewContainerRef;
        _this.el = el;
        _this.renderer = renderer;
        return _this;
    }
    /**
     * @return {?}
     */
    DynaTemplateDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (CommonsUtils.isEmptyString(this.template)) {
            this.template = this.el.nativeElement.innerHTML;
        }
        if (CommonsUtils.isEmptyString(this.template)) {
            return;
        }
        this.template = this.template.replace(/\[\[/g, '{{');
        this.template = this.template.replace(/]]/g, '}}');
        this.componentRef = DynaComponentUtils.createComponentByTemplate(this.viewContainerRef, this.template);
        /** @type {?} */
        var parent = this.el.nativeElement.parentNode;
        this.renderer.removeChild(parent, this.el.nativeElement);
    };
    /**
     * @return {?}
     */
    DynaTemplateDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnDestroy.call(this);
        if (this.componentRef) {
            this.componentRef.destroy();
            this.componentRef = null;
        }
    };
    DynaTemplateDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erDynaTemplate]',
                },] }
    ];
    /** @nocollapse */
    DynaTemplateDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    DynaTemplateDirective.propDecorators = {
        template: [{ type: Input, args: ['erDynaTemplate',] }],
        context: [{ type: Input }]
    };
    return DynaTemplateDirective;
}(BaseDirective));
export { DynaTemplateDirective };
if (false) {
    /** @type {?} */
    DynaTemplateDirective.prototype.template;
    /** @type {?} */
    DynaTemplateDirective.prototype.context;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.viewContainerRef;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DynaTemplateDirective.prototype.renderer;
}
