// @er/core/esm2015/lib/directives/movable/movable-handle.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostBinding, Input } from '@angular/core';
export class MovableHandleDirective {
    /**
     * @param {?} element
     */
    constructor(element) {
        this.element = element;
        this.movableEnabled = false;
        this.isMoving = false;
        this.isHandle = true;
    }
}
MovableHandleDirective.decorators = [
    { type: Directive, args: [{
                selector: '[movableHandle]'
            },] }
];
/** @nocollapse */
MovableHandleDirective.ctorParameters = () => [
    { type: ElementRef }
];
MovableHandleDirective.propDecorators = {
    movableEnabled: [{ type: HostBinding, args: ['class.movable-enabled',] }],
    isMoving: [{ type: HostBinding, args: ['class.movable-moving',] }],
    movableHandle: [{ type: Input }],
    isHandle: [{ type: HostBinding, args: ['class.movable-handle',] }]
};
if (false) {
    /** @type {?} */
    MovableHandleDirective.prototype.movableEnabled;
    /** @type {?} */
    MovableHandleDirective.prototype.isMoving;
    /** @type {?} */
    MovableHandleDirective.prototype.movableHandle;
    /**
     * @type {?}
     * @protected
     */
    MovableHandleDirective.prototype.isHandle;
    /** @type {?} */
    MovableHandleDirective.prototype.element;
}
