// @er/primeng/esm2015/lib/components/select-button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_SELECT_BUTTON_PROPS } from './type';
export class PngSelectButtonComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.$defaultProps = DEFAULT_SELECT_BUTTON_PROPS;
        this.templateNames = {
            item: 'item'
        };
    }
}
PngSelectButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-select-button',
                template: `
    <p-selectButton
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      [erApiData]="apiDataProps"
      [options]="data$|async|erDataItems:$props.$ext.dataItemProps"
      (onOptionClick)="updateTouch()"
    >
      <ng-template let-item>
        <er-data-item [customerTpl]="template(templateNames.item)"
                      [item]="item.data"
                      [props]="$props.$ext.dataItemProps">
        </er-data-item>
      </ng-template>  
    </p-selectButton>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngSelectButtonComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngSelectButtonComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSelectButtonComponent.prototype.templateNames;
}
