// @er/core/esm5/lib/bases/abstract/app-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BrowserOsUtils, CommonsUtils, ConfigUtils } from '@er/utils';
import { HTTP_STATE$, ROUTE_STATE$ } from '../../states/all';
import { AuthUtils } from '../../utils/auth';
import { PermissionUtils } from '../../utils/permission';
/**
 * @abstract
 */
var /**
 * @abstract
 */
AppConfigComponent = /** @class */ (function () {
    function AppConfigComponent(rootViewContainerRef) {
        this.rootViewContainerRef = rootViewContainerRef;
    }
    Object.defineProperty(AppConfigComponent.prototype, "loadingClass", {
        get: /**
         * @return {?}
         */
        function () {
            return ConfigUtils.getConfig().onLoading;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AppConfigComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var browserSupport = BrowserOsUtils.checkBrowser();
        this.resolveBrowserSupport(browserSupport);
        ConfigUtils.configApp(this.appConfig());
        setTimeout(function () {
            _this.httpState$ = HTTP_STATE$;
            _this.routeState$ = ROUTE_STATE$;
        }, 0);
        /** @type {?} */
        var user = AuthUtils.getCurrentUser();
        if (!CommonsUtils.isEmpty(user)) {
            PermissionUtils.addPermissions(user.permissions);
            PermissionUtils.addRoles(user.roles);
            if (user.roles && ConfigUtils.getConfig().auth.defaultRules) {
                PermissionUtils.addRoles(ConfigUtils.getConfig().auth.defaultRules);
            }
        }
    };
    /**
     * @param {?} browserSupport
     * @return {?}
     */
    AppConfigComponent.prototype.resolveBrowserSupport = /**
     * @param {?} browserSupport
     * @return {?}
     */
    function (browserSupport) {
        if (!browserSupport.version) {
            alert('当前浏览器版本太低，请升级浏览器!');
        }
        else if (!browserSupport.features) {
            alert('当前浏览器功能不全，请更换浏览器!');
        }
    };
    return AppConfigComponent;
}());
/**
 * @abstract
 */
export { AppConfigComponent };
if (false) {
    /** @type {?} */
    AppConfigComponent.prototype.httpState$;
    /** @type {?} */
    AppConfigComponent.prototype.routeState$;
    /** @type {?} */
    AppConfigComponent.prototype.inValidBrowser;
    /** @type {?} */
    AppConfigComponent.prototype.rootViewContainerRef;
    /**
     * @abstract
     * @return {?}
     */
    AppConfigComponent.prototype.appConfig = function () { };
}
