// @er/core/esm5/lib/bases/abstract/change-pwd.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { FormBuilder, Validators } from '@angular/forms';
import { ConfigUtils, DateUtils } from '@er/utils';
import { NgValidators } from '@er/validate';
import { AuthUtils } from '../../utils/auth/index';
import { DiUtils } from '../../utils/di/index';
import { RouteUtils } from '../../utils/nav/index';
import { AuthBaseComponent } from './auth.component';
var AbstractChangePwdComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractChangePwdComponent, _super);
    function AbstractChangePwdComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    AbstractChangePwdComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _a;
        this.form = DiUtils.get(FormBuilder).group((_a = {},
            _a[this.oldPasswordKey] = ['', Validators.required],
            _a[this.newPasswordKey] = ['', [NgValidators.required, NgValidators.minLength(6)]],
            _a.reNewPassword = ['', [NgValidators.required, NgValidators.minLength(6), NgValidators.equalTo(this.newPasswordKey)]],
            _a));
    };
    /**
     * @return {?}
     */
    AbstractChangePwdComponent.prototype.changePwd = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.error = undefined;
        /** @type {?} */
        var data = this.form.value;
        delete data['reNewPassword'];
        data['userPwdSetDate'] = DateUtils.getFormattedDateTime();
        AuthUtils.changePwd(this.getSubmitData(this.form.value))
            .subscribe(function (response) {
            if (response.success === true) {
                RouteUtils.nav(ConfigUtils.getConfig().auth.changePwdSuccessRouter);
            }
            else {
                _this.error = response.message || '密码修改失败';
            }
        });
    };
    return AbstractChangePwdComponent;
}(AuthBaseComponent));
export { AbstractChangePwdComponent };
if (false) {
    /** @type {?} */
    AbstractChangePwdComponent.prototype.oldPasswordKey;
    /** @type {?} */
    AbstractChangePwdComponent.prototype.newPasswordKey;
}
