// @er/core/esm5/lib/bases/abstract/register.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { FormBuilder, FormControl } from '@angular/forms';
import { ApiServerType } from '@er/types';
import { NgValidators } from '@er/validate';
import { DiUtils } from '../../utils/di/index';
import { NgAsyncValidator } from '../../validators/async';
import { AuthBaseComponent } from './auth.component';
var AbstractRegisterComponent = /** @class */ (function (_super) {
    tslib_1.__extends(AbstractRegisterComponent, _super);
    function AbstractRegisterComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.checkExists = false;
        _this.inFocus = false;
        _this.errorMessage = {
            duplicated: '该用户名已经被注册'
        };
        return _this;
    }
    Object.defineProperty(AbstractRegisterComponent.prototype, "isUserNameValidate", {
        get: /**
         * @return {?}
         */
        function () {
            return !this.inFocus && this.form.get(this.userToken).valid;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AbstractRegisterComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _a;
        this.form = DiUtils.get(FormBuilder).group((_a = {},
            _a[this.userToken] = new FormControl('', {
                validators: NgValidators.required,
                asyncValidators: NgAsyncValidator(this.userExistsApiOptions(), 'duplicated'),
                updateOn: 'blur'
            }),
            _a[this.password] = ['', [NgValidators.required, NgValidators.minLength(6)]],
            _a.rePassword = ['', [NgValidators.required, NgValidators.minLength(6), NgValidators.equalTo(this.password)]],
            _a));
    };
    /**
     * @return {?}
     */
    AbstractRegisterComponent.prototype.userExistsApiOptions = /**
     * @return {?}
     */
    function () {
        return {
            requestOptions: {
                serverType: ApiServerType.ER
            },
            apiEntry: 'user/exists',
            payload: function (val) {
                return { userName: val };
            }
        };
    };
    return AbstractRegisterComponent;
}(AuthBaseComponent));
export { AbstractRegisterComponent };
if (false) {
    /** @type {?} */
    AbstractRegisterComponent.prototype.checkExists;
    /** @type {?} */
    AbstractRegisterComponent.prototype.inFocus;
    /** @type {?} */
    AbstractRegisterComponent.prototype.errorMessage;
}
