// @er/primeng/esm2015/lib/components/calendar/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ControlWrapperComponent } from '@er/core';
import { DateUtils } from '@er/utils';
import { DEFAULT_CALENDAR_PROPS } from './type';
export class PngCalendarComponent extends ControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.modelFormatter = this.formatter;
        this.today = DateUtils.getDay();
        this.templateNames = {
            header: 'header',
            footer: 'footer',
            date: 'date'
        };
        this.$defaultProps = DEFAULT_CALENDAR_PROPS;
    }
    /**
     * @param {?} model
     * @return {?}
     */
    formatter(model) {
        if (model) {
            return DateUtils.toDate(model);
        }
    }
}
PngCalendarComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-calendar',
                template: `
    <p-calendar
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      [monthNavigator]="$props.monthNavigator"
      [yearNavigator]="$props.yearNavigator"
      [yearRange]="$props.yearRange"
      (onBlur)="updateTouch()"
    >
      <p-header *ngIf="template(templateNames.header)">
        <ng-container *ngTemplateOutlet="template(templateNames.header)"></ng-container>
      </p-header>

      <p-footer *ngIf="template(templateNames.footer)">
        <ng-container *ngTemplateOutlet="template(templateNames.footer)"></ng-container>
      </p-footer>
      
      <ng-template pTemplate="date" let-date>
        <ng-container *ngTemplateOutlet="template(templateNames.date, dateTpl);context:{$implicit: date}"></ng-container>
      </ng-template>
    </p-calendar>

    <ng-template #dateTpl let-date>
      <span [ngClass]="{'bg-red white': (date.day === today)}"
            style="border-radius:50%;">{{date.day}}</span>
    </ng-template>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngCalendarComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngCalendarComponent.prototype.modelFormatter;
    /** @type {?} */
    PngCalendarComponent.prototype.today;
    /** @type {?} */
    PngCalendarComponent.prototype.templateNames;
    /** @type {?} */
    PngCalendarComponent.prototype.$defaultProps;
}
