// @er/core/esm2015/lib/directives/json-viewer/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, Input } from '@angular/core';
import JSONFormatter from 'json-formatter-js';
import { BehaviorSubject, Observable } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { BaseDirective } from '../../bases/directive';
import { ObjectDiff } from './diff';
export { ObjectDiff } from './diff';
export class JsonViewerDirective extends BaseDirective {
    /**
     * @param {?} el
     * @param {?} cd
     */
    constructor(el, cd) {
        super();
        this.el = el;
        this.cd = cd;
        this.showCurrent = true;
        this.openLevel = 2;
        this.showDiff = false;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.model) {
            this.model = changes.model.currentValue;
            if (this.model) {
                Object.keys(this.model).forEach(key => {
                    if (this.model[key] instanceof BehaviorSubject) {
                        this.model[key] = this.model[key].value;
                    }
                    else if (this.model[key] instanceof Observable) {
                        ((/** @type {?} */ (this.model[key])))
                            .pipe(takeUntil(this.destroyed$))
                            .subscribe(v => this.model[key] = v);
                    }
                    else if (this.showDiff) {
                        try {
                            this.diff = ObjectDiff.diff(changes.model.previousValue, changes.model.currentValue);
                        }
                        catch (e) {
                        }
                    }
                });
            }
            this.showJson();
        }
    }
    /**
     * @return {?}
     */
    showJson() {
        this.el.nativeElement.innerHTML = '';
        if (this.showCurrent) {
            /** @type {?} */
            const formatter = new JSONFormatter(this.model || {}, this.openLevel, {
                hoverPreviewEnabled: true
            });
            try {
                /** @type {?} */
                const divEle = formatter.render();
                divEle.classList.add('json-viewer');
                this.el.nativeElement.appendChild(divEle);
            }
            catch (e) {
            }
        }
        if (this.showDiff) {
            /** @type {?} */
            const divEle = document.createElement('div');
            divEle.classList.add('diff-viewer');
            divEle.innerHTML = this.diff;
            this.el.nativeElement.appendChild(divEle);
        }
    }
}
JsonViewerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[erJsonViewer]'
            },] }
];
/** @nocollapse */
JsonViewerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: ChangeDetectorRef }
];
JsonViewerDirective.propDecorators = {
    model: [{ type: Input, args: ['erJsonViewer',] }],
    showCurrent: [{ type: Input, args: ['current',] }],
    openLevel: [{ type: Input }],
    showDiff: [{ type: Input, args: ['diff',] }]
};
if (false) {
    /** @type {?} */
    JsonViewerDirective.prototype.model;
    /** @type {?} */
    JsonViewerDirective.prototype.showCurrent;
    /** @type {?} */
    JsonViewerDirective.prototype.openLevel;
    /** @type {?} */
    JsonViewerDirective.prototype.showDiff;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.diff;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    JsonViewerDirective.prototype.cd;
}
