// @er/primeng/esm2015/lib/components/tree/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, forwardRef, Output, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent, HTTP_TAG_STATE$ } from '@er/core';
import { CommonsUtils, DataItemUtils, PinyinUtil, RegUtils, StatesUtils, TreeUtils } from '@er/utils';
import { Tree } from 'primeng/tree';
import { DEFAULT_PNG_TREE_PROPS, PNG_TREE_NODE_KEYS } from './types';
export { PNG_TREE_NODE_KEYS } from './types';
export class PngTreeComponent extends DataAwareControlWrapperComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super();
        this.cd = cd;
        this.templateNames = {
            icon: 'icon',
            node: 'node'
        };
        this.onEmptyNode = new EventEmitter();
        this.onLazyLoad = new EventEmitter();
        this.$defaultProps = DEFAULT_PNG_TREE_PROPS;
        this.modelFormatter = this.formatModel;
        this.valueParser = this.parseValue;
    }
    /**
     * @return {?}
     */
    get selection() {
        return this._selection;
    }
    /**
     * @param {?} selection
     * @return {?}
     */
    set selection(selection) {
        this._selection = selection;
        this.value = selection;
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        this.filtered = false;
        this.loading$ = HTTP_TAG_STATE$(this.$props.$id);
        if (this.$props.$ext.dataItems) {
            this.data$.next(this.$props.$ext.dataItems);
        }
        StatesUtils.create(this.$props.$id, {
            component: this
        }, true);
    }
    /**
     * @return {?}
     */
    load() {
        this.apiDataProps = Object.assign({}, this.apiDataProps);
        this.cd.markForCheck();
    }
    /**
     * @param {?} model
     * @return {?}
     */
    formatModel(model) {
        if (model) {
            this.selectedNodeValues = model;
        }
        return model;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    parseValue(value) {
        /** @type {?} */
        let selectedNodeValues;
        if (CommonsUtils.isArray(value)) {
            selectedNodeValues = CommonsUtils.flattenDeep(value).map(node => this.getSelectedNodeValues(node));
        }
        else if (CommonsUtils.isJson(value)) {
            selectedNodeValues = this.getSelectedNodeValues(value);
        }
        selectedNodeValues = CommonsUtils.flatten(selectedNodeValues);
        selectedNodeValues = CommonsUtils.unique(selectedNodeValues, this.$props.$ext.dataItemProps.idKey);
        selectedNodeValues = CommonsUtils.orderBy(selectedNodeValues, this.$props.$ext.dataItemProps.idKey, 'asc');
        if (this.$props && this.$props.$id) {
            StatesUtils.update(this.$props.$id, {
                selectedNodeValues: selectedNodeValues
            });
        }
        this.selectedNodeValues = selectedNodeValues;
        return selectedNodeValues;
    }
    /**
     * @return {?}
     */
    expandAll() {
        this.data$.value.forEach(node => {
            TreeUtils.expandNode(node, true, this.$props.$ext);
        });
    }
    /**
     * @return {?}
     */
    collapseAll() {
        this.data$.value.forEach(node => {
            TreeUtils.expandNode(node, false, this.$props.$ext);
        });
    }
    /**
     * @param {?} event
     * @return {?}
     */
    filter(event) {
        /** @type {?} */
        const data = this.applyFilter(event.target.value);
        this.data$.next(data);
        this.filtered = true;
        this.expandAll();
        //  this.cd.detectChanges();
    }
    /**
     * @param {?} filterValue
     * @return {?}
     */
    applyFilter(filterValue) {
        if (!filterValue) {
            this.clearFilter();
            return this.unFilteredValue;
        }
        else {
            filterValue = filterValue.toLowerCase();
            return TreeUtils.filter(this.unFilteredValue, (node) => {
                /** @type {?} */
                let py = node['$py'];
                if (!py) {
                    py = PinyinUtil.getFirstLetter(node.label);
                    node['$py'] = py;
                }
                /** @type {?} */
                const idKey = this.$props.$ext.dataItemProps && this.$props.$ext.dataItemProps.idKey;
                /** @type {?} */
                const sVal = ((node.data && node.data[idKey]) || '') + '';
                /** @type {?} */
                const sLab = node.label || '';
                if (/^\w+$/.test(filterValue)) {
                    return sLab.startsWith(filterValue) || sVal.startsWith(filterValue) || py.startsWith(filterValue);
                }
                else {
                    return RegUtils.test(sVal, filterValue) || RegUtils.test(py, filterValue);
                }
            });
        }
    }
    /**
     * @return {?}
     */
    clearFilter() {
        this.filterValue = '';
        this.data$.next([...this.unFilteredValue]);
        this.filtered = false;
    }
    /**
     * @param {?} node
     * @return {?}
     */
    getSelectedNodeValues(node) {
        /** @type {?} */
        const value = this.getSelectedNodeValue(node);
        return CommonsUtils.flattenDeep(value);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onNodeExpand(event) {
        this.expandedNode = event.node;
        if (!event.node.children && this.$props.$ext.lazy) {
            this.onLazyLoad.emit(this.getEventContext(event));
        }
        else {
            this.resolveNode(event.node);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onNodeSelect(event) {
        this.selectedNode = event.node && event.node[PNG_TREE_NODE_KEYS.dataKey] || event.node;
        if (this.$props && this.$props.$id) {
            StatesUtils.update(this.$props.$id, {
                selectedNode: this.selectedNode
            });
        }
        this.updateTouch();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onNodeUnSelect(event) {
        this.updateTouch();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onTreeDataReady(event) {
        /** @type {?} */
        let treeNodes = this.createTreeNodes(event);
        this.initSelection(treeNodes);
        this.unFilteredValue = treeNodes;
        if (treeNodes.length > 0) {
            /** @type {?} */
            let level = this.$props.$ext.autoExpandLevel;
            /** @type {?} */
            let nodeId;
            if (this.expandedNode) {
                /** @type {?} */
                let node = this.expandedNode;
                nodeId = node['id'];
                while (node.parent) {
                    level++;
                    node = node.parent;
                }
            }
            TreeUtils.expandNode(treeNodes[0], true, this.$props.$ext, level, 1, nodeId);
        }
        else if (treeNodes.length === 0 || (treeNodes.length === 1 && !treeNodes['children'])) {
            this.onEmptyNode.emit();
        }
        if (this.$props.$ext.filter) {
            /** @type {?} */
            const filter = this.$props.$ext.filter;
            if (CommonsUtils.isString(filter)) {
                treeNodes = this.applyFilter((/** @type {?} */ (filter)));
            }
            else {
                treeNodes = filter['method'](treeNodes);
            }
        }
        this.data$.next(treeNodes);
    }
    /**
     * @param {?} treeNodes
     * @return {?}
     */
    initSelection(treeNodes) {
        if (this._selection) {
            this.tree.selection = this._selection;
            this.tree.propagateUp((/** @type {?} */ (this._selection)), true);
            return;
        }
        if (this.selectedNodeValues && treeNodes) {
            this._selection = [];
            /** @type {?} */
            const values = CommonsUtils.getArrayValue(this.selectedNodeValues);
            for (let i = 0; i < values.length; i++) {
                /** @type {?} */
                const found = TreeUtils.findNode(treeNodes, values[i], this.$props.$ext);
                if (found) {
                    ((/** @type {?} */ (this._selection))).push(found);
                }
            }
            this.tree.selection = this._selection;
            ((/** @type {?} */ (this._selection))).forEach(node => {
                this.tree.propagateUp(node, true);
            });
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this.$props && this.$props.$id) {
            StatesUtils.deleteState(this.$props.$id);
        }
        super.ngOnDestroy();
    }
    /**
     * @private
     * @param {?} node
     * @param {?=} nodes
     * @return {?}
     */
    getSelectedNodeValue(node, nodes) {
        if (CommonsUtils.isEmpty(node)) {
            return undefined;
        }
        if (node['isVirtual']) {
            return undefined;
        }
        nodes = nodes || [];
        /** @type {?} */
        const data = node[PNG_TREE_NODE_KEYS.dataKey];
        if (CommonsUtils.isEmpty(data)) {
            return undefined;
        }
        /** @type {?} */
        let val;
        if (this.$props.$ext.dataItemProps && this.$props.$ext.dataItemProps.valueKey) {
            val = DataItemUtils.getItemValue(data, this.$props.$ext.dataItemProps.valueKey);
        }
        else {
            val = {};
            val[this.$props.$ext.dataItemProps.idKey] = data[this.$props.$ext.dataItemProps.idKey];
            val[(/** @type {?} */ (this.$props.$ext.dataItemProps.labelKey))] = data[(/** @type {?} */ (this.$props.$ext.dataItemProps.labelKey))];
            if (this.$props.$ext.dataItemProps.typeKey) {
                val[this.$props.$ext.dataItemProps.typeKey] = data[this.$props.$ext.dataItemProps.typeKey];
            }
        }
        nodes.push(val);
        if (node[PNG_TREE_NODE_KEYS.parentKey]) {
            this.getSelectedNodeValue(node[PNG_TREE_NODE_KEYS.parentKey], nodes);
        }
        return nodes;
    }
    /**
     * @private
     * @param {?} nodes
     * @return {?}
     */
    createTreeNodes(nodes) {
        if (!nodes) {
            return [];
        }
        /** @type {?} */
        let treeNodes = [...nodes];
        if (this.$props.$ext.nodesFormat === 'plainArray') {
            treeNodes = TreeUtils.transNodes(treeNodes, {
                src: this.$props.$ext.dataItemProps,
                node: PNG_TREE_NODE_KEYS
            }, this.$props.$ext.root, this.$props.$ext.lazy);
        }
        if (treeNodes[0]) {
            TreeUtils.initNode(treeNodes[0], this.$props.$ext);
        }
        /** @type {?} */
        let rootNode = {};
        if (this.$props.$ext.virtualRoot) {
            if (CommonsUtils.isString(this.$props.$ext.virtualRoot)) {
                rootNode.label = (/** @type {?} */ (this.$props.$ext.virtualRoot));
            }
            else {
                rootNode = (/** @type {?} */ (this.$props.$ext.virtualRoot));
            }
            rootNode.children = treeNodes;
            rootNode['isVirtual'] = true;
            return [rootNode];
        }
        else {
            return treeNodes;
        }
    }
    /**
     * @private
     * @param {?} node
     * @return {?}
     */
    resolveNode(node) {
        if (node) {
            TreeUtils.initNode(node, this.$props.$ext);
            if (node.children) {
                node.children.forEach(child => TreeUtils.initNode(child, this.$props.$ext));
            }
        }
    }
}
PngTreeComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-tree',
                template: "<div class=\"ui-g ui-fluid\">\n\n  <div *ngIf=\"$props.$ext.hasSearchBox\" class=\"ui-g-12\">\n     <span class=\"inner-addon\">\n       <i class=\"fa fa-search text-primary left\"></i>\n       <input (keyup)=\"filter($event)\" [(ngModel)]=\"filterValue\" pInputText placeholder=\"\u8F93\u5165\u4EE3\u7801\u6216\u540D\u79F0\u9996\u5B57\u6BCD\" title=\"\u641C\u7D22\u8282\u70B9\"/>\n       <i (click)=\"clearFilter()\" *ngIf=\"filtered\" class=\"fa fa-trash text-muted right\" title=\"\u6E05\u9664\u641C\u7D22\u5185\u5BB9\"></i>\n     </span>\n  </div>\n\n  <p-card class=\"w-100\" erAnimate=\"slideInLeft\">\n    <span class=\"pull-right\">\n          <i class=\"fa fa-arrow-circle-up mr-1\" (click)=\"collapseAll()\" title=\"\u6536\u8D77\u5168\u90E8\"></i>\n          <i (click)=\"expandAll()\" class=\"fa fa-arrow-circle-down mr-1\" title=\"\u5C55\u5F00\u5168\u90E8\"></i>\n          <i (click)=\"load()\" class=\"fa fa-refresh\" title=\"\u5237\u65B0\u6811\"></i>\n      </span>\n    <p-tree #pngTree\n            erPropsBind\n            [props]=\"$props\"\n            [context]=\"ctx\"\n            [erApiData]=\"apiDataProps\"\n            [value]=\"data$|async\"\n            [(selection)]=\"selection\"\n            [loading]=\"loading$ | async\"\n            (onDataReady)=\"onTreeDataReady($event)\"\n            (onNodeExpand)=\"onNodeExpand($event)\"\n            (onNodeSelect)=\"onNodeSelect($event)\"\n            (onNodeUnselect)=\"onNodeUnSelect($event)\"\n    >\n      <ng-template pTemplate=\"default\" let-node>\n        <ng-container *ngTemplateOutlet=\"template(templateNames.node,defaultNodeTpl);\n                                  context:{$implicit: node,props:$props}\"></ng-container>\n      </ng-template>\n\n      <ng-container *ngIf=\"template(templateNames.icon)\">\n        <ng-template let-node pTemplate=\"picture\">\n          <ng-container *ngTemplateOutlet=\"template(templateNames.icon);\n                                  context:{$implicit: node,props:$props}\"></ng-container>\n        </ng-template>\n      </ng-container>\n\n      <ng-template #defaultNodeTpl let-node let-props=\"props\">\n        <!--<input type=\"radio\" name=\"tree_radio\" *ngIf=\"props.selectionMode === 'radio' && !node['children']\"/>-->\n        <span [innerHTML]=\"node['label']\" [ngClass]=\"{'ui-state-highlight':node.matched}\"></span>\n      </ng-template>\n\n    </p-tree>\n  </p-card>\n\n</div>\n",
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngTreeComponent),
                        multi: true
                    }],
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["/deep/ li:focus{outline:green solid 1px!important}/deep/ .ui-state-highlight{background-color:#9acd32!important;color:inherit!important;font-weight:700}/deep/ .ui-treenode-content-selected .ui-state-highlight{background-color:transparent!important;font-size:1.1em!important;font-weight:700;color:red!important}"]
            }] }
];
/** @nocollapse */
PngTreeComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
PngTreeComponent.propDecorators = {
    tree: [{ type: ViewChild, args: ['pngTree',] }],
    onEmptyNode: [{ type: Output }],
    onLazyLoad: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PngTreeComponent.prototype.tree;
    /** @type {?} */
    PngTreeComponent.prototype.unFilteredValue;
    /** @type {?} */
    PngTreeComponent.prototype.loading$;
    /** @type {?} */
    PngTreeComponent.prototype.filtered;
    /** @type {?} */
    PngTreeComponent.prototype.filterValue;
    /** @type {?} */
    PngTreeComponent.prototype.expandedNode;
    /** @type {?} */
    PngTreeComponent.prototype.templateNames;
    /** @type {?} */
    PngTreeComponent.prototype.onEmptyNode;
    /** @type {?} */
    PngTreeComponent.prototype.onLazyLoad;
    /** @type {?} */
    PngTreeComponent.prototype.$defaultProps;
    /** @type {?} */
    PngTreeComponent.prototype.selectedNode;
    /** @type {?} */
    PngTreeComponent.prototype.selectedNodeValues;
    /** @type {?} */
    PngTreeComponent.prototype.modelFormatter;
    /** @type {?} */
    PngTreeComponent.prototype.valueParser;
    /**
     * @type {?}
     * @private
     */
    PngTreeComponent.prototype._selection;
    /**
     * @type {?}
     * @private
     */
    PngTreeComponent.prototype.cd;
}
