// @er/core/esm5/lib/components/query-builder/rule/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '../../../bases/component';
import { QueryBuilderComponent } from '../component';
var QueryBuilderRuleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(QueryBuilderRuleComponent, _super);
    function QueryBuilderRuleComponent(qb) {
        var _this = _super.call(this) || this;
        _this.qb = qb;
        _this.ruleChange = new EventEmitter();
        return _this;
    }
    Object.defineProperty(QueryBuilderRuleComponent.prototype, "rule", {
        get: /**
         * @return {?}
         */
        function () {
            return this._rule;
        },
        set: /**
         * @param {?} rule
         * @return {?}
         */
        function (rule) {
            this._rule = rule;
            this.ruleChange.emit(this._rule);
        },
        enumerable: true,
        configurable: true
    });
    QueryBuilderRuleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-query-builder-rule',
                    template: "<div class=\"d-flex justify-content-between align-items-center rule\">\n\n  <div class=\"badge badge-info rounded-circle\">{{ruleIndex+1}}</div>\n\n  <ng-container *ngIf=\"qb.templates[qb.templateNames.removeButton] as custumer; else defaultRemoveButton\">\n    <div [ngClass]=\"qb.getClassNames('buttonGroup', 'rightAlign')\">\n      <ng-container *ngTemplateOutlet=\"custumer; context: qb.context({$implicit:rule})\"></ng-container>\n    </div>\n  </ng-container>\n\n  <div *ngIf=\"qb.config.entities?.length > 0\" class=\"q-inline-block-display\">\n    <ng-container *ngIf=\"qb.templates[qb.templateNames.entity] as customer; else defaultEntity\">\n      <ng-container *ngTemplateOutlet=\"customer; context: qb.context({$implicit:rule})\"></ng-container>\n    </ng-container>\n  </div>\n\n  <ng-container *ngIf=\"qb.templates[qb.templateNames.field] as customer; else defaultField\">\n    <ng-container *ngTemplateOutlet=\"customer; context: qb.context({$implicit:rule})\"></ng-container>\n  </ng-container>\n\n  <ng-container *ngIf=\"qb.templates[qb.templateNames.operator] as customer; else defaultOperator\">\n    <ng-container *ngTemplateOutlet=\"customer; context: qb.context({$implicit:rule})\"></ng-container>\n  </ng-container>\n\n  <div [ngClass]=\"qb.getClassNames('inputControlSize')\">\n    <ng-container\n      *ngTemplateOutlet=\"qb.templates[qb.templateNames.input]; context: qb.context({$implicit:rule})\"></ng-container>\n  </div>\n\n</div>\n\n<ng-template #defaultRemoveButton>\n  <div [ngClass]=\"qb.getClassNames('removeButtonSize', 'rightAlign')\">\n    <button [ngClass]=\"qb.getClassNames('button', 'removeButton')\" (click)=\"qb.removeRule(rule)\"\n            [disabled]=qb.disabled>\n      <i [ngClass]=\"qb.getClassNames('removeIcon')\"></i>\n    </button>\n  </div>\n</ng-template>\n\n<ng-template #defaultEntity>\n  <div [ngClass]=\"qb.getClassNames('entityControlSize')\">\n    <select [ngClass]=\"qb.getClassNames('entityControl')\" [(ngModel)]=\"rule.entity\"\n            (change)=\"qb.changeEntity($event, rule)\" [disabled]=\"qb.disabled\">\n      <option *ngFor=\"let entity of qb.config.entities\" [ngValue]=\"entity.key\">\n        {{entity.label}}\n      </option>\n    </select>\n  </div>\n</ng-template>\n\n<ng-template #defaultField>\n  <div [ngClass]=\"qb.getClassNames('fieldControlSize')\">\n    <select [ngClass]=\"qb.getClassNames('fieldControl')\" [(ngModel)]=\"rule.field\"\n            (change)=\"qb.changeField($event, rule)\" [disabled]=\"qb.disabled\">\n      <option *ngFor=\"let field of qb.getFields(rule.entity)\" [value]=\"field.key\">\n        {{field.label}}\n      </option>\n    </select>\n  </div>\n</ng-template>\n\n<ng-template #defaultOperator>\n  <div [ngClass]=\"qb.getClassNames('operatorControlSize')\">\n    <select [ngClass]=\"qb.getClassNames('operatorControl')\" [(ngModel)]=\"rule.operator\"\n            [disabled]=\"qb.disabled\">\n      <option *ngFor=\"let operator of qb.getOperators(rule.field)\" [value]=\"operator.value\">\n        {{operator.label}}\n      </option>\n    </select>\n  </div>\n</ng-template>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".rule:hover{background-color:green}"]
                }] }
    ];
    /** @nocollapse */
    QueryBuilderRuleComponent.ctorParameters = function () { return [
        { type: QueryBuilderComponent }
    ]; };
    QueryBuilderRuleComponent.propDecorators = {
        ruleIndex: [{ type: Input }],
        ruleChange: [{ type: Output }],
        rule: [{ type: Input }]
    };
    return QueryBuilderRuleComponent;
}(BaseComponent));
export { QueryBuilderRuleComponent };
if (false) {
    /** @type {?} */
    QueryBuilderRuleComponent.prototype.ruleIndex;
    /** @type {?} */
    QueryBuilderRuleComponent.prototype.ruleChange;
    /**
     * @type {?}
     * @private
     */
    QueryBuilderRuleComponent.prototype._rule;
    /** @type {?} */
    QueryBuilderRuleComponent.prototype.qb;
}
