// @er/primeng/esm5/lib/components/notify.alert/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component } from '@angular/core';
import { BaseComponent } from '@er/core';
import { NotifyLevel, NotifyType, StateNames, UiPosition } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { MessageService } from 'primeng/api';
import { filter } from 'rxjs/operators';
/** @type {?} */
export var DEFAULT_ALERT_PROPS = {
    animation: true,
    duration: 3000,
    immutable: true,
    closable: true,
    autoZIndex: true,
    position: UiPosition.rightTop
};
var PngAlertComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngAlertComponent, _super);
    function PngAlertComponent(messageService, cd) {
        var _this = _super.call(this) || this;
        _this.messageService = messageService;
        _this.cd = cd;
        _this.alert = tslib_1.__assign({}, DEFAULT_ALERT_PROPS, ConfigUtils.getConfig().alert || {});
        _this.defaultNotifyType = ConfigUtils.getConfig().notify || NotifyType.ALERT;
        return _this;
    }
    /**
     * @return {?}
     */
    PngAlertComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.$subscriptions = StatesUtils.observe(StateNames.notify)
            .pipe(filter(function (notify) { return notify && notify.message; }), filter(function (notify) { return notify.type === NotifyType.ALERT || (!notify.type && _this.defaultNotifyType === NotifyType.ALERT); })).subscribe(function (notify) {
            /** @type {?} */
            var model = _this.getNotifyModel(notify);
            _this.messageService.add(model);
            _this.cd.detectChanges();
        });
    };
    /**
     * @param {?} notify
     * @return {?}
     */
    PngAlertComponent.prototype.getNotifyModel = /**
     * @param {?} notify
     * @return {?}
     */
    function (notify) {
        return {
            key: notify.key,
            severity: notify.level || NotifyLevel.INFO,
            summary: notify.title,
            detail: notify.message,
            sticky: notify.sticky || false,
            life: notify.duration || notify.level === NotifyLevel.ERROR ? DEFAULT_ALERT_PROPS.duration * 2 : DEFAULT_ALERT_PROPS.duration,
            closable: notify.closable || true,
            data: notify.data
        };
    };
    PngAlertComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-alert',
                    template: "\n    <p-toast [styleClass]=\"alert.styleClass\"\n             [style]=\"alert.style\"\n             [position]=\"alert.position\"\n             [autoZIndex]=\"alert.autoZIndex\"\n    ></p-toast>\n  ",
                    providers: [MessageService],
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngAlertComponent.ctorParameters = function () { return [
        { type: MessageService },
        { type: ChangeDetectorRef }
    ]; };
    return PngAlertComponent;
}(BaseComponent));
export { PngAlertComponent };
if (false) {
    /** @type {?} */
    PngAlertComponent.prototype.alert;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.defaultNotifyType;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    PngAlertComponent.prototype.cd;
}
