// @er/core/esm2015/lib/bases/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, Input, QueryList, TemplateRef } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { TypedTemplateDirective } from '../directives/typed-template/directive';
import { BaseDirective } from './directive';
/**
 * provide template & dynamic query support
 * @abstract
 */
export class BaseComponent extends BaseDirective {
    constructor() {
        super(...arguments);
        this.templates = {};
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (this.typedTemplates && this.typedTemplates.length > 0) {
            this.typedTemplates.forEach((item) => {
                if (item.type) {
                    CommonsUtils.set(this.templates, item.type, item.template);
                }
                else {
                    this.templates['_'] = item.template;
                }
            });
        }
    }
    /**
     * @param {?} outputs
     * @param {?} eventName
     * @param {?} defaultHandler
     * @param {?} event
     * @return {?}
     */
    eventHandler(outputs, eventName, defaultHandler, event) {
        if (outputs && outputs[eventName]) {
            return outputs[eventName].call(this, event);
        }
        else {
            return defaultHandler(event);
        }
    }
    /**
     * @param {?=} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    template(name, ...defaultTpls) {
        if (!name) {
            name = '_';
        }
        if (this.templates[name]) {
            return this.templates[name];
        }
        else if (defaultTpls.length > 0) {
            /** @type {?} */
            const tpls = CommonsUtils.filter(defaultTpls, defaultTpl => !!defaultTpl);
            if (tpls.length > 0) {
                return tpls[0];
            }
        }
        return undefined;
    }
    /**
     * @template C
     * @param {?} id
     * @return {?}
     */
    getComponentBySelectorId(id) {
        if (this.selectors) {
            /** @type {?} */
            const matched = this.selectors.toArray().filter(comp => comp.id === id);
            if (matched.length === 1) {
                return matched[0].hostComponent;
            }
        }
        return undefined;
    }
    /**
     * @param {?} name
     * @param {...?} defaultTpls
     * @return {?}
     */
    hasTemplate(name, ...defaultTpls) {
        return this.template(name, ...defaultTpls) !== undefined;
    }
    /**
     * @param {?} templates
     * @return {?}
     */
    hasChildren(templates) {
        return CommonsUtils.isJson(templates);
    }
    // empty hooks for dyna bind
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this['__ngAfterViewInit'] = true;
    }
    /**
     * @return {?}
     */
    ngAfterViewChecked() {
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this['__ngOnInit'] = true;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
    }
}
BaseComponent.propDecorators = {
    customerTpl: [{ type: Input }],
    typedTemplates: [{ type: ContentChildren, args: [TypedTemplateDirective,] }]
};
if (false) {
    /** @type {?} */
    BaseComponent.prototype.customerTpl;
    /** @type {?} */
    BaseComponent.prototype.typedTemplates;
    /** @type {?} */
    BaseComponent.prototype.selectors;
    /** @type {?} */
    BaseComponent.prototype.templateNames;
    /** @type {?} */
    BaseComponent.prototype.templates;
}
