// @er/core/esm2015/lib/bases/control-wrapper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HostBinding, Input } from '@angular/core';
import { NO_OP } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { WrapperComponent } from './wrappe';
/**
 * @template T
 */
export class ControlWrapperComponent extends WrapperComponent {
    constructor() {
        super(...arguments);
        this.resettable = true;
        this.disabled = false;
        this.onTouched = NO_OP;
        this.onChange = NO_OP;
    }
    /**
     * @return {?}
     */
    get opacity() {
        return this.disabled ? 0.25 : 1;
    }
    /**
     * @return {?}
     */
    get cursor() {
        return this.disabled ? 'not-allowed' : 'default';
    }
    /**
     * @return {?}
     */
    get value() {
        return this.innerValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        /** @type {?} */
        const v = this.valueParser ? this.valueParser(value) : value;
        if (!CommonsUtils.isEqual(v, this.innerValue)) {
            this.innerValue = v;
            this.onChange(v);
        }
    }
    /**
     * @return {?}
     */
    updateTouch() {
        this.onTouched();
    }
    /**
     * @param {?} model
     * @return {?}
     */
    writeValue(model) {
        /** @type {?} */
        const v = this.modelFormatter ? this.modelFormatter(model) : model;
        if (v !== this.innerValue) {
            this.innerValue = v;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        if (this.disabled) {
            this.$props['disabled'] = this.disabled;
        }
    }
    /**
     * @return {?}
     */
    reset() {
        this.value = null;
    }
}
ControlWrapperComponent.propDecorators = {
    valueParser: [{ type: Input }],
    modelFormatter: [{ type: Input }],
    resettable: [{ type: Input }],
    disabled: [{ type: Input }],
    opacity: [{ type: HostBinding, args: ['style.opacity',] }],
    cursor: [{ type: HostBinding, args: ['style.cursor',] }]
};
if (false) {
    /** @type {?} */
    ControlWrapperComponent.prototype.valueParser;
    /** @type {?} */
    ControlWrapperComponent.prototype.modelFormatter;
    /** @type {?} */
    ControlWrapperComponent.prototype.resettable;
    /** @type {?} */
    ControlWrapperComponent.prototype.disabled;
    /** @type {?} */
    ControlWrapperComponent.prototype.onTouched;
    /** @type {?} */
    ControlWrapperComponent.prototype.onChange;
    /**
     * @type {?}
     * @protected
     */
    ControlWrapperComponent.prototype.innerValue;
}
