// @er/core/esm2015/lib/bases/control.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HostBinding, Input } from '@angular/core';
import { NO_OP } from '@er/types';
import { BehaviorSubject } from 'rxjs';
import { BaseComponent } from './component';
export class ControlComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.disabled = false;
        this.disabled$ = new BehaviorSubject(false);
        this.onTouched = NO_OP;
        this.onChange = NO_OP;
    }
    /**
     * @return {?}
     */
    get opacity() {
        return this.disabled ? 0.25 : 1;
    }
    /**
     * @return {?}
     */
    get cursor() {
        return this.disabled ? 'not-allowed' : 'default';
    }
    /**
     * @return {?}
     */
    get value() {
        return this.innerValue;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        if (!value || value !== this.innerValue) {
            /** @type {?} */
            const v = this.parser ? this.parser(value) : value;
            this.innerValue = v;
            this.onChange(v);
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.defaultValue) {
            this.innerValue = this.defaultValue;
        }
    }
    /**
     * @return {?}
     */
    updateTouch() {
        this.onTouched();
    }
    /**
     * @param {?} model
     * @return {?}
     */
    writeValue(model) {
        /** @type {?} */
        const v = this.formatter ? this.formatter(model) : model;
        if (v !== this.innerValue) {
            this.innerValue = v;
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
        this.disabled$.next(isDisabled);
    }
}
ControlComponent.propDecorators = {
    parser: [{ type: Input }],
    formatter: [{ type: Input }],
    disabled: [{ type: Input }],
    opacity: [{ type: HostBinding, args: ['style.opacity',] }],
    cursor: [{ type: HostBinding, args: ['style.cursor',] }]
};
if (false) {
    /** @type {?} */
    ControlComponent.prototype.parser;
    /** @type {?} */
    ControlComponent.prototype.formatter;
    /** @type {?} */
    ControlComponent.prototype.disabled;
    /** @type {?} */
    ControlComponent.prototype.defaultValue;
    /** @type {?} */
    ControlComponent.prototype.disabled$;
    /** @type {?} */
    ControlComponent.prototype.onTouched;
    /** @type {?} */
    ControlComponent.prototype.onChange;
    /**
     * @type {?}
     * @protected
     */
    ControlComponent.prototype.innerValue;
}
