// @er/core/esm2015/lib/bases/data-aware-control-wrapper.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, ViewChild } from '@angular/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { BehaviorSubject } from 'rxjs';
import { ApiDataDirective } from '../directives/api-data/directive';
import { LogUtils } from '../utils/log';
import { ControlWrapperComponent } from './control-wrapper';
/**
 * @template T
 */
export class DataAwareControlWrapperComponent extends ControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.data$ = new BehaviorSubject(undefined);
    }
    /**
     * @return {?}
     */
    get apiDataProps() {
        return this._apiDataProps || {};
    }
    /**
     * @param {?} props
     * @return {?}
     */
    set apiDataProps(props) {
        this._apiDataProps = props;
        if (this._apiDataProps && !this._apiDataProps.data$) {
            this._apiDataProps.data$ = this.data$;
        }
    }
    /**
     * @return {?}
     */
    onPropsInit() {
        super.onPropsInit();
        this.initLocalDataItem();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    parseValue(value) {
        try {
            if (value && this.$props.$ext.dataItemProps && !CommonsUtils.isString(value)) {
                return DataItemUtils.getItemValue(value, this.$props.$ext.dataItemProps.valueKey);
            }
            return value;
        }
        catch (e) {
            LogUtils.error(this, '数据解析错误', e);
            return undefined;
        }
    }
    /**
     * @param {?} model
     * @return {?}
     */
    formatModel(model) {
        try {
            if (CommonsUtils.isJson(model) && this.$props.$ext.dataItemProps) {
                return DataItemUtils.getItemLabel(model, this.$props.$ext.dataItemProps.labelKey);
            }
            return model;
        }
        catch (e) {
            LogUtils.error(this, '数据解析错误', e);
            return undefined;
        }
    }
    /**
     * @return {?}
     */
    initLocalDataItem() {
        if (this.$props.$ext && this.$props.$ext.dataItems) {
            /** @type {?} */
            let data = this.$props.$ext.dataItems;
            data = data.map(d => d['data'] = d);
            this.data$.next(data);
        }
    }
}
DataAwareControlWrapperComponent.propDecorators = {
    apiData: [{ type: ViewChild, args: [ApiDataDirective,] }],
    apiDataProps: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    DataAwareControlWrapperComponent.prototype.data$;
    /** @type {?} */
    DataAwareControlWrapperComponent.prototype.apiData;
    /**
     * @type {?}
     * @private
     */
    DataAwareControlWrapperComponent.prototype._apiDataProps;
}
