// @er/core/esm2015/lib/bases/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter } from '@angular/core';
import { Unsubscribe } from './unsubscribe';
/**
 * provide event emit support
 * @abstract
 */
export class BaseDirective extends Unsubscribe {
    constructor() {
        super();
        this.animations = {};
        this.ctx = {
            $from: (/** @type {?} */ (this))
        };
    }
    /**
     * @return {?}
     */
    getContext() {
        /** @type {?} */
        let id;
        if (!this['$id'] && this['$props'] && this['$props'].$id) {
            id = this['$props'].$id;
            this['$id'] = id;
        }
        /** @type {?} */
        const context = {
            $from: (/** @type {?} */ (this))
        };
        return id ? Object.assign({ $id: id }, context) : context;
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    getEventContext(event) {
        return Object.assign({}, this.getContext(), event || {});
    }
    /**
     * @param {?} eventName
     * @param {?} event
     * @return {?}
     */
    emitEvent(eventName, event) {
        /** @type {?} */
        let eventEmitter = this[eventName];
        if (!eventEmitter) {
            eventEmitter = new EventEmitter();
            this[eventName] = eventEmitter;
        }
        if (eventEmitter instanceof EventEmitter) {
            return eventEmitter.emit(this.getEventContext(event));
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    isFirstChanges(changes) {
        return Object.keys(changes).every(prop => changes[prop].isFirstChange());
    }
    /**
     * @param {?=} name
     * @return {?}
     */
    animate(name) {
        return !(this.animations[name] || false);
    }
}
if (false) {
    /** @type {?} */
    BaseDirective.prototype.ctx;
    /** @type {?} */
    BaseDirective.prototype.animations;
}
