// @er/core/esm2015/lib/bases/form.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { CommonsUtils, DialogUtils } from '@er/utils';
import { Observable } from 'rxjs';
import { first } from 'rxjs/operators';
import { ApiUtils } from '../utils/api';
import { BaseComponent } from './component';
/**
 * @abstract
 */
export class BaseFormComponent extends BaseComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        super();
        this.cd = cd;
        this.options = {};
        this.buttonsStyleClass = 'text-center my-5';
        this.submitBtnStyleClass = 'btn btn-success';
        this.resetBtnStyleClass = 'btn btn-danger';
        this.initBtnStyleClass = 'btn btn-alert';
        this.onSubmit = new EventEmitter();
        this.onSubmitted = new EventEmitter();
        this.onModel = new EventEmitter();
        this.onFormValueChanges = new EventEmitter();
        this.form = new FormGroup({});
        this.templateNames = {
            header: 'header',
            buttons: 'buttons',
            footer: 'footer'
        };
        this.runtimeData = {};
        this.hasUpdated = false;
    }
    /**
     * @return {?}
     */
    get model() {
        return this._model;
    }
    /**
     * @param {?} model
     * @return {?}
     */
    set model(model) {
        this._model = CommonsUtils.omitBy(model, CommonsUtils.lodash().isUndefined);
        this.refresh();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        this.$subscriptions = this.form.valueChanges.subscribe(event => this.onFormValueChanges.emit(event));
    }
    /**
     * @return {?}
     */
    get modelId() {
        return this._modelId;
    }
    /**
     * @param {?} modelId
     * @return {?}
     */
    set modelId(modelId) {
        if (modelId !== this.modelId) {
            this._modelId = modelId;
            if (!modelId || modelId === '') {
                this.resetModel();
            }
            else if (this.apiEntry) {
                ApiUtils.getById(this.apiEntry, modelId, this.getModelOptions)
                    .subscribe(result => {
                    this.model = result;
                    this.form.updateValueAndValidity({ onlySelf: false, emitEvent: true });
                });
            }
        }
    }
    /**
     * @return {?}
     */
    get isFormValid() {
        if (!this.hasUpdated) {
            this.updateControls();
        }
        return this.form.valid;
    }
    /**
     * @return {?}
     */
    get httpTag() {
        return `http${'-' + this.formId || ''}`;
    }
    /**
     * @return {?}
     */
    get formId() {
        return this.$id;
    }
    /**
     * @return {?}
     */
    get submitModel() {
        /** @type {?} */
        let model = Object.assign({}, this.model);
        if (this.submitRootKey) {
            model = this.model[this.submitRootKey];
        }
        if (this.submitExcludeKeys) {
            model = CommonsUtils.removeKeys(model, ...CommonsUtils.getArrayValue(this.submitExcludeKeys));
        }
        Object.keys(model).forEach(key => {
            if (key.startsWith('$_')) {
                delete model[key];
            }
        });
        return model;
    }
    /**
     * @param {?} fieldName
     * @return {?}
     */
    formControl(fieldName) {
        return this.form.controls[fieldName];
    }
    /**
     * @param {?} fieldName
     * @return {?}
     */
    fieldComponent(fieldName) {
        return this.form['components'][fieldName];
    }
    /**
     * @return {?}
     */
    getFormErrorFields() {
        /** @type {?} */
        const errors = [];
        Object.keys(this.form.controls).forEach(key => {
            /** @type {?} */
            const controlErrors = this.form.get(key).errors;
            if (controlErrors != null) {
                Object.keys(controlErrors).forEach(keyError => {
                    errors.push(key); // + ', err value: ', controlErrors[keyError]);
                });
            }
        });
        return errors;
    }
    /**
     * @param {?} model
     * @return {?}
     */
    patchModel(model) {
        this.model = Object.assign({}, this.model || {}, model);
    }
    /**
     * @param {?} data
     * @return {?}
     */
    patchRuntimeData(data) {
        Object.assign(this.runtimeData, data);
    }
    /**
     * @param {?} path
     * @return {?}
     */
    isFieldValid(path) {
        /** @type {?} */
        const field = this.field(path);
        if (field) {
            return field.valid;
        }
        else {
            return undefined;
        }
    }
    /**
     * @param {?} fieldNamePath
     * @return {?}
     */
    isFieldTouched(fieldNamePath) {
        /** @type {?} */
        const field = this.field(fieldNamePath);
        if (field) {
            return field.touched;
        }
        else {
            return undefined;
        }
    }
    /**
     * @param {?=} form
     * @return {?}
     */
    updateControls(form) {
        if (!form) {
            form = this.form;
        }
        Object.keys(form.controls).forEach(field => {
            /** @type {?} */
            const control = form.get(field);
            if (control instanceof FormControl) {
                control.markAsTouched({ onlySelf: true });
                // control.updateValueAndValidity({onlySelf: true});
            }
            else if (control instanceof FormGroup) {
                this.updateControls(control);
            }
        });
        this.hasUpdated = true;
    }
    /**
     * @param {?} path
     * @return {?}
     */
    field(path) {
        return this.form.get(path);
    }
    /**
     * @return {?}
     */
    resetForm() {
        this.form.reset({});
    }
    /**
     * @param {?=} reset
     * @return {?}
     */
    resetModel(reset) {
        if (this.initModel) {
            this.model = CommonsUtils.copy(this.initModel);
        }
        else if (CommonsUtils.isJson(reset)) {
            this.model = Object.assign({}, reset);
        }
        else if (this.options['resetModel']) {
            this.options['resetModel']();
        }
        else {
            this.model = {};
        }
        this.form.reset(this.model, { emitEvent: true });
    }
    /**
     * @return {?}
     */
    setInitModel() {
        this.initModel = CommonsUtils.copy(this.model);
        if (this.options['updateInitialValue']) {
            this.options['updateInitialValue']();
        }
    }
    /**
     * @return {?}
     */
    refresh() {
        this.cd.detectChanges();
    }
    /**
     * @return {?}
     */
    submit() {
        if (!this.isFormValid) {
            DialogUtils.error('操作失败', '请填写完整信息后，再执行本操作！');
            return;
        }
        /** @type {?} */
        let result;
        if (this.beforeSubmit) {
            result = this.beforeSubmit(this.getEventContext());
            if (result instanceof Observable) {
                ((/** @type {?} */ (result)))
                    .pipe(first())
                    .subscribe(data => {
                    this.submitAfterBefore(data);
                });
            }
            else {
                this.submitAfterBefore(result);
            }
        }
        else {
            this.doSubmit();
        }
    }
    /**
     * @param {?=} submitModel
     * @return {?}
     */
    defaultSubmit(submitModel) {
        this.$subscriptions = ApiUtils.save.call(this, this.apiEntry, submitModel || this.submitModel, Object.assign({ tag: this.formId }, this.submitOptions || {}))
            .subscribe(data => {
            if (data.id && data.id !== this.model['id']) {
                this.model['id'] = data.id;
            }
            this.onSubmitted.emit(this.getEventContext({ response: data, model: this.model, submitted: submitModel || this.submitModel }));
            if (this.afterSubmit) {
                /** @type {?} */
                const result = this.afterSubmit(this.getEventContext({ submitModel: submitModel || this.submitModel, id: data.id }));
                if (result instanceof Observable) {
                    ((/** @type {?} */ (result)))
                        .pipe(first())
                        .subscribe(dat => {
                        this.doPostSubmit(dat);
                    });
                }
                else {
                    this.doPostSubmit(result);
                }
            }
            else {
                this.doPostSubmit();
            }
        });
    }
    /**
     * @param {?=} other
     * @return {?}
     */
    getEventContext(other) {
        return (Object.assign({ formlyForm: this, form: (/** @type {?} */ (this.form)), model: this.model, modelId: this.modelId, entry: this.apiEntry, entity: (/** @type {?} */ (this.entity)) }, other || {}));
    }
    /**
     * @private
     * @param {?} result
     * @return {?}
     */
    submitAfterBefore(result) {
        if (result !== false) {
            if (CommonsUtils.isJson(result) && !CommonsUtils.isEmpty(result)) {
                this.doSubmit(result);
            }
            else {
                this.doSubmit();
            }
        }
    }
    /**
     * @private
     * @param {?=} submitModel
     * @return {?}
     */
    doSubmit(submitModel) {
        if (this.onSubmit.observers.length > 0) {
            this.onSubmit.emit(this.getEventContext({ submitModel }));
        }
        else {
            this.defaultSubmit(submitModel);
        }
    }
    /**
     * @private
     * @param {?=} model
     * @return {?}
     */
    doPostSubmit(model) {
        if (model) {
            this.resetModel(model);
        }
        else if (this.resetKeys) {
            /** @type {?} */
            const keys = CommonsUtils.getArrayValue(this.resetKeys);
            keys.forEach(key => delete this.model[key]);
            delete this.model['id'];
            this.form.reset(this.model, { emitEvent: true });
        }
        else if (this.resetAfterSubmit !== false) {
            this.resetModel(this.entity && this.entity.initValue || {});
        }
    }
}
BaseFormComponent.propDecorators = {
    $id: [{ type: Input }],
    entity: [{ type: Input }],
    apiEntry: [{ type: Input }],
    options: [{ type: Input }],
    submitRootKey: [{ type: Input }],
    submitExcludeKeys: [{ type: Input }],
    resetKeys: [{ type: Input }],
    styleClass: [{ type: Input }],
    style: [{ type: Input }],
    formStyleClass: [{ type: Input }],
    formStyle: [{ type: Input }],
    buttonsStyleClass: [{ type: Input }],
    submitBtnStyleClass: [{ type: Input }],
    resetBtnStyleClass: [{ type: Input }],
    initBtnStyleClass: [{ type: Input }],
    header: [{ type: Input }],
    footer: [{ type: Input }],
    buttons: [{ type: Input }],
    getModelOptions: [{ type: Input }],
    submitOptions: [{ type: Input }],
    beforeSubmit: [{ type: Input }],
    afterSubmit: [{ type: Input }],
    valueChanges: [{ type: Input }],
    resetAfterSubmit: [{ type: Input }],
    onSubmit: [{ type: Output }],
    onSubmitted: [{ type: Output }],
    onModel: [{ type: Output }],
    onFormValueChanges: [{ type: Output }],
    model: [{ type: Input }],
    modelId: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    BaseFormComponent.prototype.$id;
    /** @type {?} */
    BaseFormComponent.prototype.entity;
    /** @type {?} */
    BaseFormComponent.prototype.apiEntry;
    /** @type {?} */
    BaseFormComponent.prototype.options;
    /** @type {?} */
    BaseFormComponent.prototype.submitRootKey;
    /** @type {?} */
    BaseFormComponent.prototype.submitExcludeKeys;
    /** @type {?} */
    BaseFormComponent.prototype.resetKeys;
    /** @type {?} */
    BaseFormComponent.prototype.styleClass;
    /** @type {?} */
    BaseFormComponent.prototype.style;
    /** @type {?} */
    BaseFormComponent.prototype.formStyleClass;
    /** @type {?} */
    BaseFormComponent.prototype.formStyle;
    /** @type {?} */
    BaseFormComponent.prototype.buttonsStyleClass;
    /** @type {?} */
    BaseFormComponent.prototype.submitBtnStyleClass;
    /** @type {?} */
    BaseFormComponent.prototype.resetBtnStyleClass;
    /** @type {?} */
    BaseFormComponent.prototype.initBtnStyleClass;
    /** @type {?} */
    BaseFormComponent.prototype.header;
    /** @type {?} */
    BaseFormComponent.prototype.footer;
    /** @type {?} */
    BaseFormComponent.prototype.buttons;
    /** @type {?} */
    BaseFormComponent.prototype.getModelOptions;
    /** @type {?} */
    BaseFormComponent.prototype.submitOptions;
    /** @type {?} */
    BaseFormComponent.prototype.beforeSubmit;
    /** @type {?} */
    BaseFormComponent.prototype.afterSubmit;
    /** @type {?} */
    BaseFormComponent.prototype.valueChanges;
    /** @type {?} */
    BaseFormComponent.prototype.resetAfterSubmit;
    /** @type {?} */
    BaseFormComponent.prototype.onSubmit;
    /** @type {?} */
    BaseFormComponent.prototype.onSubmitted;
    /** @type {?} */
    BaseFormComponent.prototype.onModel;
    /** @type {?} */
    BaseFormComponent.prototype.onFormValueChanges;
    /** @type {?} */
    BaseFormComponent.prototype.form;
    /** @type {?} */
    BaseFormComponent.prototype.templateNames;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.initModel;
    /** @type {?} */
    BaseFormComponent.prototype.runtimeData;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype.hasUpdated;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype._model;
    /**
     * @type {?}
     * @private
     */
    BaseFormComponent.prototype._modelId;
    /**
     * @type {?}
     * @protected
     */
    BaseFormComponent.prototype.cd;
}
