// @er/core/esm2015/lib/bases/roue.bind.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonsUtils } from '@er/utils';
import { ComponentUtils } from '../utils/component';
import { LogUtils } from '../utils/log';
import { RouteUtils } from '../utils/nav';
import { BaseComponent } from './component';
export class RouteBindComponent extends BaseComponent {
    /**
     * @return {?}
     */
    ngOnInit() {
        this.$subscriptions = RouteUtils.getActiveRoute().url.subscribe(url => {
            /** @type {?} */
            let activeRoute = RouteUtils.getActiveRoute();
            if (!activeRoute) {
                return;
            }
            /** @type {?} */
            const data = CommonsUtils.copy(activeRoute.snapshot.data);
            /** @type {?} */
            const params = Object.assign({}, activeRoute.snapshot.params || {}, activeRoute.snapshot.queryParams || {});
            while (activeRoute && !activeRoute.snapshot.data['$props']) {
                activeRoute = activeRoute.parent;
            }
            /** @type {?} */
            const routeData = Object.assign({}, activeRoute.snapshot.data, data);
            /** @type {?} */
            let props = CommonsUtils.copy(routeData['$resolvedProps'] || routeData['$props']);
            if (data['$propsKey']) {
                props = CommonsUtils.get(props, data['$propsKey']);
            }
            Object.keys(data).forEach(key => {
                if (CommonsUtils.isFunction(data[key])) {
                    props[key] = ((/** @type {?} */ (data[key])))(props, params, data);
                }
                else if (CommonsUtils.isString(data[key]) && data[key].startsWith(':')) {
                    /** @type {?} */
                    const param = ((/** @type {?} */ (data[key]))).substring(1);
                    if (params[param]) {
                        props[key] = params[param];
                    }
                }
                else if (key !== '$propsKey' && key !== '$propsWrapperKey' && key !== '$props') {
                    props[key] = data[key];
                }
            });
            props['$routeParams'] = params;
            LogUtils.debug(this, '路由数据', routeData);
            props = this.beforeRoutePropsBind(props);
            ComponentUtils.bind(this, props);
            this.afterRoutePropsBind();
        });
    }
    /**
     * @param {?} props
     * @return {?}
     */
    beforeRoutePropsBind(props) {
        return props;
    }
    /**
     * @return {?}
     */
    afterRoutePropsBind() {
    }
}
