// @er/core/esm2015/lib/bases/wrappe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { CommonsUtils, IdUtils } from '@er/utils';
import { BaseComponent } from './component';
/**
 * @abstract
 * @template T
 */
export class WrapperComponent extends BaseComponent {
    /**
     * @return {?}
     */
    _initProps() {
        this.$props = CommonsUtils.defaults({}, this.$props || {}, this.$defaultProps || {});
        if (!this.$props.$id) {
            this.$props.$id = IdUtils.getRandom();
        }
        if (!this.$props.$ext) {
            this.$props.$ext = {};
        }
        if (this.$props.$ext['apiDataProps']) {
            if (!this.$props.$ext['apiDataProps'].requestOptions) {
                this.$props.$ext['apiDataProps'].requestOptions = {};
            }
            CommonsUtils.defaults(this.$props.$ext['apiDataProps'].requestOptions, {
                context: this,
                tag: this.$props.$id
            });
        }
        this.onPropsInit();
    }
    /**
     * @return {?}
     */
    _propsBind() {
        this.onPropsBind();
    }
    /**
     * @return {?}
     */
    onPropsInit() { }
    /**
     * @return {?}
     */
    onPropsBind() {
    }
}
WrapperComponent.propDecorators = {
    outputContext: [{ type: Input }],
    attrs: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    WrapperComponent.prototype.$props;
    /** @type {?} */
    WrapperComponent.prototype.outputContext;
    /** @type {?} */
    WrapperComponent.prototype.attrs;
    /**
     * @type {?}
     * @protected
     */
    WrapperComponent.prototype.$defaultProps;
}
