// @er/formly-bootstrap/esm2015/lib/config/formly.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FormlyWrappers } from '@er/formly';
import { BsArrayTypeComponent } from '../field-types/array/type';
import { BsCheckboxTypeComponent } from '../field-types/basic/checkbox';
import { BsInputTypeComponent } from '../field-types/basic/input';
import { BsMultiCheckboxTypeComponent } from '../field-types/basic/multicheckbox';
import { BsTextAreaTypeComponent } from '../field-types/basic/textarea';
import { BsButtonTypeComponent } from '../field-types/button/type';
import { BsSwitchTypeComponent } from '../field-types/switch/type';
import { BsAddonsWrapperComponent } from '../field-wrappers/addons';
import { TemplateAddons } from '../run/addon';
import { BsFormlyTypes } from '../types/field-types';
import { FormlyBsWrappers } from '../types/field-wrappers';
/** @type {?} */
export const BS_FORMLY_CONFIG = {
    types: [
        {
            name: BsFormlyTypes.input,
            component: BsInputTypeComponent,
            wrappers: [FormlyWrappers.vertical]
        },
        {
            name: BsFormlyTypes.checkbox,
            component: BsCheckboxTypeComponent,
            wrappers: [FormlyWrappers.vertical],
            defaultOptions: {
                templateOptions: {
                    indeterminate: true,
                    hideLabel: true
                }
            }
        },
        {
            name: BsFormlyTypes.textArea,
            component: BsTextAreaTypeComponent,
            wrappers: [FormlyWrappers.vertical],
            defaultOptions: {
                templateOptions: {
                    cols: 1,
                    rows: 1
                }
            }
        },
        {
            name: BsFormlyTypes.multiCheckbox,
            component: BsMultiCheckboxTypeComponent,
            wrappers: [FormlyWrappers.vertical],
            defaultOptions: {
                templateOptions: {
                    options: []
                }
            }
        },
        {
            name: BsFormlyTypes.switch,
            component: BsSwitchTypeComponent,
            wrappers: [FormlyWrappers.vertical]
        },
        {
            name: BsFormlyTypes.array,
            component: BsArrayTypeComponent,
            wrappers: [FormlyWrappers.vertical]
        },
        {
            name: BsFormlyTypes.button,
            component: BsButtonTypeComponent,
            wrappers: [FormlyWrappers.vertical]
        }
    ],
    wrappers: [
        { name: FormlyBsWrappers.addon, component: BsAddonsWrapperComponent }
    ],
    manipulators: [
        { class: TemplateAddons, method: 'run' }
    ]
};
