// @er/primeng/esm5/lib/components/overlay-panel/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { OverlayPanel } from 'primeng/primeng';
var PngOverlayPanelComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngOverlayPanelComponent, _super);
    function PngOverlayPanelComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    PngOverlayPanelComponent.prototype.show = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.panel.show(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngOverlayPanelComponent.prototype.toggle = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.panel.toggle(event);
    };
    /**
     * @return {?}
     */
    PngOverlayPanelComponent.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.panel.hide();
    };
    PngOverlayPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-overlay-panel',
                    template: "<p-overlayPanel #overlayPanel [showCloseIcon]=\"true\" [dismissable]=\"false\">\n   <er-dyna-content [content]=\"content\" [context]=\"context\"></er-dyna-content>\n</p-overlayPanel>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PngOverlayPanelComponent.propDecorators = {
        content: [{ type: Input }],
        context: [{ type: Input }],
        panel: [{ type: ViewChild, args: [OverlayPanel,] }]
    };
    return PngOverlayPanelComponent;
}(BaseComponent));
export { PngOverlayPanelComponent };
if (false) {
    /** @type {?} */
    PngOverlayPanelComponent.prototype.content;
    /** @type {?} */
    PngOverlayPanelComponent.prototype.context;
    /** @type {?} */
    PngOverlayPanelComponent.prototype.panel;
}
