// @er/primeng/esm5/lib/components/rating/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_RATING_PROPS } from './type';
var PngRatingComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngRatingComponent, _super);
    function PngRatingComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.$defaultProps = DEFAULT_PNG_RATING_PROPS;
        return _this;
    }
    PngRatingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-rating',
                    template: "\n    <p-rating\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      (onRate)=\"updateTouch()\"\n    >\n    </p-rating>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngRatingComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngRatingComponent;
}(DataAwareControlWrapperComponent));
export { PngRatingComponent };
if (false) {
    /** @type {?} */
    PngRatingComponent.prototype.$defaultProps;
}
