// @er/primeng/esm5/lib/components/rating/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngRatingProps() { }
if (false) {
    /** @type {?|undefined} */
    PngRatingProps.prototype.readonly;
    /** @type {?|undefined} */
    PngRatingProps.prototype.stars;
    /** @type {?|undefined} */
    PngRatingProps.prototype.cancel;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconOnClass;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconOnStyle;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconOffClass;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconOffStyle;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconCancelClass;
    /** @type {?|undefined} */
    PngRatingProps.prototype.iconCancelStyle;
    /** @type {?|undefined} */
    PngRatingProps.prototype.onRate;
    /** @type {?|undefined} */
    PngRatingProps.prototype.onCancel;
}
/** @type {?} */
export var DEFAULT_PNG_RATING_PROPS = {
    stars: 5,
    iconOnClass: 'fa fa-star fa-2x',
    iconOffClass: 'fa fa-star-o fa-2x',
    iconCancelClass: 'fa fa-ban fa-2x red'
};
