// @er/formly-primeng/esm2015/lib/field-types/dyna/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { DynaContentComponent } from '@er/core';
import { distinctUntilChanged } from 'rxjs/operators';
import { PngFormlyFieldType } from '../../base/field-type';
/**
 * @record
 */
export function DynaProps() { }
if (false) {
    /** @type {?} */
    DynaProps.prototype.content;
    /** @type {?} */
    DynaProps.prototype.refreshOnFieldChanges;
    /** @type {?} */
    DynaProps.prototype.refreshOnFormChanges;
}
export class PngFormlyFieldDynaComponent extends PngFormlyFieldType {
    /**
     * @return {?}
     */
    ngOnInit() {
        super.ngOnInit();
        if (this.props['refreshOnForm']) {
            /** @type {?} */
            const form = this.field.formlyForm.form;
            this.subscription = form.valueChanges
                .pipe(distinctUntilChanged())
                .subscribe(_ => {
                this.comp.refresh();
            });
        }
        else if (this.props['refreshOnField']) {
            /** @type {?} */
            const formControl = this.field.formlyForm.formControl(this.props['refreshOnField']);
            this.subscription = formControl.valueChanges
                .pipe(distinctUntilChanged())
                .subscribe(_ => {
                this.comp.refresh();
            });
        }
        else {
            this.comp.refresh();
        }
    }
    /**
     * @return {?}
     */
    unsubscribe() {
        if (this.subscription) {
            this.subscription.unsubscribe();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.unsubscribe();
    }
}
PngFormlyFieldDynaComponent.decorators = [
    { type: Component, args: [{
                template: `
    <er-dyna-content [content]="props['content']" [context]="{field: field}"></er-dyna-content>
  `
            }] }
];
PngFormlyFieldDynaComponent.propDecorators = {
    comp: [{ type: ViewChild, args: [DynaContentComponent,] }]
};
if (false) {
    /** @type {?} */
    PngFormlyFieldDynaComponent.prototype.comp;
    /** @type {?} */
    PngFormlyFieldDynaComponent.prototype.subscription;
}
