// @er/utils/esm5/lib/tokenizer/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var TokenizerUtils = /** @class */ (function () {
    function TokenizerUtils() {
    }
    /**
     * @param {?} str
     * @param {?} format
     * @return {?}
     */
    TokenizerUtils.extract = /**
     * @param {?} str
     * @param {?} format
     * @return {?}
     */
    function (str, format) {
        if (str === format) {
            return new ExtractionResult(true);
        }
        /** @type {?} */
        var formatTokens = new FormatStringTokenizer().Tokenize(format);
        if (!formatTokens) {
            return new ExtractionResult(str === '');
        }
        /** @type {?} */
        var result = new ExtractionResult(true);
        for (var i = 0; i < formatTokens.length; i++) {
            /** @type {?} */
            var currentToken = formatTokens[i];
            /** @type {?} */
            var previousToken = i > 0 ? formatTokens[i - 1] : null;
            if (currentToken.Type === FormatStringTokenType.ConstantText) {
                if (i === 0) {
                    if (str.indexOf(currentToken.Text) !== 0) {
                        result.isMatch = false;
                        return result;
                    }
                    str = str.substr(currentToken.Text.length, str.length - currentToken.Text.length);
                }
                else {
                    /** @type {?} */
                    var matchIndex = str.indexOf(currentToken.Text);
                    if (matchIndex < 0) {
                        result.isMatch = false;
                        return result;
                    }
                    result.matches.push({ name: previousToken.Text, value: str.substr(0, matchIndex) });
                    str = str.substring(0, matchIndex + currentToken.Text.length);
                }
            }
        }
        /** @type {?} */
        var lastToken = formatTokens[formatTokens.length - 1];
        if (lastToken.Type === FormatStringTokenType.DynamicValue) {
            result.matches.push({ name: lastToken.Text, value: str });
        }
        return result;
    };
    /**
     * @param {?} str
     * @param {?} format
     * @return {?}
     */
    TokenizerUtils.matches = /**
     * @param {?} str
     * @param {?} format
     * @return {?}
     */
    function (str, format) {
        /** @type {?} */
        var result = TokenizerUtils.extract(str, format);
        if (!result.isMatch) {
            return [];
        }
        /** @type {?} */
        var values = [];
        for (var i = 0; i < result.matches.length; i++) {
            values.push(result.matches[i].value);
        }
        return values;
    };
    return TokenizerUtils;
}());
export { TokenizerUtils };
var ExtractionResult = /** @class */ (function () {
    function ExtractionResult(isMatch) {
        this.isMatch = isMatch;
        this.matches = [];
    }
    return ExtractionResult;
}());
export { ExtractionResult };
if (false) {
    /** @type {?} */
    ExtractionResult.prototype.isMatch;
    /** @type {?} */
    ExtractionResult.prototype.matches;
}
/** @enum {number} */
var FormatStringTokenType = {
    ConstantText: 0,
    DynamicValue: 1,
};
export { FormatStringTokenType };
FormatStringTokenType[FormatStringTokenType.ConstantText] = 'ConstantText';
FormatStringTokenType[FormatStringTokenType.DynamicValue] = 'DynamicValue';
var FormatStringToken = /** @class */ (function () {
    function FormatStringToken(text, type) {
        this.Text = text;
        this.Type = type;
    }
    return FormatStringToken;
}());
export { FormatStringToken };
if (false) {
    /** @type {?} */
    FormatStringToken.prototype.Text;
    /** @type {?} */
    FormatStringToken.prototype.Type;
}
var FormatStringTokenizer = /** @class */ (function () {
    function FormatStringTokenizer() {
    }
    /**
     * @param {?} format
     * @param {?=} includeBracketsForDynamicValues
     * @return {?}
     */
    FormatStringTokenizer.prototype.Tokenize = /**
     * @param {?} format
     * @param {?=} includeBracketsForDynamicValues
     * @return {?}
     */
    function (format, includeBracketsForDynamicValues) {
        if (includeBracketsForDynamicValues === void 0) { includeBracketsForDynamicValues = false; }
        /** @type {?} */
        var tokens = [];
        /** @type {?} */
        var currentText = '';
        /** @type {?} */
        var inDynamicValue = false;
        for (var i = 0; i < format.length; i++) {
            /** @type {?} */
            var c = format[i];
            switch (c) {
                case '{':
                    if (inDynamicValue) {
                        throw new Error('Incorrect syntax at char ' + i + '! format string can not contain agg dynamic value expression!');
                    }
                    inDynamicValue = true;
                    if (currentText.length > 0) {
                        tokens.push(new FormatStringToken(currentText, FormatStringTokenType.ConstantText));
                        currentText = '';
                    }
                    break;
                case '}':
                    if (!inDynamicValue) {
                        throw new Error(('Incorrect syntax at char ' + i + '! These is no opening brackets for the closing bracket }.'));
                    }
                    inDynamicValue = false;
                    if (currentText.length <= 0) {
                        throw new Error(('Incorrect syntax at char ' + i + '! Brackets does not containt any chars.'));
                    }
                    /** @type {?} */
                    var dynamicValue = currentText;
                    if (includeBracketsForDynamicValues) {
                        dynamicValue = '{' + dynamicValue + '}';
                    }
                    tokens.push(new FormatStringToken(dynamicValue, FormatStringTokenType.DynamicValue));
                    currentText = '';
                    break;
                default:
                    currentText += c;
                    break;
            }
        }
        if (inDynamicValue) {
            throw new Error(('There is no closing } char for an opened { char.'));
        }
        if (currentText.length > 0) {
            tokens.push(new FormatStringToken(currentText, FormatStringTokenType.ConstantText));
        }
        return tokens;
    };
    return FormatStringTokenizer;
}());
export { FormatStringTokenizer };
