// @er/presets/esm2015/lib/api-servers/mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const paramKeys = {
    sort: '_sort',
    order: '_order',
    size: '_limit',
    page: '_page',
    from: '_start',
    to: '_end'
};
/** @type {?} */
const getResponseResolver = response => {
    if (response) {
        if (response instanceof Array) {
            if (response.length === 1) {
                response = response[0];
            }
        }
        return { success: true, result: response };
    }
    else {
        return { success: false };
    }
};
const ɵ0 = getResponseResolver;
/** @type {?} */
const updateResponseResolver = response => response && response.hasOwnProperty('id')
    ? { success: true, result: response.id }
    : { failure: true, message: response };
const ɵ1 = updateResponseResolver;
/** @type {?} */
const deleteResponseResolver = response => response && Object.keys(response).length === 0
    ? { success: true, result: {} }
    : { failure: true, message: response };
const ɵ2 = deleteResponseResolver;
/** @type {?} */
export const MockServerApiRequestOptions = {};
export { ɵ0, ɵ1, ɵ2 };
