// @er/utils/esm2015/lib/es/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AggInterval, AggIntervalProps, AggMethods, AggMethodsProps, DEFAULT_ES_QUERY_CONDITION, DEFAULT_QUERY_CONDITION, ES_QUERY_OPERATORS, Order } from '@er/types';
import * as esb from 'elastic-builder';
import { RequestBodySearch } from 'elastic-builder';
import { CommonsUtils } from '../comm';
import { DialogUtils } from '../dialog';
export class EsUtils {
    /**
     * @param {?} obj
     * @return {?}
     */
    static parseEsbObject(obj) {
        if (!obj) {
            return obj;
        }
        if (CommonsUtils.isArray(obj)) {
            return obj.forEach(o => EsUtils.parseEsbObject(o));
        }
        else {
            if (CommonsUtils.isFunction(obj['toJSON'])) {
                return obj['toJSON']();
            }
            else {
                return obj;
            }
        }
    }
    /**
     * @param {?} key
     * @param {?} aggMethods
     * @return {?}
     */
    static getFieldAggs(key, aggMethods) {
        if (aggMethods) {
            /** @type {?} */
            let aggs = {};
            aggMethods.forEach(method => {
                switch (method) {
                    case AggMethodsProps.avg.key:
                        aggs = CommonsUtils.merge(aggs, esb.avgAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    case AggMethodsProps.distinct.key:
                        aggs = CommonsUtils.merge(aggs, esb.cardinalityAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    case AggMethodsProps.max.key:
                        aggs = CommonsUtils.merge(aggs, esb.maxAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    case AggMethodsProps.min.key:
                        aggs = CommonsUtils.merge(aggs, esb.minAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    case AggMethodsProps.sum.key:
                        aggs = CommonsUtils.merge(aggs, esb.sumAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    case AggMethodsProps.count.key:
                        aggs = CommonsUtils.merge(aggs, esb.valueCountAggregation(`${key}-${method}`, key).toJSON());
                        break;
                    default:
                        aggs = CommonsUtils.merge(aggs, esb.valueCountAggregation(`${key}-${method}`, key).toJSON());
                }
            });
            return aggs;
        }
        else {
            return undefined;
        }
    }
    /**
     * @param {?} query
     * @return {?}
     */
    static resolveQuery(query) {
        /** @type {?} */
        let must = [];
        /** @type {?} */
        let should = [];
        /** @type {?} */
        let not = [];
        if (query) {
            if (CommonsUtils.isArray(query)) {
                must = must.concat(query);
            }
            else if (CommonsUtils.isJson(query)) {
                if (query.hasOwnProperty('$and')) {
                    /** @type {?} */
                    const and = query['$and'];
                    if (CommonsUtils.isArray(and)) {
                        must = must.concat(and);
                    }
                    else {
                        must.push(and);
                    }
                }
                if (query.hasOwnProperty('$or')) {
                    /** @type {?} */
                    const or = query['$or'];
                    if (CommonsUtils.isArray(or)) {
                        should = should.concat(or);
                    }
                    else {
                        should.push(or);
                    }
                }
                if (query.hasOwnProperty('$not')) {
                    /** @type {?} */
                    const but = query['$not'];
                    if (CommonsUtils.isArray(but)) {
                        not = not.concat(but);
                    }
                    else {
                        not.push(but);
                    }
                }
            }
            else {
                must.push(query);
            }
        }
        return {
            must: CommonsUtils.compact(must),
            should: CommonsUtils.compact(should),
            not: CommonsUtils.compact(not)
        };
    }
    /**
     * @param {?} options
     * @return {?}
     */
    static resolveSearchFields(options) {
        /** @type {?} */
        const searchFields = CommonsUtils.getArrayValue(options.searchFields);
        if (CommonsUtils.isEmpty(searchFields)) {
            DialogUtils.error('数据错误', '没有定义查询字段');
            throw new Error('not found search fields');
        }
        /** @type {?} */
        const qs = [];
        searchFields.filter(field => !!field).forEach(field => {
            if (options.withPy) {
                if (options.withWildcard) {
                    qs.push(esb.wildcardQuery(`${field}.py`, `*${options.queryValue}*`));
                }
                else {
                    qs.push(esb.prefixQuery(`${field}.py`, options.queryValue));
                }
            }
            if (options.withZh) {
                qs.push(esb.wildcardQuery(`${field}.zh`, `*${options.queryValue}*`));
            }
            if (options.withTerm) {
                qs.push(esb.termQuery(field, options.queryValue));
            }
            if (options.withPrefix) {
                qs.push(esb.prefixQuery(field, options.queryValue));
            }
            if (options.withWildcard) {
                qs.push(esb.wildcardQuery(field, `*${options.queryValue}*`));
            }
            if (qs.length === 0) {
                qs.push(esb.matchQuery(field, options.queryValue));
            }
        });
        return qs;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    static buildQueryBody(options) {
        /** @type {?} */
        const query = esb.requestBodySearch();
        /** @type {?} */
        const boolQuery = esb.boolQuery();
        /** @type {?} */
        const queries = EsUtils.resolveQuery(options.query);
        if (options.queryValue) {
            /** @type {?} */
            const qs = EsUtils.resolveSearchFields(options);
            if (qs.length === 1) {
                queries.must.push(qs[0]);
            }
            else {
                queries.should = queries.should.concat(qs);
            }
        }
        if (queries.must.length > 0) {
            boolQuery.must(queries.must);
        }
        if (queries.should.length > 0) {
            boolQuery.should(queries.should).minimumShouldMatch(1);
        }
        if (queries.not.length > 0) {
            boolQuery.mustNot(queries.not);
        }
        query.query(boolQuery);
        if (!CommonsUtils.isEmpty(options.filter)) {
            boolQuery.filter(options.filter);
        }
        if (!(options.returnFields && options.returnFields.length === 0)) {
            /** @type {?} */
            const returnFields = options.returnFields || options.searchFields;
            query.source(returnFields);
        }
        if (options.sort) {
            /** @type {?} */
            const sorts = EsUtils.resolveSort(options.sort);
            /** @type {?} */
            const esSorts = [];
            Object.keys(sorts).forEach(key => {
                esSorts.push(esb.sort(key, sorts[key] || Order.ASC).unmappedType('string'));
            });
            query.sorts(esSorts);
        }
        if (options.aggProps) {
            /** @type {?} */
            const aggs = EsUtils.parseAggProps(options.aggProps);
            if (aggs.length > 0) {
                ((/** @type {?} */ (query))).aggs(aggs);
                ((/** @type {?} */ (query))).size(options.size || 0);
            }
        }
        if (options.from) {
            query.from(options.from);
        }
        if (options.size) {
            query.size(options.size);
        }
        return query;
    }
    /**
     * @param {?} aggProps
     * @return {?}
     */
    static parseAggProps(aggProps) {
        /** @type {?} */
        const props = CommonsUtils.getArrayValue(aggProps);
        /** @type {?} */
        const aggs = [];
        /** @type {?} */
        let agg;
        props.forEach(item => {
            /** @type {?} */
            let aggProp = item;
            if (CommonsUtils.isString(aggProp)) {
                aggProp = { field: item };
            }
            /** @type {?} */
            const aggName = aggProp.name || `${aggProp.field}-${aggProp.method || 'terms'}`;
            if (aggProp.aggs) {
                aggProp.method = AggMethods.CUSTOMER;
            }
            switch (aggProp.method) {
                case AggMethods.AVG:
                    agg = esb.avgAggregation(aggName, aggProp.field).missing('0');
                    break;
                case AggMethods.MIN:
                    agg = esb.minAggregation(aggName, aggProp.field);
                    break;
                case AggMethods.MAX:
                    agg = esb.maxAggregation(aggName, aggProp.field);
                    break;
                case AggMethods.SUM:
                    agg = esb.sumAggregation(aggName, aggProp.field).missing('0');
                    break;
                case AggMethods.DISTINCT:
                    agg = esb.cardinalityAggregation(aggName, aggProp.field);
                    break;
                case AggMethods.HISTOGRAM:
                    agg = esb.histogramAggregation(aggName, aggProp.field);
                    if (aggProp.format) {
                        ((/** @type {?} */ (agg))).interval(aggProp.interval);
                    }
                    ((/** @type {?} */ (agg))).minDocCount(aggProp.minCount || 1);
                    if (aggProp.format) {
                        ((/** @type {?} */ (agg))).format(aggProp.format);
                    }
                    if (aggProp.missing) {
                        ((/** @type {?} */ (agg))).missing(aggProp.missing);
                    }
                    break;
                case AggMethods.DATE_HISTOGRAM:
                    agg = esb.dateHistogramAggregation(aggName, aggProp.field);
                    if (aggProp.interval) {
                        ((/** @type {?} */ (agg))).interval(aggProp.interval);
                    }
                    /** @type {?} */
                    const interval = aggProp.interval || AggInterval.DAY;
                    /** @type {?} */
                    const intervalProp = AggIntervalProps[interval];
                    ((/** @type {?} */ (agg))).interval(interval);
                    ((/** @type {?} */ (agg))).minDocCount(aggProp.minCount || 1);
                    /** @type {?} */
                    const format = aggProp.format || (intervalProp && intervalProp.format);
                    if (format) {
                        ((/** @type {?} */ (agg))).format(format);
                    }
                    if (aggProp.missing) {
                        ((/** @type {?} */ (agg))).missing(aggProp.missing);
                    }
                    break;
                case AggMethods.CUSTOMER:
                    agg = CommonsUtils.getValue(aggProp.aggs, aggProp);
                    break;
                case AggMethods.COUNT:
                    agg = esb.valueCountAggregation(aggName, aggProp.field);
                    break;
                default:
                    agg = esb.termsAggregation(aggName, aggProp.field).size(aggProp.size || 10);
                    break;
            }
            aggs.push(agg);
        });
        return aggs;
    }
    /**
     * @param {?} body
     * @param {?=} options
     * @return {?}
     */
    static resolveRequestBody(body, options) {
        if (!body || typeof body === 'string') {
            return body || {};
        }
        if (body instanceof RequestBodySearch) {
            return CommonsUtils.merge({}, body.toJSON() || {}, options.payload || {}, options.body || options.params || {});
        }
        /** @type {?} */
        const _body = Object.assign({}, body);
        if (!CommonsUtils.isEmpty(_body.sort)) {
            /** @type {?} */
            const sorts = EsUtils.resolveSort(_body.sort);
            /** @type {?} */
            const esSorts = [];
            Object.keys(sorts).forEach(key => {
                esSorts.push({ [key]: { order: sorts[key] || Order.ASC, 'unmapped_type': 'string' } });
            });
            _body.sort = esSorts;
        }
        if (_body.size && _body.page) {
            _body.from = _body.page * _body.size;
        }
        if (_body.from && _body.to) {
            _body.size = _body.to - _body.from;
        }
        delete _body.to;
        delete _body.page;
        if (!CommonsUtils.isEmpty(_body.query) ||
            !CommonsUtils.isEmpty(_body.filter)) {
            /** @type {?} */
            let esQuery = {};
            if (!CommonsUtils.isEmpty(_body.query)) {
                if (CommonsUtils.isArray(_body.query)) {
                    _body.query['$and'] = _body.query;
                }
                else {
                    _body.query = EsUtils.parseEsbObject(_body.query);
                }
                /** @type {?} */
                const queryBody = _body.query;
                if (Object.keys(queryBody).length > 0) {
                    Object.keys(queryBody).forEach(key => {
                        if (key === '$and') {
                            /** @type {?} */
                            const must = EsUtils.queryResolver(queryBody, '$and');
                            if (must) {
                                esQuery['must'] = must;
                            }
                        }
                        else if (key === '$or') {
                            /** @type {?} */
                            const should = EsUtils.queryResolver(queryBody, '$or');
                            if (should) {
                                esQuery['should'] = should;
                            }
                        }
                        else if (key === '$not') {
                            /** @type {?} */
                            const must_not = EsUtils.queryResolver(queryBody, '$not');
                            if (must_not) {
                                esQuery['must_not'] = must_not;
                            }
                        }
                        else if (queryBody[key]) {
                            if (esQuery['must']) {
                                esQuery['must'].push({ [key]: queryBody[key] });
                            }
                            else {
                                esQuery['must'] = [{ [key]: queryBody[key] }];
                            }
                        }
                    });
                    esQuery = Object.assign({}, esQuery);
                }
                if (esQuery['should'] && !esQuery['minimum_should_match']) {
                    esQuery['minimum_should_match'] = 1;
                }
            }
            if (!CommonsUtils.isEmpty(_body.filter)) {
                /** @type {?} */
                const filterBody = Object.assign({}, _body.filter);
                esQuery['filter'] = [];
                if (Object.keys(filterBody).length > 0) {
                    Object.keys(filterBody).forEach(key => {
                        if (filterBody[key]) {
                            esQuery['filter'].push({ [key]: filterBody[key] });
                        }
                    });
                }
                if (esQuery['filter'].length === 0) {
                    delete esQuery['filter'];
                }
            }
            if (Object.keys(esQuery).length > 0) {
                _body.query = { bool: esQuery };
            }
        }
        if (!CommonsUtils.isEmpty(_body.aggs)) {
            /** @type {?} */
            const aggs = EsUtils.parseAggProps(_body.aggs);
            if (aggs.length > 0) {
                _body.aggs = esb.requestBodySearch().aggs(aggs).toJSON()['aggs'];
            }
        }
        Object.keys(_body).forEach(k => {
            if (CommonsUtils.isEmpty(_body[k])) {
                delete _body[k];
            }
        });
        return _body;
    }
    /**
     * @param {?} sort
     * @return {?}
     */
    static resolveSort(sort) {
        /** @type {?} */
        let sorts = {};
        if (CommonsUtils.isString(sort)) {
            sorts[(/** @type {?} */ (sort))] = Order.ASC;
        }
        else if (CommonsUtils.isArray(sort)) {
            ((/** @type {?} */ (sort))).forEach(s => {
                if (CommonsUtils.isString(s)) {
                    if (!sorts[s]) {
                        sorts[s] = Order.ASC;
                    }
                }
                else {
                    sorts = Object.assign({}, sorts, (/** @type {?} */ (s)));
                }
            });
        }
        else {
            sorts = (/** @type {?} */ (sort));
        }
        return sorts;
    }
    /**
     * @param {?} qs
     * @return {?}
     */
    static getMsearchBody(qs) {
        /** @type {?} */
        const body = qs.map(q => JSON.stringify(q));
        return body.join('\n') + '\n';
    }
    /**
     * @param {?} ruleSet
     * @return {?}
     */
    static buildQueryRuleSet(ruleSet) {
        ruleSet = (!ruleSet) ? {} : ruleSet;
        return (function parse(data) {
            if (!data['condition']) {
                data['condition'] = DEFAULT_ES_QUERY_CONDITION || DEFAULT_QUERY_CONDITION;
            }
            if (['AND', 'OR'].indexOf(data['condition'].toUpperCase()) === -1) {
                throw new Error('不识别的ElasticSearch bool query 关系(AND,OR) "{0}"'
                    .replace('{0}', data['condition']));
            }
            if (!data['rules']) {
                return {};
            }
            /** @type {?} */
            const parts = {};
            parts['add'] = function (k, v) {
                if (this.hasOwnProperty(k)) {
                    this[k].push(v);
                }
                else {
                    this[k] = [v];
                }
            };
            data['rules'].forEach(function (rule) {
                /** @type {?} */
                const get_value = (aRule) => {
                    if (aRule.data && aRule.data.transform) {
                        return aRule.data.transform.call(this, aRule.value);
                    }
                    else {
                        return aRule.value;
                    }
                };
                /** @type {?} */
                const make_query = (aRule) => {
                    /** @type {?} */
                    const operator = ES_QUERY_OPERATORS[aRule.operator];
                    if (operator === undefined) {
                        throw new Error('不识别的查询操作符 "{0}"'
                            .replace('{0}', aRule.operator));
                    }
                    /** @type {?} */
                    let part = {};
                    if (operator.inputs !== 0) {
                        /** @type {?} */
                        const keyVal = {};
                        /** @type {?} */
                        const valueMapper = operator.getValue;
                        if (valueMapper) {
                            keyVal[aRule.field] = valueMapper(get_value(aRule));
                        }
                        else {
                            keyVal[aRule.field] = get_value(aRule);
                        }
                        part[operator.operator] = keyVal;
                    }
                    if (aRule.operator === 'is_null') {
                        part = { exists: { field: aRule.field } };
                    }
                    if (data['condition'] === 'NOT') {
                        return { 'bool': { 'must_not': [part] } };
                    }
                    else {
                        return part;
                    }
                };
                /** @type {?} */
                let clause;
                if (data['condition'] === 'AND') {
                    clause = 'must';
                }
                if (data['condition'] === 'NOT') {
                    clause = 'must_not';
                }
                if (data['condition'] === 'OR') {
                    clause = 'should';
                }
                if (rule['rules'] && rule['rules'].length > 0) {
                    parts['add'](clause, parse(rule));
                }
                else {
                    parts['add'](clause, make_query(rule));
                }
            });
            delete parts['add'];
            return { 'bool': parts };
        }(ruleSet));
    }
    /**
     * @param {?} aggs
     * @return {?}
     */
    static resolveAggs(aggs) {
        /** @type {?} */
        const resolved = {};
        Object.keys(aggs).forEach(key => {
            /** @type {?} */
            const agg = aggs[key];
            resolved[key] = {};
            Object.keys(agg).forEach(k => {
                if (k === 'buckets') {
                    resolved[key] = EsUtils.resolveBuckets(agg[k]);
                }
                else if (k === 'value') {
                    resolved[key] = agg[k];
                }
                else {
                    // console.debug('esUtils', 'discard agg key:' + k);
                    // resolved[key][k] = agg[key];
                }
            });
        });
        return resolved;
    }
    /**
     * @param {?} buckets
     * @return {?}
     */
    static resolveBuckets(buckets) {
        /** @type {?} */
        const resolved = buckets.map(bucket => EsUtils.resolveBucket(bucket));
        return resolved;
    }
    /**
     * @param {?} bucket
     * @return {?}
     */
    static resolveBucket(bucket) {
        /** @type {?} */
        let resolved = {};
        Object.keys(bucket).forEach(key => {
            if (key === 'key') {
                resolved['label'] = bucket['key'];
            }
            else if (key === 'doc_count') {
                resolved['value'] = bucket['doc_count'];
            }
            else {
                if (CommonsUtils.isJson(bucket[key])) {
                    resolved = Object.assign({}, resolved, EsUtils.resolveAggs({ [key]: bucket[key] }));
                }
                else {
                    resolved[key] = bucket[key];
                }
            }
        });
        return resolved;
    }
    /**
     * @private
     * @param {?} queryBody
     * @param {?} key
     * @return {?}
     */
    static queryResolver(queryBody, key) {
        if (queryBody[key]) {
            /** @type {?} */
            let _qs = queryBody[key].filter(q => !CommonsUtils.isEmpty(q));
            if (_qs.length > 0) {
                _qs = _qs.map(q => EsUtils.parseEsbObject(q));
            }
            return (_qs.length > 0) ? _qs : undefined;
        }
    }
}
