// @er/primeng/esm5/lib/components/button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { WrapperComponent } from '@er/core';
import { DEFAULT_BUTTON_PROPS } from '@er/presets';
import { CommonsUtils } from '@er/utils';
var PngButtonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngButtonComponent, _super);
    function PngButtonComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PngButtonComponent.prototype, "isHidden", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.$props && this.$props.hidden) {
                return CommonsUtils.getValue(this.$props.hidden, this.outputContext) || false;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PngButtonComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var btnProps = CommonsUtils.copy(this.$props || {});
        if (btnProps.menu) {
            btnProps.menu.forEach(function (item) {
                if (item.command) {
                    /** @type {?} */
                    var command_1 = item.command;
                    item.command = function (event) { return _this.onMenuCommand(command_1, event); };
                }
            });
        }
        this.$props = tslib_1.__assign({}, DEFAULT_BUTTON_PROPS, btnProps);
    };
    /**
     * @param {?} command
     * @param {?} event
     * @return {?}
     */
    PngButtonComponent.prototype.onMenuCommand = /**
     * @param {?} command
     * @param {?} event
     * @return {?}
     */
    function (command, event) {
        command.call(this, tslib_1.__assign({}, event || {}, this.outputContext || {}));
    };
    /**
     * @param {?} state
     * @return {?}
     */
    PngButtonComponent.prototype.setState = /**
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.icon = state ? 'fa fa-cog fa-spin' : this.$props['icon'];
    };
    PngButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-button',
                    template: "<ng-container *ngIf=\"!isHidden\">\n  <p-splitButton *ngIf=\"$props?.menu\"\n                 erPropsBind\n                 [props]=\"$props\"\n                 [context]=\"outputContext\"\n  >\n  </p-splitButton>\n  <p-button *ngIf=\"!$props?.menu\" [title]=\"$props?.title || ''\"\n            erPropsBind\n            [props]=\"$props\"\n            [context]=\"outputContext\"\n  >\n  </p-button>\n</ng-container>\n"
                }] }
    ];
    return PngButtonComponent;
}(WrapperComponent));
export { PngButtonComponent };
if (false) {
    /** @type {?} */
    PngButtonComponent.prototype.icon;
}
