// @er/primeng/esm5/lib/components/wizard/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, EventEmitter, Input, Output, QueryList } from '@angular/core';
import { BaseComponent } from '@er/core';
import { PngStepComponent } from './step.component';
var PngWizardComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngWizardComponent, _super);
    function PngWizardComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.activeIndex = 0;
        _this.stepChange = new EventEmitter();
        _this.items = [];
        _this.templateNames = {
            header: 'header',
            footer: 'footer'
        };
        return _this;
    }
    /**
     * @return {?}
     */
    PngWizardComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        _super.prototype.ngAfterContentInit.call(this);
        this.steps.toArray().forEach(function (step, index) {
            if (!step.styleClass) {
                step.styleClass = _this.stepStyleClass;
            }
            if (!step.style) {
                step.style = _this.style;
            }
            if (index === _this.activeIndex) {
                step.active = true;
            }
            _this.items[index] = {
                label: step.label,
                command: function (event) {
                    _this.steps.toArray().forEach(function (s) { return s.active = false; });
                    step.active = true;
                    _this.activeIndex = index;
                    _this.stepChange.emit(index);
                }
            };
        });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    PngWizardComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!this.steps) {
            return;
        }
        var _loop_1 = function (prop) {
            if (prop === 'activeIndex') {
                /** @type {?} */
                var curIndex_1 = changes[prop].currentValue;
                this_1.steps.toArray().forEach(function (step, index) {
                    /** @type {?} */
                    var selected = index === curIndex_1;
                    step.active = selected;
                    if (selected) {
                        // emit currently selected label
                        // this.change.next(step.label);
                    }
                });
                this_1.stepChange.emit(curIndex_1);
            }
        };
        var this_1 = this;
        for (var prop in changes) {
            _loop_1(prop);
        }
    };
    /**
     * @return {?}
     */
    PngWizardComponent.prototype.previous = /**
     * @return {?}
     */
    function () {
        this.activeIndex--;
        this.stepChange.emit(this.activeIndex);
        this.ngOnChanges({
            activeIndex: {
                currentValue: this.activeIndex,
                previousValue: this.activeIndex + 1,
                firstChange: false,
                isFirstChange: function () { return false; }
            }
        });
    };
    /**
     * @return {?}
     */
    PngWizardComponent.prototype.next = /**
     * @return {?}
     */
    function () {
        this.activeIndex++;
        this.stepChange.emit(this.activeIndex);
        this.ngOnChanges({
            activeIndex: {
                currentValue: this.activeIndex,
                previousValue: this.activeIndex - 1,
                firstChange: false,
                isFirstChange: function () { return false; }
            }
        });
    };
    PngWizardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-wizard',
                    template: "\n    <ng-container *ngIf=\"template(templateNames.header)\">\n      <ng-container *ngTemplateOutlet=\"template(templateNames.header)\"></ng-container>\n    </ng-container>\n    <p-steps [model]=\"items\" [(activeIndex)]=\"activeIndex\"\n             [ngClass]=\"styleClass\" [ngStyle]=\"style\" [readonly]=\"false\"></p-steps>\n    <ng-content></ng-content>\n    <button pButton type=\"button\" *ngIf=\"activeIndex > 0\" class=\"mr-2\"\n            (click)=\"previous()\" icon=\"fa fa-hand-o-left\" label=\"Previous\"></button>\n    <button pButton type=\"button\" *ngIf=\"activeIndex < items.length - 1\"\n            (click)=\"next()\" icon=\"fa fa-hand-o-right\" iconPos=\"right\" label=\"Next\"></button>\n    <ng-container *ngIf=\"template(templateNames.footer)\">\n      <ng-container *ngTemplateOutlet=\"template(templateNames.footer)\"></ng-container>\n    </ng-container>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [":host::ng-deep .ui-steps ul[role=tablist]{display:flex!important;justify-content:space-around!important}"]
                }] }
    ];
    PngWizardComponent.propDecorators = {
        activeIndex: [{ type: Input }],
        styleClass: [{ type: Input }],
        style: [{ type: Input }],
        stepStyleClass: [{ type: Input }],
        stepStyle: [{ type: Input }],
        stepChange: [{ type: Output }],
        steps: [{ type: ContentChildren, args: [PngStepComponent,] }]
    };
    return PngWizardComponent;
}(BaseComponent));
export { PngWizardComponent };
if (false) {
    /** @type {?} */
    PngWizardComponent.prototype.activeIndex;
    /** @type {?} */
    PngWizardComponent.prototype.styleClass;
    /** @type {?} */
    PngWizardComponent.prototype.style;
    /** @type {?} */
    PngWizardComponent.prototype.stepStyleClass;
    /** @type {?} */
    PngWizardComponent.prototype.stepStyle;
    /** @type {?} */
    PngWizardComponent.prototype.stepChange;
    /** @type {?} */
    PngWizardComponent.prototype.items;
    /** @type {?} */
    PngWizardComponent.prototype.steps;
    /** @type {?} */
    PngWizardComponent.prototype.templateNames;
}
