// @er/primeng/esm5/lib/components/wizard/step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var PngStepComponent = /** @class */ (function () {
    function PngStepComponent() {
        this.active = false;
    }
    PngStepComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-step',
                    template: "\n    <div *ngIf=\"active\" [ngClass]=\"styleClass\" [ngStyle]=\"style\" class=\"ui-widget-content ui-corner-all step-container\">\n      <ng-content></ng-content>\n    </div>\n  ",
                    styles: ['.step-container {padding: 45px 25px 45px 25px; margin-bottom: 20px;}']
                }] }
    ];
    PngStepComponent.propDecorators = {
        styleClass: [{ type: Input }],
        style: [{ type: Input }],
        label: [{ type: Input }]
    };
    return PngStepComponent;
}());
export { PngStepComponent };
if (false) {
    /** @type {?} */
    PngStepComponent.prototype.styleClass;
    /** @type {?} */
    PngStepComponent.prototype.style;
    /** @type {?} */
    PngStepComponent.prototype.label;
    /** @type {?} */
    PngStepComponent.prototype.active;
}
