// @er/utils/esm2015/lib/url/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class UrlUtils {
    /**
     * @return {?}
     */
    static getQueryParameters() {
        return UrlUtils.getQueryParametersUsingParameters(document.location.search);
    }
    /**
     * @param {?} search
     * @return {?}
     */
    static getQueryParametersUsingParameters(search) {
        /** @type {?} */
        const r = search.replace(/(^\?)/, '').split('&').map(function (n) { return n = n.split('='), this[n[0]] = n[1], this; }.bind({}))[0];
        return r;
    }
    /**
     * @return {?}
     */
    static getInitialUrlParameters() {
        /** @type {?} */
        const questionMarkIndex = UrlUtils.initialUrl.indexOf('?');
        if (questionMarkIndex >= 0) {
            return UrlUtils.initialUrl.substr(questionMarkIndex, UrlUtils.initialUrl.length - questionMarkIndex);
        }
        return '';
    }
    /**
     * @return {?}
     */
    static getReturnUrl() {
        /** @type {?} */
        const queryStringObj = UrlUtils.getQueryParametersUsingParameters(UrlUtils.getInitialUrlParameters());
        if (queryStringObj.returnUrl) {
            return decodeURIComponent(queryStringObj.returnUrl);
        }
        return null;
    }
    /**
     * @return {?}
     */
    static getSingleSignIn() {
        /** @type {?} */
        const queryStringObj = UrlUtils.getQueryParametersUsingParameters(UrlUtils.getInitialUrlParameters());
        if (queryStringObj.ss) {
            return queryStringObj.ss;
        }
        return false;
    }
}
/**
 * The URL requested, before initial routing.
 */
UrlUtils.initialUrl = location.href;
if (false) {
    /**
     * The URL requested, before initial routing.
     * @type {?}
     */
    UrlUtils.initialUrl;
}
