// @er/core/esm5/lib/directives/text-tag/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { ColorUtils, CommonsUtils } from '@er/utils';
import { BaseDirective } from '../../bases/directive';
var TagDirective = /** @class */ (function (_super) {
    tslib_1.__extends(TagDirective, _super);
    function TagDirective(el, render) {
        var _this = _super.call(this) || this;
        _this.el = el;
        _this.render = render;
        return _this;
    }
    /**
     * @return {?}
     */
    TagDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (!CommonsUtils.isEmpty(this.color)) {
            this.render.setStyle(this.el.nativeElement, 'color', this.color);
        }
        if (!CommonsUtils.isEmpty(this.bg)) {
            this.render.setStyle(this.el.nativeElement, 'background-color', this.bg);
        }
        if (CommonsUtils.isEmpty(this.color) && CommonsUtils.isEmpty(this.bg)) {
            if (!this.text) {
                this.text = this.el.nativeElement.innerHTML;
            }
            this.render.setStyle(this.el.nativeElement, 'background', ColorUtils.getColor(this.text));
            this.render.setStyle(this.el.nativeElement, 'color', 'white');
        }
        this.render.addClass(this.el.nativeElement, 'badge-pill');
        this.render.addClass(this.el.nativeElement, 'badge');
    };
    TagDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erTag]'
                },] }
    ];
    /** @nocollapse */
    TagDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    TagDirective.propDecorators = {
        text: [{ type: Input }],
        bg: [{ type: Input }],
        color: [{ type: Input }]
    };
    return TagDirective;
}(BaseDirective));
export { TagDirective };
if (false) {
    /** @type {?} */
    TagDirective.prototype.text;
    /** @type {?} */
    TagDirective.prototype.bg;
    /** @type {?} */
    TagDirective.prototype.color;
    /**
     * @type {?}
     * @private
     */
    TagDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    TagDirective.prototype.render;
}
