// @er/primeng/esm2015/lib/components/multi-select/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_MULTI_SELECT_PROPS } from './type';
export class PngMultiSelectComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.$defaultProps = DEFAULT_PNG_MULTI_SELECT_PROPS;
    }
}
PngMultiSelectComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-multi-select',
                template: `
    <p-multiSelect
      [(ngModel)]="value"
      erPropsBind
      [props]="$props"
      [context]="ctx"
      [erApiData]="apiDataProps"
      [options]="data$|async|erDataItems:$props.$ext.dataItemProps"
      (onBlur)="updateTouch()"
    >
      <ng-template let-item pTemplate="item">
        <er-data-item [customerTpl]="template('_')"
                         [item]="item.data"
                         [props]="$props.$ext.dataItemProps">
        </er-data-item>
      </ng-template>
    </p-multiSelect>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngMultiSelectComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngMultiSelectComponent.prototype.$defaultProps;
}
