// @er/primeng/esm5/lib/components/select-button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_SELECT_BUTTON_PROPS } from './type';
var PngSelectButtonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngSelectButtonComponent, _super);
    function PngSelectButtonComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.$defaultProps = DEFAULT_SELECT_BUTTON_PROPS;
        _this.templateNames = {
            item: 'item'
        };
        return _this;
    }
    PngSelectButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-select-button',
                    template: "\n    <p-selectButton\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      [erApiData]=\"apiDataProps\"\n      [options]=\"data$|async|erDataItems:$props.$ext.dataItemProps\"\n      (onOptionClick)=\"updateTouch()\"\n    >\n      <ng-template let-item>\n        <er-data-item [customerTpl]=\"template(templateNames.item)\"\n                      [item]=\"item.data\"\n                      [props]=\"$props.$ext.dataItemProps\">\n        </er-data-item>\n      </ng-template>  \n    </p-selectButton>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngSelectButtonComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngSelectButtonComponent;
}(DataAwareControlWrapperComponent));
export { PngSelectButtonComponent };
if (false) {
    /** @type {?} */
    PngSelectButtonComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSelectButtonComponent.prototype.templateNames;
}
