// @er/utils/esm2015/lib/context/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { StateNames } from '@er/types';
import { CommonsUtils } from '../comm';
import { DateUtils } from '../date';
import { StatesUtils } from '../states';
export class ContextUtils {
    /**
     * @param {?=} date
     * @return {?}
     */
    static context(date) {
        return Object.assign({ now: DateUtils.getFormattedDateTime(), today: DateUtils.getFormattedDate(), year: DateUtils.getYear(date), lastYear: DateUtils.getYear(date) - 1, month: DateUtils.getMonth(date), lastMonth: DateUtils.getLastMonth(), day: DateUtils.getDay(date), yesterday: DateUtils.getYesterday() }, (/** @type {?} */ (StatesUtils.getValue(StateNames.user))) || {});
    }
    /**
     * @param {?} key
     * @return {?}
     */
    static get(key) {
        return ContextUtils.context()[key] || '';
    }
    /**
     * @param {?} data
     * @return {?}
     */
    static resolveVariables(data) {
        /** @type {?} */
        let resolved;
        if (CommonsUtils.isJson(data)) {
            resolved = ContextUtils.resolveJson(data);
        }
        else if (CommonsUtils.isArray(data)) {
            resolved = ContextUtils.resolveArray(data);
        }
        else if (CommonsUtils.isString(data)) {
            resolved = ContextUtils.resolveValue(data);
        }
        else {
            resolved = data;
        }
        return resolved;
    }
    /**
     * @private
     * @param {?} template
     * @return {?}
     */
    static resolveValue(template) {
        /** @type {?} */
        let resolved = template;
        if (CommonsUtils.isTemplateStr(template)) {
            resolved = CommonsUtils.templateStr(template, ContextUtils.context());
        }
        return resolved;
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    static resolveJson(data) {
        /** @type {?} */
        const resolved = {};
        for (const prop in data) {
            if (data[prop] !== undefined) {
                if (CommonsUtils.isJson(data[prop])) {
                    resolved[prop] = ContextUtils.resolveJson(data[prop]);
                }
                else if (CommonsUtils.isArray(data[prop])) {
                    resolved[prop] = ContextUtils.resolveArray(data[prop]);
                }
                else if (CommonsUtils.isString(data[prop])) {
                    resolved[prop] = ContextUtils.resolveValue(data[prop]);
                }
                else {
                    resolved[prop] = data[prop];
                }
            }
        }
        return resolved;
    }
    /**
     * @private
     * @param {?} arr
     * @return {?}
     */
    static resolveArray(arr) {
        /** @type {?} */
        const resolved = [];
        arr.forEach(a => resolved.push(ContextUtils.resolveVariables(a)));
        return resolved;
    }
}
