// @er/primeng/esm5/lib/components/slider/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils } from '@er/utils';
import { DEFAULT_PNG_SLIDER_PROPS } from './type';
var PngSliderComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngSliderComponent, _super);
    function PngSliderComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.$defaultProps = DEFAULT_PNG_SLIDER_PROPS;
        return _this;
    }
    Object.defineProperty(PngSliderComponent.prototype, "current", {
        get: /**
         * @return {?}
         */
        function () {
            return CommonsUtils.join(this.value, '——');
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    PngSliderComponent.prototype.initValue = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event.hasOwnProperty('min') && event.hasOwnProperty('max')) {
            this.min = event.min;
            this.max = event.max;
            this.value = [event.min, event.max];
        }
        else {
            this.value = event;
        }
    };
    PngSliderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-slider',
                    template: "\n    <small class=\"d-inline w-25\">\u503C\uFF1A{{current}}</small>\n    <p-slider class=\"d-inline\"\n              [erApiData]=\"apiDataProps\"\n              [(ngModel)]=\"value\"\n              erPropsBind\n              [props]=\"$props\"\n              [min]=\"min\"\n              [max]=\"max\"\n              (onChange)=\"updateTouch()\"\n              (onDataReady)=\"initValue($event)\"\n    >\n    </p-slider>\n\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngSliderComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngSliderComponent;
}(DataAwareControlWrapperComponent));
export { PngSliderComponent };
if (false) {
    /** @type {?} */
    PngSliderComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSliderComponent.prototype.min;
    /** @type {?} */
    PngSliderComponent.prototype.max;
}
