// @er/core/esm2015/lib/interceptors/error/interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpErrorResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { NotifyLevel, StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { throwError, TimeoutError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { RouteUtils } from '../../utils/nav';
export class HttpErrorHandlerInterceptor {
    /**
     * @param {?} req
     * @param {?} next
     * @return {?}
     */
    intercept(req, next) {
        return next.handle(req)
            .pipe(catchError((err) => {
            if (err instanceof HttpErrorResponse) {
                if (err.status === 401) {
                    RouteUtils.nav(ConfigUtils.getConfig().auth.loginRouter);
                }
                else if (err.status === 403) {
                    StatesUtils.create(StateNames.notify, {
                        message: '没有操作权限！',
                        level: NotifyLevel.ERROR
                    });
                }
                else {
                    this.errorHandler(err);
                }
            }
            else if (err instanceof TimeoutError) {
                StatesUtils.create(StateNames.notify, {
                    message: '网络访问超时！',
                    level: NotifyLevel.ERROR
                });
            }
            return throwError(err);
        }));
    }
    /**
     * @private
     * @param {?} err
     * @return {?}
     */
    errorHandler(err) {
        /** @type {?} */
        let message = '';
        if (err.error instanceof Error) {
            message = err.error.message;
        }
        else {
            if (err.status === 0) {
                message = '网络链接已断开';
            }
            else if (err.status === 404) {
                message = '请求地址不存在';
            }
            else if (err.status === 500) {
                message = '服务器程序错误';
            }
            else {
                message = '数据请求错误';
            }
        }
        StatesUtils.create(StateNames.notify, {
            level: NotifyLevel.ERROR,
            message: `${message},错误代码:【${err.status}】`,
            title: '网络访问失败'
        });
    }
}
HttpErrorHandlerInterceptor.decorators = [
    { type: Injectable }
];
