// @er/core/esm5/lib/directives/position/directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
var PositionDirective = /** @class */ (function () {
    function PositionDirective(el, render) {
        this.el = el;
        this.render = render;
    }
    /**
     * @return {?}
     */
    PositionDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.render.addClass(this.el.nativeElement.parentNode, 'd-flex');
        this.render.addClass(this.el.nativeElement.parentNode, 'align-items-center');
        this.render.addClass(this.el.nativeElement.parentNode, 'justify-content-center');
        this.render.addClass(this.el.nativeElement, this.positionClass || 'text-center');
    };
    PositionDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[erPosition]'
                },] }
    ];
    /** @nocollapse */
    PositionDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    PositionDirective.propDecorators = {
        positionClass: [{ type: Input, args: ['erPosition',] }]
    };
    return PositionDirective;
}());
export { PositionDirective };
if (false) {
    /** @type {?} */
    PositionDirective.prototype.positionClass;
    /**
     * @type {?}
     * @private
     */
    PositionDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    PositionDirective.prototype.render;
}
