// @er/primeng/esm2015/lib/components/input/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ControlWrapperComponent } from '@er/core';
export class PngInputComponent extends ControlWrapperComponent {
}
PngInputComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-input',
                template: `
    <div class="ui-inputgroup">
      <ng-container *ngIf="attrs?.$ext?.preIcon || attrs?.$ext?.preText">
        <span class="ui-inputgroup-addon"><i [ngClass]="attrs?.$ext?.preIcon"></i>{{attrs?.$ext?.preText}}</span>
      </ng-container>
      <png-buttons *ngIf="attrs?.$ext?.preButtons" [btns]="attrs.$ext.preButtons" [context]="ctx"></png-buttons>
      <ng-container *ngIf="attrs?.$ext?.floatLabel;else inputTpl">
       <span class="ui-float-label">
         <ng-container *ngTemplateOutlet="inputTpl"></ng-container>
         <label style="font-weight: bold;">{{attrs?.$ext.label}}</label>
       </span>
      </ng-container>
      <ng-template #inputTpl>
        <span class="w-100 inner-addon">
          <input pInputText
                 [(ngModel)]="value"
                 type="{{attrs?.type || 'text'}}"
                 erPropsBind
                 [attrs]="attrs"
                 [context]="ctx"
                 [pKeyFilter]="attrs?.$ext?.inputRule || 'alphanum'"
                 (blur)="updateTouch()"/>
          <i *ngIf="value && !disabled && resettable" title="清除输入内容" class="fa fa-trash gray right" (click)="reset()"></i>
       </span>
      </ng-template>
      <ng-container *ngIf="attrs?.$ext?.postIcon || attrs?.$ext?.postText">
        <span class="ui-inputgroup-addon"><i [ngClass]="attrs?.$ext?.postIcon"></i>{{attrs?.$ext?.postText}}</span>
      </ng-container>
      <png-buttons *ngIf="attrs?.$ext?.postButtons" [btns]="attrs.$ext.postButtons" [context]="ctx"></png-buttons>
    </div>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngInputComponent),
                        multi: true
                    }],
                styles: [`
    .ui-float-label {
      width: 100%
    }

    .ui-float-label > label {
      font-size: 1em;
      color: inherit
    }
  `]
            }] }
];
PngInputComponent.propDecorators = {
    attrs: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    PngInputComponent.prototype.attrs;
}
