// @er/core/esm5/lib/components/button/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output } from '@angular/core';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { Subject } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { BaseComponent } from '../../bases/component';
var StateButtonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(StateButtonComponent, _super);
    function StateButtonComponent(cd) {
        var _this = _super.call(this) || this;
        _this.cd = cd;
        _this.type = 'button';
        _this.hint = '';
        _this.onClick = new EventEmitter();
        _this.onFocus = new EventEmitter();
        _this.onBlur = new EventEmitter();
        _this.processIcon = 'fa fa-cog fa-spin';
        _this.processing$ = new Subject();
        return _this;
    }
    Object.defineProperty(StateButtonComponent.prototype, "processing", {
        set: /**
         * @param {?} processing
         * @return {?}
         */
        function (processing) {
            this.processing$.next(processing);
            this.cd.markForCheck(); // used for not dom event modify
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StateButtonComponent.prototype, "btnLabel", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.processing) {
                return this.stateLabel || this.label;
            }
            else {
                return this.label || '';
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StateButtonComponent.prototype, "btnTitle", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.processing) {
                return this.stateHint || '执行中...';
            }
            else {
                return this.hint || '';
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    StateButtonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.stateKey) {
            this.$subscriptions = StatesUtils.observe(StateNames.http, this.stateKey)
                .pipe(filter(function (state) { return !!state; }), map(function (state) { return state['start']; })).subscribe(function (b) { return _this.processing$.next(b); });
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    StateButtonComponent.prototype.click = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onClick.emit(tslib_1.__assign({}, event, this.context || {}));
    };
    StateButtonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-button',
                    template: "<button *ngIf=\"!hidden\"\n        [attr.type]=\"type\"\n        [ngClass]=\"styleClass\"\n        [ngStyle]=\"style\"\n        [disabled]=\"disabled || (processing$|async) \"\n        [title]=\"hint\"\n        (click)=\"click($event)\"\n        (focus)=\"emitEvent('onFocus',$event)\"\n        (blur)=\"emitEvent('onBlur',$event)\"\n>\n  <i *ngIf=\"!iconRight && !(processing$|async)\" [ngClass]=\"icon\"></i> {{btnLabel}}\n  <ng-content></ng-content>\n  <i *ngIf=\"iconRight && !(processing$|async)\" [ngClass]=\"icon\"  ></i>\n  <i *ngIf=\"processing$|async\" [ngClass]=\"processIcon\" style=\"margin-left: 5px\"></i>\n</button>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    StateButtonComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    StateButtonComponent.propDecorators = {
        type: [{ type: Input }],
        hint: [{ type: Input }],
        label: [{ type: Input }],
        icon: [{ type: Input }],
        style: [{ type: Input }],
        hidden: [{ type: Input }],
        styleClass: [{ type: Input }],
        disabled: [{ type: Input }],
        iconRight: [{ type: Input }],
        stateKey: [{ type: Input }],
        stateLabel: [{ type: Input }],
        stateHint: [{ type: Input }],
        context: [{ type: Input }],
        onClick: [{ type: Output }],
        onFocus: [{ type: Output }],
        onBlur: [{ type: Output }],
        processing: [{ type: Input }]
    };
    return StateButtonComponent;
}(BaseComponent));
export { StateButtonComponent };
if (false) {
    /** @type {?} */
    StateButtonComponent.prototype.type;
    /** @type {?} */
    StateButtonComponent.prototype.hint;
    /** @type {?} */
    StateButtonComponent.prototype.label;
    /** @type {?} */
    StateButtonComponent.prototype.icon;
    /** @type {?} */
    StateButtonComponent.prototype.style;
    /** @type {?} */
    StateButtonComponent.prototype.hidden;
    /** @type {?} */
    StateButtonComponent.prototype.styleClass;
    /** @type {?} */
    StateButtonComponent.prototype.disabled;
    /** @type {?} */
    StateButtonComponent.prototype.iconRight;
    /** @type {?} */
    StateButtonComponent.prototype.stateKey;
    /** @type {?} */
    StateButtonComponent.prototype.stateLabel;
    /** @type {?} */
    StateButtonComponent.prototype.stateHint;
    /** @type {?} */
    StateButtonComponent.prototype.context;
    /** @type {?} */
    StateButtonComponent.prototype.onClick;
    /** @type {?} */
    StateButtonComponent.prototype.onFocus;
    /** @type {?} */
    StateButtonComponent.prototype.onBlur;
    /** @type {?} */
    StateButtonComponent.prototype.processIcon;
    /** @type {?} */
    StateButtonComponent.prototype.processing$;
    /**
     * @type {?}
     * @protected
     */
    StateButtonComponent.prototype.cd;
}
