// @er/core/esm2015/lib/services/router/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { NavigationCancel, NavigationEnd, NavigationError, NavigationStart, Router } from '@angular/router';
import { StateNames } from '@er/types';
import { StatesUtils } from '@er/utils';
import { RouteUtils } from '../../utils/nav';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
export class RouterStateService {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
        router.events.subscribe((event) => {
            this.navigationInterceptor(event);
        });
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    navigationInterceptor(event) {
        /** @type {?} */
        const activeRoute = RouteUtils.getActiveRoute();
        if (event instanceof NavigationStart) {
            StatesUtils.create(StateNames.nav, {
                start: true,
                url: activeRoute.url,
                outlet: activeRoute.outlet,
                route: activeRoute
            });
        }
        if (event instanceof NavigationEnd) {
            StatesUtils.update(StateNames.nav, {
                start: false,
                status: 'success',
                url: activeRoute.url,
                outlet: activeRoute.outlet,
                route: activeRoute
            });
            StatesUtils.update(StateNames.currentRoute, { activeRoute });
            /** @type {?} */
            const title = activeRoute.data && activeRoute.data['title'];
            if (title) {
                StatesUtils.update(StateNames.appTitle, { title });
            }
        }
        if (event instanceof NavigationCancel) {
            StatesUtils.update(StateNames.nav, {
                start: false,
                status: 'cancel',
                url: activeRoute.url,
                outlet: activeRoute.outlet,
                route: activeRoute
            });
        }
        if (event instanceof NavigationError) {
            StatesUtils.update(StateNames.nav, {
                start: false,
                status: 'error',
                url: activeRoute.url,
                outlet: activeRoute.outlet,
                route: activeRoute
            });
        }
    }
}
RouterStateService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
RouterStateService.ctorParameters = () => [
    { type: Router }
];
/** @nocollapse */ RouterStateService.ngInjectableDef = i0.defineInjectable({ factory: function RouterStateService_Factory() { return new RouterStateService(i0.inject(i1.Router)); }, token: RouterStateService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    RouterStateService.prototype.router;
}
