// @er/types/esm5/lib/colors/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var UI_COLORS = {
    color1: '#e45450',
    color31: '#9f74e5',
    color32: '#ab83eb',
    color2: '#d1756b',
    color3: '#df7976',
    color4: '#fe8585',
    color29: '#8f63d7',
    color30: '#956cd7',
    color33: '#b997f0',
    color34: '#c4a3f9',
    color35: '#997cc8',
    color5: '#f4a9ae',
    color6: '#ffa897',
    color7: '#ff914e',
    color8: '#f1ad4b',
    color36: '#9279ba',
    color37: '#a08ac3',
    color9: '#facf68',
    color10: '#fdb86e',
    color11: '#fec18b',
    color12: '#9c95d0',
    color13: '#978ed7',
    color14: '#a7acf2',
    color15: '#b3abef',
    color16: '#ceabe5',
    color17: '#f06d69',
    color18: '#f47976',
    color19: '#f78b88',
    color20: '#fa9693',
    color21: '#f9a19e',
    color22: '#fcb0ae',
    color23: '#ffb9b7',
    color24: '#fc9d63',
    color25: '#fda671',
    color26: '#ffaf7e',
    color27: '#febc94',
    color28: '#fdc8a8',
    color38: '#ad9acc',
    color39: '#b8a4d8',
    color40: '#b9accf',
    color41: '#ada1c1',
    color42: '#9e94af',
    color43: '#8b809d',
    color44: '#7f768d',
    color45: '#716582'
};
/** @type {?} */
export var THEME_COLORS = {
    blue: '#007bff',
    indigo: '#6610f2',
    purple: '#6f42c1',
    pink: '#e83e8c',
    red: '#dc3545',
    orange: '#fd7e14',
    yellow: '#ffc107',
    green: '#28a745',
    teal: '#20c997',
    cyan: '#17a2b8',
    white: '#FFFFFF'
};
