// @er/primeng/esm2015/lib/work-space/profile/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { AuthUtils, RouteUtils } from '@er/core';
import { ConfigUtils } from '@er/utils';
import { PngAdminWorkSpaceComponent } from '../component';
export class WorkSpaceProfileComponent {
    /**
     * @param {?} ws
     */
    constructor(ws) {
        this.ws = ws;
        this.profileRouter = ConfigUtils.getConfig().auth.profileRouter;
        this.changePwdRouter = ConfigUtils.getConfig().auth.changePwdRouter;
        this.avatar = 'assets/layout/images/avatar.png';
    }
    /**
     * @return {?}
     */
    get user() {
        return AuthUtils.getCurrentUser();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        this.active = !this.active;
        event.preventDefault();
    }
    /**
     * @return {?}
     */
    logout() {
        AuthUtils.logout();
    }
    /**
     * @return {?}
     */
    change() {
        RouteUtils.nav(this.changePwdRouter);
    }
}
WorkSpaceProfileComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-ws-profile',
                template: "<div class=\"profile\" [ngClass]=\"{'profile-expanded':active}\">\n  <a href=\"#\" (click)=\"onClick($event)\">\n    <img class=\"profile-image\" [src]=\"user?.avatar || avatar\"/>\n    <span class=\"profile-name\">{{user?.fullName}}</span>\n    <i class=\"fa fa-fw fa-caret-down\"></i>\n    <span class=\"profile-role\">{{user?.mainRole}}</span>\n  </a>\n</div>\n\n<ul id=\"profile-menu\" class=\"layout-menu\" [@menu]=\"active ? 'visible' : 'hidden'\">\n  <li role=\"menuitem\">\n    <a href=\"#\" [routerLink]=\"profileRouter\" [attr.tabindex]=\"!active ? '-1' : null\">\n      <i class=\"fa fa-fw fa-user\"></i>\n      <span>\u4E2A\u4EBA\u4FE1\u606F</span>\n    </a>\n    <div class=\"layout-menu-tooltip\">\n      <div class=\"layout-menu-tooltip-arrow\"></div>\n      <div class=\"layout-menu-tooltip-text\"><a [routerLink]=\"profileRouter\">\u4E2A\u4EBA\u4FE1\u606F</a></div>\n    </div>\n  </li>\n  <li role=\"menuitem\">\n    <a href=\"#\" [routerLink]=\"changePwdRouter\" [attr.tabindex]=\"!active ? '-1' : null\">\n      <i class=\"fa fa-fw fa-user-secret\"></i>\n      <span>\u4FEE\u6539\u5BC6\u7801</span>\n    </a>\n    <div class=\"layout-menu-tooltip\">\n      <div class=\"layout-menu-tooltip-arrow\"></div>\n      <div class=\"layout-menu-tooltip-text\"><a [routerLink]=\"changePwdRouter\">\u4FEE\u6539\u5BC6\u7801</a></div>\n    </div>\n  </li>\n  <!--<li role=\"menuitem\">-->\n  <!--<a href=\"#\" [attr.tabindex]=\"!active ? '-1' : null\">-->\n  <!--<i class=\"fa fa-fw fa-cog\"></i>-->\n  <!--<span>\u4E2A\u6027\u8BBE\u7F6E</span>-->\n  <!--</a>-->\n  <!--<div class=\"layout-menu-tooltip\">-->\n  <!--<div class=\"layout-menu-tooltip-arrow\"></div>-->\n  <!--<div class=\"layout-menu-tooltip-text\">\u4E2A\u6027\u8BBE\u7F6E</div>-->\n  <!--</div>-->\n  <!--</li>-->\n  <li role=\"menuitem\">\n    <a href=\"#\" [attr.tabindex]=\"!active ? '-1' : null\">\n      <i class=\"fa fa-fw fa-sign-out\"></i>\n      <span (click)=\"logout()\">\u9000\u51FA\u767B\u5F55</span>\n    </a>\n    <div class=\"layout-menu-tooltip\">\n      <div class=\"layout-menu-tooltip-arrow\"></div>\n      <div class=\"layout-menu-tooltip-text\" (click)=\"logout()\">\u9000\u51FA\u767B\u5F55</div>\n    </div>\n  </li>\n</ul>\n",
                animations: [
                    trigger('menu', [
                        state('hidden', style({
                            height: '0px'
                        })),
                        state('visible', style({
                            height: '*'
                        })),
                        transition('visible => hidden', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)')),
                        transition('hidden => visible', animate('400ms cubic-bezier(0.86, 0, 0.07, 1)'))
                    ])
                ],
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
WorkSpaceProfileComponent.ctorParameters = () => [
    { type: PngAdminWorkSpaceComponent }
];
if (false) {
    /** @type {?} */
    WorkSpaceProfileComponent.prototype.active;
    /** @type {?} */
    WorkSpaceProfileComponent.prototype.profileRouter;
    /** @type {?} */
    WorkSpaceProfileComponent.prototype.changePwdRouter;
    /** @type {?} */
    WorkSpaceProfileComponent.prototype.avatar;
    /** @type {?} */
    WorkSpaceProfileComponent.prototype.ws;
}
