// @er/core/esm5/lib/services/task/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, InjectionToken, Optional } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { StateNames } from '@er/types';
import { CommonsUtils, ConfigUtils, StatesUtils } from '@er/utils';
import { timer } from 'rxjs';
import { distinctUntilChanged, filter, map } from 'rxjs/operators';
import { DiUtils } from '../../utils/di';
import * as i0 from "@angular/core";
/** @type {?} */
export var BACKGROUND_TASK_TOKEN = new InjectionToken('BACKGROUND_TASK_TOKEN');
/**
 * used to run background task globally
 */
var TaskService = /** @class */ (function () {
    function TaskService(tasks) {
        if (tasks === void 0) { tasks = []; }
        var _this = this;
        this.subs = [];
        StatesUtils.create(StateNames.appTitle, ConfigUtils.getConfig().appTitle, true);
        setTimeout(function () {
            _this.cleanStates();
            /** @type {?} */
            var _tasks = CommonsUtils.flatten(tasks);
            if (_tasks && _tasks.length > 0) {
                _tasks.forEach(function (task) { return _this.startTask((/** @type {?} */ (task))); });
            }
        }, 2000);
    }
    /**
     * @return {?}
     */
    TaskService.prototype.updateTitle = /**
     * @return {?}
     */
    function () {
        this.subs.push(StatesUtils.observe(StateNames.appTitle)
            .pipe(filter(function (state) { return state && state.hasOwnProperty('title'); }), map(function (state) { return state.title; }), distinctUntilChanged())
            .subscribe(function (title) {
            DiUtils.get(Title).setTitle(title);
        }));
    };
    /**
     * @return {?}
     */
    TaskService.prototype.cleanStates = /**
     * @return {?}
     */
    function () {
        this.subs.push(timer(30000, 30000)
            .subscribe(function () {
            /** @type {?} */
            var store = StatesUtils.getStore().value;
            Object.keys(store).forEach(function (key) {
                if (store[key]) {
                    if (!store[key].preserved && (store[key].state && store[key].state.observers.length === 0)) {
                        store = CommonsUtils.omit(store, key);
                    }
                }
            });
            StatesUtils.getStore().next(store);
        }));
    };
    /**
     * @param {?} task
     * @return {?}
     */
    TaskService.prototype.startTask = /**
     * @param {?} task
     * @return {?}
     */
    function (task) {
        if (CommonsUtils.isFunction(task)) {
            ((/** @type {?} */ (task)))();
        }
        else {
            /** @type {?} */
            var schedule_1 = (/** @type {?} */ (task));
            /** @type {?} */
            var context_1 = schedule_1.context;
            if (CommonsUtils.isFunction(schedule_1.context)) {
                context_1 = schedule_1.context();
            }
            if (schedule_1.interval) {
                this.subs.push(timer(0, schedule_1.interval)
                    .subscribe(function () {
                    schedule_1.run(context_1);
                }));
            }
            else {
                schedule_1.run(context_1);
            }
        }
    };
    /**
     * @return {?}
     */
    TaskService.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subs.forEach(function (sub) {
            sub.unsubscribe();
        });
    };
    TaskService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    TaskService.ctorParameters = function () { return [
        { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [BACKGROUND_TASK_TOKEN,] }] }
    ]; };
    /** @nocollapse */ TaskService.ngInjectableDef = i0.defineInjectable({ factory: function TaskService_Factory() { return new TaskService(i0.inject(BACKGROUND_TASK_TOKEN, 8)); }, token: TaskService, providedIn: "root" });
    return TaskService;
}());
export { TaskService };
if (false) {
    /** @type {?} */
    TaskService.prototype.subs;
}
