// @er/primeng/esm5/lib/work-space/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output, Renderer2, ViewChild } from '@angular/core';
import { BaseComponent } from '@er/core';
import { StateNames } from '@er/types';
import { ConfigUtils, StatesUtils } from '@er/utils';
import { ScrollPanel } from 'primeng/primeng';
import { filter } from 'rxjs/operators';
/** @enum {number} */
var MenuOrientation = {
    STATIC: 0,
    OVERLAY: 1,
    SLIM: 2,
    HORIZONTAL: 3,
};
MenuOrientation[MenuOrientation.STATIC] = 'STATIC';
MenuOrientation[MenuOrientation.OVERLAY] = 'OVERLAY';
MenuOrientation[MenuOrientation.SLIM] = 'SLIM';
MenuOrientation[MenuOrientation.HORIZONTAL] = 'HORIZONTAL';
var PngAdminWorkSpaceComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngAdminWorkSpaceComponent, _super);
    function PngAdminWorkSpaceComponent(renderer, cd) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.cd = cd;
        _this.menuOrientation = MenuOrientation.STATIC;
        _this.darkMenu = true;
        _this.onInitMenu = new EventEmitter();
        _this.profileMode = 'inline';
        _this.rootMenuToggleStateName = '$rootMenuToggler';
        _this.templateNames = {
            topbar: 'topbar',
            header: 'header',
            main: 'main',
            footer: 'footer',
            profile: 'profile',
            settings: 'settings',
            messages: 'messages',
            notifications: 'notifications',
            search: 'search'
        };
        return _this;
    }
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.onInitMenu.emit();
        this.$subscriptions = StatesUtils.observe(StateNames.menu, this.rootMenuToggleStateName)
            .pipe(filter(function (state) { return state !== undefined; }))
            .subscribe(function () { return _this._toggleMainMenu(); });
        this.appTitle = this.appTitle || ConfigUtils.getConfig().appTitle;
        this.appName = this.appName || ConfigUtils.getConfig().appName;
        this.appLogo = this.appLogo || ConfigUtils.getConfig().appLogo;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout(function () {
            _this.layoutMenuScrollerViewChild.moveBar();
        }, 100);
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.onLayoutClick = /**
     * @return {?}
     */
    function () {
        if (!this.topbarItemClick) {
            this.activeTopbarItem = null;
            this.topbarMenuActive = false;
        }
        if (!this.menuClick) {
            if (this.isHorizontal() || this.isSlim()) {
                this.resetMenu = true;
            }
            if (this.overlayMenuActive || this.staticMenuMobileActive) {
                this.hideOverlayMenu();
            }
            this.menuHoverActive = false;
        }
        this.topbarItemClick = false;
        this.menuClick = false;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.onMenuClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _this = this;
        this.menuClick = true;
        this.resetMenu = false;
        if (!this.isHorizontal()) {
            setTimeout(function () {
                _this.layoutMenuScrollerViewChild.moveBar();
            }, 450);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.onTopbarMenuButtonClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.topbarItemClick = true;
        this.topbarMenuActive = !this.topbarMenuActive;
        this.hideOverlayMenu();
        event.preventDefault();
    };
    /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.onTopbarItemClick = /**
     * @param {?} event
     * @param {?} item
     * @return {?}
     */
    function (event, item) {
        this.topbarItemClick = true;
        if (this.activeTopbarItem === item) {
            this.activeTopbarItem = null;
        }
        else {
            this.activeTopbarItem = item;
        }
        event.preventDefault();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.onTopbarSubItemClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.hideOverlayMenu = /**
     * @return {?}
     */
    function () {
        this.rotateMenuButton = false;
        this.overlayMenuActive = false;
        this.staticMenuMobileActive = false;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isTablet = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var width = window.innerWidth;
        return width <= 1024 && width > 640;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isDesktop = /**
     * @return {?}
     */
    function () {
        return window.innerWidth > 1024;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isMobile = /**
     * @return {?}
     */
    function () {
        return window.innerWidth <= 640;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isOverlay = /**
     * @return {?}
     */
    function () {
        return this.menuOrientation === MenuOrientation.OVERLAY;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isHorizontal = /**
     * @return {?}
     */
    function () {
        return this.menuOrientation === MenuOrientation.HORIZONTAL;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.isSlim = /**
     * @return {?}
     */
    function () {
        return this.menuOrientation === MenuOrientation.SLIM;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.changeToStaticMenu = /**
     * @return {?}
     */
    function () {
        this.menuOrientation = MenuOrientation.STATIC;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.changeToOverlayMenu = /**
     * @return {?}
     */
    function () {
        this.menuOrientation = MenuOrientation.OVERLAY;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.changeToHorizontalMenu = /**
     * @return {?}
     */
    function () {
        this.menuOrientation = MenuOrientation.HORIZONTAL;
    };
    /**
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype.changeToSlimMenu = /**
     * @return {?}
     */
    function () {
        this.menuOrientation = MenuOrientation.SLIM;
    };
    /**
     * @private
     * @return {?}
     */
    PngAdminWorkSpaceComponent.prototype._toggleMainMenu = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var menu = StatesUtils.getValue(StateNames.rootMenu);
        this.menuOrientation = menu && menu.orientation;
        this.darkMenu = menu && menu.dark;
        this.menuClick = true;
        this.rotateMenuButton = !this.rotateMenuButton;
        this.topbarMenuActive = false;
        if (this.menuOrientation === MenuOrientation.OVERLAY) {
            this.overlayMenuActive = !this.overlayMenuActive;
        }
        else {
            if (this.isDesktop()) {
                this.staticMenuDesktopInactive = !this.staticMenuDesktopInactive;
            }
            else {
                this.staticMenuMobileActive = !this.staticMenuMobileActive;
            }
        }
        this.cd.detectChanges();
    };
    PngAdminWorkSpaceComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-admin-work-space',
                    template: "<div class=\"layout-wrapper\" (click)=\"onLayoutClick()\"\n     [ngClass]=\"{'menu-layout-static': !isOverlay(),\n                'menu-layout-overlay': isOverlay(),\n                'layout-menu-overlay-active': overlayMenuActive,\n                'menu-layout-horizontal': isHorizontal(),\n                'menu-layout-slim': isSlim(),\n                'layout-menu-static-inactive': staticMenuDesktopInactive,\n                'layout-menu-static-active': staticMenuMobileActive}\">\n\n  <png-ws-topbar></png-ws-topbar>\n\n  <div class=\"layout-menu-container\" [ngClass]=\"{'layout-menu-dark':darkMenu}\" (click)=\"onMenuClick($event)\">\n    <p-scrollPanel #layoutMenuScroller [style]=\"{height: '100%' }\">\n      <div class=\"menu-scroll-content\">\n        <png-ws-profile *ngIf=\"profileMode=='inline'&&!isHorizontal()\"></png-ws-profile>\n        <png-ws-main-menu [reset]=\"resetMenu\" [httpMenuTag]=\"httpMenuTag\"></png-ws-main-menu>\n      </div>\n    </p-scrollPanel>\n  </div>\n\n  <div class=\"layout-main\">\n    <ng-container *ngTemplateOutlet=\"template(templateNames.main)\"></ng-container>\n    <router-outlet></router-outlet>\n  </div>\n\n  <div class=\"layout-mask\"></div>\n\n  <png-ws-footer></png-ws-footer>\n\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngAdminWorkSpaceComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ChangeDetectorRef }
    ]; };
    PngAdminWorkSpaceComponent.propDecorators = {
        menuOrientation: [{ type: Input }],
        darkMenu: [{ type: Input }],
        appName: [{ type: Input }],
        appTitle: [{ type: Input }],
        appLogo: [{ type: Input }],
        httpMenuTag: [{ type: Input }],
        onInitMenu: [{ type: Output }],
        layoutMenuScrollerViewChild: [{ type: ViewChild, args: ['layoutMenuScroller',] }]
    };
    return PngAdminWorkSpaceComponent;
}(BaseComponent));
export { PngAdminWorkSpaceComponent };
if (false) {
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.menuOrientation;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.darkMenu;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.appName;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.appTitle;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.appLogo;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.httpMenuTag;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.onInitMenu;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.profileMode;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.rotateMenuButton;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.topbarMenuActive;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.overlayMenuActive;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.staticMenuDesktopInactive;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.staticMenuMobileActive;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.layoutMenuScroller;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.menuClick;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.topbarItemClick;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.activeTopbarItem;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.resetMenu;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.menuHoverActive;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.rootMenuToggleStateName;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.layoutMenuScrollerViewChild;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.templateNames;
    /** @type {?} */
    PngAdminWorkSpaceComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    PngAdminWorkSpaceComponent.prototype.cd;
}
