// @er/core/esm5/lib/services/strategy/service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// https://itnext.io/cache-components-with-angular-routereusestrategy-3e4c8b174d5f
var 
// https://itnext.io/cache-components-with-angular-routereusestrategy-3e4c8b174d5f
CachedRouteReuseStrategy = /** @class */ (function () {
    function CachedRouteReuseStrategy() {
        this.storedRouteHandles = new Map();
        this.allowRetrieveCache = {};
    }
    /**
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.shouldReuseRoute = /**
     * @param {?} future
     * @param {?} curr
     * @return {?}
     */
    function (future, curr) {
        if (this.isInOutlet(future['_routerState'].url) || this.isInOutlet(curr['_routerState'].url)) {
            return future.routeConfig === curr.routeConfig;
        }
        return future.routeConfig === curr.routeConfig && (false !== (future.data && future.data.$reuse));
    };
    /**
     * @param {?} route
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.retrieve = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        return (/** @type {?} */ (this.storedRouteHandles.get(this.getPath(route))));
    };
    /**
     * @param {?} route
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.shouldAttach = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var path = this.getPath(route);
        if (this.allowRetrieveCache[path]) {
            return this.storedRouteHandles.has(this.getPath(route));
        }
        return false;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.shouldDetach = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var path = this.getPath(route);
        if (this.allowRetrieveCache.hasOwnProperty(path)) {
            return true;
        }
        return false;
    };
    /**
     * @param {?} route
     * @param {?} detachedTree
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.store = /**
     * @param {?} route
     * @param {?} detachedTree
     * @return {?}
     */
    function (route, detachedTree) {
        this.storedRouteHandles.set(this.getPath(route), detachedTree);
    };
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.isInOutlet = /**
     * @private
     * @param {?} url
     * @return {?}
     */
    function (url) {
        return /.*\(.*:.*\).*/.test(url);
    };
    /**
     * @private
     * @param {?} route
     * @return {?}
     */
    CachedRouteReuseStrategy.prototype.getPath = /**
     * @private
     * @param {?} route
     * @return {?}
     */
    function (route) {
        if (route.routeConfig !== null && route.routeConfig.path !== null) {
            return route.routeConfig.path;
        }
        return '';
    };
    return CachedRouteReuseStrategy;
}());
// https://itnext.io/cache-components-with-angular-routereusestrategy-3e4c8b174d5f
export { CachedRouteReuseStrategy };
if (false) {
    /** @type {?} */
    CachedRouteReuseStrategy.prototype.storedRouteHandles;
    /** @type {?} */
    CachedRouteReuseStrategy.prototype.allowRetrieveCache;
}
