// @er/primeng/esm5/lib/components/table/component/summary/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
import { AggMethodsProps } from '@er/types';
import { CommonsUtils } from '@er/utils';
import { PngTableComponent } from '../table/component';
var PngTableSummaryComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableSummaryComponent, _super);
    function PngTableSummaryComponent(pngTable, cd) {
        var _this = _super.call(this) || this;
        _this.pngTable = pngTable;
        _this.cd = cd;
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    PngTableSummaryComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (changes.aggValues) {
            if (changes.aggValues.currentValue) {
                this.colAggValues = {};
                Object.keys(changes.aggValues.currentValue)
                    .forEach(function (key) {
                    /** @type {?} */
                    var name = key;
                    if (key.indexOf('-') > 0) {
                        name = key.substr(0, key.indexOf('-'));
                    }
                    /** @type {?} */
                    var type = key.substr(key.indexOf('-') + 1);
                    /** @type {?} */
                    var values = _this.colAggValues[name];
                    if (!values) {
                        values = [];
                        _this.colAggValues[name] = values;
                    }
                    /** @type {?} */
                    var agg = {};
                    agg['type'] = AggMethodsProps[type] && AggMethodsProps[type].label;
                    if (CommonsUtils.isArray(_this.aggValues[key])) {
                        agg['values'] = _this.aggValues[key];
                    }
                    else {
                        agg['value'] = _this.aggValues[key];
                    }
                    values.push(agg);
                });
            }
            else {
                this.colAggValues = undefined;
            }
            this.cd.detectChanges();
        }
    };
    /**
     * @param {?} column
     * @return {?}
     */
    PngTableSummaryComponent.prototype.cellClass = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        return CommonsUtils.getValue(column.aggCellClass, column);
    };
    /**
     * @param {?} column
     * @return {?}
     */
    PngTableSummaryComponent.prototype.cellStyle = /**
     * @param {?} column
     * @return {?}
     */
    function (column) {
        return CommonsUtils.getValue(column.aggCellStyle, column);
    };
    PngTableSummaryComponent.decorators = [
        { type: Component, args: [{
                    selector: '[png-table-summary]',
                    template: "\n    <ng-container *ngIf=\"customerTpl;else default\">\n      <ng-container *ngTemplateOutlet=\"customerTpl;context:{$implicit:aggValues,columns:columns,context:pngTable}\"></ng-container>\n    </ng-container>\n    <ng-template #default>\n      <th *ngIf=\"pngTable.$props.$ext.rowExpandable\" class=\"col-expandable col-summary\"></th>\n      <th *ngIf=\"pngTable.$props.$ext.hasRowCheckbox\" class=\"col-checkbox col-summary\"></th>\n      <th *ngIf=\"pngTable.$props.$ext.hasRowRadioButton\" class=\"col-radio col-summary\"></th>\n      <ng-container *ngIf=\"colAggValues\">\n        <th *ngFor=\"let col of columns\" [width]=\"col.width\" class=\"text-right col-summary col-summary-{{col.key}}\">\n          <ng-container *ngFor=\"let agg of colAggValues[col.key]\">\n            <ng-container *ngIf=\"agg.value\">\n              <div [ngClass]=\"cellClass(col)\" style=\"font-size: 0.8em\"\n                   [ngStyle]=\"cellStyle(col)\">{{agg.type}}: {{agg.value | floor: 2}}\n              </div>\n            </ng-container>\n            <ng-container *ngIf=\"agg.values\">\n              <div [ngClass]=\"cellClass(col)\" style=\"font-size: 0.8em\"\n                   [ngStyle]=\"cellStyle(col)\"> {{agg.type}}:\n              </div>\n              <ul>\n                <ng-container *ngFor=\"let data of agg.values\">\n                  <li class=\"border-bottom\"><span class=\"\">{{data.label}}:</span><span class=\"badge badge-pill btn-info\">{{data.value}}</span></li>\n                </ng-container>\n              </ul>\n            </ng-container>\n          </ng-container>\n        </th>\n      </ng-container>\n    </ng-template>\n\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    PngTableSummaryComponent.ctorParameters = function () { return [
        { type: PngTableComponent },
        { type: ChangeDetectorRef }
    ]; };
    PngTableSummaryComponent.propDecorators = {
        columns: [{ type: Input }],
        aggValues: [{ type: Input }]
    };
    return PngTableSummaryComponent;
}(BaseComponent));
export { PngTableSummaryComponent };
if (false) {
    /** @type {?} */
    PngTableSummaryComponent.prototype.columns;
    /** @type {?} */
    PngTableSummaryComponent.prototype.aggValues;
    /** @type {?} */
    PngTableSummaryComponent.prototype.colAggValues;
    /** @type {?} */
    PngTableSummaryComponent.prototype.pngTable;
    /**
     * @type {?}
     * @protected
     */
    PngTableSummaryComponent.prototype.cd;
}
