// @er/core/esm2015/lib/components/http-status/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BaseComponent } from '../../bases/component';
import { HTTP_STATE$, HTTP_TAG_STATE$ } from '../../states/all';
export class HttpStatusComponent extends BaseComponent {
    constructor() {
        super(...arguments);
        this.loadingClass = 'loading-circle';
    }
    /**
     * @return {?}
     */
    get httpState$() {
        /** @type {?} */
        let subject;
        if (!this.tag) {
            subject = HTTP_STATE$;
        }
        else {
            subject = HTTP_TAG_STATE$(this.tag);
        }
        return subject;
    }
}
HttpStatusComponent.decorators = [
    { type: Component, args: [{
                selector: 'er-http-status',
                template: `    
      <ng-container *ngIf="httpState$|async">
        <ng-container *ngTemplateOutlet="template('_',default)"></ng-container>
        <ng-template #default>
          <div [ngClass]="loadingClass"></div>
        </ng-template>
      </ng-container>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
HttpStatusComponent.propDecorators = {
    tag: [{ type: Input }],
    loadingClass: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HttpStatusComponent.prototype.tag;
    /** @type {?} */
    HttpStatusComponent.prototype.loadingClass;
}
