// @er/primeng/esm5/lib/components/buttons/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
var PngButtonsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngButtonsComponent, _super);
    function PngButtonsComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PngButtonsComponent.prototype, "hasOverlay", {
        get: /**
         * @return {?}
         */
        function () {
            return this.btns && this.btns.some(function (btn) { return !!btn.overlay; });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @param {?} panel
     * @return {?}
     */
    PngButtonsComponent.prototype.toggle = /**
     * @param {?} event
     * @param {?} panel
     * @return {?}
     */
    function (event, panel) {
        panel.toggle(event);
    };
    PngButtonsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-buttons',
                    template: "<div [ngClass]=\"styleClass\" [ngStyle]=\"style\">\n  <ng-container *ngIf=\"btns\">\n    <div class=\"d-flex\">\n      <ng-container *ngFor=\"let btn of btns\">\n        <ng-container *ngIf=\"!btn.overlay\">\n          <png-button [outputContext]=\"context || btn.context\" [props]=\"btn\" class=\"mr-2 w-100\"\n                      erPropsBind></png-button>\n        </ng-container>\n      </ng-container>\n      <ng-container *ngIf=\"hasOverlay\">\n        <button pButton title=\"\u66F4\u591A\u64CD\u4F5C...\" icon=\"fa fa-caret-down\"\n                (click)=\"toggle($event, buttonsOverlayPanel)\"></button>\n        <p-overlayPanel #buttonsOverlayPanel [dismissable]=\"true\" appendTo=\"body\">\n        <span *ngFor=\"let btn of btns\">\n          <ng-container *ngIf=\"btn.overlay\">\n           <png-button class=\"ml-2\" erPropsBind [props]=\"btn\" [outputContext]=\"context || btn.context\"></png-button>\n          </ng-container>\n        </span>\n        </p-overlayPanel>\n      </ng-container>\n    </div>\n  </ng-container>\n</div>\n"
                }] }
    ];
    PngButtonsComponent.propDecorators = {
        context: [{ type: Input }],
        btns: [{ type: Input }],
        style: [{ type: Input }],
        styleClass: [{ type: Input }]
    };
    return PngButtonsComponent;
}(BaseComponent));
export { PngButtonsComponent };
if (false) {
    /** @type {?} */
    PngButtonsComponent.prototype.context;
    /** @type {?} */
    PngButtonsComponent.prototype.btns;
    /** @type {?} */
    PngButtonsComponent.prototype.style;
    /** @type {?} */
    PngButtonsComponent.prototype.styleClass;
}
