// @er/core/esm5/lib/pipes/tracker/pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Pipe } from '@angular/core';
import { CommonsUtils } from '@er/utils';
import { BehaviorSubject, timer } from 'rxjs';
import { mapTo, scan, startWith, switchMap, tap } from 'rxjs/operators';
var TrackerPipe = /** @class */ (function () {
    function TrackerPipe(cd) {
        this.cd = cd;
        this.currentNumber = 0;
        this.counter$ = new BehaviorSubject(0);
        //   .subscribe();
    }
    /**
     * @param {?} number
     * @param {?=} offset
     * @return {?}
     */
    TrackerPipe.prototype.transform = /**
     * @param {?} number
     * @param {?=} offset
     * @return {?}
     */
    function (number, offset) {
        var _this = this;
        if (offset === void 0) { offset = 20; }
        this.currentNumber = number - CommonsUtils.getRandomNumber(offset);
        this.counter$.next(this.currentNumber);
        this.counter$
            .pipe(switchMap(function (endRange) {
            return timer(1000, 200)
                .pipe(mapTo(_this.positiveOrNegative(endRange, _this.currentNumber)), startWith(_this.currentNumber), scan(function (acc, curr) { return acc + curr; }), tap(function (n) { return console.log(n); })
            //  takeWhile(this.takeUntil(endRange, this.currentNumber))
            );
        })).subscribe(function () { return _this.cd.detectChanges(); });
        return this.counter$;
    };
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    TrackerPipe.prototype.positiveOrNegative = /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    function (endRange, currentNumber) {
        return endRange > currentNumber ? 1 : -1;
    };
    /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    TrackerPipe.prototype.takeUntil = /**
     * @private
     * @param {?} endRange
     * @param {?} currentNumber
     * @return {?}
     */
    function (endRange, currentNumber) {
        return endRange > currentNumber
            ? function (val) { return val <= endRange; }
            : function (val) { return val >= endRange; };
    };
    TrackerPipe.decorators = [
        { type: Pipe, args: [{
                    name: 'erTracker'
                },] }
    ];
    /** @nocollapse */
    TrackerPipe.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    return TrackerPipe;
}());
export { TrackerPipe };
if (false) {
    /** @type {?} */
    TrackerPipe.prototype.currentNumber;
    /** @type {?} */
    TrackerPipe.prototype.counter$;
    /**
     * @type {?}
     * @private
     */
    TrackerPipe.prototype.cd;
}
