// @er/primeng/esm5/lib/components/table/component/column-selector/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '@er/core';
var PngTableColumnSelectorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngTableColumnSelectorComponent, _super);
    function PngTableColumnSelectorComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.onColumnsChange = new EventEmitter();
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    PngTableColumnSelectorComponent.prototype.changeColumns = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.value.forEach(function (column) { return column.hidden = false; });
        this.onColumnsChange.emit(event);
    };
    PngTableColumnSelectorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-table-column-selector',
                    template: "\n\n    <p-multiSelect [style]=\"{height: '30px', minWidth: '150px'}\"\n                   [ngClass]=\"styleClass\"\n                   [ngStyle]=\"style\"\n                   [options]=\"optionColumns\"\n                   [(ngModel)]=\"columns\"\n                   optionLabel=\"label\"\n                   [selectedItemsLabel]=\"'\u5F53\u524D\u663E\u793A{0}('+optionColumns?.length+')\u4E2A\u5217'\"\n                   defaultLabel=\"\u9009\u62E9\u663E\u793A\u5B57\u6BB5\"\n                   (onChange)=\"changeColumns($event)\"\n    >\n    </p-multiSelect>\n\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    PngTableColumnSelectorComponent.propDecorators = {
        columns: [{ type: Input }],
        optionColumns: [{ type: Input }],
        styleClass: [{ type: Input }],
        style: [{ type: Input }],
        onColumnsChange: [{ type: Output }]
    };
    return PngTableColumnSelectorComponent;
}(BaseComponent));
export { PngTableColumnSelectorComponent };
if (false) {
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.columns;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.optionColumns;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.styleClass;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.style;
    /** @type {?} */
    PngTableColumnSelectorComponent.prototype.onColumnsChange;
}
