// @er/formly-primeng/esm5/lib/field-types/slider/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { PngFormlyFieldType } from '../../base/field-type';
var PngFormlySliderComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngFormlySliderComponent, _super);
    function PngFormlySliderComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.defaultProps = {
            range: true,
            min: 0,
            max: 100,
            orientation: 'horizontal'
        };
        return _this;
    }
    PngFormlySliderComponent.decorators = [
        { type: Component, args: [{
                    template: "\n    <png-slider\n      formlyPropsBind\n      [props]=\"props\"\n      [formControl]=\"formControl\"\n    ></png-slider>\n  "
                }] }
    ];
    return PngFormlySliderComponent;
}(PngFormlyFieldType));
export { PngFormlySliderComponent };
if (false) {
    /** @type {?} */
    PngFormlySliderComponent.prototype.defaultProps;
}
