// @er/primeng/esm5/lib/components/option-item/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { BaseComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
var PngOptionItemComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngOptionItemComponent, _super);
    function PngOptionItemComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PngOptionItemComponent.prototype, "labels", {
        get: /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var _label = DataItemUtils.getItemLabel(this.item, this.props.dropDownOptionKey || this.props.labelKey || 'label');
            if (CommonsUtils.isArray(_label)) {
                return _label;
            }
            else {
                return [_label];
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PngOptionItemComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.props) {
            this.props = {};
        }
    };
    PngOptionItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-option-item',
                    template: "\n    <ng-container *ngIf=\"customerTpl;else default\">\n      <ng-container *ngTemplateOutlet=\"customerTpl;context:{$implicit:item}\"></ng-container>\n    </ng-container>\n    <ng-template #default>\n      <div class=\"d-flex option-item\" [ngClass]=\"props.itemStyleClass || 'ui-helper-clearfix'\">\n          <span class=\"label mr-3\" *ngFor=\"let label of labels\" [ngClass]=\"props.labelStyleClass\" [innerHTML]=\"label\"></span>\n          <span class=\"pull-right count\" *ngIf=\"props.showCounter && item['count']\">\n          <span [ngClass]=\"props.counterStyleClass || 'badge badge-pill badge-info'\">{{item['count']}}</span>\n        </span>\n      </div>\n    </ng-template>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".option-item{border-bottom:1px solid #d5d5d5;font-size:.8em}.option-item:hover{font-weight:700;font-size:1em}"]
                }] }
    ];
    PngOptionItemComponent.propDecorators = {
        item: [{ type: Input }],
        props: [{ type: Input }]
    };
    return PngOptionItemComponent;
}(BaseComponent));
export { PngOptionItemComponent };
if (false) {
    /** @type {?} */
    PngOptionItemComponent.prototype.item;
    /** @type {?} */
    PngOptionItemComponent.prototype.props;
}
