// @er/primeng/esm2015/lib/components/check-box/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { CommonsUtils, DataItemUtils } from '@er/utils';
import { DEFAULT_CHECK_BOX_PROPS } from './type';
export class PngCheckBoxComponent extends DataAwareControlWrapperComponent {
    constructor() {
        super(...arguments);
        this.valueParser = this.parseValue;
        this.modelFormatter = this.formatModel;
        this.$defaultProps = DEFAULT_CHECK_BOX_PROPS;
    }
    /**
     * @return {?}
     */
    get chkValue() {
        return this._chkValue;
    }
    /**
     * @param {?} values
     * @return {?}
     */
    set chkValue(values) {
        this._chkValue = values;
        /** @type {?} */
        const data = this.data$.value;
        if (CommonsUtils.isArray(data)) {
            /** @type {?} */
            const objectValues = [];
            values.forEach(value => {
                /** @type {?} */
                const matched = ((/** @type {?} */ (data))).filter(v => v[this.dataKey] === value);
                if (matched.length === 1) {
                    objectValues.push(DataItemUtils.getItemValue(matched[0], this.valueKey));
                }
            });
            this.value = objectValues.length > 0 ? objectValues : null;
        }
        else {
            this.value = values;
        }
    }
    /**
     * @return {?}
     */
    get dataKey() {
        return CommonsUtils.get(this.$props, '$ext.dataItemProps.idKey', 'value');
    }
    /**
     * @return {?}
     */
    get valueKey() {
        return CommonsUtils.get(this.$props, '$ext.dataItemProps.valueKey', 'value');
    }
    /**
     * @param {?} model
     * @return {?}
     */
    formatModel(model) {
        if (this.$props && !this.$props.label && model) {
            /** @type {?} */
            const models = CommonsUtils.getArrayValue(model);
            this._chkValue = [];
            models.forEach(m => {
                if (CommonsUtils.isJson(m)) {
                    this._chkValue.push(DataItemUtils.getItemValue(m, this.dataKey));
                }
                else {
                    this._chkValue.push(m);
                }
            });
            return models;
        }
        else {
            return model;
        }
    }
}
PngCheckBoxComponent.decorators = [
    { type: Component, args: [{
                selector: 'png-check-box',
                template: `
    <ng-container *ngIf="$props.label;else multiTpl">
      <p-checkbox
        class="m-2"
        [(ngModel)]="value"
        erPropsBind
        [props]="$props"
        [context]="ctx"
        binary="true"
        (onChange)="updateTouch()"
      >
      </p-checkbox>
    </ng-container>
    <ng-template #multiTpl>
      <div class="d-flex pb-2"
           [ngClass]="$props.$ext['inline']===false?'flex-column':'flex-wrap'"
           [erApiData]="apiDataProps"
      >
        <ng-container *ngFor="let option of data$|async|erDataItems:$props.$ext.dataItemProps;let i=index">
          <span class="d-flex align-items-center">
             <p-checkbox
               class="m-2"
               [(ngModel)]="chkValue"
               [attr.index]="i"
               erPropsBind
               [props]="$props"
               [context]="ctx"
               [value]="option.data|erValue:dataKey"
               (onChange)="updateTouch()"
             >
             </p-checkbox>
             <span class="mr-4" [innerHTML]="option.label|erSafeHtml"></span>
          </span>
        </ng-container>
      </div>
    </ng-template>
  `,
                providers: [{
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => PngCheckBoxComponent),
                        multi: true
                    }]
            }] }
];
if (false) {
    /** @type {?} */
    PngCheckBoxComponent.prototype.valueParser;
    /** @type {?} */
    PngCheckBoxComponent.prototype.modelFormatter;
    /** @type {?} */
    PngCheckBoxComponent.prototype.$defaultProps;
    /**
     * @type {?}
     * @private
     */
    PngCheckBoxComponent.prototype._chkValue;
}
