// @er/primeng/esm5/lib/components/select/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, forwardRef, ViewChild } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { Dropdown } from 'primeng/primeng';
import { DEFAULT_PNG_SELECT_PROPS } from './type';
var PngSelectComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngSelectComponent, _super);
    function PngSelectComponent(cd) {
        var _this = _super.call(this) || this;
        _this.cd = cd;
        _this.$defaultProps = DEFAULT_PNG_SELECT_PROPS;
        _this.templateNames = {
            group: 'group',
            item: 'item'
        };
        return _this;
    }
    /**
     * @return {?}
     */
    PngSelectComponent.prototype.onPropsInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.onPropsInit.call(this);
        if (!this.$props.group && this.$props.$ext && this.$props.$ext.dataItemProps && this.$props.$ext.dataItemProps.groupKey) {
            this.$props.group = true;
        }
    };
    PngSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-select',
                    template: "\n    <p-dropdown\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      [erApiData]=\"apiDataProps\"\n      [options]=\"data$|async|erDataItems:$props?.$ext?.dataItemProps\"\n      (onBlur)=\"updateTouch()\"\n    >\n      <ng-template let-group pTemplate=\"group\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.group,groupTpl);context:{$implicit:group}\"></ng-container>\n      </ng-template>\n      <ng-template let-group #groupTpl>\n        <span *ngIf=\"group\" style=\"font-size: 1em\"><i class=\"fa fa-group text-primary\"></i> {{group.label}}</span>\n      </ng-template>\n      <ng-template let-item pTemplate=\"item\">\n        <er-data-item [customerTpl]=\"template(templateNames.item)\"\n                      [item]=\"item.data\"\n                      [props]=\"$props.$ext.dataItemProps\">\n        </er-data-item>\n      </ng-template>\n    </p-dropdown>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngSelectComponent; }),
                            multi: true
                        }],
                    styles: ["/deep/ .ui-dropdown-item{border-bottom:1px solid #a8a8a8!important;font-size:.8em;color:#000!important}/deep/ .ui-dropdown-item:hover{background-color:transparent!important;font-weight:700;color:red!important;font-size:1em}"]
                }] }
    ];
    /** @nocollapse */
    PngSelectComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    PngSelectComponent.propDecorators = {
        dropDown: [{ type: ViewChild, args: [Dropdown,] }]
    };
    return PngSelectComponent;
}(DataAwareControlWrapperComponent));
export { PngSelectComponent };
if (false) {
    /** @type {?} */
    PngSelectComponent.prototype.$defaultProps;
    /** @type {?} */
    PngSelectComponent.prototype.templateNames;
    /** @type {?} */
    PngSelectComponent.prototype.dropDown;
    /**
     * @type {?}
     * @private
     */
    PngSelectComponent.prototype.cd;
}
