// @er/primeng/esm5/lib/components/select/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngSelectProps() { }
if (false) {
    /** @type {?|undefined} */
    PngSelectProps.prototype.dataKey;
    /** @type {?|undefined} */
    PngSelectProps.prototype.autoWidth;
    /** @type {?|undefined} */
    PngSelectProps.prototype.editable;
    /** @type {?|undefined} */
    PngSelectProps.prototype.filter;
    /** @type {?|undefined} */
    PngSelectProps.prototype.filterBy;
    /** @type {?|undefined} */
    PngSelectProps.prototype.showClear;
    /** @type {?|undefined} */
    PngSelectProps.prototype.placeholder;
    /** @type {?|undefined} */
    PngSelectProps.prototype.group;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onClick;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onChange;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onFocus;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onBlur;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onShow;
    /** @type {?|undefined} */
    PngSelectProps.prototype.onHide;
}
/** @type {?} */
export var DEFAULT_PNG_SELECT_PROPS = {
    autoWidth: false,
    placeholder: '请选择',
    showClear: true,
    editable: false,
    $ext: {}
};
