// @er/types/esm2015/lib/http/index.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function HttpClientRequestOptions() { }
if (false) {
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.body;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.headers;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.params;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.observe;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.reportProgress;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.responseType;
    /** @type {?|undefined} */
    HttpClientRequestOptions.prototype.withCredentials;
}
/** @enum {string} */
const HttpMethod = {
    GET: 'Get',
    POST: 'Post',
    DELETE: 'Delete',
    OPTIONS: 'Options',
    PUT: 'Put',
    HEAD: 'Head',
    PATCH: 'Patch',
};
export { HttpMethod };
/** @enum {string} */
const HttpResponseType = {
    ARRAY: 'arraybuffer',
    BLOB: 'blob',
    JSON: 'json',
    TEXT: 'text',
};
export { HttpResponseType };
/** @enum {string} */
const HttpContentType = {
    FORM: 'application/x-www-form-urlencoded',
    MULTIPART: 'multipart/form-data',
    JSON: 'application/json',
    TEXT: 'text/plain',
};
export { HttpContentType };
/** @enum {string} */
const HttpRequestPayloadType = {
    PARAMS: 'params',
    BODY: 'body',
};
export { HttpRequestPayloadType };
/**
 * @record
 */
export function HttpRequestOptions() { }
if (false) {
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.tag;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.uri;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.serverType;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.method;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.headers;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.params;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.body;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.payload;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.payloadType;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.contentType;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.responseType;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.observe;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.reportProgress;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.withCredentials;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.retry;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.dataPath;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.notifySuccess;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.alertSuccess;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.successMessage;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.notifyFailure;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.alertFailure;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.failureMessage;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.stateName;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.context;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.skipRootHttpNotify;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.onRequest;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.onResponse;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.onError;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.onSuccess;
    /** @type {?|undefined} */
    HttpRequestOptions.prototype.onFailure;
    /* Skipping unhandled member: [extended: string]: any;*/
}
