// @er/core/esm5/lib/components/layout/left-right/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { BaseComponent } from '../../../bases/component';
var LeftRightContainerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LeftRightContainerComponent, _super);
    function LeftRightContainerComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.leftStyleClass = 'col-3';
        _this.rightStyleClass = 'col-9';
        _this.switchable = true;
        _this.onLeftHide = new EventEmitter();
        _this.templateNames = {
            left: 'left',
            right: 'right'
        };
        return _this;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    LeftRightContainerComponent.prototype.reLayout = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.hideLeft = event;
        this.onLeftHide.emit(event);
    };
    LeftRightContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'er-left-right',
                    template: "\n    <div class=\"d-flex\">\n      <div *ngIf=\"!hideLeft\" class=\"w-25 shadow-lg\">\n        <ng-container *ngTemplateOutlet=\"template(templateNames.left)\"></ng-container>\n      </div>\n      <div class=\"flex-grow-1\">\n        <div *ngIf=\"switchable\">\n          <i class=\"fa fa-2x  text-primary\" [title]=\"hideLeft?'\u5C55\u5F00\u5DE6\u4FA7\u5185\u5BB9':'\u6536\u8D77\u5DE6\u4FA7\u5185\u5BB9'\"\n             [ngClass]=\"hideLeft?'fa-arrow-circle-right':'fa-arrow-circle-left'\"\n             (click)=\"reLayout(!hideLeft)\"></i>\n        </div>\n        <ng-container *ngTemplateOutlet=\"template(templateNames.right)\"></ng-container>\n      </div>\n    </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    LeftRightContainerComponent.propDecorators = {
        leftStyleClass: [{ type: Input }],
        rightStyleClass: [{ type: Input }],
        hideLeft: [{ type: Input }],
        switchable: [{ type: Input }],
        onLeftHide: [{ type: Output }]
    };
    return LeftRightContainerComponent;
}(BaseComponent));
export { LeftRightContainerComponent };
if (false) {
    /** @type {?} */
    LeftRightContainerComponent.prototype.leftStyleClass;
    /** @type {?} */
    LeftRightContainerComponent.prototype.rightStyleClass;
    /** @type {?} */
    LeftRightContainerComponent.prototype.hideLeft;
    /** @type {?} */
    LeftRightContainerComponent.prototype.switchable;
    /** @type {?} */
    LeftRightContainerComponent.prototype.onLeftHide;
    /** @type {?} */
    LeftRightContainerComponent.prototype.templateNames;
}
