// @er/primeng/esm5/lib/components/multi-select/component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { DataAwareControlWrapperComponent } from '@er/core';
import { DEFAULT_PNG_MULTI_SELECT_PROPS } from './type';
var PngMultiSelectComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PngMultiSelectComponent, _super);
    function PngMultiSelectComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.$defaultProps = DEFAULT_PNG_MULTI_SELECT_PROPS;
        return _this;
    }
    PngMultiSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: 'png-multi-select',
                    template: "\n    <p-multiSelect\n      [(ngModel)]=\"value\"\n      erPropsBind\n      [props]=\"$props\"\n      [context]=\"ctx\"\n      [erApiData]=\"apiDataProps\"\n      [options]=\"data$|async|erDataItems:$props.$ext.dataItemProps\"\n      (onBlur)=\"updateTouch()\"\n    >\n      <ng-template let-item pTemplate=\"item\">\n        <er-data-item [customerTpl]=\"template('_')\"\n                         [item]=\"item.data\"\n                         [props]=\"$props.$ext.dataItemProps\">\n        </er-data-item>\n      </ng-template>\n    </p-multiSelect>\n  ",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return PngMultiSelectComponent; }),
                            multi: true
                        }]
                }] }
    ];
    return PngMultiSelectComponent;
}(DataAwareControlWrapperComponent));
export { PngMultiSelectComponent };
if (false) {
    /** @type {?} */
    PngMultiSelectComponent.prototype.$defaultProps;
}
