// @er/primeng/esm5/lib/components/multi-select/type.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PngMultiSelectProps() { }
if (false) {
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.scrollHeight;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.defaultLabel;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.panelStyle;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.panelStyleClass;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.inputId;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.disabled;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.filter;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.filterPlaceHolder;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.overlayVisible;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.tabindex;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.appendTo;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.dataKey;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.name;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.displaySelectedLabel;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.maxSelectedLabels;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.selectionLimit;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.selectedItemsLabel;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.showToggleAll;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.resetFilterOnHide;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.dropdownIcon;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.optionLabel;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.showHeader;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.autoZIndex;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.baseZIndex;
    /** @type {?|undefined} */
    PngMultiSelectProps.prototype.showTransitionOptions;
}
/** @type {?} */
export var DEFAULT_PNG_MULTI_SELECT_PROPS = {
    defaultLabel: '请选择',
    selectedItemsLabel: '选中{0} 项',
    filter: false,
    showToggleAll: true,
    $ext: {
        dataItemProps: {
            itemStyleClass: 'd-inline-flex ml-2'
        }
    }
};
